// matrix-react-sdk/lib/components/views/room_settings/RoomPublishSetting.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _partials = require("matrix-js-sdk/src/@types/partials");

var _LabelledToggleSwitch = _interopRequireDefault(require("../elements/LabelledToggleSwitch"));

var _languageHandler = require("../../../languageHandler");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Directory = _interopRequireDefault(require("../../../customisations/Directory"));

var _dec, _class;

let RoomPublishSetting = (_dec = (0, _replaceableComponent.replaceableComponent)("views.room_settings.RoomPublishSetting"), _dec(_class = class RoomPublishSetting extends _react.default.PureComponent {
  constructor(props, context) {
    super(props, context);
    (0, _defineProperty2.default)(this, "onRoomPublishChange", e => {
      const valueBefore = this.state.isRoomPublished;
      const newValue = !valueBefore;
      this.setState({
        isRoomPublished: newValue
      });

      const client = _MatrixClientPeg.MatrixClientPeg.get();

      client.setRoomDirectoryVisibility(this.props.roomId, newValue ? _partials.Visibility.Public : _partials.Visibility.Private).catch(() => {
        // Roll back the local echo on the change
        this.setState({
          isRoomPublished: valueBefore
        });
      });
    });
    this.state = {
      isRoomPublished: false
    };
  }

  componentDidMount() {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    client.getRoomDirectoryVisibility(this.props.roomId).then(result => {
      this.setState({
        isRoomPublished: result.visibility === 'public'
      });
    });
  }

  render() {
    var _DirectoryCustomisati;

    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const enabled = ((_DirectoryCustomisati = _Directory.default.requireCanonicalAliasAccessToPublish) === null || _DirectoryCustomisati === void 0 ? void 0 : _DirectoryCustomisati.call(_Directory.default)) === false || this.props.canSetCanonicalAlias;
    return /*#__PURE__*/_react.default.createElement(_LabelledToggleSwitch.default, {
      value: this.state.isRoomPublished,
      onChange: this.onRoomPublishChange,
      disabled: !enabled,
      label: (0, _languageHandler._t)("Publish this room to the public in %(domain)s's room directory?", {
        domain: client.getDomain()
      })
    });
  }

}) || _class);
exports.default = RoomPublishSetting;
