// matrix-react-sdk/lib/components/views/messages/DateSeparator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _DateUtils = require("../../../DateUtils");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

function getDaysArray() {
  return [(0, _languageHandler._t)('Sunday'), (0, _languageHandler._t)('Monday'), (0, _languageHandler._t)('Tuesday'), (0, _languageHandler._t)('Wednesday'), (0, _languageHandler._t)('Thursday'), (0, _languageHandler._t)('Friday'), (0, _languageHandler._t)('Saturday')];
}

let DateSeparator = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.DateSeparator"), _dec(_class = class DateSeparator extends _react.default.Component {
  getLabel() {
    const date = new Date(this.props.ts); // During the time the archive is being viewed, a specific day might not make sense, so we return the full date

    if (this.props.forExport) return (0, _DateUtils.formatFullDateNoTime)(date);
    const today = new Date();
    const yesterday = new Date();
    const days = getDaysArray();
    yesterday.setDate(today.getDate() - 1);

    if (date.toDateString() === today.toDateString()) {
      return (0, _languageHandler._t)('Today');
    } else if (date.toDateString() === yesterday.toDateString()) {
      return (0, _languageHandler._t)('Yesterday');
    } else if (today.getTime() - date.getTime() < 6 * 24 * 60 * 60 * 1000) {
      return days[date.getDay()];
    } else {
      return (0, _DateUtils.formatFullDateNoTime)(date);
    }
  }

  render() {
    // ARIA treats <hr/>s as separators, here we abuse them slightly so manually treat this entire thing as one
    // tab-index=-1 to allow it to be focusable but do not add tab stop for it, primarily for screen readers
    return /*#__PURE__*/_react.default.createElement("h2", {
      className: "mx_DateSeparator",
      role: "separator",
      tabIndex: -1,
      "aria-label": this.getLabel()
    }, /*#__PURE__*/_react.default.createElement("hr", {
      role: "none"
    }), /*#__PURE__*/_react.default.createElement("div", {
      "aria-hidden": "true"
    }, this.getLabel()), /*#__PURE__*/_react.default.createElement("hr", {
      role: "none"
    }));
  }

}) || _class);
exports.default = DateSeparator;
