// matrix-react-sdk/lib/components/views/messages/DownloadActionButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _RovingTabIndex = require("../../../accessibility/RovingTabIndex");

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _classnames = _interopRequireDefault(require("classnames"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _FileDownloader = require("../../../utils/FileDownloader");

var _dec, _class;

let DownloadActionButton = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.DownloadActionButton"), _dec(_class = class DownloadActionButton extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "downloader", new _FileDownloader.FileDownloader());
    (0, _defineProperty2.default)(this, "onDownloadClick", async () => {
      if (this.state.loading) return;

      if (this.props.mediaEventHelperGet().media.isEncrypted) {
        this.setState({
          tooltip: (0, _languageHandler._td)("Decrypting")
        });
      }

      this.setState({
        loading: true
      });

      if (this.state.blob) {
        // Cheat and trigger a download, again.
        return this.doDownload();
      }

      const blob = await this.props.mediaEventHelperGet().sourceBlob.value;
      this.setState({
        blob
      });
      await this.doDownload();
    });
    this.state = {
      loading: false,
      tooltip: (0, _languageHandler._td)("Downloading")
    };
  }

  async doDownload() {
    await this.downloader.download({
      blob: this.state.blob,
      name: this.props.mediaEventHelperGet().fileName
    });
    this.setState({
      loading: false
    });
  }

  render() {
    let spinner;

    if (this.state.loading) {
      spinner = /*#__PURE__*/_react.default.createElement(_Spinner.default, {
        w: 18,
        h: 18
      });
    }

    const classes = (0, _classnames.default)({
      'mx_MessageActionBar_maskButton': true,
      'mx_MessageActionBar_downloadButton': true,
      'mx_MessageActionBar_downloadSpinnerButton': !!spinner
    });
    return /*#__PURE__*/_react.default.createElement(_RovingTabIndex.RovingAccessibleTooltipButton, {
      className: classes,
      title: spinner ? (0, _languageHandler._t)(this.state.tooltip) : (0, _languageHandler._t)("Download"),
      onClick: this.onDownloadClick,
      disabled: !!spinner
    }, spinner);
  }

}) || _class);
exports.default = DownloadActionButton;
