// matrix-react-sdk/lib/components/views/messages/MImageReplyBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _MImageBody = _interopRequireDefault(require("./MImageBody"));

var _FileUtils = require("../../../utils/FileUtils");

var _SenderProfile = _interopRequireDefault(require("./SenderProfile"));

var _event = require("matrix-js-sdk/src/@types/event");

var _languageHandler = require("../../../languageHandler");

/*
Copyright 2020-2021 Tulir Asokan <tulir@maunium.net>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const FORCED_IMAGE_HEIGHT = 44;

class MImageReplyBody extends _MImageBody.default {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", ev => {
      ev.preventDefault();
    });
  }

  wrapImage(contentUrl, children) {
    return children;
  } // Don't show "Download this_file.png ..."


  getFileBody() {
    const sticker = this.props.mxEvent.getType() === _event.EventType.Sticker;

    return (0, _FileUtils.presentableTextForFile)(this.props.mxEvent.getContent(), sticker ? (0, _languageHandler._t)("Sticker") : (0, _languageHandler._t)("Image"), !sticker);
  }

  render() {
    if (this.state.error !== null) {
      return super.render();
    }

    const content = this.props.mxEvent.getContent();
    const contentUrl = this.getContentUrl();
    const thumbnail = this.messageContent(contentUrl, this.getThumbUrl(), content, FORCED_IMAGE_HEIGHT);
    const fileBody = this.getFileBody();

    const sender = /*#__PURE__*/_react.default.createElement(_SenderProfile.default, {
      mxEvent: this.props.mxEvent,
      enableFlair: false
    });

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MImageReplyBody"
    }, thumbnail, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MImageReplyBody_info"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MImageReplyBody_sender"
    }, sender), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MImageReplyBody_filename"
    }, fileBody)));
  }

}

exports.default = MImageReplyBody;
