// matrix-react-sdk/lib/components/views/messages/MjolnirBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let MjolnirBody = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.MjolnirBody"), _dec(_class = class MjolnirBody extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onAllowClick", e => {
      e.preventDefault();
      e.stopPropagation();
      const key = `mx_mjolnir_render_${this.props.mxEvent.getRoomId()}__${this.props.mxEvent.getId()}`;
      localStorage.setItem(key, "true");
      this.props.onMessageAllowed();
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MjolnirBody"
    }, /*#__PURE__*/_react.default.createElement("i", null, (0, _languageHandler._t)("You have ignored this user, so their message is hidden. <a>Show anyways.</a>", {}, {
      a: sub => /*#__PURE__*/_react.default.createElement("a", {
        href: "#",
        onClick: this.onAllowClick
      }, sub)
    })));
  }

}) || _class);
exports.default = MjolnirBody;
