// matrix-react-sdk/lib/components/views/messages/RoomCreate.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _Permalinks = require("../../../utils/permalinks/Permalinks");

var _languageHandler = require("../../../languageHandler");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _EventTileBubble = _interopRequireDefault(require("./EventTileBubble"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let RoomCreate = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.RoomCreate"), _dec(_class = class RoomCreate extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onLinkClicked", e => {
      e.preventDefault();
      const predecessor = this.props.mxEvent.getContent()['predecessor'];

      _dispatcher.default.dispatch({
        action: _actions.Action.ViewRoom,
        event_id: predecessor['event_id'],
        highlighted: true,
        room_id: predecessor['room_id']
      });
    });
  }

  render() {
    const predecessor = this.props.mxEvent.getContent()['predecessor'];

    if (predecessor === undefined) {
      return /*#__PURE__*/_react.default.createElement("div", null); // We should never have been instantiated in this case
    }

    const prevRoom = _MatrixClientPeg.MatrixClientPeg.get().getRoom(predecessor['room_id']);

    const permalinkCreator = new _Permalinks.RoomPermalinkCreator(prevRoom, predecessor['room_id']);
    permalinkCreator.load();
    const predecessorPermalink = permalinkCreator.forEvent(predecessor['event_id']);

    const link = /*#__PURE__*/_react.default.createElement("a", {
      href: predecessorPermalink,
      onClick: this.onLinkClicked
    }, (0, _languageHandler._t)("Click here to see older messages."));

    return /*#__PURE__*/_react.default.createElement(_EventTileBubble.default, {
      className: "mx_CreateEvent",
      title: (0, _languageHandler._t)("This room is a continuation of another conversation."),
      subtitle: link
    });
  }

}) || _class);
exports.default = RoomCreate;
