// matrix-react-sdk/lib/components/views/messages/TileErrorBoundary.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _languageHandler = require("../../../languageHandler");

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BugReportDialog = _interopRequireDefault(require("../dialogs/BugReportDialog"));

var _dec, _class;

let TileErrorBoundary = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.TileErrorBoundary"), _dec(_class = class TileErrorBoundary extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onBugReport", () => {
      _Modal.default.createTrackedDialog('Bug Report Dialog', '', _BugReportDialog.default, {
        label: 'react-soft-crash-tile',
        error: this.state.error
      });
    });
    this.state = {
      error: null
    };
  }

  static getDerivedStateFromError(error) {
    // Side effects are not permitted here, so we only update the state so
    // that the next render shows an error message.
    return {
      error
    };
  }

  render() {
    if (this.state.error) {
      const {
        mxEvent
      } = this.props;
      const classes = {
        mx_EventTile: true,
        mx_EventTile_info: true,
        mx_EventTile_content: true,
        mx_EventTile_tileError: true
      };
      let submitLogsButton;

      if (_SdkConfig.default.get().bug_report_endpoint_url) {
        submitLogsButton = /*#__PURE__*/_react.default.createElement("a", {
          onClick: this.onBugReport,
          href: "#"
        }, (0, _languageHandler._t)("Submit logs"));
      }

      return /*#__PURE__*/_react.default.createElement("div", {
        className: (0, _classnames.default)(classes)
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_EventTile_line"
      }, /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Can't load this message"), mxEvent && ` (${mxEvent.getType()})`, submitLogsButton)));
    }

    return this.props.children;
  }

}) || _class);
exports.default = TileErrorBoundary;
