// matrix-react-sdk/lib/stores/widgets/StopGapWidgetDriver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StopGapWidgetDriver = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _matrixWidgetApi = require("matrix-widget-api");

var _iterables = require("../../utils/iterables");

var _MatrixClientPeg = require("../../MatrixClientPeg");

var _ActiveRoomObserver = _interopRequireDefault(require("../../ActiveRoomObserver"));

var _Modal = _interopRequireDefault(require("../../Modal"));

var _WidgetOpenIDPermissionsDialog = _interopRequireDefault(require("../../components/views/dialogs/WidgetOpenIDPermissionsDialog"));

var _WidgetCapabilitiesPromptDialog = _interopRequireDefault(require("../../components/views/dialogs/WidgetCapabilitiesPromptDialog"));

var _WidgetPermissions = require("../../customisations/WidgetPermissions");

var _WidgetPermissionStore = require("./WidgetPermissionStore");

var _WidgetType = require("../../widgets/WidgetType");

var _event = require("matrix-js-sdk/src/@types/event");

var _effects = require("../../effects");

var _utils = require("../../effects/utils");

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _Permalinks = require("../../utils/permalinks/Permalinks");

var _logger = require("matrix-js-sdk/src/logger");

var _SettingsStore = _interopRequireDefault(require("../../settings/SettingsStore"));

/*
 * Copyright 2020 - 2021 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// TODO: Purge this from the universe
function getRememberedCapabilitiesForWidget(widget) {
  return JSON.parse(localStorage.getItem(`widget_${widget.id}_approved_caps`) || "[]");
}

function setRememberedCapabilitiesForWidget(widget, caps) {
  localStorage.setItem(`widget_${widget.id}_approved_caps`, JSON.stringify(caps));
}

class StopGapWidgetDriver extends _matrixWidgetApi.WidgetDriver {
  // TODO: Refactor widgetKind into the Widget class
  constructor(allowedCapabilities, forWidget, forWidgetKind, inRoomId) {
    super(); // Always allow screenshots to be taken because it's a client-induced flow. The widget can't
    // spew screenshots at us and can't request screenshots of us, so it's up to us to provide the
    // button if the widget says it supports screenshots.

    this.forWidget = forWidget;
    this.forWidgetKind = forWidgetKind;
    this.inRoomId = inRoomId;
    (0, _defineProperty2.default)(this, "allowedCapabilities", void 0);
    this.allowedCapabilities = new Set([...allowedCapabilities, _matrixWidgetApi.MatrixCapabilities.Screenshots, _matrixWidgetApi.MatrixCapabilities.RequiresClient]); // Grant the permissions that are specific to given widget types

    if (_WidgetType.WidgetType.JITSI.matches(this.forWidget.type) && forWidgetKind === _matrixWidgetApi.WidgetKind.Room) {
      this.allowedCapabilities.add(_matrixWidgetApi.MatrixCapabilities.AlwaysOnScreen);
    } else if (_WidgetType.WidgetType.STICKERPICKER.matches(this.forWidget.type) && forWidgetKind === _matrixWidgetApi.WidgetKind.Account) {
      const stickerSendingCap = _matrixWidgetApi.WidgetEventCapability.forRoomEvent(_matrixWidgetApi.EventDirection.Send, _event.EventType.Sticker).raw;

      this.allowedCapabilities.add(_matrixWidgetApi.MatrixCapabilities.StickerSending); // legacy as far as MSC2762 is concerned

      this.allowedCapabilities.add(stickerSendingCap); // Auto-approve the legacy visibility capability. We send it regardless of capability.
      // Widgets don't technically need to request this capability, but Scalar still does.

      this.allowedCapabilities.add("visibility");
    }
  }

  async validateCapabilities(requested) {
    // Check to see if any capabilities aren't automatically accepted (such as sticker pickers
    // allowing stickers to be sent). If there are excess capabilities to be approved, the user
    // will be prompted to accept them.
    const diff = (0, _iterables.iterableDiff)(requested, this.allowedCapabilities);
    const missing = new Set(diff.removed); // "removed" is "in A (requested) but not in B (allowed)"

    const allowedSoFar = new Set(this.allowedCapabilities);
    getRememberedCapabilitiesForWidget(this.forWidget).forEach(cap => {
      allowedSoFar.add(cap);
      missing.delete(cap);
    });

    if (_WidgetPermissions.WidgetPermissionCustomisations.preapproveCapabilities) {
      const approved = await _WidgetPermissions.WidgetPermissionCustomisations.preapproveCapabilities(this.forWidget, requested);

      if (approved) {
        approved.forEach(cap => {
          allowedSoFar.add(cap);
          missing.delete(cap);
        });
      }
    } // TODO: Do something when the widget requests new capabilities not yet asked for


    let rememberApproved = false;

    if (missing.size > 0) {
      try {
        const [result] = await _Modal.default.createTrackedDialog('Approve Widget Caps', '', _WidgetCapabilitiesPromptDialog.default, {
          requestedCapabilities: missing,
          widget: this.forWidget,
          widgetKind: this.forWidgetKind
        }).finished;
        (result.approved || []).forEach(cap => allowedSoFar.add(cap));
        rememberApproved = result.remember;
      } catch (e) {
        _logger.logger.error("Non-fatal error getting capabilities: ", e);
      }
    }

    const allAllowed = new Set((0, _iterables.iterableUnion)(allowedSoFar, requested));

    if (rememberApproved) {
      setRememberedCapabilitiesForWidget(this.forWidget, Array.from(allAllowed));
    }

    return allAllowed;
  }

  async sendEvent(eventType, content, stateKey = null, targetRoomId = null) {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const roomId = targetRoomId || _ActiveRoomObserver.default.activeRoomId;
    if (!client || !roomId) throw new Error("Not in a room or not attached to a client");
    let r = null; // eslint-disable-line camelcase

    if (stateKey !== null) {
      // state event
      r = await client.sendStateEvent(roomId, eventType, content, stateKey);
    } else if (eventType === _event.EventType.RoomRedaction) {
      // special case: extract the `redacts` property and call redact
      r = await client.redactEvent(roomId, content['redacts']);
    } else {
      // message event
      r = await client.sendEvent(roomId, eventType, content);

      if (eventType === _event.EventType.RoomMessage) {
        _effects.CHAT_EFFECTS.forEach(effect => {
          if ((0, _utils.containsEmoji)(content, effect.emojis)) {
            // For initial threads launch, chat effects are disabled
            // see #19731
            const isNotThread = content["m.relates_to"].rel_type !== _event.RelationType.Thread;

            if (!_SettingsStore.default.getValue("feature_thread") || isNotThread) {
              _dispatcher.default.dispatch({
                action: `effects.${effect.command}`
              });
            }
          }
        });
      }
    }

    return {
      roomId,
      eventId: r.event_id
    };
  }

  pickRooms(roomIds = null) {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    if (!client) throw new Error("Not attached to a client");
    const targetRooms = roomIds ? roomIds.includes(_matrixWidgetApi.Symbols.AnyRoom) ? client.getVisibleRooms() : roomIds.map(r => client.getRoom(r)) : [client.getRoom(_ActiveRoomObserver.default.activeRoomId)];
    return targetRooms.filter(r => !!r);
  }

  async readRoomEvents(eventType, msgtype, limitPerRoom, roomIds = null) {
    limitPerRoom = limitPerRoom > 0 ? Math.min(limitPerRoom, Number.MAX_SAFE_INTEGER) : Number.MAX_SAFE_INTEGER; // relatively arbitrary

    const rooms = this.pickRooms(roomIds);
    const allResults = [];

    for (const room of rooms) {
      const results = [];
      const events = room.getLiveTimeline().getEvents(); // timelines are most recent last

      for (let i = events.length - 1; i > 0; i--) {
        if (results.length >= limitPerRoom) break;
        const ev = events[i];
        if (ev.getType() !== eventType || ev.isState()) continue;
        if (eventType === _event.EventType.RoomMessage && msgtype && msgtype !== ev.getContent()['msgtype']) continue;
        results.push(ev);
      }

      results.forEach(e => allResults.push(e.getEffectiveEvent()));
    }

    return allResults;
  }

  async readStateEvents(eventType, stateKey, limitPerRoom, roomIds = null) {
    limitPerRoom = limitPerRoom > 0 ? Math.min(limitPerRoom, Number.MAX_SAFE_INTEGER) : Number.MAX_SAFE_INTEGER; // relatively arbitrary

    const rooms = this.pickRooms(roomIds);
    const allResults = [];

    for (const room of rooms) {
      const results = [];
      const state = room.currentState.events.get(eventType);

      if (state) {
        if (stateKey === "" || !!stateKey) {
          const forKey = state.get(stateKey);
          if (forKey) results.push(forKey);
        } else {
          results.push(...Array.from(state.values()));
        }
      }

      results.slice(0, limitPerRoom).forEach(e => allResults.push(e.getEffectiveEvent()));
    }

    return allResults;
  }

  async askOpenID(observer) {
    const oidcState = _WidgetPermissionStore.WidgetPermissionStore.instance.getOIDCState(this.forWidget, this.forWidgetKind, this.inRoomId);

    const getToken = () => {
      return _MatrixClientPeg.MatrixClientPeg.get().getOpenIdToken();
    };

    if (oidcState === _WidgetPermissionStore.OIDCState.Denied) {
      return observer.update({
        state: _matrixWidgetApi.OpenIDRequestState.Blocked
      });
    }

    if (oidcState === _WidgetPermissionStore.OIDCState.Allowed) {
      return observer.update({
        state: _matrixWidgetApi.OpenIDRequestState.Allowed,
        token: await getToken()
      });
    }

    observer.update({
      state: _matrixWidgetApi.OpenIDRequestState.PendingUserConfirmation
    });

    _Modal.default.createTrackedDialog("OpenID widget permissions", '', _WidgetOpenIDPermissionsDialog.default, {
      widget: this.forWidget,
      widgetKind: this.forWidgetKind,
      inRoomId: this.inRoomId,
      onFinished: async confirm => {
        if (!confirm) {
          return observer.update({
            state: _matrixWidgetApi.OpenIDRequestState.Blocked
          });
        }

        return observer.update({
          state: _matrixWidgetApi.OpenIDRequestState.Allowed,
          token: await getToken()
        });
      }
    });
  }

  async navigate(uri) {
    const localUri = (0, _Permalinks.tryTransformPermalinkToLocalHref)(uri);

    if (!localUri || localUri === uri) {
      // parse failure can lead to an unmodified URL
      throw new Error("Failed to transform URI");
    }

    window.location.hash = localUri; // it'll just be a fragment
  }

}

exports.StopGapWidgetDriver = StopGapWidgetDriver;
