// matrix-react-sdk/lib/components/views/dialogs/security/ConfirmDestroyCrossSigningDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../../languageHandler");

var _replaceableComponent = require("../../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("../BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../../elements/DialogButtons"));

var _dec, _class;

let ConfirmDestroyCrossSigningDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.security.ConfirmDestroyCrossSigningDialog"), _dec(_class = class ConfirmDestroyCrossSigningDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onConfirm", () => {
      this.props.onFinished(true);
    });
    (0, _defineProperty2.default)(this, "onDecline", () => {
      this.props.onFinished(false);
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_ConfirmDestroyCrossSigningDialog",
      hasCancel: true,
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Destroy cross-signing keys?")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ConfirmDestroyCrossSigningDialog_content"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Deleting cross-signing keys is permanent. " + "Anyone you have verified with will see security alerts. " + "You almost certainly don't want to do this, unless " + "you've lost every device you can cross-sign from."))), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("Clear cross-signing keys"),
      onPrimaryButtonClick: this.onConfirm,
      primaryButtonClass: "danger",
      cancelButton: (0, _languageHandler._t)("Cancel"),
      onCancel: this.onDecline
    }));
  }

}) || _class);
exports.default = ConfirmDestroyCrossSigningDialog;
