// matrix-react-sdk/lib/components/views/dialogs/security/SetupEncryptionDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _SetupEncryptionBody = _interopRequireDefault(require("../../../structures/auth/SetupEncryptionBody"));

var _BaseDialog = _interopRequireDefault(require("../BaseDialog"));

var _languageHandler = require("../../../../languageHandler");

var _SetupEncryptionStore = require("../../../../stores/SetupEncryptionStore");

var _replaceableComponent = require("../../../../utils/replaceableComponent");

var _dec, _class;

function iconFromPhase(phase) {
  if (phase === _SetupEncryptionStore.Phase.Done) {
    return require("../../../../../res/img/e2e/verified.svg");
  } else {
    return require("../../../../../res/img/e2e/warning.svg");
  }
}

let SetupEncryptionDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.security.SetupEncryptionDialog"), _dec(_class = class SetupEncryptionDialog extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "store", void 0);
    (0, _defineProperty2.default)(this, "onStoreUpdate", () => {
      this.setState({
        icon: iconFromPhase(this.store.phase)
      });
    });
    this.store = _SetupEncryptionStore.SetupEncryptionStore.sharedInstance();
    this.state = {
      icon: iconFromPhase(this.store.phase)
    };
  }

  componentDidMount() {
    this.store.on("update", this.onStoreUpdate);
  }

  componentWillUnmount() {
    this.store.removeListener("update", this.onStoreUpdate);
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      headerImage: this.state.icon,
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Verify this session")
    }, /*#__PURE__*/_react.default.createElement(_SetupEncryptionBody.default, {
      onFinished: this.props.onFinished
    }));
  }

}) || _class);
exports.default = SetupEncryptionDialog;
