// matrix-react-sdk/lib/components/views/audio_messages/PlaybackClock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Clock = _interopRequireDefault(require("./Clock"));

var _Playback = require("../../../audio/Playback");

var _AsyncStore = require("../../../stores/AsyncStore");

var _dec, _class;

let PlaybackClock = (
/**
 * A clock for a playback of a recording.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.PlaybackClock"), _dec(_class = class PlaybackClock extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onPlaybackUpdate", ev => {
      // Convert Decoding -> Stopped because we don't care about the distinction here
      if (ev === _Playback.PlaybackState.Decoding) ev = _Playback.PlaybackState.Stopped;
      this.setState({
        playbackPhase: ev
      });
    });
    (0, _defineProperty2.default)(this, "onTimeUpdate", time => {
      this.setState({
        seconds: time[0],
        durationSeconds: time[1]
      });
    });
    this.state = {
      seconds: this.props.playback.clockInfo.timeSeconds,
      // we track the duration on state because we won't really know what the clip duration
      // is until the first time update, and as a PureComponent we are trying to dedupe state
      // updates as much as possible. This is just the easiest way to avoid a forceUpdate() or
      // member property to track "did we get a duration".
      durationSeconds: this.props.playback.clockInfo.durationSeconds,
      playbackPhase: _Playback.PlaybackState.Stopped // assume not started, so full clock

    };
    this.props.playback.on(_AsyncStore.UPDATE_EVENT, this.onPlaybackUpdate);
    this.props.playback.clockInfo.liveData.onUpdate(this.onTimeUpdate);
  }

  render() {
    let seconds = this.state.seconds;

    if (this.state.playbackPhase === _Playback.PlaybackState.Stopped) {
      if (Number.isFinite(this.props.defaultDisplaySeconds)) {
        seconds = this.props.defaultDisplaySeconds;
      } else {
        seconds = this.state.durationSeconds;
      }
    }

    return /*#__PURE__*/_react.default.createElement(_Clock.default, {
      seconds: seconds
    });
  }

}) || _class);
exports.default = PlaybackClock;
