// matrix-react-sdk/lib/components/views/audio_messages/AudioPlayerBase.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Playback = require("../../../audio/Playback");

var _react = _interopRequireDefault(require("react"));

var _AsyncStore = require("../../../stores/AsyncStore");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _languageHandler = require("../../../languageHandler");

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class;

let AudioPlayerBase = (_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.AudioPlayerBase"), _dec(_class = class AudioPlayerBase extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onPlaybackUpdate", ev => {
      this.setState({
        playbackPhase: ev
      });
    });
    this.state = {
      playbackPhase: _Playback.PlaybackState.Decoding // default assumption

    }; // We don't need to de-register: the class handles this for us internally

    this.props.playback.on(_AsyncStore.UPDATE_EVENT, this.onPlaybackUpdate); // Don't wait for the promise to complete - it will emit a progress update when it
    // is done, and it's not meant to take long anyhow.

    this.props.playback.prepare().catch(e => {
      _logger.logger.error("Error processing audio file:", e);

      this.setState({
        error: true
      });
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, this.renderComponent(), this.state.error && /*#__PURE__*/_react.default.createElement("div", {
      className: "text-warning"
    }, (0, _languageHandler._t)("Error downloading audio")));
  }

}) || _class);
exports.default = AudioPlayerBase;
