// matrix-react-sdk/lib/components/views/audio_messages/LiveRecordingClock.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Clock = _interopRequireDefault(require("./Clock"));

var _MarkedExecution = require("../../../utils/MarkedExecution");

var _dec, _class;

let LiveRecordingClock = (
/**
 * A clock for a live recording.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.LiveRecordingClock"), _dec(_class = class LiveRecordingClock extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "seconds", 0);
    (0, _defineProperty2.default)(this, "scheduledUpdate", new _MarkedExecution.MarkedExecution(() => this.updateClock(), () => requestAnimationFrame(() => this.scheduledUpdate.trigger())));
    this.state = {
      seconds: 0
    };
  }

  componentDidMount() {
    this.props.recorder.liveData.onUpdate(update => {
      this.seconds = update.timeSeconds;
      this.scheduledUpdate.mark();
    });
  }

  updateClock() {
    this.setState({
      seconds: this.seconds
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_Clock.default, {
      seconds: this.state.seconds
    });
  }

}) || _class);
exports.default = LiveRecordingClock;
