// matrix-react-sdk/lib/components/views/audio_messages/LiveRecordingWaveform.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _VoiceRecording = require("../../../audio/VoiceRecording");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _arrays = require("../../../utils/arrays");

var _Waveform = _interopRequireDefault(require("./Waveform"));

var _MarkedExecution = require("../../../utils/MarkedExecution");

var _dec, _class, _class2, _temp;

let LiveRecordingWaveform = (
/**
 * A waveform which shows the waveform of a live recording
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.LiveRecordingWaveform"), _dec(_class = (_temp = _class2 = class LiveRecordingWaveform extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "waveform", []);
    (0, _defineProperty2.default)(this, "scheduledUpdate", new _MarkedExecution.MarkedExecution(() => this.updateWaveform(), () => requestAnimationFrame(() => this.scheduledUpdate.trigger())));
    this.state = {
      waveform: (0, _arrays.arraySeed)(0, _VoiceRecording.RECORDING_PLAYBACK_SAMPLES)
    };
  }

  componentDidMount() {
    this.props.recorder.liveData.onUpdate(update => {
      // The incoming data is between zero and one, so we don't need to clamp/rescale it.
      this.waveform = (0, _arrays.arrayFastResample)(Array.from(update.waveform), _VoiceRecording.RECORDING_PLAYBACK_SAMPLES);
      this.scheduledUpdate.mark();
    });
  }

  updateWaveform() {
    this.setState({
      waveform: this.waveform
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_Waveform.default, {
      relHeights: this.state.waveform
    });
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  progress: 1
}), _temp)) || _class);
exports.default = LiveRecordingWaveform;
