// matrix-react-sdk/lib/components/views/audio_messages/PlaybackWaveform.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _arrays = require("../../../utils/arrays");

var _Waveform = _interopRequireDefault(require("./Waveform"));

var _Playback = require("../../../audio/Playback");

var _numbers = require("../../../utils/numbers");

var _dec, _class;

let PlaybackWaveform = (
/**
 * A waveform which shows the waveform of a previously recorded recording
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.PlaybackWaveform"), _dec(_class = class PlaybackWaveform extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onWaveformUpdate", waveform => {
      this.setState({
        heights: this.toHeights(waveform)
      });
    });
    (0, _defineProperty2.default)(this, "onTimeUpdate", time => {
      // Track percentages to a general precision to avoid over-waking the component.
      const progress = Number((0, _numbers.percentageOf)(time[0], 0, time[1]).toFixed(3));
      this.setState({
        progress
      });
    });
    this.state = {
      heights: this.toHeights(this.props.playback.waveform),
      progress: 0 // default no progress

    };
    this.props.playback.waveformData.onUpdate(this.onWaveformUpdate);
    this.props.playback.clockInfo.liveData.onUpdate(this.onTimeUpdate);
  }

  toHeights(waveform) {
    const seed = (0, _arrays.arraySeed)(0, _Playback.PLAYBACK_WAVEFORM_SAMPLES);
    return (0, _arrays.arrayTrimFill)(waveform, _Playback.PLAYBACK_WAVEFORM_SAMPLES, seed);
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_Waveform.default, {
      relHeights: this.state.heights,
      progress: this.state.progress
    });
  }

}) || _class);
exports.default = PlaybackWaveform;
