// matrix-react-sdk/lib/components/views/audio_messages/PlayPauseButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _AccessibleTooltipButton = _interopRequireDefault(require("../elements/AccessibleTooltipButton"));

var _languageHandler = require("../../../languageHandler");

var _Playback = require("../../../audio/Playback");

var _classnames = _interopRequireDefault(require("classnames"));

const _excluded = ["playback", "playbackPhase"];

var _dec, _class;

let PlayPauseButton = (
/**
 * Displays a play/pause button (activating the play/pause function of the recorder)
 * to be displayed in reference to a recording.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.PlayPauseButton"), _dec(_class = class PlayPauseButton extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onClick", () => {
      // noinspection JSIgnoredPromiseFromCall
      this.toggleState();
    });
  }

  async toggleState() {
    await this.props.playback.toggle();
  }

  render() {
    const _this$props = this.props,
          {
      playback,
      playbackPhase
    } = _this$props,
          restProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    const isPlaying = playback.isPlaying;
    const isDisabled = playbackPhase === _Playback.PlaybackState.Decoding;
    const classes = (0, _classnames.default)('mx_PlayPauseButton', {
      'mx_PlayPauseButton_play': !isPlaying,
      'mx_PlayPauseButton_pause': isPlaying,
      'mx_PlayPauseButton_disabled': isDisabled
    });
    return /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, (0, _extends2.default)({
      className: classes,
      title: isPlaying ? (0, _languageHandler._t)("Pause") : (0, _languageHandler._t)("Play"),
      onClick: this.onClick,
      disabled: isDisabled
    }, restProps));
  }

}) || _class);
exports.default = PlayPauseButton;
