// matrix-react-sdk/lib/components/views/audio_messages/RecordingPlayback.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _PlayPauseButton = _interopRequireDefault(require("./PlayPauseButton"));

var _PlaybackClock = _interopRequireDefault(require("./PlaybackClock"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _EventTile = require("../rooms/EventTile");

var _PlaybackWaveform = _interopRequireDefault(require("./PlaybackWaveform"));

var _AudioPlayerBase = _interopRequireDefault(require("./AudioPlayerBase"));

var _dec, _class;

let RecordingPlayback = (_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.RecordingPlayback"), _dec(_class = class RecordingPlayback extends _AudioPlayerBase.default {
  get isWaveformable() {
    return this.props.tileShape !== _EventTile.TileShape.Notif && this.props.tileShape !== _EventTile.TileShape.FileGrid && this.props.tileShape !== _EventTile.TileShape.Pinned;
  }

  renderComponent() {
    const shapeClass = !this.isWaveformable ? 'mx_VoiceMessagePrimaryContainer_noWaveform' : '';
    return /*#__PURE__*/_react.default.createElement("div", {
      className: 'mx_MediaBody mx_VoiceMessagePrimaryContainer ' + shapeClass
    }, /*#__PURE__*/_react.default.createElement(_PlayPauseButton.default, {
      playback: this.props.playback,
      playbackPhase: this.state.playbackPhase
    }), /*#__PURE__*/_react.default.createElement(_PlaybackClock.default, {
      playback: this.props.playback
    }), this.isWaveformable && /*#__PURE__*/_react.default.createElement(_PlaybackWaveform.default, {
      playback: this.props.playback
    }));
  }

}) || _class);
exports.default = RecordingPlayback;
