// matrix-react-sdk/lib/components/views/audio_messages/Waveform.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _classnames = _interopRequireDefault(require("classnames"));

var _dec, _class, _class2, _temp;

let Waveform = (
/**
 * A simple waveform component. This renders bars (centered vertically) for each
 * height provided in the component properties. Updating the properties will update
 * the rendered waveform.
 *
 * For CSS purposes, a mx_Waveform_bar_100pct class is added when the bar should be
 * "filled", as a demonstration of the progress property.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.audio_messages.Waveform"), _dec(_class = (_temp = _class2 = class Waveform extends _react.default.PureComponent {
  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Waveform"
    }, this.props.relHeights.map((h, i) => {
      const progress = this.props.progress;
      const isCompleteBar = i / this.props.relHeights.length <= progress && progress > 0;
      const classes = (0, _classnames.default)({
        'mx_Waveform_bar': true,
        'mx_Waveform_bar_100pct': isCompleteBar
      });
      return /*#__PURE__*/_react.default.createElement("span", {
        key: i,
        style: {
          "--barHeight": h
        },
        className: classes
      });
    }));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  progress: 1
}), _temp)) || _class);
exports.default = Waveform;
