// matrix-react-sdk/lib/toasts/AnalyticsToast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showToast = exports.hideToast = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../SdkConfig"));

var _dispatcher = _interopRequireDefault(require("../dispatcher/dispatcher"));

var _Analytics = _interopRequireDefault(require("../Analytics"));

var _AccessibleButton = _interopRequireDefault(require("../components/views/elements/AccessibleButton"));

var _GenericToast = _interopRequireDefault(require("../components/views/toasts/GenericToast"));

var _ToastStore = _interopRequireDefault(require("../stores/ToastStore"));

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const onAccept = () => {
  _dispatcher.default.dispatch({
    action: 'accept_cookies'
  });
};

const onReject = () => {
  _dispatcher.default.dispatch({
    action: "reject_cookies"
  });
};

const onUsageDataClicked = () => {
  _Analytics.default.showDetailsModal();
};

const TOAST_KEY = "analytics";

const showToast = policyUrl => {
  const brand = _SdkConfig.default.get().brand;

  _ToastStore.default.sharedInstance().addOrReplaceToast({
    key: TOAST_KEY,
    title: (0, _languageHandler._t)("Help us improve %(brand)s", {
      brand
    }),
    props: {
      description: (0, _languageHandler._t)("Send <UsageDataLink>anonymous usage data</UsageDataLink> which helps us improve %(brand)s. " + "This will use a <PolicyLink>cookie</PolicyLink>.", {
        brand
      }, {
        "UsageDataLink": sub => /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
          kind: "link",
          onClick: onUsageDataClicked
        }, sub),
        // XXX: We need to link to the page that explains our cookies
        "PolicyLink": sub => policyUrl ? /*#__PURE__*/_react.default.createElement("a", {
          target: "_blank",
          href: policyUrl
        }, sub) : sub
      }),
      acceptLabel: (0, _languageHandler._t)("Yes"),
      onAccept,
      rejectLabel: (0, _languageHandler._t)("No"),
      onReject
    },
    component: _GenericToast.default,
    className: "mx_AnalyticsToast",
    priority: 10
  });
};

exports.showToast = showToast;

const hideToast = () => {
  _ToastStore.default.sharedInstance().dismissToast(TOAST_KEY);
};

exports.hideToast = hideToast;
