// matrix-react-sdk/lib/toasts/UpdateToast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showToast = exports.hideToast = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../SdkConfig"));

var _GenericToast = _interopRequireDefault(require("../components/views/toasts/GenericToast"));

var _ToastStore = _interopRequireDefault(require("../stores/ToastStore"));

var _QuestionDialog = _interopRequireDefault(require("../components/views/dialogs/QuestionDialog"));

var _ChangelogDialog = _interopRequireDefault(require("../components/views/dialogs/ChangelogDialog"));

var _PlatformPeg = _interopRequireDefault(require("../PlatformPeg"));

var _Modal = _interopRequireDefault(require("../Modal"));

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const TOAST_KEY = "update";
/*
 * Check a version string is compatible with the Changelog
 * dialog ([element-version]-react-[react-sdk-version]-js-[js-sdk-version])
 */

function checkVersion(ver) {
  const parts = ver.split('-');
  return parts.length === 5 && parts[1] === 'react' && parts[3] === 'js';
}

function installUpdate() {
  _PlatformPeg.default.get().installUpdate();
}

const showToast = (version, newVersion, releaseNotes) => {
  function onReject() {
    _PlatformPeg.default.get().deferUpdate(newVersion);
  }

  let onAccept;
  let acceptLabel = (0, _languageHandler._t)("What's new?");

  if (releaseNotes) {
    onAccept = () => {
      _Modal.default.createTrackedDialog('Display release notes', '', _QuestionDialog.default, {
        title: (0, _languageHandler._t)("What's New"),
        description: /*#__PURE__*/_react.default.createElement("pre", null, releaseNotes),
        button: (0, _languageHandler._t)("Update"),
        onFinished: update => {
          if (update && _PlatformPeg.default.get()) {
            _PlatformPeg.default.get().installUpdate();
          }
        }
      });
    };
  } else if (checkVersion(version) && checkVersion(newVersion)) {
    onAccept = () => {
      _Modal.default.createTrackedDialog('Display Changelog', '', _ChangelogDialog.default, {
        version,
        newVersion,
        onFinished: update => {
          if (update && _PlatformPeg.default.get()) {
            _PlatformPeg.default.get().installUpdate();
          }
        }
      });
    };
  } else {
    onAccept = installUpdate;
    acceptLabel = (0, _languageHandler._t)("Update");
  }

  const brand = _SdkConfig.default.get().brand;

  _ToastStore.default.sharedInstance().addOrReplaceToast({
    key: TOAST_KEY,
    title: (0, _languageHandler._t)("Update %(brand)s", {
      brand
    }),
    props: {
      description: (0, _languageHandler._t)("New version of %(brand)s is available", {
        brand
      }),
      acceptLabel,
      onAccept,
      rejectLabel: (0, _languageHandler._t)("Dismiss"),
      onReject
    },
    component: _GenericToast.default,
    priority: 20
  });
};

exports.showToast = showToast;

const hideToast = () => {
  _ToastStore.default.sharedInstance().dismissToast(TOAST_KEY);
};

exports.hideToast = hideToast;
