// matrix-react-sdk/lib/components/views/spaces/SpaceBasicSettings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SpaceAvatar = void 0;

var _react = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _Field = _interopRequireDefault(require("../elements/Field"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const SpaceAvatar = ({
  avatarUrl,
  avatarDisabled = false,
  setAvatar
}) => {
  const avatarUploadRef = (0, _react.useRef)();
  const [avatar, setAvatarDataUrl] = (0, _react.useState)(avatarUrl); // avatar data url cache

  let avatarSection;

  if (avatarDisabled) {
    if (avatar) {
      avatarSection = /*#__PURE__*/_react.default.createElement("img", {
        className: "mx_SpaceBasicSettings_avatar",
        src: avatar,
        alt: ""
      });
    } else {
      avatarSection = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SpaceBasicSettings_avatar"
      });
    }
  } else {
    if (avatar) {
      avatarSection = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        className: "mx_SpaceBasicSettings_avatar",
        onClick: () => {
          var _avatarUploadRef$curr;

          return (_avatarUploadRef$curr = avatarUploadRef.current) === null || _avatarUploadRef$curr === void 0 ? void 0 : _avatarUploadRef$curr.click();
        },
        element: "img",
        src: avatar,
        alt: ""
      }), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        onClick: () => {
          avatarUploadRef.current.value = "";
          setAvatarDataUrl(undefined);
          setAvatar(undefined);
        },
        kind: "link",
        className: "mx_SpaceBasicSettings_avatar_remove",
        "aria-label": (0, _languageHandler._t)("Delete avatar")
      }, (0, _languageHandler._t)("Delete")));
    } else {
      avatarSection = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SpaceBasicSettings_avatar",
        onClick: () => {
          var _avatarUploadRef$curr2;

          return (_avatarUploadRef$curr2 = avatarUploadRef.current) === null || _avatarUploadRef$curr2 === void 0 ? void 0 : _avatarUploadRef$curr2.click();
        }
      }), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        onClick: () => {
          var _avatarUploadRef$curr3;

          return (_avatarUploadRef$curr3 = avatarUploadRef.current) === null || _avatarUploadRef$curr3 === void 0 ? void 0 : _avatarUploadRef$curr3.click();
        },
        kind: "link",
        "aria-label": (0, _languageHandler._t)("Upload avatar")
      }, (0, _languageHandler._t)("Upload")));
    }
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceBasicSettings_avatarContainer"
  }, avatarSection, /*#__PURE__*/_react.default.createElement("input", {
    type: "file",
    ref: avatarUploadRef,
    onChange: e => {
      var _e$target$files;

      if (!((_e$target$files = e.target.files) !== null && _e$target$files !== void 0 && _e$target$files.length)) return;
      const file = e.target.files[0];
      setAvatar(file);
      const reader = new FileReader();

      reader.onload = ev => {
        setAvatarDataUrl(ev.target.result);
      };

      reader.readAsDataURL(file);
    },
    accept: "image/*"
  }));
};

exports.SpaceAvatar = SpaceAvatar;

const SpaceBasicSettings = ({
  avatarUrl,
  avatarDisabled = false,
  setAvatar,
  name = "",
  nameDisabled = false,
  setName,
  topic = "",
  topicDisabled = false,
  setTopic
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceBasicSettings"
  }, /*#__PURE__*/_react.default.createElement(SpaceAvatar, {
    avatarUrl: avatarUrl,
    avatarDisabled: avatarDisabled,
    setAvatar: setAvatar
  }), /*#__PURE__*/_react.default.createElement(_Field.default, {
    name: "spaceName",
    label: (0, _languageHandler._t)("Name"),
    autoFocus: true,
    value: name,
    onChange: ev => setName(ev.target.value),
    disabled: nameDisabled
  }), /*#__PURE__*/_react.default.createElement(_Field.default, {
    name: "spaceTopic",
    element: "textarea",
    label: (0, _languageHandler._t)("Description"),
    value: topic,
    onChange: ev => setTopic(ev.target.value),
    rows: 3,
    disabled: topicDisabled
  }));
};

var _default = SpaceBasicSettings;
exports.default = _default;
