// matrix-react-sdk/lib/stores/room-list/filters/SpaceFilterCondition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpaceFilterCondition = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _events = require("events");

var _IFilterCondition = require("./IFilterCondition");

var _SpaceStore = _interopRequireDefault(require("../../spaces/SpaceStore"));

var _spaces = require("../../spaces");

var _sets = require("../../../utils/sets");

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * A filter condition for the room list which reveals rooms which
 * are a member of a given space or if no space is selected shows:
 *  + Orphaned rooms (ones not in any space you are a part of)
 *  + All DMs
 */
class SpaceFilterCondition extends _events.EventEmitter {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "roomIds", new Set());
    (0, _defineProperty2.default)(this, "space", _spaces.MetaSpace.Home);
    (0, _defineProperty2.default)(this, "onStoreUpdate", async () => {
      const beforeRoomIds = this.roomIds; // clone the set as it may be mutated by the space store internally

      this.roomIds = new Set(_SpaceStore.default.instance.getSpaceFilteredRoomIds(this.space));

      if ((0, _sets.setHasDiff)(beforeRoomIds, this.roomIds)) {
        this.emit(_IFilterCondition.FILTER_CHANGED); // XXX: Room List Store has a bug where updates to the pre-filter during a local echo of a
        // tags transition seem to be ignored, so refire in the next tick to work around it

        setImmediate(() => {
          this.emit(_IFilterCondition.FILTER_CHANGED);
        });
      }
    });
  }

  get kind() {
    return _IFilterCondition.FilterKind.Prefilter;
  }

  isVisible(room) {
    return this.roomIds.has(room.roomId);
  }

  updateSpace(space) {
    _SpaceStore.default.instance.off(this.space, this.onStoreUpdate);

    _SpaceStore.default.instance.on(this.space = space, this.onStoreUpdate);

    this.onStoreUpdate(); // initial update from the change to the space
  }

  destroy() {
    _SpaceStore.default.instance.off(this.space, this.onStoreUpdate);
  }

}

exports.SpaceFilterCondition = SpaceFilterCondition;
