// matrix-react-sdk/lib/settings/watchers/FontWatcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FontWatcher = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _SettingsStore = _interopRequireDefault(require("../SettingsStore"));

var _units = require("../../utils/units");

var _actions = require("../../dispatcher/actions");

var _SettingLevel = require("../SettingLevel");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class FontWatcher {
  // Externally we tell the user the font is size 15. Internally we use 10.
  constructor() {
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "onAction", payload => {
      if (payload.action === _actions.Action.UpdateFontSize) {
        this.setRootFontSize(payload.size);
      } else if (payload.action === _actions.Action.UpdateSystemFont) {
        this.setSystemFont(payload);
      }
    });
    (0, _defineProperty2.default)(this, "setRootFontSize", size => {
      const fontSize = Math.max(Math.min(FontWatcher.MAX_SIZE, size), FontWatcher.MIN_SIZE);

      if (fontSize !== size) {
        _SettingsStore.default.setValue("baseFontSize", null, _SettingLevel.SettingLevel.DEVICE, fontSize);
      }

      document.querySelector(":root").style.fontSize = (0, _units.toPx)(fontSize);
    });
    (0, _defineProperty2.default)(this, "setSystemFont", ({
      useSystemFont,
      font
    }) => {
      document.body.style.fontFamily = useSystemFont ? font : "";
    });
    this.dispatcherRef = null;
  }

  start() {
    this.setRootFontSize(_SettingsStore.default.getValue("baseFontSize"));
    this.setSystemFont({
      useSystemFont: _SettingsStore.default.getValue("useSystemFont"),
      font: _SettingsStore.default.getValue("systemFont")
    });
    this.dispatcherRef = _dispatcher.default.register(this.onAction);
  }

  stop() {
    _dispatcher.default.unregister(this.dispatcherRef);
  }

}

exports.FontWatcher = FontWatcher;
(0, _defineProperty2.default)(FontWatcher, "MIN_SIZE", 8);
(0, _defineProperty2.default)(FontWatcher, "MAX_SIZE", 15);
(0, _defineProperty2.default)(FontWatcher, "SIZE_DIFF", 5);
