// matrix-react-sdk/lib/components/views/avatars/GroupAvatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _BaseAvatar = _interopRequireDefault(require("./BaseAvatar"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Media = require("../../../customisations/Media");

const _excluded = ["groupId", "groupAvatarUrl", "groupName"];

var _dec, _class, _class2, _temp;

let GroupAvatar = (_dec = (0, _replaceableComponent.replaceableComponent)("views.avatars.GroupAvatar"), _dec(_class = (_temp = _class2 = class GroupAvatar extends _react.default.Component {
  getGroupAvatarUrl() {
    if (!this.props.groupAvatarUrl) return null;
    return (0, _Media.mediaFromMxc)(this.props.groupAvatarUrl).getThumbnailOfSourceHttp(this.props.width, this.props.height, this.props.resizeMethod);
  }

  render() {
    // extract the props we use from props so we can pass any others through
    // should consider adding this as a global rule in js-sdk?

    /* eslint @typescript-eslint/no-unused-vars: ["error", { "ignoreRestSiblings": true }] */
    const _this$props = this.props,
          {
      groupId,
      groupAvatarUrl,
      groupName
    } = _this$props,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    return /*#__PURE__*/_react.default.createElement(_BaseAvatar.default, (0, _extends2.default)({
      name: groupName || this.props.groupId[1],
      idName: this.props.groupId,
      url: this.getGroupAvatarUrl()
    }, otherProps));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  width: 36,
  height: 36,
  resizeMethod: 'crop'
}), _temp)) || _class);
exports.default = GroupAvatar;
