// matrix-react-sdk/lib/components/views/avatars/WidgetAvatar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _BaseAvatar = _interopRequireDefault(require("./BaseAvatar"));

var _Media = require("../../../customisations/Media");

const _excluded = ["app", "className", "width", "height"];

const WidgetAvatar = _ref => {
  let {
    app,
    className,
    width = 20,
    height = 20
  } = _ref,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  let iconUrls = [require("../../../../res/img/element-icons/room/default_app.svg")]; // heuristics for some better icons until Widgets support their own icons

  if (app.type.includes("jitsi")) {
    iconUrls = [require("../../../../res/img/element-icons/room/default_video.svg")];
  } else if (app.type.includes("meeting") || app.type.includes("calendar")) {
    iconUrls = [require("../../../../res/img/element-icons/room/default_cal.svg")];
  } else if (app.type.includes("pad") || app.type.includes("doc") || app.type.includes("calc")) {
    iconUrls = [require("../../../../res/img/element-icons/room/default_doc.svg")];
  } else if (app.type.includes("clock")) {
    iconUrls = [require("../../../../res/img/element-icons/room/default_clock.svg")];
  }

  return /*#__PURE__*/_react.default.createElement(_BaseAvatar.default, (0, _extends2.default)({}, props, {
    name: app.id,
    className: (0, _classnames.default)("mx_WidgetAvatar", className) // MSC2765
    ,
    url: app.avatar_url ? (0, _Media.mediaFromMxc)(app.avatar_url).getSquareThumbnailHttp(20) : undefined,
    urls: iconUrls,
    width: width,
    height: height
  }));
};

var _default = WidgetAvatar;
exports.default = _default;
