// matrix-react-sdk/lib/components/views/right_panel/BaseCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Group = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _AutoHideScrollbar = _interopRequireDefault(require("../../structures/AutoHideScrollbar"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const Group = ({
  className,
  title,
  children
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)("mx_BaseCard_Group", className)
  }, /*#__PURE__*/_react.default.createElement("h1", null, title), children);
};

exports.Group = Group;

const BaseCard = ({
  closeLabel,
  onClose,
  className,
  header,
  footer,
  withoutScrollContainer,
  previousPhase,
  previousPhaseLabel,
  children,
  refireParams
}) => {
  let backButton;

  if (previousPhase) {
    const onBackClick = () => {
      _dispatcher.default.dispatch({
        action: _actions.Action.SetRightPanelPhase,
        phase: previousPhase,
        refireParams: refireParams
      });
    };

    const label = previousPhaseLabel ?? (0, _languageHandler._t)("Back");
    backButton = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_BaseCard_back",
      onClick: onBackClick,
      title: label
    });
  }

  let closeButton;

  if (onClose) {
    closeButton = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_BaseCard_close",
      onClick: onClose,
      title: closeLabel || (0, _languageHandler._t)("Close")
    });
  }

  if (!withoutScrollContainer) {
    children = /*#__PURE__*/_react.default.createElement(_AutoHideScrollbar.default, null, children);
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)("mx_BaseCard", className)
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_BaseCard_header"
  }, backButton, closeButton, header), children, footer && /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_BaseCard_footer"
  }, footer));
};

var _default = BaseCard;
exports.default = _default;
