// matrix-react-sdk/lib/components/views/right_panel/EncryptionInfo.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.PendingActionSpinner = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

/*
Copyright 2019, 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const PendingActionSpinner = ({
  text
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_EncryptionInfo_spinner"
  }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null), text);
};

exports.PendingActionSpinner = PendingActionSpinner;

const EncryptionInfo = ({
  waitingForOtherParty,
  waitingForNetwork,
  member,
  onStartVerification,
  isRoomEncrypted,
  inDialog,
  isSelfVerification
}) => {
  let content;

  if (waitingForOtherParty && isSelfVerification) {
    content = /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("To proceed, please accept the verification request on your other login."));
  } else if (waitingForOtherParty || waitingForNetwork) {
    let text;

    if (waitingForOtherParty) {
      text = (0, _languageHandler._t)("Waiting for %(displayName)s to accept…", {
        displayName: member.displayName || member.name || member.userId
      });
    } else {
      text = (0, _languageHandler._t)("Accepting…");
    }

    content = /*#__PURE__*/_react.default.createElement(PendingActionSpinner, {
      text: text
    });
  } else {
    content = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      kind: "primary",
      className: "mx_UserInfo_wideButton",
      onClick: onStartVerification
    }, (0, _languageHandler._t)("Start Verification"));
  }

  let description;

  if (isRoomEncrypted) {
    description = /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Messages in this room are end-to-end encrypted.")), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Your messages are secured and only you and the recipient have " + "the unique keys to unlock them.")));
  } else {
    description = /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Messages in this room are not end-to-end encrypted.")), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("In encrypted rooms, your messages are secured and only you and the recipient have " + "the unique keys to unlock them.")));
  }

  if (inDialog) {
    return content;
  }

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_UserInfo_container"
  }, /*#__PURE__*/_react.default.createElement("h3", null, (0, _languageHandler._t)("Encryption")), description), /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_UserInfo_container"
  }, /*#__PURE__*/_react.default.createElement("h3", null, (0, _languageHandler._t)("Verify User")), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("For extra security, verify this user by checking a one-time code on both of your devices.")), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("To be secure, do this in person or use a trusted way to communicate.")), content)));
};

var _default = EncryptionInfo;
exports.default = _default;
