// matrix-react-sdk/lib/components/views/right_panel/HeaderButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Analytics = _interopRequireDefault(require("../../../Analytics"));

var _AccessibleTooltipButton = _interopRequireDefault(require("../elements/AccessibleTooltipButton"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

const _excluded = ["isHighlighted", "onClick", "analytics", "name", "title"];

var _dec, _class;

let HeaderButton = ( // TODO: replace this, the composer buttons and the right panel buttons with a unified representation
_dec = (0, _replaceableComponent.replaceableComponent)("views.right_panel.HeaderButton"), _dec(_class = class HeaderButton extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", () => {
      _Analytics.default.trackEvent(...this.props.analytics);

      this.props.onClick();
    });
  }

  render() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _this$props = this.props,
          {
      isHighlighted,
      onClick,
      analytics,
      name,
      title
    } = _this$props,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    const classes = (0, _classnames.default)({
      mx_RightPanel_headerButton: true,
      mx_RightPanel_headerButton_highlight: isHighlighted,
      [`mx_RightPanel_${name}`]: true
    });
    return /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, (0, _extends2.default)({}, props, {
      "aria-selected": isHighlighted,
      role: "tab",
      title: title,
      className: classes,
      onClick: this.onClick
    }));
  }

}) || _class);
exports.default = HeaderButton;
