// matrix-react-sdk/lib/components/views/right_panel/TimelineCard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _BaseCard = _interopRequireDefault(require("./BaseCard"));

var _MessageComposer = _interopRequireDefault(require("../rooms/MessageComposer"));

var _Layout = require("../../../settings/enums/Layout");

var _TimelinePanel = _interopRequireDefault(require("../../structures/TimelinePanel"));

var _RoomContext = _interopRequireDefault(require("../../../contexts/RoomContext"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

let TimelineCard = (_dec = (0, _replaceableComponent.replaceableComponent)("structures.TimelineCard"), _dec(_class = (_temp = _class2 = class TimelineCard extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "renderTimelineCardHeader", () => {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_TimelineCard__header"
      }, /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Chat")));
    });
    this.state = {};
  }

  render() {
    var _this$props$initialEv;

    return /*#__PURE__*/_react.default.createElement(_BaseCard.default, {
      className: "mx_ThreadPanel mx_TimelineCard",
      onClose: this.props.onClose,
      withoutScrollContainer: true,
      header: this.renderTimelineCardHeader()
    }, /*#__PURE__*/_react.default.createElement(_TimelinePanel.default, {
      showReadReceipts: false // TODO: RR's cause issues with limited horizontal space
      ,
      manageReadReceipts: true,
      manageReadMarkers: false // No RM support in the TimelineCard
      ,
      sendReadReceiptOnLoad: true,
      timelineSet: this.props.room.getUnfilteredTimelineSet(),
      showUrlPreview: true,
      layout: _Layout.Layout.Group,
      hideThreadedMessages: false,
      hidden: false,
      showReactions: true,
      className: "mx_RoomView_messagePanel mx_GroupLayout",
      permalinkCreator: this.props.permalinkCreator,
      membersLoaded: true,
      editState: this.state.editState,
      eventId: (_this$props$initialEv = this.props.initialEvent) === null || _this$props$initialEv === void 0 ? void 0 : _this$props$initialEv.getId(),
      resizeNotifier: this.props.resizeNotifier
    }), /*#__PURE__*/_react.default.createElement(_MessageComposer.default, {
      room: this.props.room,
      resizeNotifier: this.props.resizeNotifier,
      replyToEvent: this.state.replyToEvent,
      permalinkCreator: this.props.permalinkCreator,
      e2eStatus: this.props.e2eStatus,
      compact: true
    }));
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _RoomContext.default), _temp)) || _class);
exports.default = TimelineCard;
