// matrix-react-sdk/lib/components/views/elements/AccessibleTooltipButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _AccessibleButton = _interopRequireDefault(require("./AccessibleButton"));

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

const _excluded = ["title", "tooltip", "children", "tooltipClassName", "forceHide", "yOffset", "alignment"];

var _dec, _class;

let AccessibleTooltipButton = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.AccessibleTooltipButton"), _dec(_class = class AccessibleTooltipButton extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onMouseOver", () => {
      if (this.props.forceHide) return;
      this.setState({
        hover: true
      });
    });
    (0, _defineProperty2.default)(this, "onMouseLeave", () => {
      this.setState({
        hover: false
      });
    });
    this.state = {
      hover: false
    };
  }

  componentDidUpdate(prevProps) {
    if (!prevProps.forceHide && this.props.forceHide && this.state.hover) {
      this.setState({
        hover: false
      });
    }
  }

  render() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _this$props = this.props,
          {
      title,
      tooltip,
      children,
      tooltipClassName,
      forceHide,
      yOffset,
      alignment
    } = _this$props,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded);

    const tip = this.state.hover && /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
      tooltipClassName: tooltipClassName,
      label: tooltip || title,
      yOffset: yOffset,
      alignment: alignment
    });

    return /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, (0, _extends2.default)({}, props, {
      onMouseOver: this.onMouseOver,
      onMouseLeave: this.onMouseLeave,
      "aria-label": title
    }), children, this.props.label, (tooltip || title) && tip);
  }

}) || _class);
exports.default = AccessibleTooltipButton;
