// matrix-react-sdk/lib/components/views/elements/ActionButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _AccessibleButton = _interopRequireDefault(require("./AccessibleButton"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _Analytics = _interopRequireDefault(require("../../../Analytics"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

var _dec, _class, _class2, _temp;

let ActionButton = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.ActionButton"), _dec(_class = (_temp = _class2 = class ActionButton extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onClick", ev => {
      ev.stopPropagation();

      _Analytics.default.trackEvent('Action Button', 'click', this.props.action);

      _dispatcher.default.dispatch({
        action: this.props.action
      });
    });
    (0, _defineProperty2.default)(this, "onMouseEnter", () => {
      if (this.props.tooltip) this.setState({
        showTooltip: true
      });

      if (this.props.mouseOverAction) {
        _dispatcher.default.dispatch({
          action: this.props.mouseOverAction
        });
      }
    });
    (0, _defineProperty2.default)(this, "onMouseLeave", () => {
      this.setState({
        showTooltip: false
      });
    });
    this.state = {
      showTooltip: false
    };
  }

  render() {
    let tooltip;

    if (this.state.showTooltip) {
      tooltip = /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
        className: "mx_RoleButton_tooltip",
        label: this.props.label
      });
    }

    const icon = this.props.iconPath ? /*#__PURE__*/_react.default.createElement("img", {
      src: this.props.iconPath,
      width: this.props.size,
      height: this.props.size
    }) : undefined;
    const classNames = ["mx_RoleButton"];

    if (this.props.className) {
      classNames.push(this.props.className);
    }

    return /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: classNames.join(" "),
      onClick: this.onClick,
      onMouseEnter: this.onMouseEnter,
      onMouseLeave: this.onMouseLeave,
      "aria-label": this.props.label
    }, icon, tooltip, this.props.children);
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  size: "25",
  tooltip: false
}), _temp)) || _class);
exports.default = ActionButton;
