// matrix-react-sdk/lib/components/views/elements/DesktopBuildsNotice.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WarningKind = void 0;
exports.default = DesktopBuildsNotice;

var _react = _interopRequireDefault(require("react"));

var _EventIndexPeg = _interopRequireDefault(require("../../../indexing/EventIndexPeg"));

var _languageHandler = require("../../../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _UserSettingsDialog = require("../dialogs/UserSettingsDialog");

var _logger = require("matrix-js-sdk/src/logger");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
let WarningKind;
exports.WarningKind = WarningKind;

(function (WarningKind) {
  WarningKind[WarningKind["Files"] = 0] = "Files";
  WarningKind[WarningKind["Search"] = 1] = "Search";
})(WarningKind || (exports.WarningKind = WarningKind = {}));

function DesktopBuildsNotice({
  isRoomEncrypted,
  kind
}) {
  if (!isRoomEncrypted) return null;
  if (_EventIndexPeg.default.get()) return null;

  if (_EventIndexPeg.default.error) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (0, _languageHandler._t)("Message search initialisation failed, check <a>your settings</a> for more information", {}, {
      a: sub => /*#__PURE__*/_react.default.createElement("a", {
        onClick: evt => {
          evt.preventDefault();

          _dispatcher.default.dispatch({
            action: _actions.Action.ViewUserSettings,
            initialTabId: _UserSettingsDialog.UserTab.Security
          });
        }
      }, sub)
    }));
  }

  const {
    desktopBuilds,
    brand
  } = _SdkConfig.default.get();

  let text = null;
  let logo = null;

  if (desktopBuilds.available) {
    logo = /*#__PURE__*/_react.default.createElement("img", {
      src: desktopBuilds.logo
    });

    switch (kind) {
      case WarningKind.Files:
        text = (0, _languageHandler._t)("Use the <a>Desktop app</a> to see all encrypted files", {}, {
          a: sub => /*#__PURE__*/_react.default.createElement("a", {
            href: desktopBuilds.url,
            target: "_blank",
            rel: "noreferrer noopener"
          }, sub)
        });
        break;

      case WarningKind.Search:
        text = (0, _languageHandler._t)("Use the <a>Desktop app</a> to search encrypted messages", {}, {
          a: sub => /*#__PURE__*/_react.default.createElement("a", {
            href: desktopBuilds.url,
            target: "_blank",
            rel: "noreferrer noopener"
          }, sub)
        });
        break;
    }
  } else {
    switch (kind) {
      case WarningKind.Files:
        text = (0, _languageHandler._t)("This version of %(brand)s does not support viewing some encrypted files", {
          brand
        });
        break;

      case WarningKind.Search:
        text = (0, _languageHandler._t)("This version of %(brand)s does not support searching encrypted messages", {
          brand
        });
        break;
    }
  } // for safety


  if (!text) {
    _logger.logger.warn("Unknown desktop builds warning kind: ", kind);

    return null;
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_DesktopBuildsNotice"
  }, logo, /*#__PURE__*/_react.default.createElement("span", null, text));
}
