// matrix-react-sdk/lib/components/views/elements/DialogButtons.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

let DialogButtons = (
/**
 * Basic container for buttons in modal dialogs.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.DialogButtons"), _dec(_class = (_temp = _class2 = class DialogButtons extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onCancelClick", event => {
      this.props.onCancel(event);
    });
  }

  render() {
    let primaryButtonClassName = "mx_Dialog_primary";

    if (this.props.primaryButtonClass) {
      primaryButtonClassName += " " + this.props.primaryButtonClass;
    }

    let cancelButton;

    if (this.props.cancelButton || this.props.hasCancel) {
      cancelButton = /*#__PURE__*/_react.default.createElement("button", {
        // important: the default type is 'submit' and this button comes before the
        // primary in the DOM so will get form submissions unless we make it not a submit.
        type: "button",
        onClick: this.onCancelClick,
        className: this.props.cancelButtonClass,
        disabled: this.props.disabled
      }, this.props.cancelButton || (0, _languageHandler._t)("Cancel"));
    }

    let additive = null;

    if (this.props.additive) {
      additive = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_Dialog_buttons_additive"
      }, this.props.additive);
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_buttons"
    }, additive, cancelButton, this.props.children, /*#__PURE__*/_react.default.createElement("button", {
      type: this.props.primaryIsSubmit ? 'submit' : 'button',
      className: primaryButtonClassName,
      onClick: this.props.onPrimaryButtonClick,
      autoFocus: this.props.focus,
      disabled: this.props.disabled || this.props.primaryDisabled
    }, this.props.primaryButton));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  hasCancel: true,
  disabled: false
}), _temp)) || _class);
exports.default = DialogButtons;
