// matrix-react-sdk/lib/components/views/elements/DNDTagTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = DNDTagTile;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _TagTile = _interopRequireDefault(require("./TagTile"));

var _react = _interopRequireDefault(require("react"));

var _ContextMenu = _interopRequireWildcard(require("../../structures/ContextMenu"));

var sdk = _interopRequireWildcard(require("../../../index"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/* eslint new-cap: "off" */

/*
Copyright 2017 New Vector Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
function DNDTagTile(props) {
  const [menuDisplayed, handle, openMenu, closeMenu] = (0, _ContextMenu.useContextMenu)();
  let contextMenu = null;

  if (menuDisplayed && handle.current) {
    const elementRect = handle.current.getBoundingClientRect();
    const TagTileContextMenu = sdk.getComponent('context_menus.TagTileContextMenu');
    contextMenu = /*#__PURE__*/_react.default.createElement(_ContextMenu.default, (0, _extends2.default)({}, (0, _ContextMenu.toRightOf)(elementRect), {
      onFinished: closeMenu
    }), /*#__PURE__*/_react.default.createElement(TagTileContextMenu, {
      tag: props.tag,
      onFinished: closeMenu,
      index: props.index
    }));
  }

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_TagTile.default, (0, _extends2.default)({}, props, {
    contextMenuButtonRef: handle,
    menuDisplayed: menuDisplayed,
    openMenu: openMenu
  })), contextMenu);
}
