// matrix-react-sdk/lib/components/views/elements/Draggable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let Draggable = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.Draggable"), _dec(_class = class Draggable extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onMouseDown", event => {
      this.setState({
        location: {
          currentX: event.clientX,
          currentY: event.clientY
        }
      });
      document.addEventListener("mousemove", this.state.onMouseMove);
      document.addEventListener("mouseup", this.state.onMouseUp);
    });
    (0, _defineProperty2.default)(this, "onMouseUp", event => {
      document.removeEventListener("mousemove", this.state.onMouseMove);
      document.removeEventListener("mouseup", this.state.onMouseUp);
      this.props.onMouseUp(event);
    });
    this.state = {
      onMouseMove: this.onMouseMove.bind(this),
      onMouseUp: this.onMouseUp.bind(this),
      location: {
        currentX: 0,
        currentY: 0
      }
    };
  }

  onMouseMove(event) {
    const newLocation = this.props.dragFunc(this.state.location, event);
    this.setState({
      location: newLocation
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: this.props.className,
      onMouseDown: this.onMouseDown.bind(this)
    });
  }

}) || _class);
exports.default = Draggable;
