// matrix-react-sdk/lib/components/views/elements/ImageView.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleTooltipButton = _interopRequireDefault(require("./AccessibleTooltipButton"));

var _Keyboard = require("../../../Keyboard");

var _reactFocusLock = _interopRequireDefault(require("react-focus-lock"));

var _MemberAvatar = _interopRequireDefault(require("../avatars/MemberAvatar"));

var _ContextMenuTooltipButton = require("../../../accessibility/context_menu/ContextMenuTooltipButton");

var _MessageContextMenu = _interopRequireDefault(require("../context_menus/MessageContextMenu"));

var _ContextMenu = require("../../structures/ContextMenu");

var _MessageTimestamp = _interopRequireDefault(require("../messages/MessageTimestamp"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _DateUtils = require("../../../DateUtils");

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Mouse = require("../../../utils/Mouse");

var _UIStore = _interopRequireDefault(require("../../../stores/UIStore"));

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// Max scale to keep gaps around the image
const MAX_SCALE = 0.95; // This is used for the buttons

const ZOOM_STEP = 0.10; // This is used for mouse wheel events

const ZOOM_COEFFICIENT = 0.0025; // If we have moved only this much we can zoom

const ZOOM_DISTANCE = 10; // Height of mx_ImageView_panel

const getPanelHeight = () => {
  const value = getComputedStyle(document.documentElement).getPropertyValue("--image-view-panel-height"); // Return the value as a number without the unit

  return parseInt(value.slice(0, value.length - 2));
};

let ImageView = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.ImageView"), _dec(_class = class ImageView extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "contextMenuButton", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "focusLock", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "imageWrapper", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "image", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "initX", 0);
    (0, _defineProperty2.default)(this, "initY", 0);
    (0, _defineProperty2.default)(this, "previousX", 0);
    (0, _defineProperty2.default)(this, "previousY", 0);
    (0, _defineProperty2.default)(this, "animatingLoading", false);
    (0, _defineProperty2.default)(this, "imageIsLoaded", false);
    (0, _defineProperty2.default)(this, "imageLoaded", () => {
      // First, we calculate the zoom, so that the image has the same size as
      // the thumbnail
      const {
        thumbnailInfo
      } = this.props;

      if (thumbnailInfo !== null && thumbnailInfo !== void 0 && thumbnailInfo.width) {
        this.setState({
          zoom: thumbnailInfo.width / this.image.current.naturalWidth
        });
      } // Once the zoom is set, we the image is considered loaded and we can
      // start animating it into the center of the screen


      this.imageIsLoaded = true;
      this.animatingLoading = true;
      this.setZoomAndRotation();
      this.setState({
        translationX: 0,
        translationY: 0
      }); // Once the position is set, there is no need to animate anymore

      this.animatingLoading = false;
    });
    (0, _defineProperty2.default)(this, "recalculateZoom", () => {
      this.setZoomAndRotation();
    });
    (0, _defineProperty2.default)(this, "setZoomAndRotation", inputRotation => {
      const image = this.image.current;
      const imageWrapper = this.imageWrapper.current;
      const rotation = inputRotation ?? this.state.rotation;
      const imageIsNotFlipped = rotation % 180 === 0; // If the image is rotated take it into account

      const width = imageIsNotFlipped ? image.naturalWidth : image.naturalHeight;
      const height = imageIsNotFlipped ? image.naturalHeight : image.naturalWidth;
      const zoomX = imageWrapper.clientWidth / width;
      const zoomY = imageWrapper.clientHeight / height; // If the image is smaller in both dimensions set its the zoom to 1 to
      // display it in its original size

      if (zoomX >= 1 && zoomY >= 1) {
        this.setState({
          zoom: 1,
          minZoom: 1,
          maxZoom: 1,
          rotation: rotation
        });
        return;
      } // We set minZoom to the min of the zoomX and zoomY to avoid overflow in
      // any direction. We also multiply by MAX_SCALE to get a gap around the
      // image by default


      const minZoom = Math.min(zoomX, zoomY) * MAX_SCALE; // If zoom is smaller than minZoom don't go below that value

      const zoom = this.state.zoom <= this.state.minZoom ? minZoom : this.state.zoom;
      this.setState({
        minZoom: minZoom,
        maxZoom: 1,
        rotation: rotation,
        zoom: zoom
      });
    });
    (0, _defineProperty2.default)(this, "onWheel", ev => {
      if (ev.target === this.image.current) {
        ev.stopPropagation();
        ev.preventDefault();
        const {
          deltaY
        } = (0, _Mouse.normalizeWheelEvent)(ev); // Zoom in on the point on the image targeted by the cursor

        this.zoomDelta(-deltaY * ZOOM_COEFFICIENT, ev.offsetX, ev.offsetY);
      }
    });
    (0, _defineProperty2.default)(this, "onZoomInClick", () => {
      this.zoomDelta(ZOOM_STEP);
    });
    (0, _defineProperty2.default)(this, "onZoomOutClick", () => {
      this.zoomDelta(-ZOOM_STEP);
    });
    (0, _defineProperty2.default)(this, "onKeyDown", ev => {
      if (ev.key === _Keyboard.Key.ESCAPE) {
        ev.stopPropagation();
        ev.preventDefault();
        this.props.onFinished();
      }
    });
    (0, _defineProperty2.default)(this, "onRotateCounterClockwiseClick", () => {
      const cur = this.state.rotation;
      this.setZoomAndRotation(cur - 90);
    });
    (0, _defineProperty2.default)(this, "onRotateClockwiseClick", () => {
      const cur = this.state.rotation;
      this.setZoomAndRotation(cur + 90);
    });
    (0, _defineProperty2.default)(this, "onDownloadClick", () => {
      const a = document.createElement("a");
      a.href = this.props.src;
      a.download = this.props.name;
      a.target = "_blank";
      a.rel = "noreferrer noopener";
      a.click();
    });
    (0, _defineProperty2.default)(this, "onOpenContextMenu", () => {
      this.setState({
        contextMenuDisplayed: true
      });
    });
    (0, _defineProperty2.default)(this, "onCloseContextMenu", () => {
      this.setState({
        contextMenuDisplayed: false
      });
    });
    (0, _defineProperty2.default)(this, "onPermalinkClicked", ev => {
      // This allows the permalink to be opened in a new tab/window or copied as
      // matrix.to, but also for it to enable routing within Element when clicked.
      ev.preventDefault();

      _dispatcher.default.dispatch({
        action: _actions.Action.ViewRoom,
        event_id: this.props.mxEvent.getId(),
        highlighted: true,
        room_id: this.props.mxEvent.getRoomId()
      });

      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "onStartMoving", ev => {
      ev.stopPropagation();
      ev.preventDefault(); // Don't do anything if we pressed any
      // other button than the left one

      if (ev.button !== 0) return; // Zoom in if we are completely zoomed out

      if (this.state.zoom === this.state.minZoom) {
        this.zoom(this.state.maxZoom, ev.nativeEvent.offsetX, ev.nativeEvent.offsetY);
        return;
      }

      this.setState({
        moving: true
      });
      this.previousX = this.state.translationX;
      this.previousY = this.state.translationY;
      this.initX = ev.pageX - this.state.translationX;
      this.initY = ev.pageY - this.state.translationY;
    });
    (0, _defineProperty2.default)(this, "onMoving", ev => {
      ev.stopPropagation();
      ev.preventDefault();
      if (!this.state.moving) return;
      this.setState({
        translationX: ev.pageX - this.initX,
        translationY: ev.pageY - this.initY
      });
    });
    (0, _defineProperty2.default)(this, "onEndMoving", () => {
      // Zoom out if we haven't moved much
      if (this.state.moving === true && Math.abs(this.state.translationX - this.previousX) < ZOOM_DISTANCE && Math.abs(this.state.translationY - this.previousY) < ZOOM_DISTANCE) {
        this.zoom(this.state.minZoom);
        this.initX = 0;
        this.initY = 0;
      }

      this.setState({
        moving: false
      });
    });
    const {
      thumbnailInfo: _thumbnailInfo
    } = this.props;
    this.state = {
      zoom: 0,
      // We default to 0 and override this in imageLoaded once we have naturalSize
      minZoom: MAX_SCALE,
      maxZoom: MAX_SCALE,
      rotation: 0,
      translationX: (_thumbnailInfo === null || _thumbnailInfo === void 0 ? void 0 : _thumbnailInfo.positionX) + (_thumbnailInfo === null || _thumbnailInfo === void 0 ? void 0 : _thumbnailInfo.width) / 2 - _UIStore.default.instance.windowWidth / 2 ?? 0,
      translationY: (_thumbnailInfo === null || _thumbnailInfo === void 0 ? void 0 : _thumbnailInfo.positionY) + (_thumbnailInfo === null || _thumbnailInfo === void 0 ? void 0 : _thumbnailInfo.height) / 2 - _UIStore.default.instance.windowHeight / 2 - getPanelHeight() / 2 ?? 0,
      moving: false,
      contextMenuDisplayed: false
    };
  } // XXX: Refs to functional components


  componentDidMount() {
    // We have to use addEventListener() because the listener
    // needs to be passive in order to work with Chromium
    this.focusLock.current.addEventListener('wheel', this.onWheel, {
      passive: false
    }); // We want to recalculate zoom whenever the window's size changes

    window.addEventListener("resize", this.recalculateZoom); // After the image loads for the first time we want to calculate the zoom

    this.image.current.addEventListener("load", this.imageLoaded);
  }

  componentWillUnmount() {
    this.focusLock.current.removeEventListener('wheel', this.onWheel);
    window.removeEventListener("resize", this.recalculateZoom);
    this.image.current.removeEventListener("load", this.imageLoaded);
  }

  zoomDelta(delta, anchorX, anchorY) {
    this.zoom(this.state.zoom + delta, anchorX, anchorY);
  }

  zoom(zoomLevel, anchorX, anchorY) {
    const oldZoom = this.state.zoom;
    const newZoom = Math.min(zoomLevel, this.state.maxZoom);

    if (newZoom <= this.state.minZoom) {
      // Zoom out fully
      this.setState({
        zoom: this.state.minZoom,
        translationX: 0,
        translationY: 0
      });
    } else if (typeof anchorX !== "number" && typeof anchorY !== "number") {
      // Zoom relative to the center of the view
      this.setState({
        zoom: newZoom,
        translationX: this.state.translationX * newZoom / oldZoom,
        translationY: this.state.translationY * newZoom / oldZoom
      });
    } else {
      // Zoom relative to the given point on the image.
      // First we need to figure out the offset of the anchor point
      // relative to the center of the image, accounting for rotation.
      let offsetX;
      let offsetY; // The modulo operator can return negative values for some
      // rotations, so we have to do some extra work to normalize it

      switch ((this.state.rotation % 360 + 360) % 360) {
        case 0:
          offsetX = this.image.current.clientWidth / 2 - anchorX;
          offsetY = this.image.current.clientHeight / 2 - anchorY;
          break;

        case 90:
          offsetX = anchorY - this.image.current.clientHeight / 2;
          offsetY = this.image.current.clientWidth / 2 - anchorX;
          break;

        case 180:
          offsetX = anchorX - this.image.current.clientWidth / 2;
          offsetY = anchorY - this.image.current.clientHeight / 2;
          break;

        case 270:
          offsetX = this.image.current.clientHeight / 2 - anchorY;
          offsetY = anchorX - this.image.current.clientWidth / 2;
      } // Apply the zoom and offset


      this.setState({
        zoom: newZoom,
        translationX: this.state.translationX + (newZoom - oldZoom) * offsetX,
        translationY: this.state.translationY + (newZoom - oldZoom) * offsetY
      });
    }
  }

  renderContextMenu() {
    let contextMenu = null;

    if (this.state.contextMenuDisplayed) {
      contextMenu = /*#__PURE__*/_react.default.createElement(_MessageContextMenu.default, (0, _extends2.default)({}, (0, _ContextMenu.aboveLeftOf)(this.contextMenuButton.current.getBoundingClientRect()), {
        mxEvent: this.props.mxEvent,
        permalinkCreator: this.props.permalinkCreator,
        onFinished: this.onCloseContextMenu,
        onCloseDialog: this.props.onFinished
      }));
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, contextMenu);
  }

  render() {
    const showEventMeta = !!this.props.mxEvent;
    const zoomingDisabled = this.state.maxZoom === this.state.minZoom;
    let transitionClassName;
    if (this.animatingLoading) transitionClassName = "mx_ImageView_image_animatingLoading";else if (this.state.moving || !this.imageIsLoaded) transitionClassName = "";else transitionClassName = "mx_ImageView_image_animating";
    let cursor;
    if (this.state.moving) cursor = "grabbing";else if (zoomingDisabled) cursor = "default";else if (this.state.zoom === this.state.minZoom) cursor = "zoom-in";else cursor = "zoom-out";
    const rotationDegrees = this.state.rotation + "deg";
    const zoom = this.state.zoom;
    const translatePixelsX = this.state.translationX + "px";
    const translatePixelsY = this.state.translationY + "px"; // The order of the values is important!
    // First, we translate and only then we rotate, otherwise
    // we would apply the translation to an already rotated
    // image causing it translate in the wrong direction.

    const style = {
      cursor: cursor,
      transform: `translateX(${translatePixelsX})
                        translateY(${translatePixelsY})
                        scale(${zoom})
                        rotate(${rotationDegrees})`
    };
    let info;

    if (showEventMeta) {
      const mxEvent = this.props.mxEvent;

      const showTwelveHour = _SettingsStore.default.getValue("showTwelveHourTimestamps");

      let permalink = "#";

      if (this.props.permalinkCreator) {
        permalink = this.props.permalinkCreator.forEvent(this.props.mxEvent.getId());
      }

      const senderName = mxEvent.sender ? mxEvent.sender.name : mxEvent.getSender();

      const sender = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_ImageView_info_sender"
      }, senderName);

      const messageTimestamp = /*#__PURE__*/_react.default.createElement("a", {
        href: permalink,
        onClick: this.onPermalinkClicked,
        "aria-label": (0, _DateUtils.formatFullDate)(new Date(this.props.mxEvent.getTs()), showTwelveHour, false)
      }, /*#__PURE__*/_react.default.createElement(_MessageTimestamp.default, {
        showFullDate: true,
        showTwelveHour: showTwelveHour,
        ts: mxEvent.getTs(),
        showSeconds: false
      }));

      const avatar = /*#__PURE__*/_react.default.createElement(_MemberAvatar.default, {
        member: mxEvent.sender,
        fallbackUserId: mxEvent.getSender(),
        width: 32,
        height: 32,
        viewUserOnClick: true
      });

      info = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_ImageView_info_wrapper"
      }, avatar, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_ImageView_info"
      }, sender, messageTimestamp));
    } else {
      // If there is no event - we're viewing an avatar, we set
      // an empty div here, since the panel uses space-between
      // and we want the same placement of elements
      info = /*#__PURE__*/_react.default.createElement("div", null);
    }

    let contextMenuButton;

    if (this.props.mxEvent) {
      contextMenuButton = /*#__PURE__*/_react.default.createElement(_ContextMenuTooltipButton.ContextMenuTooltipButton, {
        className: "mx_ImageView_button mx_ImageView_button_more",
        title: (0, _languageHandler._t)("Options"),
        onClick: this.onOpenContextMenu,
        inputRef: this.contextMenuButton,
        isExpanded: this.state.contextMenuDisplayed
      });
    }

    let zoomOutButton;
    let zoomInButton;

    if (!zoomingDisabled) {
      zoomOutButton = /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
        className: "mx_ImageView_button mx_ImageView_button_zoomOut",
        title: (0, _languageHandler._t)("Zoom out"),
        onClick: this.onZoomOutClick
      });
      zoomInButton = /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
        className: "mx_ImageView_button mx_ImageView_button_zoomIn",
        title: (0, _languageHandler._t)("Zoom in"),
        onClick: this.onZoomInClick
      });
    }

    return /*#__PURE__*/_react.default.createElement(_reactFocusLock.default, {
      returnFocus: true,
      lockProps: {
        onKeyDown: this.onKeyDown,
        role: "dialog"
      },
      className: "mx_ImageView",
      ref: this.focusLock
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageView_panel"
    }, info, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageView_toolbar"
    }, zoomOutButton, zoomInButton, /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      className: "mx_ImageView_button mx_ImageView_button_rotateCCW",
      title: (0, _languageHandler._t)("Rotate Left"),
      onClick: this.onRotateCounterClockwiseClick
    }), /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      className: "mx_ImageView_button mx_ImageView_button_rotateCW",
      title: (0, _languageHandler._t)("Rotate Right"),
      onClick: this.onRotateClockwiseClick
    }), /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      className: "mx_ImageView_button mx_ImageView_button_download",
      title: (0, _languageHandler._t)("Download"),
      onClick: this.onDownloadClick
    }), contextMenuButton, /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      className: "mx_ImageView_button mx_ImageView_button_close",
      title: (0, _languageHandler._t)("Close"),
      onClick: this.props.onFinished
    }), this.renderContextMenu())), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageView_image_wrapper",
      ref: this.imageWrapper,
      onMouseDown: this.props.onFinished,
      onMouseMove: this.onMoving,
      onMouseUp: this.onEndMoving,
      onMouseLeave: this.onEndMoving
    }, /*#__PURE__*/_react.default.createElement("img", {
      src: this.props.src,
      style: style,
      alt: this.props.name,
      ref: this.image,
      className: `mx_ImageView_image ${transitionClassName}`,
      draggable: true,
      onMouseDown: this.onStartMoving
    })));
  }

}) || _class);
exports.default = ImageView;
