// matrix-react-sdk/lib/components/views/elements/InviteReason.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _classnames = _interopRequireDefault(require("classnames"));

var _react = _interopRequireDefault(require("react"));

var _HtmlUtils = require("../../../HtmlUtils");

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let InviteReason = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.InviteReason"), _dec(_class = class InviteReason extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onViewClick", () => {
      this.setState({
        hidden: false
      });
    });
    this.state = {
      // We hide the reason for invitation by default, since it can be a
      // vector for spam/harassment.
      hidden: true
    };
  }

  render() {
    const classes = (0, _classnames.default)({
      "mx_InviteReason": true,
      "mx_InviteReason_hidden": this.state.hidden
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classes
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_InviteReason_reason"
    }, this.props.htmlReason ? (0, _HtmlUtils.sanitizedHtmlNode)(this.props.htmlReason) : this.props.reason), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_InviteReason_view",
      onClick: this.onViewClick
    }, (0, _languageHandler._t)("View message")));
  }

}) || _class);
exports.default = InviteReason;
