// matrix-react-sdk/lib/components/views/elements/Spinner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

/*
Copyright 2015-2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class Spinner extends _react.default.PureComponent {
  render() {
    const {
      w,
      h,
      message
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Spinner"
    }, message && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Spinner_Msg"
    }, message), "\xA0"), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Spinner_icon",
      style: {
        width: w,
        height: h
      },
      "aria-label": (0, _languageHandler._t)("Loading...")
    }));
  }

}

exports.default = Spinner;
(0, _defineProperty2.default)(Spinner, "defaultProps", {
  w: 32,
  h: 32
});
