// matrix-react-sdk/lib/components/views/elements/StyledRadioButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

const _excluded = ["children", "className", "disabled", "outlined", "childrenInLabel"];

var _dec, _class, _class2, _temp;

let StyledRadioButton = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.StyledRadioButton"), _dec(_class = (_temp = _class2 = class StyledRadioButton extends _react.default.PureComponent {
  render() {
    const _this$props = this.props,
          {
      children,
      className,
      disabled,
      outlined,
      childrenInLabel
    } = _this$props,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);

    const _className = (0, _classnames.default)('mx_StyledRadioButton', className, {
      "mx_StyledRadioButton_disabled": disabled,
      "mx_StyledRadioButton_enabled": !disabled,
      "mx_StyledRadioButton_checked": this.props.checked,
      "mx_StyledRadioButton_outlined": outlined
    });

    const radioButton = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({
      type: "radio",
      disabled: disabled
    }, otherProps)), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null)));

    if (childrenInLabel) {
      return /*#__PURE__*/_react.default.createElement("label", {
        className: _className
      }, radioButton, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_StyledRadioButton_content"
      }, children), /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_StyledRadioButton_spacer"
      }));
    } else {
      return /*#__PURE__*/_react.default.createElement("div", {
        className: _className
      }, /*#__PURE__*/_react.default.createElement("label", {
        className: "mx_StyledRadioButton_innerLabel"
      }, radioButton), /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_StyledRadioButton_content"
      }, children), /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_StyledRadioButton_spacer"
      }));
    }
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  className: '',
  childrenInLabel: true
}), _temp)) || _class);
exports.default = StyledRadioButton;
