// matrix-react-sdk/lib/components/views/elements/SyntaxHighlight.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _highlight = _interopRequireDefault(require("highlight.js"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let SyntaxHighlight = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.SyntaxHighlight"), _dec(_class = class SyntaxHighlight extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "el", null);
    (0, _defineProperty2.default)(this, "ref", el => {
      this.el = el;
      this.componentDidUpdate();
    });
  } // componentDidUpdate used here for reusability


  componentDidUpdate() {
    if (this.el) _highlight.default.highlightElement(this.el);
  } // call componentDidUpdate because _ref is fired on initial render
  // which does not fire componentDidUpdate


  render() {
    const {
      className,
      children
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("pre", {
      className: `${className} mx_SyntaxHighlight`,
      ref: this.ref
    }, /*#__PURE__*/_react.default.createElement("code", null, children));
  }

}) || _class);
exports.default = SyntaxHighlight;
