// matrix-react-sdk/lib/components/views/elements/UserTagTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _GroupFilterOrderStore = _interopRequireDefault(require("../../../stores/GroupFilterOrderStore"));

var _AccessibleTooltipButton = _interopRequireDefault(require("./AccessibleTooltipButton"));

var _classnames = _interopRequireDefault(require("classnames"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let UserTagTile = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.UserTagTile"), _dec(_class = class UserTagTile extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "tagStoreRef", void 0);
    (0, _defineProperty2.default)(this, "onTagStoreUpdate", () => {
      const selected = _GroupFilterOrderStore.default.getSelectedTags().length === 0;
      this.setState({
        selected
      });
    });
    (0, _defineProperty2.default)(this, "onTileClick", ev => {
      ev.preventDefault();
      ev.stopPropagation(); // Deselect all tags

      _dispatcher.default.dispatch({
        action: "deselect_tags"
      });
    });
    this.state = {
      selected: _GroupFilterOrderStore.default.getSelectedTags().length === 0
    };
  }

  componentDidMount() {
    this.tagStoreRef = _GroupFilterOrderStore.default.addListener(this.onTagStoreUpdate);
  }

  componentWillUnmount() {
    this.tagStoreRef.remove();
  }

  render() {
    // XXX: We reuse TagTile classes for ease of demonstration - we should probably generify
    // TagTile instead if we continue to use this component.
    const className = (0, _classnames.default)({
      mx_TagTile: true,
      mx_TagTile_prototype: true,
      mx_TagTile_selected_prototype: this.state.selected,
      mx_TagTile_home: true
    });
    return /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      className: className,
      onClick: this.onTileClick,
      title: (0, _languageHandler._t)("Home")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_TagTile_avatar"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_TagTile_homeIcon"
    })));
  }

}) || _class);
exports.default = UserTagTile;
