// matrix-react-sdk/lib/components/views/voip/CallViewSidebar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _VideoFeed = _interopRequireDefault(require("./VideoFeed"));

var _classnames = _interopRequireDefault(require("classnames"));

/*
Copyright 2021 Šimon Brandner <simon.bra.ag@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class CallViewSidebar extends _react.default.Component {
  render() {
    const feeds = this.props.feeds.map(feed => {
      return /*#__PURE__*/_react.default.createElement(_VideoFeed.default, {
        key: feed.stream.id,
        feed: feed,
        call: this.props.call,
        primary: false,
        pipMode: this.props.pipMode
      });
    });
    const className = (0, _classnames.default)("mx_CallViewSidebar", {
      mx_CallViewSidebar_pipMode: this.props.pipMode
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className
    }, feeds);
  }

}

exports.default = CallViewSidebar;
