// matrix-react-sdk/lib/components/views/voip/DialPad.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const BUTTONS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '*', '0', '#'];
const BUTTON_LETTERS = ['', 'ABC', 'DEF', 'GHI', 'JKL', 'MNO', 'PQRS', 'TUV', 'WXYZ', '', '+', ''];
var DialPadButtonKind;

(function (DialPadButtonKind) {
  DialPadButtonKind[DialPadButtonKind["Digit"] = 0] = "Digit";
  DialPadButtonKind[DialPadButtonKind["Dial"] = 1] = "Dial";
})(DialPadButtonKind || (DialPadButtonKind = {}));

class DialPadButton extends React.PureComponent {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", ev => {
      this.props.onButtonPress(this.props.digit, ev);
    });
  }

  render() {
    switch (this.props.kind) {
      case DialPadButtonKind.Digit:
        return /*#__PURE__*/React.createElement(_AccessibleButton.default, {
          className: "mx_DialPad_button",
          onClick: this.onClick
        }, this.props.digit, /*#__PURE__*/React.createElement("div", {
          className: "mx_DialPad_buttonSubText"
        }, this.props.digitSubtext));

      case DialPadButtonKind.Dial:
        return /*#__PURE__*/React.createElement(_AccessibleButton.default, {
          className: "mx_DialPad_button mx_DialPad_dialButton",
          onClick: this.onClick
        });
    }
  }

}

let Dialpad = (_dec = (0, _replaceableComponent.replaceableComponent)("views.voip.DialPad"), _dec(_class = class Dialpad extends React.PureComponent {
  render() {
    const buttonNodes = [];

    for (let i = 0; i < BUTTONS.length; i++) {
      const button = BUTTONS[i];
      const digitSubtext = BUTTON_LETTERS[i];
      buttonNodes.push( /*#__PURE__*/React.createElement(DialPadButton, {
        key: button,
        kind: DialPadButtonKind.Digit,
        digit: button,
        digitSubtext: digitSubtext,
        onButtonPress: this.props.onDigitPress
      }));
    }

    if (this.props.hasDial) {
      buttonNodes.push( /*#__PURE__*/React.createElement(DialPadButton, {
        key: "dial",
        kind: DialPadButtonKind.Dial,
        onButtonPress: this.props.onDialPress
      }));
    }

    return /*#__PURE__*/React.createElement("div", {
      className: "mx_DialPad"
    }, buttonNodes);
  }

}) || _class);
exports.default = Dialpad;
