// matrix-react-sdk/lib/components/views/spaces/SpaceTreeLevel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SpaceItem = exports.SpaceButton = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _RoomAvatar = _interopRequireDefault(require("../avatars/RoomAvatar"));

var _SpaceStore = _interopRequireDefault(require("../../../stores/spaces/SpaceStore"));

var _SpaceTreeLevelLayoutStore = _interopRequireDefault(require("../../../stores/spaces/SpaceTreeLevelLayoutStore"));

var _NotificationBadge = _interopRequireDefault(require("../rooms/NotificationBadge"));

var _languageHandler = require("../../../languageHandler");

var _ContextMenuTooltipButton = require("../../../accessibility/context_menu/ContextMenuTooltipButton");

var _ContextMenu = require("../../structures/ContextMenu");

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _StaticNotificationState = require("../../../stores/notifications/StaticNotificationState");

var _NotificationColor = require("../../../stores/notifications/NotificationColor");

var _KeyBindingsManager = require("../../../KeyBindingsManager");

var _SpaceContextMenu = _interopRequireDefault(require("../context_menus/SpaceContextMenu"));

var _AccessibleTooltipButton = _interopRequireDefault(require("../elements/AccessibleTooltipButton"));

var _RovingTabIndex = require("../../../accessibility/RovingTabIndex");

const _excluded = ["space", "spaceKey", "className", "selected", "label", "contextMenuTooltip", "notificationState", "avatarSize", "isNarrow", "children", "ContextMenuComponent"],
      _excluded2 = ["space", "activeSpaces", "isNested", "isPanelCollapsed", "onExpand", "parents", "innerRef", "dragHandleProps"],
      _excluded3 = ["tabIndex"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const SpaceButton = _ref => {
  let {
    space,
    spaceKey,
    className,
    selected,
    label,
    contextMenuTooltip,
    notificationState,
    avatarSize,
    isNarrow,
    children,
    ContextMenuComponent
  } = _ref,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  const [menuDisplayed, ref, openMenu, closeMenu] = (0, _ContextMenu.useContextMenu)();
  const [onFocus, isActive, handle] = (0, _RovingTabIndex.useRovingTabIndex)(ref);
  const tabIndex = isActive ? 0 : -1;

  let avatar = /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceButton_avatarPlaceholder"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceButton_icon"
  }));

  if (space) {
    avatar = /*#__PURE__*/_react.default.createElement(_RoomAvatar.default, {
      width: avatarSize,
      height: avatarSize,
      room: space
    });
  }

  let notifBadge;

  if (notificationState) {
    let ariaLabel = (0, _languageHandler._t)("Jump to first unread room.");

    if ((space === null || space === void 0 ? void 0 : space.getMyMembership()) === "invite") {
      ariaLabel = (0, _languageHandler._t)("Jump to first invite.");
    }

    notifBadge = /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SpacePanel_badgeContainer"
    }, /*#__PURE__*/_react.default.createElement(_NotificationBadge.default, {
      onClick: () => _SpaceStore.default.instance.setActiveRoomInSpace(spaceKey ?? space.roomId),
      forceCount: false,
      notification: notificationState,
      "aria-label": ariaLabel,
      tabIndex: tabIndex,
      showUnsentTooltip: true
    }));
  }

  let contextMenu;

  if (menuDisplayed && ContextMenuComponent) {
    var _handle$current;

    contextMenu = /*#__PURE__*/_react.default.createElement(ContextMenuComponent, (0, _extends2.default)({}, (0, _ContextMenu.toRightOf)((_handle$current = handle.current) === null || _handle$current === void 0 ? void 0 : _handle$current.getBoundingClientRect(), 0), {
      space: space,
      onFinished: closeMenu
    }));
  }

  return /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, (0, _extends2.default)({}, props, {
    className: (0, _classnames.default)("mx_SpaceButton", className, {
      mx_SpaceButton_active: selected,
      mx_SpaceButton_hasMenuOpen: menuDisplayed,
      mx_SpaceButton_narrow: isNarrow
    }),
    title: label,
    onClick: spaceKey ? () => _SpaceStore.default.instance.setActiveSpace(spaceKey) : props.onClick,
    onContextMenu: openMenu,
    forceHide: !isNarrow || menuDisplayed,
    inputRef: handle,
    tabIndex: tabIndex,
    onFocus: onFocus
  }), children, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceButton_selectionWrapper"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceButton_avatarWrapper"
  }, avatar, notifBadge), !isNarrow && /*#__PURE__*/_react.default.createElement("span", {
    className: "mx_SpaceButton_name"
  }, label), ContextMenuComponent && /*#__PURE__*/_react.default.createElement(_ContextMenuTooltipButton.ContextMenuTooltipButton, {
    className: "mx_SpaceButton_menuButton",
    onClick: openMenu,
    title: contextMenuTooltip,
    isExpanded: menuDisplayed
  }), contextMenu));
};

exports.SpaceButton = SpaceButton;

class SpaceItem extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "buttonRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "onSpaceUpdate", () => {
      this.setState({
        childSpaces: this.childSpaces
      });
    });
    (0, _defineProperty2.default)(this, "toggleCollapse", evt => {
      if (this.props.onExpand && this.isCollapsed) {
        this.props.onExpand();
      }

      const newCollapsedState = !this.isCollapsed;

      _SpaceTreeLevelLayoutStore.default.instance.setSpaceCollapsedState(this.props.space.roomId, this.props.parents, newCollapsedState);

      this.setState({
        collapsed: newCollapsedState
      }); // don't bubble up so encapsulating button for space
      // doesn't get triggered

      evt.stopPropagation();
    });
    (0, _defineProperty2.default)(this, "onKeyDown", ev => {
      var _this$state$childSpac;

      let handled = true;
      const action = (0, _KeyBindingsManager.getKeyBindingsManager)().getRoomListAction(ev);
      const hasChildren = (_this$state$childSpac = this.state.childSpaces) === null || _this$state$childSpac === void 0 ? void 0 : _this$state$childSpac.length;

      switch (action) {
        case _KeyBindingsManager.RoomListAction.CollapseSection:
          if (hasChildren && !this.isCollapsed) {
            this.toggleCollapse(ev);
          } else {
            var _this$buttonRef, _this$buttonRef$curre, _this$buttonRef$curre2;

            const parentItem = (_this$buttonRef = this.buttonRef) === null || _this$buttonRef === void 0 ? void 0 : (_this$buttonRef$curre = _this$buttonRef.current) === null || _this$buttonRef$curre === void 0 ? void 0 : (_this$buttonRef$curre2 = _this$buttonRef$curre.parentElement) === null || _this$buttonRef$curre2 === void 0 ? void 0 : _this$buttonRef$curre2.parentElement;
            const parentButton = parentItem === null || parentItem === void 0 ? void 0 : parentItem.previousElementSibling;
            parentButton === null || parentButton === void 0 ? void 0 : parentButton.focus();
          }

          break;

        case _KeyBindingsManager.RoomListAction.ExpandSection:
          if (hasChildren) {
            if (this.isCollapsed) {
              this.toggleCollapse(ev);
            } else {
              var _this$buttonRef2, _this$buttonRef2$curr, _firstSpaceItemChild$;

              const childLevel = (_this$buttonRef2 = this.buttonRef) === null || _this$buttonRef2 === void 0 ? void 0 : (_this$buttonRef2$curr = _this$buttonRef2.current) === null || _this$buttonRef2$curr === void 0 ? void 0 : _this$buttonRef2$curr.nextElementSibling;
              const firstSpaceItemChild = childLevel === null || childLevel === void 0 ? void 0 : childLevel.querySelector(".mx_SpaceItem");
              firstSpaceItemChild === null || firstSpaceItemChild === void 0 ? void 0 : (_firstSpaceItemChild$ = firstSpaceItemChild.querySelector(".mx_SpaceButton")) === null || _firstSpaceItemChild$ === void 0 ? void 0 : _firstSpaceItemChild$.focus();
            }
          }

          break;

        default:
          handled = false;
      }

      if (handled) {
        ev.stopPropagation();
        ev.preventDefault();
      }
    });
    (0, _defineProperty2.default)(this, "onClick", ev => {
      ev.preventDefault();
      ev.stopPropagation();

      _SpaceStore.default.instance.setActiveSpace(this.props.space.roomId);
    });

    const collapsed = _SpaceTreeLevelLayoutStore.default.instance.getSpaceCollapsedState(props.space.roomId, this.props.parents, !props.isNested // default to collapsed for root items
    );

    this.state = {
      collapsed: collapsed,
      childSpaces: this.childSpaces
    };

    _SpaceStore.default.instance.on(this.props.space.roomId, this.onSpaceUpdate);
  }

  componentWillUnmount() {
    _SpaceStore.default.instance.off(this.props.space.roomId, this.onSpaceUpdate);
  }

  get childSpaces() {
    return _SpaceStore.default.instance.getChildSpaces(this.props.space.roomId).filter(s => {
      var _this$props$parents;

      return !((_this$props$parents = this.props.parents) !== null && _this$props$parents !== void 0 && _this$props$parents.has(s.roomId));
    });
  }

  get isCollapsed() {
    return this.state.collapsed || this.props.isPanelCollapsed;
  }

  render() {
    var _this$state$childSpac2, _this$state$childSpac3;

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _this$props = this.props,
          {
      space,
      activeSpaces,
      isNested,
      isPanelCollapsed,
      onExpand,
      parents,
      innerRef,
      dragHandleProps
    } = _this$props,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded2);
    const collapsed = this.isCollapsed;
    const itemClasses = (0, _classnames.default)(this.props.className, {
      "mx_SpaceItem": true,
      "mx_SpaceItem_narrow": isPanelCollapsed,
      "collapsed": collapsed,
      "hasSubSpaces": (_this$state$childSpac2 = this.state.childSpaces) === null || _this$state$childSpac2 === void 0 ? void 0 : _this$state$childSpac2.length
    });
    const isInvite = space.getMyMembership() === "invite";
    const notificationState = isInvite ? _StaticNotificationState.StaticNotificationState.forSymbol("!", _NotificationColor.NotificationColor.Red) : _SpaceStore.default.instance.getNotificationState(space.roomId);
    const hasChildren = (_this$state$childSpac3 = this.state.childSpaces) === null || _this$state$childSpac3 === void 0 ? void 0 : _this$state$childSpac3.length;
    let childItems;

    if (hasChildren && !collapsed) {
      childItems = /*#__PURE__*/_react.default.createElement(SpaceTreeLevel, {
        spaces: this.state.childSpaces,
        activeSpaces: activeSpaces,
        isNested: true,
        parents: new Set(parents).add(space.roomId)
      });
    }

    const toggleCollapseButton = hasChildren ? /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_SpaceButton_toggleCollapse",
      onClick: this.toggleCollapse,
      tabIndex: -1,
      "aria-label": collapsed ? (0, _languageHandler._t)("Expand") : (0, _languageHandler._t)("Collapse")
    }) : null; // eslint-disable-next-line @typescript-eslint/no-unused-vars

    const _ref2 = dragHandleProps || {},
          {
      tabIndex
    } = _ref2,
          restDragHandleProps = (0, _objectWithoutProperties2.default)(_ref2, _excluded3);

    return /*#__PURE__*/_react.default.createElement("li", (0, _extends2.default)({}, otherProps, {
      className: itemClasses,
      ref: innerRef,
      "aria-expanded": hasChildren ? !collapsed : undefined,
      role: "treeitem"
    }), /*#__PURE__*/_react.default.createElement(SpaceButton, (0, _extends2.default)({}, restDragHandleProps, {
      space: space,
      className: isInvite ? "mx_SpaceButton_invite" : undefined,
      selected: activeSpaces.includes(space.roomId),
      label: space.name,
      contextMenuTooltip: (0, _languageHandler._t)("Space options"),
      notificationState: notificationState,
      isNarrow: isPanelCollapsed,
      avatarSize: isNested ? 24 : 32,
      onClick: this.onClick,
      onKeyDown: this.onKeyDown,
      ContextMenuComponent: this.props.space.getMyMembership() === "join" ? _SpaceContextMenu.default : undefined
    }), toggleCollapseButton), childItems);
  }

}

exports.SpaceItem = SpaceItem;
(0, _defineProperty2.default)(SpaceItem, "contextType", _MatrixClientContext.default);

const SpaceTreeLevel = ({
  spaces,
  activeSpaces,
  isNested,
  parents
}) => {
  return /*#__PURE__*/_react.default.createElement("ul", {
    className: "mx_SpaceTreeLevel",
    role: "group"
  }, spaces.map(s => {
    return /*#__PURE__*/_react.default.createElement(SpaceItem, {
      key: s.roomId,
      activeSpaces: activeSpaces,
      space: s,
      isNested: isNested,
      parents: parents
    });
  }));
};

var _default = SpaceTreeLevel;
exports.default = _default;
