// matrix-react-sdk/lib/stores/room-list/filters/CommunityFilterCondition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommunityFilterCondition = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _IFilterCondition = require("./IFilterCondition");

var _events = require("events");

var _GroupStore = _interopRequireDefault(require("../../GroupStore"));

var _DMRoomMap = _interopRequireDefault(require("../../../utils/DMRoomMap"));

var _sets = require("../../../utils/sets");

/*
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * A filter condition for the room list which reveals rooms which
 * are a member of a given community.
 */
class CommunityFilterCondition extends _events.EventEmitter {
  constructor(community) {
    super();
    this.community = community;
    (0, _defineProperty2.default)(this, "roomIds", new Set());
    (0, _defineProperty2.default)(this, "userIds", new Set());
    (0, _defineProperty2.default)(this, "onStoreUpdate", async () => {
      // We don't actually know if the room list changed for the community, so just check it again.
      const beforeRoomIds = this.roomIds;
      this.roomIds = new Set((await _GroupStore.default.getGroupRooms(this.community.groupId)).map(r => r.roomId));
      const beforeUserIds = this.userIds;
      this.userIds = new Set((await _GroupStore.default.getGroupMembers(this.community.groupId)).map(u => u.userId));

      if ((0, _sets.setHasDiff)(beforeRoomIds, this.roomIds) || (0, _sets.setHasDiff)(beforeUserIds, this.userIds)) {
        this.emit(_IFilterCondition.FILTER_CHANGED);
      }
    });

    _GroupStore.default.on("update", this.onStoreUpdate); // noinspection JSIgnoredPromiseFromCall


    this.onStoreUpdate(); // trigger a false update to seed the store
  }

  get kind() {
    return _IFilterCondition.FilterKind.Prefilter;
  }

  isVisible(room) {
    return this.roomIds.has(room.roomId) || this.userIds.has(_DMRoomMap.default.shared().getUserIdForRoomId(room.roomId));
  }

  destroy() {
    _GroupStore.default.off("update", this.onStoreUpdate);
  }

}

exports.CommunityFilterCondition = CommunityFilterCondition;
