// matrix-react-sdk/lib/stores/room-list/filters/NameFilterCondition.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NameFilterCondition = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _IFilterCondition = require("./IFilterCondition");

var _events = require("events");

var _utils = require("matrix-js-sdk/src/utils");

var _lodash = require("lodash");

/*
Copyright 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * A filter condition for the room list which reveals rooms of a particular
 * name, or associated name (like a room alias).
 */
class NameFilterCondition extends _events.EventEmitter {
  constructor() {
    super();
    (0, _defineProperty2.default)(this, "_search", "");
    (0, _defineProperty2.default)(this, "callUpdate", (0, _lodash.throttle)(() => {
      this.emit(_IFilterCondition.FILTER_CHANGED);
    }, 200, {
      trailing: true,
      leading: true
    }));
  }

  get kind() {
    return _IFilterCondition.FilterKind.Runtime;
  }

  get search() {
    return this._search;
  }

  set search(val) {
    this._search = val;
    this.callUpdate();
  }

  isVisible(room) {
    const lcFilter = this.search.toLowerCase();

    if (this.search[0] === '#') {
      // Try and find rooms by alias
      if (room.getCanonicalAlias() && room.getCanonicalAlias().toLowerCase().startsWith(lcFilter)) {
        return true;
      }

      if (room.getAltAliases().some(a => a.toLowerCase().startsWith(lcFilter))) {
        return true;
      }
    }

    if (!room.name) return false; // should realistically not happen: the js-sdk always calculates a name

    return this.matches(room.normalizedName);
  }

  matches(normalizedName) {
    return normalizedName.includes((0, _utils.normalize)(this.search));
  }

}

exports.NameFilterCondition = NameFilterCondition;
