// matrix-react-sdk/lib/components/views/voip/AudioFeed.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _callFeed = require("matrix-js-sdk/src/webrtc/callFeed");

var _logger = require("matrix-js-sdk/src/logger");

var _MediaDeviceHandler = _interopRequireWildcard(require("../../../MediaDeviceHandler"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2021 Šimon Brandner <simon.bra.ag@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class AudioFeed extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "element", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "onAudioOutputChanged", audioOutput => {
      const element = this.element.current;

      if (audioOutput) {
        try {
          // This seems quite unreliable in Chrome, although I haven't yet managed to make a jsfiddle where
          // it fails.
          // It seems reliable if you set the sink ID after setting the srcObject and then set the sink ID
          // back to the default after the call is over - Dave
          element.setSinkId(audioOutput);
        } catch (e) {
          _logger.logger.error("Couldn't set requested audio output device: using default", e);

          _logger.logger.warn("Couldn't set requested audio output device: using default", e);
        }
      }
    });
    (0, _defineProperty2.default)(this, "onNewStream", () => {
      this.setState({
        audioMuted: this.props.feed.isAudioMuted()
      });
      this.playMedia();
    });
    this.state = {
      audioMuted: this.props.feed.isAudioMuted()
    };
  }

  componentDidMount() {
    _MediaDeviceHandler.default.instance.addListener(_MediaDeviceHandler.MediaDeviceHandlerEvent.AudioOutputChanged, this.onAudioOutputChanged);

    this.props.feed.addListener(_callFeed.CallFeedEvent.NewStream, this.onNewStream);
    this.playMedia();
  }

  componentWillUnmount() {
    _MediaDeviceHandler.default.instance.removeListener(_MediaDeviceHandler.MediaDeviceHandlerEvent.AudioOutputChanged, this.onAudioOutputChanged);

    this.props.feed.removeListener(_callFeed.CallFeedEvent.NewStream, this.onNewStream);
    this.stopMedia();
  }

  async playMedia() {
    const element = this.element.current;
    if (!element) return;
    this.onAudioOutputChanged(_MediaDeviceHandler.default.getAudioOutput());
    element.muted = false;
    element.srcObject = this.props.feed.stream;
    element.autoplay = true;

    try {
      // A note on calling methods on media elements:
      // We used to have queues per media element to serialise all calls on those elements.
      // The reason given for this was that load() and play() were racing. However, we now
      // never call load() explicitly so this seems unnecessary. However, serialising every
      // operation was causing bugs where video would not resume because some play command
      // had got stuck and all media operations were queued up behind it. If necessary, we
      // should serialise the ones that need to be serialised but then be able to interrupt
      // them with another load() which will cancel the pending one, but since we don't call
      // load() explicitly, it shouldn't be a problem. - Dave
      await element.load();
    } catch (e) {
      _logger.logger.info("Failed to play media element with feed", this.props.feed, e);
    }
  }

  stopMedia() {
    const element = this.element.current;
    if (!element) return;
    element.pause();
    element.src = null; // As per comment in componentDidMount, setting the sink ID back to the
    // default once the call is over makes setSinkId work reliably. - Dave
    // Since we are not using the same element anymore, the above doesn't
    // seem to be necessary - Šimon
  }

  render() {
    // Do not render the audio element if there is no audio track
    if (this.state.audioMuted) return null;
    return /*#__PURE__*/_react.default.createElement("audio", {
      ref: this.element
    });
  }

}

exports.default = AudioFeed;
