// matrix-react-sdk/lib/components/views/voip/AudioFeedArrayForCall.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _AudioFeed = _interopRequireDefault(require("./AudioFeed"));

var _call = require("matrix-js-sdk/src/webrtc/call");

/*
Copyright 2021 Šimon Brandner <simon.bra.ag@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class AudioFeedArrayForCall extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onFeedsChanged", () => {
      this.setState({
        feeds: this.props.call.getRemoteFeeds()
      });
    });
    this.state = {
      feeds: this.props.call.getRemoteFeeds()
    };
  }

  componentDidMount() {
    this.props.call.addListener(_call.CallEvent.FeedsChanged, this.onFeedsChanged);
  }

  componentWillUnmount() {
    this.props.call.removeListener(_call.CallEvent.FeedsChanged, this.onFeedsChanged);
  }

  render() {
    return this.state.feeds.map((feed, i) => {
      return /*#__PURE__*/_react.default.createElement(_AudioFeed.default, {
        feed: feed,
        key: i
      });
    });
  }

}

exports.default = AudioFeedArrayForCall;
