// matrix-react-sdk/lib/accessibility/context_menu/StyledMenuItemCheckbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledMenuItemCheckbox = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _Keyboard = require("../../Keyboard");

var _StyledCheckbox = _interopRequireDefault(require("../../components/views/elements/StyledCheckbox"));

const _excluded = ["children", "label", "onChange", "onClose"];

// Semantic component for representing a styled role=menuitemcheckbox
const StyledMenuItemCheckbox = _ref => {
  let {
    children,
    label,
    onChange,
    onClose
  } = _ref,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);

  const onKeyDown = e => {
    if (e.key === _Keyboard.Key.ENTER || e.key === _Keyboard.Key.SPACE) {
      e.stopPropagation();
      e.preventDefault();
      onChange(); // Implements https://www.w3.org/TR/wai-aria-practices/#keyboard-interaction-12

      if (e.key === _Keyboard.Key.ENTER) {
        onClose();
      }
    }
  };

  const onKeyUp = e => {
    // prevent the input default handler as we handle it on keydown to match
    // https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-2/menubar-2.html
    if (e.key === _Keyboard.Key.SPACE || e.key === _Keyboard.Key.ENTER) {
      e.stopPropagation();
      e.preventDefault();
    }
  };

  return /*#__PURE__*/_react.default.createElement(_StyledCheckbox.default, (0, _extends2.default)({}, props, {
    role: "menuitemcheckbox",
    tabIndex: -1,
    "aria-label": label,
    onChange: onChange,
    onKeyDown: onKeyDown,
    onKeyUp: onKeyUp
  }), children);
};

exports.StyledMenuItemCheckbox = StyledMenuItemCheckbox;
