// matrix-react-sdk/lib/async-components/views/dialogs/eventindex/DisableEventIndexDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _BaseDialog = _interopRequireDefault(require("../../../../components/views/dialogs/BaseDialog"));

var _Spinner = _interopRequireDefault(require("../../../../components/views/elements/Spinner"));

var _DialogButtons = _interopRequireDefault(require("../../../../components/views/elements/DialogButtons"));

var _dispatcher = _interopRequireDefault(require("../../../../dispatcher/dispatcher"));

var _languageHandler = require("../../../../languageHandler");

var _SettingsStore = _interopRequireDefault(require("../../../../settings/SettingsStore"));

var _EventIndexPeg = _interopRequireDefault(require("../../../../indexing/EventIndexPeg"));

var _actions = require("../../../../dispatcher/actions");

var _SettingLevel = require("../../../../settings/SettingLevel");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/*
 * Allows the user to disable the Event Index.
 */
class DisableEventIndexDialog extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onDisable", async () => {
      this.setState({
        disabling: true
      });
      await _SettingsStore.default.setValue('enableEventIndexing', null, _SettingLevel.SettingLevel.DEVICE, false);
      await _EventIndexPeg.default.deleteEventIndex();
      this.props.onFinished(true);

      _dispatcher.default.fire(_actions.Action.ViewUserSettings);
    });
    this.state = {
      disabling: false
    };
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Are you sure?")
    }, (0, _languageHandler._t)("If disabled, messages from encrypted rooms won't appear in search results."), this.state.disabling ? /*#__PURE__*/_react.default.createElement(_Spinner.default, null) : /*#__PURE__*/_react.default.createElement("div", null), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)('Disable'),
      onPrimaryButtonClick: this.onDisable,
      primaryButtonClass: "danger",
      cancelButtonClass: "warning",
      onCancel: this.props.onFinished,
      disabled: this.state.disabling
    }));
  }

}

exports.default = DisableEventIndexDialog;
