// matrix-react-sdk/lib/dispatcher/dispatch-actions/threads.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dispatchShowThreadsPanelEvent = exports.dispatchShowThreadEvent = void 0;

var _RightPanelStorePhases = require("../../stores/RightPanelStorePhases");

var _actions = require("../actions");

var _dispatcher = _interopRequireDefault(require("../dispatcher"));

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const dispatchShowThreadEvent = (rootEvent, initialEvent, highlighted) => {
  _dispatcher.default.dispatch({
    action: _actions.Action.SetRightPanelPhase,
    phase: _RightPanelStorePhases.RightPanelPhases.ThreadView,
    refireParams: {
      event: rootEvent,
      initialEvent,
      highlighted
    }
  });
};

exports.dispatchShowThreadEvent = dispatchShowThreadEvent;

const dispatchShowThreadsPanelEvent = () => {
  _dispatcher.default.dispatch({
    action: _actions.Action.SetRightPanelPhase,
    phase: _RightPanelStorePhases.RightPanelPhases.ThreadPanel
  });
};

exports.dispatchShowThreadsPanelEvent = dispatchShowThreadsPanelEvent;
