// matrix-react-sdk/lib/stores/widgets/ElementWidgetActions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElementWidgetActions = void 0;

/*
 * Copyright 2020 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
let ElementWidgetActions;
/**
 * @deprecated Use MSC2931 instead
 */

exports.ElementWidgetActions = ElementWidgetActions;

(function (ElementWidgetActions) {
  ElementWidgetActions["ClientReady"] = "im.vector.ready";
  ElementWidgetActions["HangupCall"] = "im.vector.hangup";
  ElementWidgetActions["StartLiveStream"] = "im.vector.start_live_stream";
  ElementWidgetActions["OpenIntegrationManager"] = "integration_manager_open";
  ElementWidgetActions["ViewRoom"] = "io.element.view_room";
})(ElementWidgetActions || (exports.ElementWidgetActions = ElementWidgetActions = {}));
