// matrix-react-sdk/lib/stores/widgets/WidgetMessagingStore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WidgetMessagingStore = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _AsyncStoreWithClient = require("../AsyncStoreWithClient");

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _maps = require("../../utils/maps");

/*
 * Copyright 2020 The Matrix.org Foundation C.I.C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Temporary holding store for widget messaging instances. This is eventually
 * going to be merged with a more complete WidgetStore, but for now it's
 * easiest to split this into a single place.
 */
class WidgetMessagingStore extends _AsyncStoreWithClient.AsyncStoreWithClient {
  // TODO: Fix uniqueness problem (widget IDs are not unique across the whole app)
  // <widget ID, ClientWidgetAPi>
  constructor() {
    super(_dispatcher.default);
    (0, _defineProperty2.default)(this, "widgetMap", new _maps.EnhancedMap());
  }

  static get instance() {
    return WidgetMessagingStore.internalInstance;
  }

  async onAction(payload) {// nothing to do
  }

  async onReady() {
    // just in case
    this.widgetMap.clear();
  }

  storeMessaging(widget, widgetApi) {
    this.stopMessaging(widget);
    this.widgetMap.set(widget.id, widgetApi);
  }

  stopMessaging(widget) {
    var _this$widgetMap$remov;

    (_this$widgetMap$remov = this.widgetMap.remove(widget.id)) === null || _this$widgetMap$remov === void 0 ? void 0 : _this$widgetMap$remov.stop();
  }

  getMessaging(widget) {
    return this.widgetMap.get(widget.id);
  }
  /**
   * Stops the widget messaging instance for a given widget ID.
   * @param {string} widgetId The widget ID.
   * @deprecated Widget IDs are not globally unique.
   */


  stopMessagingById(widgetId) {
    var _this$widgetMap$remov2;

    (_this$widgetMap$remov2 = this.widgetMap.remove(widgetId)) === null || _this$widgetMap$remov2 === void 0 ? void 0 : _this$widgetMap$remov2.stop();
  }
  /**
   * Gets the widget messaging class for a given widget ID.
   * @param {string} widgetId The widget ID.
   * @returns {ClientWidgetApi} The widget API, or a falsey value if not found.
   * @deprecated Widget IDs are not globally unique.
   */


  getMessagingForId(widgetId) {
    return this.widgetMap.get(widgetId);
  }

}

exports.WidgetMessagingStore = WidgetMessagingStore;
(0, _defineProperty2.default)(WidgetMessagingStore, "internalInstance", new WidgetMessagingStore());
