// matrix-react-sdk/lib/components/views/messages/MStickerBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _MImageBody = _interopRequireDefault(require("./MImageBody"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ContentMessages = require("../../../ContentMessages");

var _Tooltip = _interopRequireDefault(require("../elements/Tooltip"));

var _dec, _class;

let MStickerBody = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.MStickerBody"), _dec(_class = class MStickerBody extends _MImageBody.default {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", ev => {
      ev.preventDefault();

      if (!this.state.showImage) {
        this.showImage();
      }
    });
  }

  // MStickerBody doesn't need a wrapping `<a href=...>`, but it does need extra padding
  // which is added by mx_MStickerBody_wrapper
  wrapImage(contentUrl, children) {
    let onClick = null;

    if (!this.state.showImage) {
      onClick = this.onClick;
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MStickerBody_wrapper",
      onClick: onClick
    }, " ", children, " ");
  } // Placeholder to show in place of the sticker image if
  // img onLoad hasn't fired yet.


  getPlaceholder(width, height) {
    var _this$props$mxEvent$g;

    if ((_this$props$mxEvent$g = this.props.mxEvent.getContent().info) !== null && _this$props$mxEvent$g !== void 0 && _this$props$mxEvent$g[_ContentMessages.BLURHASH_FIELD]) return super.getPlaceholder(width, height);
    return /*#__PURE__*/_react.default.createElement("img", {
      src: require("../../../../res/img/icons-show-stickers.svg"),
      width: "75",
      height: "75"
    });
  } // Tooltip to show on mouse over


  getTooltip() {
    const content = this.props.mxEvent && this.props.mxEvent.getContent();
    if (!content || !content.body || !content.info || !content.info.w) return null;
    return /*#__PURE__*/_react.default.createElement("div", {
      style: {
        left: content.info.w + 'px'
      },
      className: "mx_MStickerBody_tooltip"
    }, /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
      label: content.body
    }));
  } // Don't show "Download this_file.png ..."


  getFileBody() {
    return null;
  }

}) || _class);
exports.default = MStickerBody;
