// matrix-react-sdk/lib/components/views/messages/MVoiceMessageBody.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _InlineSpinner = _interopRequireDefault(require("../elements/InlineSpinner"));

var _languageHandler = require("../../../languageHandler");

var _RecordingPlayback = _interopRequireDefault(require("../audio_messages/RecordingPlayback"));

var _MAudioBody = _interopRequireDefault(require("./MAudioBody"));

var _MFileBody = _interopRequireDefault(require("./MFileBody"));

var _dec, _class;

let MVoiceMessageBody = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.MVoiceMessageBody"), _dec(_class = class MVoiceMessageBody extends _MAudioBody.default {
  // A voice message is an audio file but rendered in a special way.
  render() {
    if (this.state.error) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: "mx_MVoiceMessageBody"
      }, /*#__PURE__*/_react.default.createElement("img", {
        src: require("../../../../res/img/warning.svg"),
        width: "16",
        height: "16"
      }), (0, _languageHandler._t)("Error processing voice message"));
    }

    if (!this.state.playback) {
      return /*#__PURE__*/_react.default.createElement("span", {
        className: "mx_MVoiceMessageBody"
      }, /*#__PURE__*/_react.default.createElement(_InlineSpinner.default, null));
    } // At this point we should have a playable state


    return /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_MVoiceMessageBody"
    }, /*#__PURE__*/_react.default.createElement(_RecordingPlayback.default, {
      playback: this.state.playback,
      tileShape: this.props.tileShape
    }), this.props.tileShape && /*#__PURE__*/_react.default.createElement(_MFileBody.default, (0, _extends2.default)({}, this.props, {
      showGenericPlaceholder: false
    })));
  }

}) || _class);
exports.default = MVoiceMessageBody;
