// matrix-react-sdk/lib/components/views/messages/ReactionsRowButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _languageHandler = require("../../../languageHandler");

var _FormattingUtils = require("../../../utils/FormattingUtils");

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ReactionsRowButtonTooltip = _interopRequireDefault(require("./ReactionsRowButtonTooltip"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _dec, _class, _class2, _temp;

let ReactionsRowButton = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.ReactionsRowButton"), _dec(_class = (_temp = _class2 = class ReactionsRowButton extends _react.default.PureComponent {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      tooltipRendered: false,
      tooltipVisible: false
    });
    (0, _defineProperty2.default)(this, "onClick", () => {
      const {
        mxEvent,
        myReactionEvent,
        content
      } = this.props;

      if (myReactionEvent) {
        this.context.redactEvent(mxEvent.getRoomId(), myReactionEvent.getId());
      } else {
        this.context.sendEvent(mxEvent.getRoomId(), "m.reaction", {
          "m.relates_to": {
            "rel_type": "m.annotation",
            "event_id": mxEvent.getId(),
            "key": content
          }
        });

        _dispatcher.default.dispatch({
          action: "message_sent"
        });
      }
    });
    (0, _defineProperty2.default)(this, "onMouseOver", () => {
      this.setState({
        // To avoid littering the DOM with a tooltip for every reaction,
        // only render it on first use.
        tooltipRendered: true,
        tooltipVisible: true
      });
    });
    (0, _defineProperty2.default)(this, "onMouseLeave", () => {
      this.setState({
        tooltipVisible: false
      });
    });
  }

  render() {
    const {
      mxEvent,
      content,
      count,
      reactionEvents,
      myReactionEvent
    } = this.props;
    const classes = (0, _classnames.default)({
      mx_ReactionsRowButton: true,
      mx_ReactionsRowButton_selected: !!myReactionEvent
    });
    let tooltip;

    if (this.state.tooltipRendered) {
      tooltip = /*#__PURE__*/_react.default.createElement(_ReactionsRowButtonTooltip.default, {
        mxEvent: this.props.mxEvent,
        content: content,
        reactionEvents: reactionEvents,
        visible: this.state.tooltipVisible
      });
    }

    const room = this.context.getRoom(mxEvent.getRoomId());
    let label;

    if (room) {
      const senders = [];

      for (const reactionEvent of reactionEvents) {
        const member = room.getMember(reactionEvent.getSender());
        senders.push((member === null || member === void 0 ? void 0 : member.name) || reactionEvent.getSender());
      }

      const reactors = (0, _FormattingUtils.formatCommaSeparatedList)(senders, 6);

      if (content) {
        label = (0, _languageHandler._t)("%(reactors)s reacted with %(content)s", {
          reactors,
          content
        });
      } else {
        label = reactors;
      }
    }

    const isPeeking = room.getMyMembership() !== "join";
    return /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: classes,
      "aria-label": label,
      onClick: this.onClick,
      disabled: isPeeking,
      onMouseOver: this.onMouseOver,
      onMouseLeave: this.onMouseLeave
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_ReactionsRowButton_content",
      "aria-hidden": "true"
    }, content), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_ReactionsRowButton_count",
      "aria-hidden": "true"
    }, count), tooltip);
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _MatrixClientContext.default), _temp)) || _class);
exports.default = ReactionsRowButton;
