// matrix-react-sdk/lib/components/views/messages/ReactionsRowButtonTooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _HtmlUtils = require("../../../HtmlUtils");

var _languageHandler = require("../../../languageHandler");

var _FormattingUtils = require("../../../utils/FormattingUtils");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Tooltip = _interopRequireDefault(require("../elements/Tooltip"));

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _dec, _class, _class2, _temp;

let ReactionsRowButtonTooltip = (_dec = (0, _replaceableComponent.replaceableComponent)("views.messages.ReactionsRowButtonTooltip"), _dec(_class = (_temp = _class2 = class ReactionsRowButtonTooltip extends _react.default.PureComponent {
  render() {
    const {
      content,
      reactionEvents,
      mxEvent,
      visible
    } = this.props;
    const room = this.context.getRoom(mxEvent.getRoomId());
    let tooltipLabel;

    if (room) {
      const senders = [];

      for (const reactionEvent of reactionEvents) {
        const member = room.getMember(reactionEvent.getSender());
        const name = member ? member.name : reactionEvent.getSender();
        senders.push(name);
      }

      const shortName = (0, _HtmlUtils.unicodeToShortcode)(content);
      tooltipLabel = /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("<reactors/><reactedWith>reacted with %(shortName)s</reactedWith>", {
        shortName
      }, {
        reactors: () => {
          return /*#__PURE__*/_react.default.createElement("div", {
            className: "mx_Tooltip_title"
          }, (0, _FormattingUtils.formatCommaSeparatedList)(senders, 6));
        },
        reactedWith: sub => {
          if (!shortName) {
            return null;
          }

          return /*#__PURE__*/_react.default.createElement("div", {
            className: "mx_Tooltip_sub"
          }, sub);
        }
      }));
    }

    let tooltip;

    if (tooltipLabel) {
      tooltip = /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
        visible: visible,
        label: tooltipLabel
      });
    }

    return tooltip;
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _MatrixClientContext.default), _temp)) || _class);
exports.default = ReactionsRowButtonTooltip;
