// matrix-react-sdk/lib/ActiveRoomObserver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ActiveRoomObserver = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _RoomViewStore = _interopRequireDefault(require("./stores/RoomViewStore"));

var _logger = require("matrix-js-sdk/src/logger");

/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
  constructor() {
    (0, _defineProperty2.default)(this, "listeners", {});
    (0, _defineProperty2.default)(this, "_activeRoomId", _RoomViewStore.default.getRoomId());
    (0, _defineProperty2.default)(this, "roomStoreToken", void 0);
    (0, _defineProperty2.default)(this, "onRoomViewStoreUpdate", () => {
      // emit for the old room ID
      if (this._activeRoomId) this.emit(this._activeRoomId, false); // update our cache

      this._activeRoomId = _RoomViewStore.default.getRoomId(); // and emit for the new one

      if (this._activeRoomId) this.emit(this._activeRoomId, true);
    });
    // TODO: We could self-destruct when the last listener goes away, or at least stop listening.
    this.roomStoreToken = _RoomViewStore.default.addListener(this.onRoomViewStoreUpdate);
  }

  get activeRoomId() {
    return this._activeRoomId;
  }

  addListener(roomId, listener) {
    if (!this.listeners[roomId]) this.listeners[roomId] = [];
    this.listeners[roomId].push(listener);
  }

  removeListener(roomId, listener) {
    if (this.listeners[roomId]) {
      const i = this.listeners[roomId].indexOf(listener);

      if (i > -1) {
        this.listeners[roomId].splice(i, 1);
      }
    } else {
      _logger.logger.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
    }
  }

  emit(roomId, isActive) {
    if (!this.listeners[roomId]) return;

    for (const l of this.listeners[roomId]) {
      l.call(null, isActive);
    }
  }

}

exports.ActiveRoomObserver = ActiveRoomObserver;

if (window.mxActiveRoomObserver === undefined) {
  window.mxActiveRoomObserver = new ActiveRoomObserver();
}

var _default = window.mxActiveRoomObserver;
exports.default = _default;
