// matrix-react-sdk/lib/AsyncWrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var sdk = _interopRequireWildcard(require("./index"));

var _languageHandler = require("./languageHandler");

var _logger = require("matrix-js-sdk/src/logger");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2015-2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Wrap an asynchronous loader function with a react component which shows a
 * spinner until the real component loads.
 */
class AsyncWrapper extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "unmounted", false);
    (0, _defineProperty2.default)(this, "state", {
      component: null,
      error: null
    });
    (0, _defineProperty2.default)(this, "onWrapperCancelClick", () => {
      this.props.onFinished(false);
    });
  }

  componentDidMount() {
    // XXX: temporary logging to try to diagnose
    // https://github.com/vector-im/element-web/issues/3148
    _logger.logger.log('Starting load of AsyncWrapper for modal');

    this.props.prom.then(result => {
      if (this.unmounted) return; // Take the 'default' member if it's there, then we support
      // passing in just an import()ed module, since ES6 async import
      // always returns a module *namespace*.

      const component = result.default ? result.default : result;
      this.setState({
        component
      });
    }).catch(e => {
      _logger.logger.warn('AsyncWrapper promise failed', e);

      this.setState({
        error: e
      });
    });
  }

  componentWillUnmount() {
    this.unmounted = true;
  }

  render() {
    if (this.state.component) {
      const Component = this.state.component;
      return /*#__PURE__*/_react.default.createElement(Component, this.props);
    } else if (this.state.error) {
      // FIXME: Using an import will result in test failures
      const BaseDialog = sdk.getComponent('views.dialogs.BaseDialog');
      const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
      return /*#__PURE__*/_react.default.createElement(BaseDialog, {
        onFinished: this.props.onFinished,
        title: (0, _languageHandler._t)("Error")
      }, (0, _languageHandler._t)("Unable to load! Check your network connectivity and try again."), /*#__PURE__*/_react.default.createElement(DialogButtons, {
        primaryButton: (0, _languageHandler._t)("Dismiss"),
        onPrimaryButtonClick: this.onWrapperCancelClick,
        hasCancel: false
      }));
    } else {
      // show a spinner until the component is loaded.
      const Spinner = sdk.getComponent("elements.Spinner");
      return /*#__PURE__*/_react.default.createElement(Spinner, null);
    }
  }

}

exports.default = AsyncWrapper;
