// matrix-react-sdk/lib/CountlyAnalytics.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _randomstring = require("matrix-js-sdk/src/randomstring");

var _utils = require("matrix-js-sdk/src/utils");

var _languageHandler = require("./languageHandler");

var _PlatformPeg = _interopRequireDefault(require("./PlatformPeg"));

var _SdkConfig = _interopRequireDefault(require("./SdkConfig"));

var _MatrixClientPeg = require("./MatrixClientPeg");

var _RoomViewStore = _interopRequireDefault(require("./stores/RoomViewStore"));

var _actions = require("./dispatcher/actions");

var _logger = require("matrix-js-sdk/src/logger");

const _excluded = ["count"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const INACTIVITY_TIME = 20; // seconds

const HEARTBEAT_INTERVAL = 5000; // ms

const SESSION_UPDATE_INTERVAL = 60; // seconds

const MAX_PENDING_EVENTS = 1000;
var Orientation;
/* eslint-disable camelcase */

(function (Orientation) {
  Orientation["Landscape"] = "landscape";
  Orientation["Portrait"] = "portrait";
})(Orientation || (Orientation = {}));

/* eslint-enable camelcase */
const hashHex = async input => {
  const buf = new TextEncoder().encode(input);
  const digestBuf = await window.crypto.subtle.digest("sha-256", buf);
  return [...new Uint8Array(digestBuf)].map(b => b.toString(16).padStart(2, "0")).join("");
};

const knownScreens = new Set(["register", "login", "forgot_password", "soft_logout", "new", "settings", "welcome", "home", "start", "directory", "start_sso", "start_cas", "groups", "complete_security", "post_registration", "room", "user", "group"]);

// Apply fn to all hash path parts after the 1st one
async function getViewData(anonymous = true) {
  const rand = (0, _randomstring.randomString)(8);
  const {
    origin,
    hash
  } = window.location;
  let {
    pathname
  } = window.location; // Redact paths which could contain unexpected PII

  if (origin.startsWith('file://')) {
    pathname = `/<redacted_${rand}>/`; // XXX: inject rand because Count.ly doesn't like X->X transitions
  }

  let [_, screen, ...parts] = hash.split("/");

  if (!knownScreens.has(screen)) {
    screen = `<redacted_${rand}>`;
  }

  for (let i = 0; i < parts.length; i++) {
    parts[i] = anonymous ? `<redacted_${rand}>` : await hashHex(parts[i]);
  }

  const hashStr = `${_}/${screen}/${parts.join("/")}`;
  const url = origin + pathname + hashStr;
  const meta = {};
  let name = "$/" + hash;

  switch (screen) {
    case "room":
      {
        name = "view_room";

        const roomId = _RoomViewStore.default.getRoomId();

        name += " " + parts[0]; // XXX: workaround Count.ly missing X->X transitions

        meta["room_id"] = parts[0];
        Object.assign(meta, getRoomStats(roomId));
        break;
      }
  }

  return {
    name,
    url,
    meta
  };
}

const getRoomStats = roomId => {
  var _room$currentState$ge, _room$currentState$ge2;

  const cli = _MatrixClientPeg.MatrixClientPeg.get();

  const room = cli === null || cli === void 0 ? void 0 : cli.getRoom(roomId);
  return {
    "num_users": room === null || room === void 0 ? void 0 : room.getJoinedMemberCount(),
    "is_encrypted": cli === null || cli === void 0 ? void 0 : cli.isRoomEncrypted(roomId),
    // eslint-disable-next-line camelcase
    "is_public": (room === null || room === void 0 ? void 0 : (_room$currentState$ge = room.currentState.getStateEvents("m.room.join_rules", "")) === null || _room$currentState$ge === void 0 ? void 0 : (_room$currentState$ge2 = _room$currentState$ge.getContent()) === null || _room$currentState$ge2 === void 0 ? void 0 : _room$currentState$ge2.join_rule) === "public"
  };
}; // async wrapper for regex-powered String.prototype.replace


const strReplaceAsync = async (str, regex, fn) => {
  const promises = []; // dry-run to calculate the replace values

  str.replace(regex, (...args) => {
    promises.push(fn(...args));
    return "";
  });
  const values = await Promise.all(promises);
  return str.replace(regex, () => values.shift());
};

class CountlyAnalytics {
  constructor() {
    (0, _defineProperty2.default)(this, "baseUrl", null);
    (0, _defineProperty2.default)(this, "appKey", null);
    (0, _defineProperty2.default)(this, "userKey", null);
    (0, _defineProperty2.default)(this, "anonymous", void 0);
    (0, _defineProperty2.default)(this, "appPlatform", void 0);
    (0, _defineProperty2.default)(this, "appVersion", "unknown");
    (0, _defineProperty2.default)(this, "initTime", CountlyAnalytics.getTimestamp());
    (0, _defineProperty2.default)(this, "firstPage", true);
    (0, _defineProperty2.default)(this, "heartbeatIntervalId", void 0);
    (0, _defineProperty2.default)(this, "activityIntervalId", void 0);
    (0, _defineProperty2.default)(this, "trackTime", true);
    (0, _defineProperty2.default)(this, "lastBeat", void 0);
    (0, _defineProperty2.default)(this, "storedDuration", 0);
    (0, _defineProperty2.default)(this, "lastView", void 0);
    (0, _defineProperty2.default)(this, "lastViewTime", 0);
    (0, _defineProperty2.default)(this, "lastViewStoredDuration", 0);
    (0, _defineProperty2.default)(this, "sessionStarted", false);
    (0, _defineProperty2.default)(this, "heartbeatEnabled", false);
    (0, _defineProperty2.default)(this, "inactivityCounter", 0);
    (0, _defineProperty2.default)(this, "pendingEvents", []);
    (0, _defineProperty2.default)(this, "lastMsTs", 0);
    (0, _defineProperty2.default)(this, "getOrientation", () => {
      return window.matchMedia("(orientation: landscape)").matches ? Orientation.Landscape : Orientation.Portrait;
    });
    (0, _defineProperty2.default)(this, "reportOrientation", () => {
      this.track("[CLY]_orientation", {
        mode: this.getOrientation()
      });
    });
    (0, _defineProperty2.default)(this, "endSession", () => {
      if (this.sessionStarted) {
        window.removeEventListener("resize", this.reportOrientation);
        this.reportViewDuration();
        this.request({
          end_session: 1,
          session_duration: CountlyAnalytics.getTimestamp() - this.lastBeat
        });
      }

      this.sessionStarted = false;
    });
    (0, _defineProperty2.default)(this, "onVisibilityChange", () => {
      if (document.hidden) {
        this.stopTime();
      } else {
        this.startTime();
      }
    });
    (0, _defineProperty2.default)(this, "onUserActivity", () => {
      if (this.inactivityCounter >= INACTIVITY_TIME) {
        this.startTime();
      }

      this.inactivityCounter = 0;
    });
  }

  static get instance() {
    return CountlyAnalytics.internalInstance;
  }

  get disabled() {
    return !this.baseUrl;
  }

  canEnable() {
    var _config$countly, _config$countly2;

    const config = _SdkConfig.default.get();

    return Boolean(navigator.doNotTrack !== "1" && (config === null || config === void 0 ? void 0 : (_config$countly = config.countly) === null || _config$countly === void 0 ? void 0 : _config$countly.url) && (config === null || config === void 0 ? void 0 : (_config$countly2 = config.countly) === null || _config$countly2 === void 0 ? void 0 : _config$countly2.appKey));
  }

  async changeUserKey(userKey, merge = false) {
    const oldUserKey = this.userKey;
    this.userKey = userKey;

    if (oldUserKey && merge) {
      await this.request({
        old_device_id: oldUserKey
      });
    }
  }

  async enable(anonymous = true) {
    if (!this.disabled && this.anonymous === anonymous) return;
    if (!this.canEnable()) return;

    if (!this.disabled) {
      // flush request queue as our userKey is going to change, no need to await it
      this.request();
    }

    const config = _SdkConfig.default.get();

    this.baseUrl = new URL("/i", config.countly.url);
    this.appKey = config.countly.appKey;
    this.anonymous = anonymous;

    if (anonymous) {
      await this.changeUserKey((0, _randomstring.randomString)(64));
    } else {
      await this.changeUserKey(await hashHex(_MatrixClientPeg.MatrixClientPeg.get().getUserId()), true);
    }

    const platform = _PlatformPeg.default.get();

    this.appPlatform = platform.getHumanReadableName();

    try {
      this.appVersion = await platform.getAppVersion();
    } catch (e) {
      _logger.logger.warn("Failed to get app version, using 'unknown'");
    } // start heartbeat


    this.heartbeatIntervalId = setInterval(this.heartbeat.bind(this), HEARTBEAT_INTERVAL);
    this.trackSessions();
    this.trackErrors();
  }

  async disable() {
    if (this.disabled) return;
    await this.track("Opt-Out");
    this.endSession();
    window.clearInterval(this.heartbeatIntervalId);
    window.clearTimeout(this.activityIntervalId);
    this.baseUrl = null; // remove listeners bound in trackSessions()

    window.removeEventListener("beforeunload", this.endSession);
    window.removeEventListener("unload", this.endSession);
    window.removeEventListener("visibilitychange", this.onVisibilityChange);
    window.removeEventListener("mousemove", this.onUserActivity);
    window.removeEventListener("click", this.onUserActivity);
    window.removeEventListener("keydown", this.onUserActivity);
    window.removeEventListener("scroll", this.onUserActivity);
  }

  reportFeedback(rating, comment) {
    this.track("[CLY]_star_rating", {
      rating,
      comment
    }, null, {}, true);
  }

  trackPageChange(generationTimeMs) {
    if (this.disabled) return; // TODO use generationTimeMs

    this.trackPageView();
  }

  async trackPageView() {
    this.reportViewDuration();
    await (0, _utils.sleep)(0); // XXX: we sleep here because otherwise we get the old hash and not the new one

    const viewData = await getViewData(this.anonymous);
    const page = viewData.name;
    this.lastView = page;
    this.lastViewTime = CountlyAnalytics.getTimestamp();

    const segments = _objectSpread(_objectSpread({}, viewData.meta), {}, {
      name: page,
      visit: 1,
      domain: window.location.hostname,
      view: viewData.url,
      segment: this.appPlatform,
      start: this.firstPage
    });

    if (this.firstPage) {
      this.firstPage = false;
    }

    this.track("[CLY]_view", segments);
  }

  static getTimestamp() {
    return Math.floor(new Date().getTime() / 1000);
  } // store the last ms timestamp returned
  // we do this to prevent the ts from ever decreasing in the case of system time changing


  getMsTimestamp() {
    const ts = new Date().getTime();

    if (this.lastMsTs >= ts) {
      // increment ts as to keep our data points well-ordered
      this.lastMsTs++;
    } else {
      this.lastMsTs = ts;
    }

    return this.lastMsTs;
  }

  async recordError(err, fatal = false) {
    if (this.disabled || this.anonymous) return;
    let error = "";

    if (typeof err === "object") {
      if (typeof err.stack !== "undefined") {
        error = err.stack;
      } else {
        if (typeof err.name !== "undefined") {
          error += err.name + ":";
        }

        if (typeof err.message !== "undefined") {
          error += err.message + "\n";
        }

        if (typeof err.fileName !== "undefined") {
          error += "in " + err.fileName + "\n";
        }

        if (typeof err.lineNumber !== "undefined") {
          error += "on " + err.lineNumber;
        }

        if (typeof err.columnNumber !== "undefined") {
          error += ":" + err.columnNumber;
        }
      }
    } else {
      error = err + "";
    } // sanitize the error from identifiers


    error = await strReplaceAsync(error, /([!@+#]).+?:[\w:.]+/g, async (substring, glyph) => {
      return glyph + (await hashHex(substring.substring(1)));
    });
    const metrics = this.getMetrics();
    const ob = {
      _resolution: metrics === null || metrics === void 0 ? void 0 : metrics._resolution,
      _error: error,
      _app_version: this.appVersion,
      _run: CountlyAnalytics.getTimestamp() - this.initTime,
      _nonfatal: !fatal,
      _view: this.lastView
    };

    if (typeof navigator.onLine !== "undefined") {
      ob._online = navigator.onLine;
    }

    ob._background = document.hasFocus();
    this.request({
      crash: JSON.stringify(ob)
    });
  }

  trackErrors() {
    //override global uncaught error handler
    window.onerror = (msg, url, line, col, err) => {
      if (typeof err !== "undefined") {
        this.recordError(err, false);
      } else {
        let error = "";

        if (typeof msg !== "undefined") {
          error += msg + "\n";
        }

        if (typeof url !== "undefined") {
          error += "at " + url;
        }

        if (typeof line !== "undefined") {
          error += ":" + line;
        }

        if (typeof col !== "undefined") {
          error += ":" + col;
        }

        error += "\n";

        try {
          const stack = []; // eslint-disable-next-line no-caller

          let f = arguments.callee.caller;

          while (f) {
            stack.push(f.name);
            f = f.caller;
          }

          error += stack.join("\n");
        } catch (ex) {//silent error
        }

        this.recordError(error, false);
      }
    };

    window.addEventListener('unhandledrejection', event => {
      var _event$reason;

      this.recordError(new Error(`Unhandled rejection (reason: ${((_event$reason = event.reason) === null || _event$reason === void 0 ? void 0 : _event$reason.stack) || event.reason}).`), true);
    });
  }

  heartbeat() {
    const args = {}; // extend session if needed

    if (this.sessionStarted && this.trackTime) {
      const last = CountlyAnalytics.getTimestamp();

      if (last - this.lastBeat >= SESSION_UPDATE_INTERVAL) {
        args.session_duration = last - this.lastBeat;
        this.lastBeat = last;
      }
    } // process event queue


    if (this.pendingEvents.length > 0 || args.session_duration) {
      this.request(args);
    }
  }

  async request(args = {}) {
    const request = _objectSpread(_objectSpread({
      app_key: this.appKey,
      device_id: this.userKey
    }, this.getTimeParams()), args);

    if (this.pendingEvents.length > 0) {
      const EVENT_BATCH_SIZE = 10;
      const events = this.pendingEvents.splice(0, EVENT_BATCH_SIZE);
      request.events = JSON.stringify(events);
    }

    const params = new URLSearchParams(request);

    try {
      await window.fetch(this.baseUrl.toString(), {
        method: "POST",
        mode: "no-cors",
        cache: "no-cache",
        redirect: "follow",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded"
        },
        body: params
      });
    } catch (e) {
      _logger.logger.error("Analytics error: ", e);
    }
  }

  getTimeParams() {
    const date = new Date();
    return {
      timestamp: this.getMsTimestamp(),
      hour: date.getHours(),
      dow: date.getDay()
    };
  }

  queue(args) {
    const {
      count = 1
    } = args,
          rest = (0, _objectWithoutProperties2.default)(args, _excluded);

    const ev = _objectSpread(_objectSpread(_objectSpread({}, this.getTimeParams()), rest), {}, {
      count,
      platform: this.appPlatform,
      app_version: this.appVersion
    });

    this.pendingEvents.push(ev);

    if (this.pendingEvents.length > MAX_PENDING_EVENTS) {
      this.pendingEvents.shift();
    }
  }

  startTime() {
    if (!this.trackTime) {
      this.trackTime = true;
      this.lastBeat = CountlyAnalytics.getTimestamp() - this.storedDuration;
      this.lastViewTime = CountlyAnalytics.getTimestamp() - this.lastViewStoredDuration;
      this.lastViewStoredDuration = 0;
    }
  }

  stopTime() {
    if (this.trackTime) {
      this.trackTime = false;
      this.storedDuration = CountlyAnalytics.getTimestamp() - this.lastBeat;
      this.lastViewStoredDuration = CountlyAnalytics.getTimestamp() - this.lastViewTime;
    }
  }

  getMetrics() {
    if (this.anonymous) return undefined;
    const metrics = {}; // getting app version

    metrics._app_version = this.appVersion;
    metrics._ua = navigator.userAgent; // getting resolution

    if (screen.width && screen.height) {
      metrics._resolution = `${screen.width}x${screen.height}`;
    } // getting density ratio


    if (window.devicePixelRatio) {
      metrics._density = window.devicePixelRatio;
    } // getting locale


    metrics._locale = (0, _languageHandler.getCurrentLanguage)();
    return metrics;
  }

  async beginSession(heartbeat = true) {
    if (!this.sessionStarted) {
      this.reportOrientation();
      window.addEventListener("resize", this.reportOrientation);
      this.lastBeat = CountlyAnalytics.getTimestamp();
      this.sessionStarted = true;
      this.heartbeatEnabled = heartbeat;
      const userDetails = {
        custom: {
          "home_server": _MatrixClientPeg.MatrixClientPeg.get() && _MatrixClientPeg.MatrixClientPeg.getHomeserverName(),
          // TODO hash?
          "anonymous": this.anonymous
        }
      };
      const request = {
        begin_session: 1,
        user_details: JSON.stringify(userDetails)
      };
      const metrics = this.getMetrics();

      if (metrics) {
        request.metrics = JSON.stringify(metrics);
      }

      await this.request(request);
    }
  }

  reportViewDuration() {
    if (this.lastView) {
      this.track("[CLY]_view", {
        name: this.lastView
      }, null, {
        dur: this.trackTime ? CountlyAnalytics.getTimestamp() - this.lastViewTime : this.lastViewStoredDuration
      });
      this.lastView = null;
    }
  }

  trackSessions() {
    this.beginSession();
    this.startTime();
    window.addEventListener("beforeunload", this.endSession);
    window.addEventListener("unload", this.endSession);
    window.addEventListener("visibilitychange", this.onVisibilityChange);
    window.addEventListener("mousemove", this.onUserActivity);
    window.addEventListener("click", this.onUserActivity);
    window.addEventListener("keydown", this.onUserActivity); // Using the passive option to not block the main thread
    // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners

    window.addEventListener("scroll", this.onUserActivity, {
      passive: true
    });
    this.activityIntervalId = setInterval(() => {
      this.inactivityCounter++;

      if (this.inactivityCounter >= INACTIVITY_TIME) {
        this.stopTime();
      }
    }, 60000);
  }

  trackBeginInvite(roomId) {
    this.track("begin_invite", {}, roomId);
  }

  trackSendInvite(startTime, roomId, qty) {
    this.track("send_invite", {}, roomId, {
      dur: CountlyAnalytics.getTimestamp() - startTime,
      sum: qty
    });
  }

  async trackRoomCreate(startTime, roomId) {
    if (this.disabled) return;
    let endTime = CountlyAnalytics.getTimestamp();

    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    if (!cli.getRoom(roomId)) {
      await new Promise(resolve => {
        const handler = room => {
          if (room.roomId === roomId) {
            cli.off("Room", handler);
            resolve();
          }
        };

        cli.on("Room", handler);
      });
      endTime = CountlyAnalytics.getTimestamp();
    }

    this.track("create_room", {}, roomId, {
      dur: endTime - startTime
    });
  }

  trackRoomJoin(startTime, roomId, type) {
    this.track(_actions.Action.JoinRoom, {
      type
    }, roomId, {
      dur: CountlyAnalytics.getTimestamp() - startTime
    });
  }

  async trackSendMessage(startTime, // eslint-disable-next-line camelcase
  sendPromise, roomId, isEdit, isReply, content) {
    if (this.disabled) return;

    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    const room = cli.getRoom(roomId);
    const eventId = (await sendPromise).event_id;
    let endTime = CountlyAnalytics.getTimestamp();

    if (!room.findEventById(eventId)) {
      await new Promise(resolve => {
        const handler = ev => {
          if (ev.getId() === eventId) {
            room.off("Room.localEchoUpdated", handler);
            resolve();
          }
        };

        room.on("Room.localEchoUpdated", handler);
      });
      endTime = CountlyAnalytics.getTimestamp();
    }

    this.track("send_message", {
      is_edit: isEdit,
      is_reply: isReply,
      msgtype: content.msgtype,
      format: content.format
    }, roomId, {
      dur: endTime - startTime
    });
  }

  trackStartCall(roomId, isVideo = false, isJitsi = false) {
    this.track("start_call", {
      is_video: isVideo,
      is_jitsi: isJitsi
    }, roomId);
  }

  trackJoinCall(roomId, isVideo = false, isJitsi = false) {
    this.track("join_call", {
      is_video: isVideo,
      is_jitsi: isJitsi
    }, roomId);
  }

  trackRoomDirectoryBegin() {
    this.track("room_directory");
  }

  trackRoomDirectory(startTime) {
    this.track("room_directory_done", {}, null, {
      dur: CountlyAnalytics.getTimestamp() - startTime
    });
  }

  trackRoomDirectorySearch(numResults, query) {
    this.track("room_directory_search", {
      query_length: query.length,
      query_num_words: query.split(" ").length
    }, null, {
      sum: numResults
    });
  }

  async track(key, segments, roomId, args, anonymous = false) {
    if (this.disabled && !anonymous) return;
    let segmentation = segments || {};

    if (roomId) {
      segmentation = _objectSpread(_objectSpread({
        room_id: await hashHex(roomId)
      }, getRoomStats(roomId)), segments);
    }

    this.queue(_objectSpread({
      key,
      count: 1,
      segmentation
    }, args)); // if this event can be sent anonymously and we are disabled then dispatch it right away

    if (this.disabled && anonymous) {
      await this.request({
        device_id: (0, _randomstring.randomString)(64)
      });
    }
  }

} // expose on window for easy access from the console


exports.default = CountlyAnalytics;
(0, _defineProperty2.default)(CountlyAnalytics, "internalInstance", new CountlyAnalytics());
window.mxCountlyAnalytics = CountlyAnalytics;
