// matrix-react-sdk/lib/groups.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GroupRoomType = exports.GroupMemberType = void 0;
exports.groupMemberFromApiObject = groupMemberFromApiObject;
exports.groupRoomFromApiObject = groupRoomFromApiObject;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _languageHandler = require("./languageHandler");

/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const GroupMemberType = _propTypes.default.shape({
  userId: _propTypes.default.string.isRequired,
  displayname: _propTypes.default.string,
  avatarUrl: _propTypes.default.string
});

exports.GroupMemberType = GroupMemberType;

const GroupRoomType = _propTypes.default.shape({
  displayname: _propTypes.default.string,
  name: _propTypes.default.string,
  roomId: _propTypes.default.string.isRequired,
  canonicalAlias: _propTypes.default.string,
  avatarUrl: _propTypes.default.string
});

exports.GroupRoomType = GroupRoomType;

function groupMemberFromApiObject(apiObject) {
  return {
    userId: apiObject.user_id,
    displayname: apiObject.displayname,
    avatarUrl: apiObject.avatar_url,
    isPrivileged: apiObject.is_privileged
  };
}

function groupRoomFromApiObject(apiObject) {
  return {
    displayname: apiObject.name || apiObject.canonical_alias || (0, _languageHandler._t)("Unnamed Room"),
    name: apiObject.name,
    roomId: apiObject.room_id,
    canonicalAlias: apiObject.canonical_alias,
    avatarUrl: apiObject.avatar_url,
    topic: apiObject.topic,
    numJoinedMembers: apiObject.num_joined_members,
    worldReadable: apiObject.world_readable,
    guestCanJoin: apiObject.guest_can_join,
    isPublic: apiObject.is_public !== false
  };
}
