// matrix-react-sdk/lib/Registration.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SAFE_LOCALPART_REGEX = void 0;
exports.startAnyRegistrationFlow = startAnyRegistrationFlow;

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("./dispatcher/dispatcher"));

var _Modal = _interopRequireDefault(require("./Modal"));

var _languageHandler = require("./languageHandler");

var _QuestionDialog = _interopRequireDefault(require("./components/views/dialogs/QuestionDialog"));

/*
Copyright 2018 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Utility code for registering with a homeserver
 * Note that this is currently *not* used by the actual
 * registration code.
 */
// Regex for what a "safe" or "Matrix-looking" localpart would be.
// TODO: Update as needed for https://github.com/matrix-org/matrix-doc/issues/1514
const SAFE_LOCALPART_REGEX = /^[a-z0-9=_\-./]+$/;
/**
 * Starts either the ILAG or full registration flow, depending
 * on what the HS supports
 *
 * @param {object} options
 * @param {bool} options.go_home_on_cancel
 *     If true, goes to the home page if the user cancels the action
 * @param {bool} options.go_welcome_on_cancel
 *     If true, goes to the welcome page if the user cancels the action
 * @param {bool} options.screen_after
 *     If present the screen to redirect to after a successful login or register.
 */

exports.SAFE_LOCALPART_REGEX = SAFE_LOCALPART_REGEX;

async function startAnyRegistrationFlow( // eslint-disable-next-line camelcase
options) {
  if (options === undefined) options = {};

  const modal = _Modal.default.createTrackedDialog('Registration required', '', _QuestionDialog.default, {
    hasCancelButton: true,
    quitOnly: true,
    title: (0, _languageHandler._t)("Sign In or Create Account"),
    description: (0, _languageHandler._t)("Use your account or create a new one to continue."),
    button: (0, _languageHandler._t)("Create Account"),
    extraButtons: [/*#__PURE__*/_react.default.createElement("button", {
      key: "start_login",
      onClick: () => {
        modal.close();

        _dispatcher.default.dispatch({
          action: 'start_login',
          screenAfterLogin: options.screen_after
        });
      }
    }, (0, _languageHandler._t)('Sign In'))],
    onFinished: proceed => {
      if (proceed) {
        _dispatcher.default.dispatch({
          action: 'start_registration',
          screenAfterLogin: options.screen_after
        });
      } else if (options.go_home_on_cancel) {
        _dispatcher.default.dispatch({
          action: 'view_home_page'
        });
      } else if (options.go_welcome_on_cancel) {
        _dispatcher.default.dispatch({
          action: 'view_welcome_page'
        });
      }
    }
  });
}
