// matrix-react-sdk/lib/Rooms.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getDisplayAliasForAliasSet = getDisplayAliasForAliasSet;
exports.getDisplayAliasForRoom = getDisplayAliasForRoom;
exports.guessAndSetDMRoom = guessAndSetDMRoom;
exports.looksLikeDirectMessageRoom = looksLikeDirectMessageRoom;
exports.roomContextDetailsText = roomContextDetailsText;
exports.setDMRoom = setDMRoom;

var _MatrixClientPeg = require("./MatrixClientPeg");

var _Alias = _interopRequireDefault(require("./customisations/Alias"));

var _DMRoomMap = _interopRequireDefault(require("./utils/DMRoomMap"));

var _SpaceStore = _interopRequireDefault(require("./stores/spaces/SpaceStore"));

var _languageHandler = require("./languageHandler");

/*
Copyright 2015 - 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

/**
 * Given a room object, return the alias we should use for it,
 * if any. This could be the canonical alias if one exists, otherwise
 * an alias selected arbitrarily but deterministically from the list
 * of aliases. Otherwise return null;
 *
 * @param {Object} room The room object
 * @returns {string} A display alias for the given room
 */
function getDisplayAliasForRoom(room) {
  return getDisplayAliasForAliasSet(room.getCanonicalAlias(), room.getAltAliases());
} // The various display alias getters should all feed through this one path so
// there's a single place to change the logic.


function getDisplayAliasForAliasSet(canonicalAlias, altAliases) {
  if (_Alias.default.getDisplayAliasForAliasSet) {
    return _Alias.default.getDisplayAliasForAliasSet(canonicalAlias, altAliases);
  }

  return canonicalAlias || (altAliases === null || altAliases === void 0 ? void 0 : altAliases[0]);
}

function looksLikeDirectMessageRoom(room, myUserId) {
  const myMembership = room.getMyMembership();
  const me = room.getMember(myUserId);

  if (myMembership == "join" || myMembership === "ban" || me && me.isKicked()) {
    // Used to split rooms via tags
    const tagNames = Object.keys(room.tags); // Used for 1:1 direct chats
    // Show 1:1 chats in seperate "Direct Messages" section as long as they haven't
    // been moved to a different tag section

    const totalMemberCount = room.currentState.getJoinedMemberCount() + room.currentState.getInvitedMemberCount();

    if (totalMemberCount === 2 && !tagNames.length) {
      return true;
    }
  }

  return false;
}

function guessAndSetDMRoom(room, isDirect) {
  let newTarget;

  if (isDirect) {
    const guessedUserId = guessDMRoomTargetId(room, _MatrixClientPeg.MatrixClientPeg.get().getUserId());
    newTarget = guessedUserId;
  } else {
    newTarget = null;
  }

  return setDMRoom(room.roomId, newTarget);
}
/**
 * Marks or unmarks the given room as being as a DM room.
 * @param {string} roomId The ID of the room to modify
 * @param {string} userId The user ID of the desired DM
 room target user or null to un-mark
 this room as a DM room
 * @returns {object} A promise
 */


async function setDMRoom(roomId, userId) {
  if (_MatrixClientPeg.MatrixClientPeg.get().isGuest()) return;

  const mDirectEvent = _MatrixClientPeg.MatrixClientPeg.get().getAccountData('m.direct');

  let dmRoomMap = {};
  if (mDirectEvent !== undefined) dmRoomMap = mDirectEvent.getContent(); // remove it from the lists of any others users
  // (it can only be a DM room for one person)

  for (const thisUserId of Object.keys(dmRoomMap)) {
    const roomList = dmRoomMap[thisUserId];

    if (thisUserId != userId) {
      const indexOfRoom = roomList.indexOf(roomId);

      if (indexOfRoom > -1) {
        roomList.splice(indexOfRoom, 1);
      }
    }
  } // now add it, if it's not already there


  if (userId) {
    const roomList = dmRoomMap[userId] || [];

    if (roomList.indexOf(roomId) == -1) {
      roomList.push(roomId);
    }

    dmRoomMap[userId] = roomList;
  }

  await _MatrixClientPeg.MatrixClientPeg.get().setAccountData('m.direct', dmRoomMap);
}
/**
 * Given a room, estimate which of its members is likely to
 * be the target if the room were a DM room and return that user.
 *
 * @param {Object} room Target room
 * @param {string} myUserId User ID of the current user
 * @returns {string} User ID of the user that the room is probably a DM with
 */


function guessDMRoomTargetId(room, myUserId) {
  let oldestTs;
  let oldestUser; // Pick the joined user who's been here longest (and isn't us),

  for (const user of room.getJoinedMembers()) {
    if (user.userId == myUserId) continue;

    if (oldestTs === undefined || user.events.member && user.events.member.getTs() < oldestTs) {
      oldestUser = user;
      oldestTs = user.events.member.getTs();
    }
  }

  if (oldestUser) return oldestUser.userId; // if there are no joined members other than us, use the oldest member

  for (const user of room.currentState.getMembers()) {
    if (user.userId == myUserId) continue;

    if (oldestTs === undefined || user.events.member && user.events.member.getTs() < oldestTs) {
      oldestUser = user;
      oldestTs = user.events.member.getTs();
    }
  }

  if (oldestUser === undefined) return myUserId;
  return oldestUser.userId;
}

function roomContextDetailsText(room) {
  if (room.isSpaceRoom()) return undefined;

  const dmPartner = _DMRoomMap.default.shared().getUserIdForRoomId(room.roomId);

  if (dmPartner) {
    var _room$getMember;

    return (_room$getMember = room.getMember(dmPartner)) === null || _room$getMember === void 0 ? void 0 : _room$getMember.rawDisplayName;
  }

  const [parent, ...otherParents] = _SpaceStore.default.instance.getKnownParents(room.roomId);

  if (parent) {
    return (0, _languageHandler._t)("%(spaceName)s and %(count)s others", {
      spaceName: room.client.getRoom(parent).name,
      count: otherParents.length
    });
  }

  return room.getCanonicalAlias();
}
