// matrix-react-sdk/lib/components/views/settings/AvatarSetting.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _react = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _classnames = _interopRequireDefault(require("classnames"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const AvatarSetting = ({
  avatarUrl,
  avatarAltText,
  avatarName,
  uploadAvatar,
  removeAvatar
}) => {
  const [isHovering, setIsHovering] = (0, _react.useState)(false);
  const hoveringProps = {
    onMouseEnter: () => setIsHovering(true),
    onMouseLeave: () => setIsHovering(false)
  };

  let avatarElement = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, (0, _extends2.default)({
    element: "div",
    onClick: uploadAvatar,
    className: "mx_AvatarSetting_avatarPlaceholder"
  }, hoveringProps));

  if (avatarUrl) {
    avatarElement = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, (0, _extends2.default)({
      element: "img",
      src: avatarUrl,
      alt: avatarAltText,
      "aria-label": avatarAltText,
      onClick: uploadAvatar
    }, hoveringProps));
  }

  let uploadAvatarBtn;

  if (uploadAvatar) {
    // insert an empty div to be the host for a css mask containing the upload.svg
    uploadAvatarBtn = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, (0, _extends2.default)({
      onClick: uploadAvatar,
      className: "mx_AvatarSetting_uploadButton"
    }, hoveringProps));
  }

  let removeAvatarBtn;

  if (avatarUrl && removeAvatar) {
    removeAvatarBtn = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: removeAvatar,
      kind: "link_sm"
    }, (0, _languageHandler._t)("Remove"));
  }

  const avatarClasses = (0, _classnames.default)({
    "mx_AvatarSetting_avatar": true,
    "mx_AvatarSetting_avatar_hovering": isHovering && uploadAvatar
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    className: avatarClasses
  }, avatarElement, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_AvatarSetting_hover"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_AvatarSetting_hoverBg"
  }), /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Upload"))), uploadAvatarBtn, removeAvatarBtn);
};

var _default = AvatarSetting;
exports.default = _default;
