// matrix-react-sdk/lib/autocomplete/AutocompleteProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _RoomContext = require("../contexts/RoomContext");

/*
Copyright 2016 Aviral Dasgupta
Copyright 2017 Vector Creations Ltd
Copyright 2017, 2018 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class AutocompleteProvider {
  constructor({
    commandRegex,
    forcedCommandRegex,
    renderingType
  }) {
    (0, _defineProperty2.default)(this, "commandRegex", void 0);
    (0, _defineProperty2.default)(this, "forcedCommandRegex", void 0);
    (0, _defineProperty2.default)(this, "renderingType", _RoomContext.TimelineRenderingType.Room);

    if (commandRegex) {
      if (!commandRegex.global) {
        throw new Error('commandRegex must have global flag set');
      }

      this.commandRegex = commandRegex;
    }

    if (forcedCommandRegex) {
      if (!forcedCommandRegex.global) {
        throw new Error('forcedCommandRegex must have global flag set');
      }

      this.forcedCommandRegex = forcedCommandRegex;
    }

    if (renderingType) {
      this.renderingType = renderingType;
    }
  }

  destroy() {// stub
  }
  /**
   * Of the matched commands in the query, returns the first that contains or is contained by the selection, or null.
   * @param {string} query The query string
   * @param {ISelectionRange} selection Selection to search
   * @param {boolean} force True if the user is forcing completion
   * @return {object} { command, range } where both objects fields are null if no match
   */


  getCurrentCommand(query, selection, force = false) {
    let commandRegex = this.commandRegex;

    if (force && this.shouldForceComplete()) {
      commandRegex = this.forcedCommandRegex || /\S+/g;
    }

    if (!commandRegex) {
      return null;
    }

    commandRegex.lastIndex = 0;
    let match;

    while ((match = commandRegex.exec(query)) !== null) {
      const start = match.index;
      const end = start + match[0].length;

      if (selection.start <= end && selection.end >= start) {
        return {
          command: match,
          range: {
            start,
            end
          }
        };
      }
    }

    return {
      command: null,
      range: {
        start: -1,
        end: -1
      }
    };
  }

  // Whether we should provide completions even if triggered forcefully, without a sigil.
  shouldForceComplete() {
    return false;
  }

}

exports.default = AutocompleteProvider;
