// matrix-react-sdk/lib/autocomplete/CommandProvider.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../languageHandler");

var _AutocompleteProvider = _interopRequireDefault(require("./AutocompleteProvider"));

var _QueryMatcher = _interopRequireDefault(require("./QueryMatcher"));

var _Components = require("./Components");

var _SlashCommands = require("../SlashCommands");

/*
Copyright 2016 Aviral Dasgupta
Copyright 2017 Vector Creations Ltd
Copyright 2017 New Vector Ltd
Copyright 2018 Michael Telatynski <7t3chguy@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const COMMAND_RE = /(^\/\w*)(?: .*)?/g;

class CommandProvider extends _AutocompleteProvider.default {
  constructor(room, renderingType) {
    super({
      commandRegex: COMMAND_RE,
      renderingType
    });
    (0, _defineProperty2.default)(this, "matcher", void 0);
    this.matcher = new _QueryMatcher.default(_SlashCommands.Commands, {
      keys: ['command', 'args', 'description'],
      funcs: [({
        aliases
      }) => aliases.join(" ")],
      // aliases
      context: renderingType
    });
  }

  async getCompletions(query, selection, force, limit = -1) {
    const {
      command,
      range
    } = this.getCurrentCommand(query, selection);
    if (!command) return [];
    let matches = []; // check if the full match differs from the first word (i.e. returns false if the command has args)

    if (command[0] !== command[1]) {
      // The input looks like a command with arguments, perform exact match
      const name = command[1].substr(1); // strip leading `/`

      if (_SlashCommands.CommandMap.has(name) && _SlashCommands.CommandMap.get(name).isEnabled()) {
        // some commands, namely `me` don't suit having the usage shown whilst typing their arguments
        if (_SlashCommands.CommandMap.get(name).hideCompletionAfterSpace) return [];
        matches = [_SlashCommands.CommandMap.get(name)];
      }
    } else {
      if (query === '/') {
        // If they have just entered `/` show everything
        // We exclude the limit on purpose to have a comprehensive list
        matches = _SlashCommands.Commands;
      } else {
        // otherwise fuzzy match against all of the fields
        matches = this.matcher.match(command[1], limit);
      }
    }

    return matches.filter(cmd => {
      const display = !cmd.renderingTypes || cmd.renderingTypes.includes(this.renderingType);
      return cmd.isEnabled() && display;
    }).map(result => {
      let completion = result.getCommand() + ' ';
      const usedAlias = result.aliases.find(alias => `/${alias}` === command[1]); // If the command (or an alias) is the same as the one they entered, we don't want to discard their arguments

      if (usedAlias || result.getCommand() === command[1]) {
        completion = command[0];
      }

      return {
        completion,
        type: "command",
        component: /*#__PURE__*/_react.default.createElement(_Components.TextualCompletion, {
          title: `/${usedAlias || result.command}`,
          subtitle: result.args,
          description: (0, _languageHandler._t)(result.description)
        }),
        range
      };
    });
  }

  getName() {
    return '*️⃣ ' + (0, _languageHandler._t)('Commands');
  }

  renderCompletions(completions) {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Autocomplete_Completion_container_pill",
      role: "presentation",
      "aria-label": (0, _languageHandler._t)("Command Autocomplete")
    }, completions);
  }

}

exports.default = CommandProvider;
