// matrix-react-sdk/lib/components/views/groups/GroupRoomTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var sdk = _interopRequireWildcard(require("../../../index"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _groups = require("../../../groups");

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Media = require("../../../customisations/Media");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let GroupRoomTile = (_dec = (0, _replaceableComponent.replaceableComponent)("views.groups.GroupRoomTile"), _dec(_class = (_temp = _class2 = class GroupRoomTile extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", e => {
      _dispatcher.default.dispatch({
        action: 'view_group_room',
        groupId: this.props.groupId,
        groupRoomId: this.props.groupRoom.roomId
      });
    });
  }

  render() {
    const BaseAvatar = sdk.getComponent('avatars.BaseAvatar');
    const AccessibleButton = sdk.getComponent('elements.AccessibleButton');
    const avatarUrl = this.props.groupRoom.avatarUrl ? (0, _Media.mediaFromMxc)(this.props.groupRoom.avatarUrl).getSquareThumbnailHttp(36) : null;

    const av = /*#__PURE__*/_react.default.createElement(BaseAvatar, {
      name: this.props.groupRoom.displayname,
      width: 36,
      height: 36,
      url: avatarUrl
    });

    return /*#__PURE__*/_react.default.createElement(AccessibleButton, {
      className: "mx_GroupRoomTile",
      onClick: this.onClick
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_GroupRoomTile_avatar"
    }, av), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_GroupRoomTile_name"
    }, this.props.groupRoom.displayname));
  }

}, (0, _defineProperty2.default)(_class2, "propTypes", {
  groupId: _propTypes.default.string.isRequired,
  groupRoom: _groups.GroupRoomType.isRequired
}), (0, _defineProperty2.default)(_class2, "contextType", _MatrixClientContext.default), _temp)) || _class);
var _default = GroupRoomTile;
exports.default = _default;
