// matrix-react-sdk/lib/components/views/groups/GroupPublicityToggle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var sdk = _interopRequireWildcard(require("../../../index"));

var _GroupStore = _interopRequireDefault(require("../../../stores/GroupStore"));

var _ToggleSwitch = _interopRequireDefault(require("../elements/ToggleSwitch"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let GroupPublicityToggle = (_dec = (0, _replaceableComponent.replaceableComponent)("views.groups.GroupPublicityTile"), _dec(_class = (_temp = _class2 = class GroupPublicityToggle extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      busy: false,
      ready: false,
      isGroupPublicised: false // assume false as <ToggleSwitch /> expects a boolean

    });
    (0, _defineProperty2.default)(this, "_onPublicityToggle", () => {
      this.setState({
        busy: true,
        // Optimistic early update
        isGroupPublicised: !this.state.isGroupPublicised
      });

      _GroupStore.default.setGroupPublicity(this.props.groupId, !this.state.isGroupPublicised).then(() => {
        this.setState({
          busy: false
        });
      });
    });
  }

  componentDidMount() {
    this._initGroupStore(this.props.groupId);
  }

  _initGroupStore(groupId) {
    this._groupStoreToken = _GroupStore.default.registerListener(groupId, () => {
      this.setState({
        isGroupPublicised: Boolean(_GroupStore.default.getGroupPublicity(groupId)),
        ready: _GroupStore.default.isStateReady(groupId, _GroupStore.default.STATE_KEY.Summary)
      });
    });
  }

  componentWillUnmount() {
    if (this._groupStoreToken) this._groupStoreToken.unregister();
  }

  render() {
    const GroupTile = sdk.getComponent('groups.GroupTile');
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_GroupPublicity_toggle"
    }, /*#__PURE__*/_react.default.createElement(GroupTile, {
      groupId: this.props.groupId,
      showDescription: false,
      avatarHeight: 40
    }), /*#__PURE__*/_react.default.createElement(_ToggleSwitch.default, {
      checked: this.state.isGroupPublicised,
      disabled: !this.state.ready || this.state.busy,
      onChange: this._onPublicityToggle
    }));
  }

}, (0, _defineProperty2.default)(_class2, "propTypes", {
  groupId: _propTypes.default.string.isRequired
}), _temp)) || _class);
exports.default = GroupPublicityToggle;
