// matrix-react-sdk/lib/components/views/elements/AppTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _url = _interopRequireDefault(require("url"));

var _react = _interopRequireWildcard(require("react"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _AccessibleButton = _interopRequireDefault(require("./AccessibleButton"));

var _languageHandler = require("../../../languageHandler");

var _AppPermission = _interopRequireDefault(require("./AppPermission"));

var _AppWarning = _interopRequireDefault(require("./AppWarning"));

var _Spinner = _interopRequireDefault(require("./Spinner"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _ActiveWidgetStore = _interopRequireDefault(require("../../../stores/ActiveWidgetStore"));

var _classnames = _interopRequireDefault(require("classnames"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _ContextMenu = require("../../structures/ContextMenu");

var _PersistedElement = _interopRequireWildcard(require("./PersistedElement"));

var _WidgetType = require("../../../widgets/WidgetType");

var _StopGapWidget = require("../../../stores/widgets/StopGapWidget");

var _ElementWidgetActions = require("../../../stores/widgets/ElementWidgetActions");

var _matrixWidgetApi = require("matrix-widget-api");

var _WidgetContextMenu = _interopRequireDefault(require("../context_menus/WidgetContextMenu"));

var _WidgetAvatar = _interopRequireDefault(require("../avatars/WidgetAvatar"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _WidgetLayoutStore = require("../../../stores/widgets/WidgetLayoutStore");

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let AppTile = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.AppTile"), _dec(_class = (_temp = _class2 = class AppTile extends _react.default.Component {
  // ref to the iframe (callback style)
  constructor(_props) {
    super(_props); // The key used for PersistedElement

    (0, _defineProperty2.default)(this, "contextMenuButton", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "iframe", void 0);
    (0, _defineProperty2.default)(this, "allowedWidgetsWatchRef", void 0);
    (0, _defineProperty2.default)(this, "persistKey", void 0);
    (0, _defineProperty2.default)(this, "sgWidget", void 0);
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "hasPermissionToLoad", props => {
      if (this.usingLocalWidget()) return true;
      if (!props.room) return true; // user widgets always have permissions

      const currentlyAllowedWidgets = _SettingsStore.default.getValue("allowedWidgets", props.room.roomId);

      if (currentlyAllowedWidgets[props.app.eventId] === undefined) {
        return props.userId === props.creatorUserId;
      }

      return !!currentlyAllowedWidgets[props.app.eventId];
    });
    (0, _defineProperty2.default)(this, "onAllowedWidgetsChange", () => {
      const hasPermissionToLoad = this.hasPermissionToLoad(this.props);

      if (this.state.hasPermissionToLoad && !hasPermissionToLoad) {
        // Force the widget to be non-persistent (able to be deleted/forgotten)
        _ActiveWidgetStore.default.instance.destroyPersistentWidget(this.props.app.id);

        _PersistedElement.default.destroyElement(this.persistKey);

        if (this.sgWidget) this.sgWidget.stop();
      }

      this.setState({
        hasPermissionToLoad
      });
    });
    (0, _defineProperty2.default)(this, "iframeRefChange", ref => {
      this.iframe = ref;

      if (ref) {
        try {
          if (this.sgWidget) {
            this.sgWidget.start(ref);
          }
        } catch (e) {
          _logger.logger.error("Failed to start widget", e);
        }
      } else {
        this.resetWidget(this.props);
      }
    });
    (0, _defineProperty2.default)(this, "onWidgetPreparing", () => {
      this.setState({
        loading: false
      });
    });
    (0, _defineProperty2.default)(this, "onWidgetReady", () => {
      if (_WidgetType.WidgetType.JITSI.matches(this.props.app.type)) {
        this.sgWidget.widgetApi.transport.send(_ElementWidgetActions.ElementWidgetActions.ClientReady, {});
      }
    });
    (0, _defineProperty2.default)(this, "onWidgetCapabilitiesNotified", () => {
      this.setState({
        requiresClient: this.sgWidget.widgetApi.hasCapability(_matrixWidgetApi.MatrixCapabilities.RequiresClient)
      });
    });
    (0, _defineProperty2.default)(this, "onAction", payload => {
      if (payload.widgetId === this.props.app.id) {
        switch (payload.action) {
          case 'm.sticker':
            if (this.sgWidget.widgetApi.hasCapability(_matrixWidgetApi.MatrixCapabilities.StickerSending)) {
              _dispatcher.default.dispatch({
                action: 'post_sticker_message',
                data: payload.data
              });

              _dispatcher.default.dispatch({
                action: 'stickerpicker_close'
              });
            } else {
              _logger.logger.warn('Ignoring sticker message. Invalid capability');
            }

            break;
        }
      }
    });
    (0, _defineProperty2.default)(this, "grantWidgetPermission", () => {
      const roomId = this.props.room.roomId;

      _logger.logger.info("Granting permission for widget to load: " + this.props.app.eventId);

      const current = _SettingsStore.default.getValue("allowedWidgets", roomId);

      current[this.props.app.eventId] = true;

      const level = _SettingsStore.default.firstSupportedLevel("allowedWidgets");

      _SettingsStore.default.setValue("allowedWidgets", roomId, level, current).then(() => {
        this.setState({
          hasPermissionToLoad: true
        }); // Fetch a token for the integration manager, now that we're allowed to

        this.startWidget();
      }).catch(err => {
        _logger.logger.error(err); // We don't really need to do anything about this - the user will just hit the button again.

      });
    });
    (0, _defineProperty2.default)(this, "onPopoutWidgetClick", () => {
      // Ensure Jitsi conferences are closed on pop-out, to not confuse the user to join them
      // twice from the same computer, which Jitsi can have problems with (audio echo/gain-loop).
      if (_WidgetType.WidgetType.JITSI.matches(this.props.app.type)) {
        this.endWidgetActions().then(() => {
          if (this.iframe) {
            // Reload iframe
            this.iframe.src = this.sgWidget.embedUrl;
          }
        });
      } // Using Object.assign workaround as the following opens in a new window instead of a new tab.
      // window.open(this._getPopoutUrl(), '_blank', 'noopener=yes');


      Object.assign(document.createElement('a'), {
        target: '_blank',
        href: this.sgWidget.popoutUrl,
        rel: 'noreferrer noopener'
      }).click();
    });
    (0, _defineProperty2.default)(this, "onMaxMinWidgetClick", () => {
      const targetContainer = _WidgetLayoutStore.WidgetLayoutStore.instance.isInContainer(this.props.room, this.props.app, _WidgetLayoutStore.Container.Center) ? _WidgetLayoutStore.Container.Right : _WidgetLayoutStore.Container.Center;

      _WidgetLayoutStore.WidgetLayoutStore.instance.moveToContainer(this.props.room, this.props.app, targetContainer);
    });
    (0, _defineProperty2.default)(this, "onContextMenuClick", () => {
      this.setState({
        menuDisplayed: true
      });
    });
    (0, _defineProperty2.default)(this, "closeContextMenu", () => {
      this.setState({
        menuDisplayed: false
      });
    });
    this.persistKey = (0, _PersistedElement.getPersistKey)(this.props.app.id);

    try {
      this.sgWidget = new _StopGapWidget.StopGapWidget(this.props);
      this.sgWidget.on("preparing", this.onWidgetPreparing);
      this.sgWidget.on("ready", this.onWidgetReady); // emits when the capabilites have been setup or changed

      this.sgWidget.on("capabilitiesNotified", this.onWidgetCapabilitiesNotified);
    } catch (e) {
      _logger.logger.log("Failed to construct widget", e);

      this.sgWidget = null;
    }

    this.state = this.getNewState(_props);
    this.allowedWidgetsWatchRef = _SettingsStore.default.watchSetting("allowedWidgets", null, this.onAllowedWidgetsChange);
  } // This is a function to make the impact of calling SettingsStore slightly less


  /**
   * Set initial component state when the App wUrl (widget URL) is being updated.
   * Component props *must* be passed (rather than relying on this.props).
   * @param  {Object} newProps The new properties of the component
   * @return {Object} Updated component state to be set with setState
   */
  getNewState(newProps) {
    return {
      initialising: true,
      // True while we are mangling the widget URL
      // True while the iframe content is loading
      loading: this.props.waitForIframeLoad && !_PersistedElement.default.isMounted(this.persistKey),
      // Assume that widget has permission to load if we are the user who
      // added it to the room, or if explicitly granted by the user
      hasPermissionToLoad: this.hasPermissionToLoad(newProps),
      error: null,
      menuDisplayed: false,
      widgetPageTitle: this.props.widgetPageTitle,
      // requiresClient is initially set to true. This avoids the broken state of the popout
      // button being visible (for an instance) and then disappearing when the widget is loaded.
      // requiresClient <-> hide the popout button
      requiresClient: true
    };
  }

  isMixedContent() {
    const parentContentProtocol = window.location.protocol;

    const u = _url.default.parse(this.props.app.url);

    const childContentProtocol = u.protocol;

    if (parentContentProtocol === 'https:' && childContentProtocol !== 'https:') {
      _logger.logger.warn("Refusing to load mixed-content app:", parentContentProtocol, childContentProtocol, window.location, this.props.app.url);

      return true;
    }

    return false;
  }

  componentDidMount() {
    // Only fetch IM token on mount if we're showing and have permission to load
    if (this.sgWidget && this.state.hasPermissionToLoad) {
      this.startWidget();
    } // Widget action listeners


    this.dispatcherRef = _dispatcher.default.register(this.onAction);
  }

  componentWillUnmount() {
    // Widget action listeners
    if (this.dispatcherRef) _dispatcher.default.unregister(this.dispatcherRef); // if it's not remaining on screen, get rid of the PersistedElement container

    if (!_ActiveWidgetStore.default.instance.getWidgetPersistence(this.props.app.id)) {
      _ActiveWidgetStore.default.instance.destroyPersistentWidget(this.props.app.id);

      _PersistedElement.default.destroyElement(this.persistKey);
    }

    if (this.sgWidget) {
      this.sgWidget.stop();
    }

    _SettingsStore.default.unwatchSetting(this.allowedWidgetsWatchRef);
  }

  resetWidget(newProps) {
    if (this.sgWidget) {
      this.sgWidget.stop();
    }

    try {
      this.sgWidget = new _StopGapWidget.StopGapWidget(newProps);
      this.sgWidget.on("preparing", this.onWidgetPreparing);
      this.sgWidget.on("ready", this.onWidgetReady);
      this.startWidget();
    } catch (e) {
      _logger.logger.error("Failed to construct widget", e);

      this.sgWidget = null;
    }
  }

  startWidget() {
    this.sgWidget.prepare().then(() => {
      this.setState({
        initialising: false
      });
    });
  }

  // TODO: [REACT-WARNING] Replace with appropriate lifecycle event
  // eslint-disable-next-line @typescript-eslint/naming-convention
  UNSAFE_componentWillReceiveProps(nextProps) {
    // eslint-disable-line camelcase
    if (nextProps.app.url !== this.props.app.url) {
      this.getNewState(nextProps);

      if (this.state.hasPermissionToLoad) {
        this.resetWidget(nextProps);
      }
    }

    if (nextProps.widgetPageTitle !== this.props.widgetPageTitle) {
      this.setState({
        widgetPageTitle: nextProps.widgetPageTitle
      });
    }
  }
  /**
   * Ends all widget interaction, such as cancelling calls and disabling webcams.
   * @private
   * @returns {Promise<*>} Resolves when the widget is terminated, or timeout passed.
   */


  async endWidgetActions() {
    // widget migration dev note: async to maintain signature
    // HACK: This is a really dirty way to ensure that Jitsi cleans up
    // its hold on the webcam. Without this, the widget holds a media
    // stream open, even after death. See https://github.com/vector-im/element-web/issues/7351
    if (this.iframe) {
      // In practice we could just do `+= ''` to trick the browser
      // into thinking the URL changed, however I can foresee this
      // being optimized out by a browser. Instead, we'll just point
      // the iframe at a page that is reasonably safe to use in the
      // event the iframe doesn't wink away.
      // This is relative to where the Element instance is located.
      this.iframe.src = 'about:blank';
    }

    if (_WidgetType.WidgetType.JITSI.matches(this.props.app.type)) {
      _dispatcher.default.dispatch({
        action: 'hangup_conference'
      });
    } // Delete the widget from the persisted store for good measure.


    _PersistedElement.default.destroyElement(this.persistKey);

    _ActiveWidgetStore.default.instance.destroyPersistentWidget(this.props.app.id);

    if (this.sgWidget) this.sgWidget.stop({
      forceDestroy: true
    });
  }

  formatAppTileName() {
    let appTileName = "No name";

    if (this.props.app.name && this.props.app.name.trim()) {
      appTileName = this.props.app.name.trim();
    }

    return appTileName;
  }
  /**
   * Whether we're using a local version of the widget rather than loading the
   * actual widget URL
   * @returns {bool} true If using a local version of the widget
   */


  usingLocalWidget() {
    return _WidgetType.WidgetType.JITSI.matches(this.props.app.type);
  }

  getTileTitle() {
    const name = this.formatAppTileName();

    const titleSpacer = /*#__PURE__*/_react.default.createElement("span", null, "\xA0-\xA0");

    let title = '';

    if (this.state.widgetPageTitle && this.state.widgetPageTitle !== this.formatAppTileName()) {
      title = this.state.widgetPageTitle;
    }

    return /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement(_WidgetAvatar.default, {
      app: this.props.app
    }), /*#__PURE__*/_react.default.createElement("b", null, name), /*#__PURE__*/_react.default.createElement("span", null, title ? titleSpacer : '', title));
  } // TODO replace with full screen interactions


  render() {
    let appTileBody; // Note that there is advice saying allow-scripts shouldn't be used with allow-same-origin
    // because that would allow the iframe to programmatically remove the sandbox attribute, but
    // this would only be for content hosted on the same origin as the element client: anything
    // hosted on the same origin as the client will get the same access as if you clicked
    // a link to it.

    const sandboxFlags = "allow-forms allow-popups allow-popups-to-escape-sandbox " + "allow-same-origin allow-scripts allow-presentation"; // Additional iframe feature pemissions
    // (see - https://sites.google.com/a/chromium.org/dev/Home/chromium-security/deprecating-permissions-in-cross-origin-iframes and https://wicg.github.io/feature-policy/)

    const iframeFeatures = "microphone; camera; encrypted-media; autoplay; display-capture; clipboard-write;";
    const appTileBodyClass = 'mx_AppTileBody' + (this.props.miniMode ? '_mini  ' : ' ');
    const appTileBodyStyles = {};

    if (this.props.pointerEvents) {
      appTileBodyStyles['pointer-events'] = this.props.pointerEvents;
    }

    const loadingElement = /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_AppLoading_spinner_fadeIn"
    }, /*#__PURE__*/_react.default.createElement(_Spinner.default, {
      message: (0, _languageHandler._t)("Loading...")
    }));

    if (this.sgWidget === null) {
      appTileBody = /*#__PURE__*/_react.default.createElement("div", {
        className: appTileBodyClass,
        style: appTileBodyStyles
      }, /*#__PURE__*/_react.default.createElement(_AppWarning.default, {
        errorMsg: (0, _languageHandler._t)("Error loading Widget")
      }));
    } else if (!this.state.hasPermissionToLoad) {
      // only possible for room widgets, can assert this.props.room here
      const isEncrypted = _MatrixClientPeg.MatrixClientPeg.get().isRoomEncrypted(this.props.room.roomId);

      appTileBody = /*#__PURE__*/_react.default.createElement("div", {
        className: appTileBodyClass,
        style: appTileBodyStyles
      }, /*#__PURE__*/_react.default.createElement(_AppPermission.default, {
        roomId: this.props.room.roomId,
        creatorUserId: this.props.creatorUserId,
        url: this.sgWidget.embedUrl,
        isRoomEncrypted: isEncrypted,
        onPermissionGranted: this.grantWidgetPermission
      }));
    } else if (this.state.initialising) {
      appTileBody = /*#__PURE__*/_react.default.createElement("div", {
        className: appTileBodyClass + (this.state.loading ? 'mx_AppLoading' : ''),
        style: appTileBodyStyles
      }, loadingElement);
    } else {
      if (this.isMixedContent()) {
        appTileBody = /*#__PURE__*/_react.default.createElement("div", {
          className: appTileBodyClass,
          style: appTileBodyStyles
        }, /*#__PURE__*/_react.default.createElement(_AppWarning.default, {
          errorMsg: (0, _languageHandler._t)("Error - Mixed content")
        }));
      } else {
        appTileBody = /*#__PURE__*/_react.default.createElement("div", {
          className: appTileBodyClass + (this.state.loading ? 'mx_AppLoading' : ''),
          style: appTileBodyStyles
        }, this.state.loading && loadingElement, /*#__PURE__*/_react.default.createElement("iframe", {
          allow: iframeFeatures,
          ref: this.iframeRefChange,
          src: this.sgWidget.embedUrl,
          allowFullScreen: true,
          sandbox: sandboxFlags
        }));

        if (!this.props.userWidget) {
          // All room widgets can theoretically be allowed to remain on screen, so we
          // wrap them all in a PersistedElement from the get-go. If we wait, the iframe
          // will be re-mounted later, which means the widget has to start over, which is
          // bad.
          // Also wrap the PersistedElement in a div to fix the height, otherwise
          // AppTile's border is in the wrong place
          appTileBody = /*#__PURE__*/_react.default.createElement("div", {
            className: "mx_AppTile_persistedWrapper"
          }, /*#__PURE__*/_react.default.createElement(_PersistedElement.default, {
            persistKey: this.persistKey
          }, appTileBody));
        }
      }
    }

    let appTileClasses;

    if (this.props.miniMode) {
      appTileClasses = {
        mx_AppTile_mini: true
      };
    } else if (this.props.fullWidth) {
      appTileClasses = {
        mx_AppTileFullWidth: true
      };
    } else {
      appTileClasses = {
        mx_AppTile: true
      };
    }

    appTileClasses = (0, _classnames.default)(appTileClasses);
    let contextMenu;

    if (this.state.menuDisplayed) {
      contextMenu = /*#__PURE__*/_react.default.createElement(_WidgetContextMenu.default, (0, _extends2.default)({}, (0, _ContextMenu.aboveLeftOf)(this.contextMenuButton.current.getBoundingClientRect(), null), {
        app: this.props.app,
        onFinished: this.closeContextMenu,
        showUnpin: !this.props.userWidget,
        userWidget: this.props.userWidget,
        onEditClick: this.props.onEditClick,
        onDeleteClick: this.props.onDeleteClick
      }));
    }

    let maxMinButton;

    if (_SettingsStore.default.getValue("feature_maximised_widgets")) {
      const widgetIsMaximised = _WidgetLayoutStore.WidgetLayoutStore.instance.isInContainer(this.props.room, this.props.app, _WidgetLayoutStore.Container.Center);

      maxMinButton = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        className: "mx_AppTileMenuBar_iconButton" + (widgetIsMaximised ? " mx_AppTileMenuBar_iconButton_minWidget" : " mx_AppTileMenuBar_iconButton_maxWidget"),
        title: widgetIsMaximised ? (0, _languageHandler._t)('Close') : (0, _languageHandler._t)('Maximise widget'),
        onClick: this.onMaxMinWidgetClick
      });
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: appTileClasses,
      id: this.props.app.id
    }, this.props.showMenubar && /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_AppTileMenuBar"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_AppTileMenuBarTitle",
      style: {
        pointerEvents: this.props.handleMinimisePointerEvents ? 'all' : "none"
      }
    }, this.props.showTitle && this.getTileTitle()), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_AppTileMenuBarWidgets"
    }, maxMinButton, this.props.showPopout && !this.state.requiresClient && /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_AppTileMenuBar_iconButton mx_AppTileMenuBar_iconButton_popout",
      title: (0, _languageHandler._t)('Popout widget'),
      onClick: this.onPopoutWidgetClick
    }), /*#__PURE__*/_react.default.createElement(_ContextMenu.ContextMenuButton, {
      className: "mx_AppTileMenuBar_iconButton mx_AppTileMenuBar_iconButton_menu",
      label: (0, _languageHandler._t)("Options"),
      isExpanded: this.state.menuDisplayed,
      inputRef: this.contextMenuButton,
      onClick: this.onContextMenuClick
    }))), appTileBody), contextMenu);
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  waitForIframeLoad: true,
  showMenubar: true,
  showTitle: true,
  showPopout: true,
  handleMinimisePointerEvents: false,
  userWidget: false,
  miniMode: false
}), _temp)) || _class);
exports.default = AppTile;
