// matrix-react-sdk/lib/components/views/elements/DirectorySearchBox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _AccessibleButton = _interopRequireDefault(require("./AccessibleButton"));

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let DirectorySearchBox = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.DirectorySearchBox"), _dec(_class = class DirectorySearchBox extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "input", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "onClearClick", () => {
      this.setState({
        value: ''
      });

      if (this.input.current) {
        this.input.current.focus();

        if (this.props.onClear) {
          this.props.onClear();
        }
      }
    });
    (0, _defineProperty2.default)(this, "onChange", ev => {
      if (!this.input.current) return;
      this.setState({
        value: ev.target.value
      });

      if (this.props.onChange) {
        this.props.onChange(ev.target.value);
      }
    });
    (0, _defineProperty2.default)(this, "onKeyUp", ev => {
      if (ev.key == 'Enter' && this.props.showJoinButton) {
        if (this.props.onJoinClick) {
          this.props.onJoinClick(this.state.value);
        }
      }
    });
    (0, _defineProperty2.default)(this, "onJoinButtonClick", () => {
      if (this.props.onJoinClick) {
        this.props.onJoinClick(this.state.value);
      }
    });
    this.state = {
      value: this.props.initialText || ''
    };
  }

  render() {
    const searchboxClasses = {
      mx_DirectorySearchBox: true
    };
    searchboxClasses[this.props.className] = true;
    let joinButton;

    if (this.props.showJoinButton) {
      joinButton = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        className: "mx_DirectorySearchBox_joinButton",
        onClick: this.onJoinButtonClick
      }, (0, _languageHandler._t)("Join"));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: `mx_DirectorySearchBox ${this.props.className} mx_textinput`
    }, /*#__PURE__*/_react.default.createElement("input", {
      type: "text",
      name: "dirsearch",
      value: this.state.value,
      className: "mx_textinput_icon mx_textinput_search",
      ref: this.input,
      onChange: this.onChange,
      onKeyUp: this.onKeyUp,
      placeholder: this.props.placeholder,
      autoFocus: true
    }), joinButton, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_DirectorySearchBox_clear",
      onClick: this.onClearClick
    }));
  }

}) || _class);
exports.default = DirectorySearchBox;
