// matrix-react-sdk/lib/components/views/elements/EventTilePreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _event = require("matrix-js-sdk/src/models/event");

var Avatar = _interopRequireWildcard(require("../../../Avatar"));

var _EventTile = _interopRequireDefault(require("../rooms/EventTile"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _Layout = require("../../../settings/enums/Layout");

var _UIFeature = require("../../../settings/UIFeature");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Spinner = _interopRequireDefault(require("./Spinner"));

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const AVATAR_SIZE = 32;
let EventTilePreview = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.EventTilePreview"), _dec(_class = class EventTilePreview extends _react.default.Component {
  constructor(props) {
    super(props);
    this.state = {
      message: props.message
    };
  }

  fakeEvent({
    message
  }) {
    // Fake it till we make it

    /* eslint-disable quote-props */
    const rawEvent = {
      type: "m.room.message",
      sender: this.props.userId,
      content: {
        "m.new_content": {
          msgtype: "m.text",
          body: message,
          displayname: this.props.displayName,
          avatar_url: this.props.avatarUrl
        },
        msgtype: "m.text",
        body: message,
        displayname: this.props.displayName,
        avatar_url: this.props.avatarUrl
      },
      unsigned: {
        age: 97
      },
      event_id: "$9999999999999999999999999999999999999999999",
      room_id: "!999999999999999999:example.org"
    };
    const event = new _event.MatrixEvent(rawEvent);
    /* eslint-enable quote-props */
    // Fake it more

    event.sender = {
      name: this.props.displayName || this.props.userId,
      rawDisplayName: this.props.displayName,
      userId: this.props.userId,
      getAvatarUrl: (..._) => {
        return Avatar.avatarUrlForUser({
          avatarUrl: this.props.avatarUrl
        }, AVATAR_SIZE, AVATAR_SIZE, "crop");
      },
      getMxcAvatarUrl: () => this.props.avatarUrl
    };
    return event;
  }

  render() {
    const className = (0, _classnames.default)(this.props.className, {
      "mx_IRCLayout": this.props.layout == _Layout.Layout.IRC,
      "mx_GroupLayout": this.props.layout == _Layout.Layout.Group,
      "mx_EventTilePreview_loader": !this.props.userId
    });
    if (!this.props.userId) return /*#__PURE__*/_react.default.createElement("div", {
      className: className
    }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
    const event = this.fakeEvent(this.state);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className
    }, /*#__PURE__*/_react.default.createElement(_EventTile.default, {
      mxEvent: event,
      layout: this.props.layout,
      enableFlair: _SettingsStore.default.getValue(_UIFeature.UIFeature.Flair),
      as: "div"
    }));
  }

}) || _class);
exports.default = EventTilePreview;
