// matrix-react-sdk/lib/components/views/elements/InfoTooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.InfoTooltipKind = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _Tooltip = _interopRequireWildcard(require("./Tooltip"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let InfoTooltipKind;
exports.InfoTooltipKind = InfoTooltipKind;

(function (InfoTooltipKind) {
  InfoTooltipKind["Info"] = "info";
  InfoTooltipKind["Warning"] = "warning";
})(InfoTooltipKind || (exports.InfoTooltipKind = InfoTooltipKind = {}));

let InfoTooltip = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.InfoTooltip"), _dec(_class = class InfoTooltip extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onMouseOver", () => {
      this.setState({
        hover: true
      });
    });
    (0, _defineProperty2.default)(this, "onMouseLeave", () => {
      this.setState({
        hover: false
      });
    });
    this.state = {
      hover: false
    };
  }

  render() {
    const {
      tooltip,
      children,
      tooltipClassName,
      className,
      kind
    } = this.props;
    const title = (0, _languageHandler._t)("Information");
    const iconClassName = kind !== InfoTooltipKind.Warning ? "mx_InfoTooltip_icon_info" : "mx_InfoTooltip_icon_warning"; // Tooltip are forced on the right for a more natural feel to them on info icons

    const tip = this.state.hover ? /*#__PURE__*/_react.default.createElement(_Tooltip.default, {
      className: "mx_InfoTooltip_container",
      tooltipClassName: (0, _classnames.default)("mx_InfoTooltip_tooltip", tooltipClassName),
      label: tooltip || title,
      alignment: _Tooltip.Alignment.Right
    }) : /*#__PURE__*/_react.default.createElement("div", null);
    return /*#__PURE__*/_react.default.createElement("div", {
      onMouseOver: this.onMouseOver,
      onMouseLeave: this.onMouseLeave,
      className: (0, _classnames.default)("mx_InfoTooltip", className)
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: (0, _classnames.default)("mx_InfoTooltip_icon", iconClassName),
      "aria-label": title
    }), children, tip);
  }

}) || _class);
exports.default = InfoTooltip;
