// matrix-react-sdk/lib/components/views/elements/InlineSpinner.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

let InlineSpinner = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.InlineSpinner"), _dec(_class = (_temp = _class2 = class InlineSpinner extends _react.default.PureComponent {
  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_InlineSpinner"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_InlineSpinner_icon mx_Spinner_icon",
      style: {
        width: this.props.w,
        height: this.props.h
      },
      "aria-label": (0, _languageHandler._t)("Loading...")
    }, this.props.children));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  w: 16,
  h: 16
}), _temp)) || _class);
exports.default = InlineSpinner;
