// matrix-react-sdk/lib/components/views/elements/IRCTimelineProfileResizer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _Draggable = _interopRequireDefault(require("./Draggable"));

var _SettingLevel = require("../../../settings/SettingLevel");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let IRCTimelineProfileResizer = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.IRCTimelineProfileResizer"), _dec(_class = class IRCTimelineProfileResizer extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "dragFunc", (location, event) => {
      const offset = event.clientX - location.currentX;
      const newWidth = this.state.width + offset; // If we're trying to go smaller than min width, don't.

      if (newWidth < this.props.minWidth) {
        return location;
      }

      if (newWidth > this.props.maxWidth) {
        return location;
      }

      this.setState({
        width: newWidth
      });
      this.updateCSSWidth.bind(this)(newWidth);
      return {
        currentX: event.clientX,
        currentY: location.currentY
      };
    });
    this.state = {
      width: _SettingsStore.default.getValue("ircDisplayNameWidth", this.props.roomId),
      IRCLayoutRoot: null
    };
  }

  componentDidMount() {
    this.setState({
      IRCLayoutRoot: document.querySelector(".mx_IRCLayout")
    }, () => this.updateCSSWidth(this.state.width));
  }

  updateCSSWidth(newWidth) {
    this.state.IRCLayoutRoot.style.setProperty("--name-width", newWidth + "px");
  }

  onMoueUp(event) {
    if (this.props.roomId) {
      _SettingsStore.default.setValue("ircDisplayNameWidth", this.props.roomId, _SettingLevel.SettingLevel.ROOM_DEVICE, this.state.width);
    }
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_Draggable.default, {
      className: "mx_ProfileResizer",
      dragFunc: this.dragFunc.bind(this),
      onMouseUp: this.onMoueUp.bind(this)
    });
  }

}) || _class);
exports.default = IRCTimelineProfileResizer;
