// matrix-react-sdk/lib/components/views/elements/LabelledToggleSwitch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _ToggleSwitch = _interopRequireDefault(require("./ToggleSwitch"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let LabelledToggleSwitch = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.LabelledToggleSwitch"), _dec(_class = class LabelledToggleSwitch extends _react.default.PureComponent {
  render() {
    // This is a minimal version of a SettingsFlag
    let firstPart = /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsFlag_label"
    }, this.props.label);

    let secondPart = /*#__PURE__*/_react.default.createElement(_ToggleSwitch.default, {
      checked: this.props.value,
      disabled: this.props.disabled,
      onChange: this.props.onChange,
      "aria-label": this.props.label
    });

    if (this.props.toggleInFront) {
      const temp = firstPart;
      firstPart = secondPart;
      secondPart = temp;
    }

    const classes = `mx_SettingsFlag ${this.props.className || ""}`;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: classes
    }, firstPart, secondPart);
  }

}) || _class);
exports.default = LabelledToggleSwitch;
