// matrix-react-sdk/lib/components/views/elements/RoomAliasField.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _Validation = _interopRequireDefault(require("./Validation"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Field = _interopRequireDefault(require("./Field"));

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let RoomAliasField = ( // Controlled form component wrapping Field for inputting a room alias scoped to a given domain
_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.RoomAliasField"), _dec(_class = class RoomAliasField extends _react.default.PureComponent {
  constructor(props, context) {
    super(props, context);
    (0, _defineProperty2.default)(this, "fieldRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "onChange", ev => {
      if (this.props.onChange) {
        this.props.onChange(this.asFullAlias(ev.target.value));
      }
    });
    (0, _defineProperty2.default)(this, "onValidate", async fieldState => {
      const result = await this.validationRules(fieldState);
      this.setState({
        isValid: result.valid
      });
      return result;
    });
    (0, _defineProperty2.default)(this, "validationRules", (0, _Validation.default)({
      rules: [{
        key: "safeLocalpart",
        test: async ({
          value
        }) => {
          if (!value) {
            return true;
          }

          const fullAlias = this.asFullAlias(value); // XXX: FIXME https://github.com/matrix-org/matrix-doc/issues/668

          return !value.includes("#") && !value.includes(":") && !value.includes(",") && encodeURI(fullAlias) === fullAlias;
        },
        invalid: () => (0, _languageHandler._t)("Some characters not allowed")
      }, {
        key: "required",
        test: async ({
          value,
          allowEmpty
        }) => allowEmpty || !!value,
        invalid: () => (0, _languageHandler._t)("Please provide an address")
      }, {
        key: "taken",
        final: true,
        test: async ({
          value
        }) => {
          if (!value) {
            return true;
          }

          const client = _MatrixClientPeg.MatrixClientPeg.get();

          try {
            await client.getRoomIdForAlias(this.asFullAlias(value)); // we got a room id, so the alias is taken

            return false;
          } catch (err) {
            // any server error code will do,
            // either it M_NOT_FOUND or the alias is invalid somehow,
            // in which case we don't want to show the invalid message
            return !!err.errcode;
          }
        },
        valid: () => (0, _languageHandler._t)("This address is available to use"),
        invalid: () => (0, _languageHandler._t)("This address is already in use")
      }]
    }));
    this.state = {
      isValid: true
    };
  }

  asFullAlias(localpart) {
    return `#${localpart}:${this.props.domain}`;
  }

  render() {
    const poundSign = /*#__PURE__*/_react.default.createElement("span", null, "#");

    const aliasPostfix = ":" + this.props.domain;

    const domain = /*#__PURE__*/_react.default.createElement("span", {
      title: aliasPostfix
    }, aliasPostfix);

    const maxlength = 255 - this.props.domain.length - 2; // 2 for # and :

    return /*#__PURE__*/_react.default.createElement(_Field.default, {
      label: this.props.label || (0, _languageHandler._t)("Room address"),
      className: "mx_RoomAliasField",
      prefixComponent: poundSign,
      postfixComponent: domain,
      ref: this.fieldRef,
      onValidate: this.onValidate,
      placeholder: this.props.placeholder || (0, _languageHandler._t)("e.g. my-room"),
      onChange: this.onChange,
      value: this.props.value.substring(1, this.props.value.length - this.props.domain.length - 1),
      maxLength: maxlength,
      disabled: this.props.disabled,
      autoComplete: "off",
      onKeyDown: this.props.onKeyDown
    });
  }

  get isValid() {
    return this.state.isValid;
  }

  validate(options) {
    var _this$fieldRef$curren;

    return (_this$fieldRef$curren = this.fieldRef.current) === null || _this$fieldRef$curren === void 0 ? void 0 : _this$fieldRef$curren.validate(options);
  }

  focus() {
    var _this$fieldRef$curren2;

    (_this$fieldRef$curren2 = this.fieldRef.current) === null || _this$fieldRef$curren2 === void 0 ? void 0 : _this$fieldRef$curren2.focus();
  }

}) || _class);
exports.default = RoomAliasField;
