// matrix-react-sdk/lib/components/views/elements/Spoiler.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let Spoiler = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.Spoiler"), _dec(_class = class Spoiler extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "toggleVisible", e => {
      if (!this.state.visible) {
        // we are un-blurring, we don't want this click to propagate to potential child pills
        e.preventDefault();
        e.stopPropagation();
      }

      this.setState({
        visible: !this.state.visible
      });
    });
    this.state = {
      visible: false
    };
  }

  render() {
    const reason = this.props.reason ? /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_EventTile_spoiler_reason"
    }, "(" + this.props.reason + ")") : null; // react doesn't allow appending a DOM node as child.
    // as such, we pass the this.props.contentHtml instead and then set the raw
    // HTML content. This is secure as the contents have already been parsed previously

    return /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_EventTile_spoiler" + (this.state.visible ? " visible" : ""),
      onClick: this.toggleVisible
    }, reason, "\xA0", /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_EventTile_spoiler_content",
      dangerouslySetInnerHTML: {
        __html: this.props.contentHtml
      }
    }));
  }

}) || _class);
exports.default = Spoiler;
