// matrix-react-sdk/lib/components/views/elements/StyledCheckbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CheckboxStyle = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _randomstring = require("matrix-js-sdk/src/randomstring");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _classnames = _interopRequireDefault(require("classnames"));

const _excluded = ["children", "className", "kind"];

var _dec, _class, _class2, _temp;

let CheckboxStyle;
exports.CheckboxStyle = CheckboxStyle;

(function (CheckboxStyle) {
  CheckboxStyle["Solid"] = "solid";
  CheckboxStyle["Outline"] = "outline";
})(CheckboxStyle || (exports.CheckboxStyle = CheckboxStyle = {}));

let StyledCheckbox = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.StyledCheckbox"), _dec(_class = (_temp = _class2 = class StyledCheckbox extends _react.default.PureComponent {
  constructor(props) {
    super(props); // 56^10 so unlikely chance of collision.

    (0, _defineProperty2.default)(this, "id", void 0);
    this.id = "checkbox_" + (0, _randomstring.randomString)(10);
  }

  render() {
    /* eslint @typescript-eslint/no-unused-vars: ["error", { "ignoreRestSiblings": true }] */
    const _this$props = this.props,
          {
      children,
      className,
      kind = CheckboxStyle.Solid
    } = _this$props,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    const newClassName = (0, _classnames.default)("mx_Checkbox", className, {
      "mx_Checkbox_hasKind": kind,
      [`mx_Checkbox_kind_${kind}`]: kind
    });
    return /*#__PURE__*/_react.default.createElement("span", {
      className: newClassName
    }, /*#__PURE__*/_react.default.createElement("input", (0, _extends2.default)({
      id: this.id
    }, otherProps, {
      type: "checkbox"
    })), /*#__PURE__*/_react.default.createElement("label", {
      htmlFor: this.id
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Checkbox_background"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Checkbox_checkmark"
    })), /*#__PURE__*/_react.default.createElement("div", null, this.props.children)));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  className: ""
}), _temp)) || _class);
exports.default = StyledCheckbox;
