// matrix-react-sdk/lib/components/views/elements/TagComposer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Field = _interopRequireDefault(require("./Field"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("./AccessibleButton"));

var _dec, _class;

let TagComposer = (
/**
 * A simple, controlled, composer for entering string tags. Contains a simple
 * input, add button, and per-tag remove button.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.TagComposer"), _dec(_class = class TagComposer extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onInputChange", ev => {
      this.setState({
        newTag: ev.target.value
      });
    });
    (0, _defineProperty2.default)(this, "onAdd", ev => {
      ev.preventDefault();
      if (!this.state.newTag) return;
      this.props.onAdd(this.state.newTag);
      this.setState({
        newTag: ""
      });
    });
    this.state = {
      newTag: ""
    };
  }

  onRemove(tag) {
    // We probably don't need to proxy this, but for
    // sanity of `this` we'll do so anyways.
    this.props.onRemove(tag);
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_TagComposer"
    }, /*#__PURE__*/_react.default.createElement("form", {
      className: "mx_TagComposer_input",
      onSubmit: this.onAdd
    }, /*#__PURE__*/_react.default.createElement(_Field.default, {
      value: this.state.newTag,
      onChange: this.onInputChange,
      label: this.props.label || (0, _languageHandler._t)("Keyword"),
      placeholder: this.props.placeholder || (0, _languageHandler._t)("New keyword"),
      disabled: this.props.disabled,
      autoComplete: "off"
    }), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: this.onAdd,
      kind: "primary",
      disabled: this.props.disabled
    }, (0, _languageHandler._t)("Add"))), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_TagComposer_tags"
    }, this.props.tags.map((t, i) => /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_TagComposer_tag",
      key: i
    }, /*#__PURE__*/_react.default.createElement("span", null, t), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: this.onRemove.bind(this, t),
      disabled: this.props.disabled
    })))));
  }

}) || _class);
exports.default = TagComposer;
