// matrix-react-sdk/lib/components/views/elements/TextWithTooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

const _excluded = ["class", "children", "tooltip", "tooltipClass", "tooltipProps"];

var _dec, _class;

let TextWithTooltip = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.TextWithTooltip"), _dec(_class = class TextWithTooltip extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onMouseOver", () => {
      this.setState({
        hover: true
      });
    });
    (0, _defineProperty2.default)(this, "onMouseLeave", () => {
      this.setState({
        hover: false
      });
    });
    this.state = {
      hover: false
    };
  }

  render() {
    const _this$props = this.props,
          {
      class: className,
      children,
      tooltip,
      tooltipClass,
      tooltipProps
    } = _this$props,
          props = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    return /*#__PURE__*/_react.default.createElement("span", (0, _extends2.default)({}, props, {
      onMouseOver: this.onMouseOver,
      onMouseLeave: this.onMouseLeave,
      onClick: this.props.onClick,
      className: className
    }), children, this.state.hover && /*#__PURE__*/_react.default.createElement(_Tooltip.default, (0, _extends2.default)({}, tooltipProps, {
      label: tooltip,
      tooltipClassName: tooltipClass,
      className: "mx_TextWithTooltip_tooltip"
    })));
  }

}) || _class);
exports.default = TextWithTooltip;
