// matrix-react-sdk/lib/components/views/elements/Tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Alignment = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _reactDom = _interopRequireDefault(require("react-dom"));

var _classnames = _interopRequireDefault(require("classnames"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _UIStore = _interopRequireDefault(require("../../../stores/UIStore"));

var _dec, _class, _class2, _temp;

const MIN_TOOLTIP_HEIGHT = 25;
let Alignment;
exports.Alignment = Alignment;

(function (Alignment) {
  Alignment[Alignment["Natural"] = 0] = "Natural";
  Alignment[Alignment["Left"] = 1] = "Left";
  Alignment[Alignment["Right"] = 2] = "Right";
  Alignment[Alignment["Top"] = 3] = "Top";
  Alignment[Alignment["Bottom"] = 4] = "Bottom";
})(Alignment || (exports.Alignment = Alignment = {}));

let Tooltip = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.Tooltip"), _dec(_class = (_temp = _class2 = class Tooltip extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "tooltipContainer", void 0);
    (0, _defineProperty2.default)(this, "tooltip", void 0);
    (0, _defineProperty2.default)(this, "parent", void 0);
    (0, _defineProperty2.default)(this, "renderTooltip", () => {
      // Add the parent's position to the tooltips, so it's correctly
      // positioned, also taking into account any window zoom
      // NOTE: The additional 6 pixels for the left position, is to take account of the
      // tooltips chevron
      const style = this.updatePosition({}); // Hide the entire container when not visible. This prevents flashing of the tooltip
      // if it is not meant to be visible on first mount.

      style.display = this.props.visible ? "block" : "none";
      const tooltipClasses = (0, _classnames.default)("mx_Tooltip", this.props.tooltipClassName, {
        "mx_Tooltip_visible": this.props.visible,
        "mx_Tooltip_invisible": !this.props.visible
      });

      const tooltip = /*#__PURE__*/_react.default.createElement("div", {
        className: tooltipClasses,
        style: style
      }, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_Tooltip_chevron"
      }), this.props.label); // Render the tooltip manually, as we wish it not to be rendered within the parent


      this.tooltip = _reactDom.default.render(tooltip, this.tooltipContainer);
    });
  }

  // Create a wrapper for the tooltip outside the parent and attach it to the body element
  componentDidMount() {
    this.tooltipContainer = document.createElement("div");
    this.tooltipContainer.className = "mx_Tooltip_wrapper";
    document.body.appendChild(this.tooltipContainer);
    window.addEventListener('scroll', this.renderTooltip, {
      passive: true,
      capture: true
    });
    this.parent = _reactDom.default.findDOMNode(this).parentNode;
    this.renderTooltip();
  }

  componentDidUpdate() {
    this.renderTooltip();
  } // Remove the wrapper element, as the tooltip has finished using it


  componentWillUnmount() {
    _reactDom.default.unmountComponentAtNode(this.tooltipContainer);

    document.body.removeChild(this.tooltipContainer);
    window.removeEventListener('scroll', this.renderTooltip, {
      capture: true
    });
  }

  updatePosition(style) {
    const parentBox = this.parent.getBoundingClientRect();
    let offset = 0;

    if (parentBox.height > MIN_TOOLTIP_HEIGHT) {
      offset = Math.floor((parentBox.height - MIN_TOOLTIP_HEIGHT) / 2);
    } else {
      // The tooltip is larger than the parent height: figure out what offset
      // we need so that we're still centered.
      offset = Math.floor(parentBox.height - MIN_TOOLTIP_HEIGHT);
    }

    const width = _UIStore.default.instance.windowWidth;
    const baseTop = parentBox.top - 2 + this.props.yOffset + window.pageYOffset;
    const top = baseTop + offset;
    const right = width - parentBox.right - window.pageXOffset - 16;
    const left = parentBox.right + window.pageXOffset + 6;
    const horizontalCenter = parentBox.right - window.pageXOffset - parentBox.width / 2;

    switch (this.props.alignment) {
      case Alignment.Natural:
        if (parentBox.right > width / 2) {
          style.right = right;
          style.top = top;
          break;
        }

      // fall through to Right

      case Alignment.Right:
        style.left = left;
        style.top = top;
        break;

      case Alignment.Left:
        style.right = right;
        style.top = top;
        break;

      case Alignment.Top:
        style.top = baseTop - 16;
        style.left = horizontalCenter;
        break;

      case Alignment.Bottom:
        style.top = baseTop + parentBox.height;
        style.left = horizontalCenter;
        break;
    }

    return style;
  }

  render() {
    // Render a placeholder
    return /*#__PURE__*/_react.default.createElement("div", {
      className: this.props.className
    });
  }

}, (0, _defineProperty2.default)(_class2, "Alignment", Alignment), (0, _defineProperty2.default)(_class2, "defaultProps", {
  visible: true,
  yOffset: 0,
  alignment: Alignment.Natural
}), _temp)) || _class);
exports.default = Tooltip;
