// matrix-react-sdk/lib/components/views/elements/TruncatedList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

let TruncatedList = (_dec = (0, _replaceableComponent.replaceableComponent)("views.elements.TruncatedList"), _dec(_class = (_temp = _class2 = class TruncatedList extends _react.default.Component {
  getChildren(start, end) {
    if (this.props.getChildren && this.props.getChildCount) {
      return this.props.getChildren(start, end);
    } else {
      // XXX: I'm not sure why anything would pass null into this, it seems
      // like a bizzare case to handle, but I'm preserving the behaviour.
      // (see commit 38d5c7d5c5d5a34dc16ef5d46278315f5c57f542)
      return _react.default.Children.toArray(this.props.children).filter(c => {
        return c != null;
      }).slice(start, end);
    }
  }

  getChildCount() {
    if (this.props.getChildren && this.props.getChildCount) {
      return this.props.getChildCount();
    } else {
      return _react.default.Children.toArray(this.props.children).filter(c => {
        return c != null;
      }).length;
    }
  }

  render() {
    let overflowNode = null;
    const totalChildren = this.getChildCount();
    let upperBound = totalChildren;

    if (this.props.truncateAt >= 0) {
      const overflowCount = totalChildren - this.props.truncateAt;

      if (overflowCount > 1) {
        overflowNode = this.props.createOverflowElement(overflowCount, totalChildren);
        upperBound = this.props.truncateAt;
      }
    }

    const childNodes = this.getChildren(0, upperBound);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: this.props.className
    }, childNodes, overflowNode);
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  truncateAt: 2,

  createOverflowElement(overflowCount, totalCount) {
    return /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("And %(count)s more...", {
      count: overflowCount
    }));
  }

}), _temp)) || _class);
exports.default = TruncatedList;
