// matrix-react-sdk/lib/components/views/settings/ChangeDisplayName.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _EditableTextContainer = _interopRequireDefault(require("../elements/EditableTextContainer"));

var _dec, _class;

let ChangeDisplayName = (_dec = (0, _replaceableComponent.replaceableComponent)("views.settings.ChangeDisplayName"), _dec(_class = class ChangeDisplayName extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "getDisplayName", async () => {
      const cli = _MatrixClientPeg.MatrixClientPeg.get();

      try {
        const res = await cli.getProfileInfo(cli.getUserId());
        return res.displayname;
      } catch (e) {
        throw new Error("Failed to fetch display name");
      }
    });
    (0, _defineProperty2.default)(this, "changeDisplayName", newDisplayname => {
      const cli = _MatrixClientPeg.MatrixClientPeg.get();

      return cli.setDisplayName(newDisplayname).catch(function () {
        throw new Error("Failed to set display name");
      });
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_EditableTextContainer.default, {
      getInitialValue: this.getDisplayName,
      placeholder: (0, _languageHandler._t)("No display name"),
      blurToSubmit: true,
      onSubmit: this.changeDisplayName
    });
  }

}) || _class);
exports.default = ChangeDisplayName;
