// matrix-react-sdk/lib/components/views/settings/ChangePassword.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Field = _interopRequireDefault(require("../elements/Field"));

var _react = _interopRequireDefault(require("react"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _Validation = _interopRequireDefault(require("../elements/Validation"));

var _languageHandler = require("../../../languageHandler");

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _PassphraseField = _interopRequireDefault(require("../auth/PassphraseField"));

var _CountlyAnalytics = _interopRequireDefault(require("../../../CountlyAnalytics"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _RegistrationForm = require("../auth/RegistrationForm");

var _SetEmailDialog = _interopRequireDefault(require("../dialogs/SetEmailDialog"));

var _QuestionDialog = _interopRequireDefault(require("../dialogs/QuestionDialog"));

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const FIELD_OLD_PASSWORD = 'field_old_password';
const FIELD_NEW_PASSWORD = 'field_new_password';
const FIELD_NEW_PASSWORD_CONFIRM = 'field_new_password_confirm';
var Phase;

(function (Phase) {
  Phase["Edit"] = "edit";
  Phase["Uploading"] = "uploading";
  Phase["Error"] = "error";
})(Phase || (Phase = {}));

let ChangePassword = (_dec = (0, _replaceableComponent.replaceableComponent)("views.settings.ChangePassword"), _dec(_class = (_temp = _class2 = class ChangePassword extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onExportE2eKeysClicked", () => {
      _Modal.default.createTrackedDialogAsync('Export E2E Keys', 'Change Password', Promise.resolve().then(() => _interopRequireWildcard(require('../../../async-components/views/dialogs/security/ExportE2eKeysDialog'))), {
        matrixClient: _MatrixClientPeg.MatrixClientPeg.get()
      });
    });
    (0, _defineProperty2.default)(this, "onChangeOldPassword", ev => {
      this.setState({
        oldPassword: ev.target.value
      });
    });
    (0, _defineProperty2.default)(this, "onOldPasswordValidate", async fieldState => {
      const result = await this.validateOldPasswordRules(fieldState);
      this.markFieldValid(FIELD_OLD_PASSWORD, result.valid);
      return result;
    });
    (0, _defineProperty2.default)(this, "validateOldPasswordRules", (0, _Validation.default)({
      rules: [{
        key: "required",
        test: ({
          value,
          allowEmpty
        }) => allowEmpty || !!value,
        invalid: () => (0, _languageHandler._t)("Passwords can't be empty")
      }]
    }));
    (0, _defineProperty2.default)(this, "onChangeNewPassword", ev => {
      this.setState({
        newPassword: ev.target.value
      });
    });
    (0, _defineProperty2.default)(this, "onNewPasswordValidate", result => {
      this.markFieldValid(FIELD_NEW_PASSWORD, result.valid);
    });
    (0, _defineProperty2.default)(this, "onChangeNewPasswordConfirm", ev => {
      this.setState({
        newPasswordConfirm: ev.target.value
      });
    });
    (0, _defineProperty2.default)(this, "onNewPasswordConfirmValidate", async fieldState => {
      const result = await this.validatePasswordConfirmRules(fieldState);
      this.markFieldValid(FIELD_NEW_PASSWORD_CONFIRM, result.valid);
      return result;
    });
    (0, _defineProperty2.default)(this, "validatePasswordConfirmRules", (0, _Validation.default)({
      rules: [{
        key: "required",
        test: ({
          value,
          allowEmpty
        }) => allowEmpty || !!value,
        invalid: () => (0, _languageHandler._t)("Confirm password")
      }, {
        key: "match",

        test({
          value
        }) {
          return !value || value === this.state.newPassword;
        },

        invalid: () => (0, _languageHandler._t)("Passwords don't match")
      }]
    }));
    (0, _defineProperty2.default)(this, "onClickChange", async ev => {
      ev.preventDefault();
      const allFieldsValid = await this.verifyFieldsBeforeSubmit();

      if (!allFieldsValid) {
        _CountlyAnalytics.default.instance.track("onboarding_registration_submit_failed");

        return;
      }

      const oldPassword = this.state.oldPassword;
      const newPassword = this.state.newPassword;
      const confirmPassword = this.state.newPasswordConfirm;
      const err = this.checkPassword(oldPassword, newPassword, confirmPassword);

      if (err) {
        this.props.onError(err);
      } else {
        this.onChangePassword(oldPassword, newPassword);
      }
    });
    this.state = {
      fieldValid: {},
      phase: Phase.Edit,
      oldPassword: "",
      newPassword: "",
      newPasswordConfirm: ""
    };
  }

  onChangePassword(oldPassword, newPassword) {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    if (!this.props.confirm) {
      this.changePassword(cli, oldPassword, newPassword);
      return;
    }

    _Modal.default.createTrackedDialog('Change Password', '', _QuestionDialog.default, {
      title: (0, _languageHandler._t)("Warning!"),
      description: /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)('Changing password will currently reset any end-to-end encryption keys on all sessions, ' + 'making encrypted chat history unreadable, unless you first export your room keys ' + 'and re-import them afterwards. ' + 'In future this will be improved.'), ' ', /*#__PURE__*/_react.default.createElement("a", {
        href: "https://github.com/vector-im/element-web/issues/2671",
        target: "_blank",
        rel: "noreferrer noopener"
      }, "https://github.com/vector-im/element-web/issues/2671")),
      button: (0, _languageHandler._t)("Continue"),
      extraButtons: [/*#__PURE__*/_react.default.createElement("button", {
        key: "exportRoomKeys",
        className: "mx_Dialog_primary",
        onClick: this.onExportE2eKeysClicked
      }, (0, _languageHandler._t)('Export E2E room keys'))],
      onFinished: confirmed => {
        if (confirmed) {
          this.changePassword(cli, oldPassword, newPassword);
        }
      }
    });
  }

  changePassword(cli, oldPassword, newPassword) {
    const authDict = {
      type: 'm.login.password',
      identifier: {
        type: 'm.id.user',
        user: cli.credentials.userId
      },
      // TODO: Remove `user` once servers support proper UIA
      // See https://github.com/matrix-org/synapse/issues/5665
      user: cli.credentials.userId,
      password: oldPassword
    };
    this.setState({
      phase: Phase.Uploading
    });
    cli.setPassword(authDict, newPassword).then(() => {
      if (this.props.shouldAskForEmail) {
        return this.optionallySetEmail().then(confirmed => {
          this.props.onFinished({
            didSetEmail: confirmed
          });
        });
      } else {
        this.props.onFinished();
      }
    }, err => {
      this.props.onError(err);
    }).finally(() => {
      this.setState({
        phase: Phase.Edit,
        oldPassword: "",
        newPassword: "",
        newPasswordConfirm: ""
      });
    });
  }

  checkPassword(oldPass, newPass, confirmPass) {
    if (newPass !== confirmPass) {
      return {
        error: (0, _languageHandler._t)("New passwords don't match")
      };
    } else if (!newPass || newPass.length === 0) {
      return {
        error: (0, _languageHandler._t)("Passwords can't be empty")
      };
    }
  }

  optionallySetEmail() {
    // Ask for an email otherwise the user has no way to reset their password
    const modal = _Modal.default.createTrackedDialog('Do you want to set an email address?', '', _SetEmailDialog.default, {
      title: (0, _languageHandler._t)('Do you want to set an email address?')
    });

    return modal.finished.then(([confirmed]) => confirmed);
  }

  markFieldValid(fieldID, valid) {
    const {
      fieldValid
    } = this.state;
    fieldValid[fieldID] = valid;
    this.setState({
      fieldValid
    });
  }

  async verifyFieldsBeforeSubmit() {
    // Blur the active element if any, so we first run its blur validation,
    // which is less strict than the pass we're about to do below for all fields.
    const activeElement = document.activeElement;

    if (activeElement) {
      activeElement.blur();
    }

    const fieldIDsInDisplayOrder = [FIELD_OLD_PASSWORD, FIELD_NEW_PASSWORD, FIELD_NEW_PASSWORD_CONFIRM]; // Run all fields with stricter validation that no longer allows empty
    // values for required fields.

    for (const fieldID of fieldIDsInDisplayOrder) {
      const field = this[fieldID];

      if (!field) {
        continue;
      } // We must wait for these validations to finish before queueing
      // up the setState below so our setState goes in the queue after
      // all the setStates from these validate calls (that's how we
      // know they've finished).


      await field.validate({
        allowEmpty: false
      });
    } // Validation and state updates are async, so we need to wait for them to complete
    // first. Queue a `setState` callback and wait for it to resolve.


    await new Promise(resolve => this.setState({}, resolve));

    if (this.allFieldsValid()) {
      return true;
    }

    const invalidField = this.findFirstInvalidField(fieldIDsInDisplayOrder);

    if (!invalidField) {
      return true;
    } // Focus the first invalid field and show feedback in the stricter mode
    // that no longer allows empty values for required fields.


    invalidField.focus();
    invalidField.validate({
      allowEmpty: false,
      focused: true
    });
    return false;
  }

  allFieldsValid() {
    const keys = Object.keys(this.state.fieldValid);

    for (let i = 0; i < keys.length; ++i) {
      if (!this.state.fieldValid[keys[i]]) {
        return false;
      }
    }

    return true;
  }

  findFirstInvalidField(fieldIDs) {
    for (const fieldID of fieldIDs) {
      if (!this.state.fieldValid[fieldID] && this[fieldID]) {
        return this[fieldID];
      }
    }

    return null;
  }

  render() {
    const rowClassName = this.props.rowClassName;
    const buttonClassName = this.props.buttonClassName;

    switch (this.state.phase) {
      case Phase.Edit:
        return /*#__PURE__*/_react.default.createElement("form", {
          className: this.props.className,
          onSubmit: this.onClickChange
        }, /*#__PURE__*/_react.default.createElement("div", {
          className: rowClassName
        }, /*#__PURE__*/_react.default.createElement(_Field.default, {
          ref: field => this[FIELD_OLD_PASSWORD] = field,
          type: "password",
          label: (0, _languageHandler._t)('Current password'),
          value: this.state.oldPassword,
          onChange: this.onChangeOldPassword,
          onValidate: this.onOldPasswordValidate
        })), /*#__PURE__*/_react.default.createElement("div", {
          className: rowClassName
        }, /*#__PURE__*/_react.default.createElement(_PassphraseField.default, {
          fieldRef: field => this[FIELD_NEW_PASSWORD] = field,
          type: "password",
          label: "New Password",
          minScore: _RegistrationForm.PASSWORD_MIN_SCORE,
          value: this.state.newPassword,
          autoFocus: this.props.autoFocusNewPasswordInput,
          onChange: this.onChangeNewPassword,
          onValidate: this.onNewPasswordValidate,
          autoComplete: "new-password"
        })), /*#__PURE__*/_react.default.createElement("div", {
          className: rowClassName
        }, /*#__PURE__*/_react.default.createElement(_Field.default, {
          ref: field => this[FIELD_NEW_PASSWORD_CONFIRM] = field,
          type: "password",
          label: (0, _languageHandler._t)("Confirm password"),
          value: this.state.newPasswordConfirm,
          onChange: this.onChangeNewPasswordConfirm,
          onValidate: this.onNewPasswordConfirmValidate,
          autoComplete: "new-password"
        })), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
          className: buttonClassName,
          kind: this.props.buttonKind,
          onClick: this.onClickChange
        }, this.props.buttonLabel || (0, _languageHandler._t)('Change Password')));

      case Phase.Uploading:
        return /*#__PURE__*/_react.default.createElement("div", {
          className: "mx_Dialog_content"
        }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
    }
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  onFinished() {},

  onError() {},

  confirm: true
}), _temp)) || _class);
exports.default = ChangePassword;
