// matrix-react-sdk/lib/components/views/settings/ImageSizePanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _StyledRadioButton = _interopRequireDefault(require("../elements/StyledRadioButton"));

var _languageHandler = require("../../../languageHandler");

var _SettingLevel = require("../../../settings/SettingLevel");

var _ImageSize = require("../../../settings/enums/ImageSize");

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class ImageSizePanel extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onSizeChange", ev => {
      const newSize = ev.target.value;
      this.setState({
        size: newSize
      }); // noinspection JSIgnoredPromiseFromCall

      _SettingsStore.default.setValue("Images.size", null, _SettingLevel.SettingLevel.ACCOUNT, newSize);
    });
    this.state = {
      size: _SettingsStore.default.getValue("Images.size")
    };
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_section mx_ImageSizePanel"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subheading"
    }, (0, _languageHandler._t)("Image size in the timeline")), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageSizePanel_radios"
    }, /*#__PURE__*/_react.default.createElement("label", null, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageSizePanel_size mx_ImageSizePanel_sizeDefault"
    }), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
      name: "image_size",
      value: _ImageSize.ImageSize.Normal,
      checked: this.state.size === _ImageSize.ImageSize.Normal,
      onChange: this.onSizeChange
    }, (0, _languageHandler._t)("Default"))), /*#__PURE__*/_react.default.createElement("label", null, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ImageSizePanel_size mx_ImageSizePanel_sizeLarge"
    }), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
      name: "image_size",
      value: _ImageSize.ImageSize.Large,
      checked: this.state.size === _ImageSize.ImageSize.Large,
      onChange: this.onSizeChange
    }, (0, _languageHandler._t)("Large")))));
  }

}

exports.default = ImageSizePanel;
