// matrix-react-sdk/lib/components/views/settings/SecureBackupPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _languageHandler = require("../../../languageHandler");

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _WellKnownUtils = require("../../../utils/WellKnownUtils");

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _QuestionDialog = _interopRequireDefault(require("../dialogs/QuestionDialog"));

var _RestoreKeyBackupDialog = _interopRequireDefault(require("../dialogs/security/RestoreKeyBackupDialog"));

var _SecurityManager = require("../../../SecurityManager");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let SecureBackupPanel = (_dec = (0, _replaceableComponent.replaceableComponent)("views.settings.SecureBackupPanel"), _dec(_class = class SecureBackupPanel extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "unmounted", false);
    (0, _defineProperty2.default)(this, "onKeyBackupSessionsRemaining", sessionsRemaining => {
      this.setState({
        sessionsRemaining
      });
    });
    (0, _defineProperty2.default)(this, "onKeyBackupStatus", () => {
      // This just loads the current backup status rather than forcing
      // a re-check otherwise we risk causing infinite loops
      this.loadBackupStatus();
    });
    (0, _defineProperty2.default)(this, "startNewBackup", () => {
      _Modal.default.createTrackedDialogAsync('Key Backup', 'Key Backup', Promise.resolve().then(() => _interopRequireWildcard(require('../../../async-components/views/dialogs/security/CreateKeyBackupDialog'))), {
        onFinished: () => {
          this.loadBackupStatus();
        }
      }, null,
      /* priority = */
      false,
      /* static = */
      true);
    });
    (0, _defineProperty2.default)(this, "deleteBackup", () => {
      _Modal.default.createTrackedDialog('Delete Backup', '', _QuestionDialog.default, {
        title: (0, _languageHandler._t)('Delete Backup'),
        description: (0, _languageHandler._t)("Are you sure? You will lose your encrypted messages if your " + "keys are not backed up properly."),
        button: (0, _languageHandler._t)('Delete Backup'),
        danger: true,
        onFinished: proceed => {
          if (!proceed) return;
          this.setState({
            loading: true
          });

          _MatrixClientPeg.MatrixClientPeg.get().deleteKeyBackupVersion(this.state.backupInfo.version).then(() => {
            this.loadBackupStatus();
          });
        }
      });
    });
    (0, _defineProperty2.default)(this, "restoreBackup", async () => {
      _Modal.default.createTrackedDialog('Restore Backup', '', _RestoreKeyBackupDialog.default, null, null,
      /* priority = */
      false,
      /* static = */
      true);
    });
    (0, _defineProperty2.default)(this, "resetSecretStorage", async () => {
      this.setState({
        error: null
      });

      try {
        await (0, _SecurityManager.accessSecretStorage)(async () => {},
        /* forceReset = */
        true);
      } catch (e) {
        _logger.logger.error("Error resetting secret storage", e);

        if (this.unmounted) return;
        this.setState({
          error: e
        });
      }

      if (this.unmounted) return;
      this.loadBackupStatus();
    });
    this.state = {
      loading: true,
      error: null,
      backupKeyStored: null,
      backupKeyCached: null,
      backupKeyWellFormed: null,
      secretStorageKeyInAccount: null,
      secretStorageReady: null,
      backupInfo: null,
      backupSigStatus: null,
      sessionsRemaining: 0
    };
  }

  componentDidMount() {
    this.checkKeyBackupStatus();

    _MatrixClientPeg.MatrixClientPeg.get().on('crypto.keyBackupStatus', this.onKeyBackupStatus);

    _MatrixClientPeg.MatrixClientPeg.get().on('crypto.keyBackupSessionsRemaining', this.onKeyBackupSessionsRemaining);
  }

  componentWillUnmount() {
    this.unmounted = true;

    if (_MatrixClientPeg.MatrixClientPeg.get()) {
      _MatrixClientPeg.MatrixClientPeg.get().removeListener('crypto.keyBackupStatus', this.onKeyBackupStatus);

      _MatrixClientPeg.MatrixClientPeg.get().removeListener('crypto.keyBackupSessionsRemaining', this.onKeyBackupSessionsRemaining);
    }
  }

  async checkKeyBackupStatus() {
    this.getUpdatedDiagnostics();

    try {
      const {
        backupInfo,
        trustInfo
      } = await _MatrixClientPeg.MatrixClientPeg.get().checkKeyBackup();
      this.setState({
        loading: false,
        error: null,
        backupInfo,
        backupSigStatus: trustInfo
      });
    } catch (e) {
      _logger.logger.log("Unable to fetch check backup status", e);

      if (this.unmounted) return;
      this.setState({
        loading: false,
        error: e,
        backupInfo: null,
        backupSigStatus: null
      });
    }
  }

  async loadBackupStatus() {
    this.setState({
      loading: true
    });
    this.getUpdatedDiagnostics();

    try {
      const backupInfo = await _MatrixClientPeg.MatrixClientPeg.get().getKeyBackupVersion();
      const backupSigStatus = await _MatrixClientPeg.MatrixClientPeg.get().isKeyBackupTrusted(backupInfo);
      if (this.unmounted) return;
      this.setState({
        loading: false,
        error: null,
        backupInfo,
        backupSigStatus
      });
    } catch (e) {
      _logger.logger.log("Unable to fetch key backup status", e);

      if (this.unmounted) return;
      this.setState({
        loading: false,
        error: e,
        backupInfo: null,
        backupSigStatus: null
      });
    }
  }

  async getUpdatedDiagnostics() {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    const secretStorage = cli.crypto.secretStorage;
    const backupKeyStored = !!(await cli.isKeyBackupKeyStored());
    const backupKeyFromCache = await cli.crypto.getSessionBackupPrivateKey();
    const backupKeyCached = !!backupKeyFromCache;
    const backupKeyWellFormed = backupKeyFromCache instanceof Uint8Array;
    const secretStorageKeyInAccount = await secretStorage.hasKey();
    const secretStorageReady = await cli.isSecretStorageReady();
    if (this.unmounted) return;
    this.setState({
      backupKeyStored,
      backupKeyCached,
      backupKeyWellFormed,
      secretStorageKeyInAccount,
      secretStorageReady
    });
  }

  render() {
    const {
      loading,
      error,
      backupKeyStored,
      backupKeyCached,
      backupKeyWellFormed,
      secretStorageKeyInAccount,
      secretStorageReady,
      backupInfo,
      backupSigStatus,
      sessionsRemaining
    } = this.state;
    let statusDescription;
    let extraDetailsTableRows;
    let extraDetails;
    const actions = [];

    if (error) {
      statusDescription = /*#__PURE__*/_react.default.createElement("div", {
        className: "error"
      }, (0, _languageHandler._t)("Unable to load key backup status"));
    } else if (loading) {
      statusDescription = /*#__PURE__*/_react.default.createElement(_Spinner.default, null);
    } else if (backupInfo) {
      let restoreButtonCaption = (0, _languageHandler._t)("Restore from Backup");

      if (_MatrixClientPeg.MatrixClientPeg.get().getKeyBackupEnabled()) {
        statusDescription = /*#__PURE__*/_react.default.createElement("p", null, "\u2705 ", (0, _languageHandler._t)("This session is backing up your keys. "));
      } else {
        statusDescription = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("This session is <b>not backing up your keys</b>, " + "but you do have an existing backup you can restore from " + "and add to going forward.", {}, {
          b: sub => /*#__PURE__*/_react.default.createElement("b", null, sub)
        })), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Connect this session to key backup before signing out to avoid " + "losing any keys that may only be on this session.")));
        restoreButtonCaption = (0, _languageHandler._t)("Connect this session to Key Backup");
      }

      let uploadStatus;

      if (!_MatrixClientPeg.MatrixClientPeg.get().getKeyBackupEnabled()) {
        // No upload status to show when backup disabled.
        uploadStatus = "";
      } else if (sessionsRemaining > 0) {
        uploadStatus = /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("Backing up %(sessionsRemaining)s keys...", {
          sessionsRemaining
        }), " ", /*#__PURE__*/_react.default.createElement("br", null));
      } else {
        uploadStatus = /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("All keys backed up"), " ", /*#__PURE__*/_react.default.createElement("br", null));
      }

      let backupSigStatuses = backupSigStatus.sigs.map((sig, i) => {
        const deviceName = sig.device ? sig.device.getDisplayName() || sig.device.deviceId : null;

        const validity = sub => /*#__PURE__*/_react.default.createElement("span", {
          className: sig.valid ? 'mx_SecureBackupPanel_sigValid' : 'mx_SecureBackupPanel_sigInvalid'
        }, sub);

        const verify = sub => /*#__PURE__*/_react.default.createElement("span", {
          className: sig.device && sig.deviceTrust.isVerified() ? 'mx_SecureBackupPanel_deviceVerified' : 'mx_SecureBackupPanel_deviceNotVerified'
        }, sub);

        const device = sub => /*#__PURE__*/_react.default.createElement("span", {
          className: "mx_SecureBackupPanel_deviceName"
        }, deviceName);

        const fromThisDevice = sig.device && sig.device.getFingerprint() === _MatrixClientPeg.MatrixClientPeg.get().getDeviceEd25519Key();

        const fromThisUser = sig.crossSigningId && sig.deviceId === _MatrixClientPeg.MatrixClientPeg.get().getCrossSigningId();

        let sigStatus;

        if (sig.valid && fromThisUser) {
          sigStatus = (0, _languageHandler._t)("Backup has a <validity>valid</validity> signature from this user", {}, {
            validity
          });
        } else if (!sig.valid && fromThisUser) {
          sigStatus = (0, _languageHandler._t)("Backup has a <validity>invalid</validity> signature from this user", {}, {
            validity
          });
        } else if (sig.crossSigningId) {
          sigStatus = (0, _languageHandler._t)("Backup has a signature from <verify>unknown</verify> user with ID %(deviceId)s", {
            deviceId: sig.deviceId
          }, {
            verify
          });
        } else if (!sig.device) {
          sigStatus = (0, _languageHandler._t)("Backup has a signature from <verify>unknown</verify> session with ID %(deviceId)s", {
            deviceId: sig.deviceId
          }, {
            verify
          });
        } else if (sig.valid && fromThisDevice) {
          sigStatus = (0, _languageHandler._t)("Backup has a <validity>valid</validity> signature from this session", {}, {
            validity
          });
        } else if (!sig.valid && fromThisDevice) {
          // it can happen...
          sigStatus = (0, _languageHandler._t)("Backup has an <validity>invalid</validity> signature from this session", {}, {
            validity
          });
        } else if (sig.valid && sig.deviceTrust.isVerified()) {
          sigStatus = (0, _languageHandler._t)("Backup has a <validity>valid</validity> signature from " + "<verify>verified</verify> session <device></device>", {}, {
            validity,
            verify,
            device
          });
        } else if (sig.valid && !sig.deviceTrust.isVerified()) {
          sigStatus = (0, _languageHandler._t)("Backup has a <validity>valid</validity> signature from " + "<verify>unverified</verify> session <device></device>", {}, {
            validity,
            verify,
            device
          });
        } else if (!sig.valid && sig.deviceTrust.isVerified()) {
          sigStatus = (0, _languageHandler._t)("Backup has an <validity>invalid</validity> signature from " + "<verify>verified</verify> session <device></device>", {}, {
            validity,
            verify,
            device
          });
        } else if (!sig.valid && !sig.deviceTrust.isVerified()) {
          sigStatus = (0, _languageHandler._t)("Backup has an <validity>invalid</validity> signature from " + "<verify>unverified</verify> session <device></device>", {}, {
            validity,
            verify,
            device
          });
        }

        return /*#__PURE__*/_react.default.createElement("div", {
          key: i
        }, sigStatus);
      });

      if (backupSigStatus.sigs.length === 0) {
        backupSigStatuses = (0, _languageHandler._t)("Backup is not signed by any of your sessions");
      }

      let trustedLocally;

      if (backupSigStatus.trusted_locally) {
        trustedLocally = (0, _languageHandler._t)("This backup is trusted because it has been restored on this session");
      }

      extraDetailsTableRows = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Backup version:")), /*#__PURE__*/_react.default.createElement("td", null, backupInfo.version)), /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Algorithm:")), /*#__PURE__*/_react.default.createElement("td", null, backupInfo.algorithm)));
      extraDetails = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, uploadStatus, /*#__PURE__*/_react.default.createElement("div", null, backupSigStatuses), /*#__PURE__*/_react.default.createElement("div", null, trustedLocally));
      actions.push( /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        key: "restore",
        kind: "primary",
        onClick: this.restoreBackup
      }, restoreButtonCaption));

      if (!(0, _WellKnownUtils.isSecureBackupRequired)()) {
        actions.push( /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
          key: "delete",
          kind: "danger",
          onClick: this.deleteBackup
        }, (0, _languageHandler._t)("Delete Backup")));
      }
    } else {
      statusDescription = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Your keys are <b>not being backed up from this session</b>.", {}, {
        b: sub => /*#__PURE__*/_react.default.createElement("b", null, sub)
      })), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Back up your keys before signing out to avoid losing them.")));
      actions.push( /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        key: "setup",
        kind: "primary",
        onClick: this.startNewBackup
      }, (0, _languageHandler._t)("Set up")));
    }

    if (secretStorageKeyInAccount) {
      actions.push( /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        key: "reset",
        kind: "danger",
        onClick: this.resetSecretStorage
      }, (0, _languageHandler._t)("Reset")));
    }

    let backupKeyWellFormedText = "";

    if (backupKeyCached) {
      backupKeyWellFormedText = ", ";

      if (backupKeyWellFormed) {
        backupKeyWellFormedText += (0, _languageHandler._t)("well formed");
      } else {
        backupKeyWellFormedText += (0, _languageHandler._t)("unexpected type");
      }
    }

    let actionRow;

    if (actions.length) {
      actionRow = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SecureBackupPanel_buttonRow"
      }, actions);
    }

    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Back up your encryption keys with your account data in case you " + "lose access to your sessions. Your keys will be secured with a " + "unique Security Key.")), statusDescription, /*#__PURE__*/_react.default.createElement("details", null, /*#__PURE__*/_react.default.createElement("summary", null, (0, _languageHandler._t)("Advanced")), /*#__PURE__*/_react.default.createElement("table", {
      className: "mx_SecureBackupPanel_statusList"
    }, /*#__PURE__*/_react.default.createElement("tbody", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Backup key stored:")), /*#__PURE__*/_react.default.createElement("td", null, backupKeyStored === true ? (0, _languageHandler._t)("in secret storage") : (0, _languageHandler._t)("not stored"))), /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Backup key cached:")), /*#__PURE__*/_react.default.createElement("td", null, backupKeyCached ? (0, _languageHandler._t)("cached locally") : (0, _languageHandler._t)("not found locally"), backupKeyWellFormedText)), /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Secret storage public key:")), /*#__PURE__*/_react.default.createElement("td", null, secretStorageKeyInAccount ? (0, _languageHandler._t)("in account data") : (0, _languageHandler._t)("not found"))), /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", null, (0, _languageHandler._t)("Secret storage:")), /*#__PURE__*/_react.default.createElement("td", null, secretStorageReady ? (0, _languageHandler._t)("ready") : (0, _languageHandler._t)("not ready"))), extraDetailsTableRows)), extraDetails), actionRow);
  }

}) || _class);
exports.default = SecureBackupPanel;
