// matrix-react-sdk/lib/components/views/settings/SetIntegrationManager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _IntegrationManagers = require("../../../integrations/IntegrationManagers");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _SettingLevel = require("../../../settings/SettingLevel");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ToggleSwitch = _interopRequireDefault(require("../elements/ToggleSwitch"));

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class;

let SetIntegrationManager = (_dec = (0, _replaceableComponent.replaceableComponent)("views.settings.SetIntegrationManager"), _dec(_class = class SetIntegrationManager extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onProvisioningToggled", () => {
      const current = this.state.provisioningEnabled;

      _SettingsStore.default.setValue("integrationProvisioning", null, _SettingLevel.SettingLevel.ACCOUNT, !current).catch(err => {
        _logger.logger.error("Error changing integration manager provisioning");

        _logger.logger.error(err);

        this.setState({
          provisioningEnabled: current
        });
      });

      this.setState({
        provisioningEnabled: !current
      });
    });

    const currentManager = _IntegrationManagers.IntegrationManagers.sharedInstance().getPrimaryManager();

    this.state = {
      currentManager,
      provisioningEnabled: _SettingsStore.default.getValue("integrationProvisioning")
    };
  }

  render() {
    const currentManager = this.state.currentManager;
    let managerName;
    let bodyText;

    if (currentManager) {
      managerName = `(${currentManager.name})`;
      bodyText = (0, _languageHandler._t)("Use an integration manager <b>(%(serverName)s)</b> to manage bots, widgets, " + "and sticker packs.", {
        serverName: currentManager.name
      }, {
        b: sub => /*#__PURE__*/_react.default.createElement("b", null, sub)
      });
    } else {
      bodyText = (0, _languageHandler._t)("Use an integration manager to manage bots, widgets, and sticker packs.");
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SetIntegrationManager"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_heading"
    }, /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Manage integrations")), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subheading"
    }, managerName), /*#__PURE__*/_react.default.createElement(_ToggleSwitch.default, {
      checked: this.state.provisioningEnabled,
      disabled: false,
      onChange: this.onProvisioningToggled
    })), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subsectionText"
    }, bodyText, /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement("br", null), (0, _languageHandler._t)("Integration managers receive configuration data, and can modify widgets, " + "send room invites, and set power levels on your behalf.")));
  }

}) || _class);
exports.default = SetIntegrationManager;
