// matrix-react-sdk/lib/components/views/settings/UpdateCheckButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _BasePlatform = require("../../../BasePlatform");

var _PlatformPeg = _interopRequireDefault(require("../../../PlatformPeg"));

var _useDispatcher = require("../../../hooks/useDispatcher");

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _languageHandler = require("../../../languageHandler");

var _InlineSpinner = _interopRequireDefault(require("../../../components/views/elements/InlineSpinner"));

var _AccessibleButton = _interopRequireDefault(require("../../../components/views/elements/AccessibleButton"));

const _excluded = ["action"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function installUpdate() {
  _PlatformPeg.default.get().installUpdate();
}

function getStatusText(status, errorDetail) {
  switch (status) {
    case _BasePlatform.UpdateCheckStatus.Error:
      return (0, _languageHandler._t)('Error encountered (%(errorDetail)s).', {
        errorDetail
      });

    case _BasePlatform.UpdateCheckStatus.Checking:
      return (0, _languageHandler._t)('Checking for an update...');

    case _BasePlatform.UpdateCheckStatus.NotAvailable:
      return (0, _languageHandler._t)('No update available.');

    case _BasePlatform.UpdateCheckStatus.Downloading:
      return (0, _languageHandler._t)('Downloading update...');

    case _BasePlatform.UpdateCheckStatus.Ready:
      return (0, _languageHandler._t)("New version available. <a>Update now.</a>", {}, {
        a: sub => /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
          kind: "link",
          onClick: installUpdate
        }, sub)
      });
  }
}

const doneStatuses = [_BasePlatform.UpdateCheckStatus.Ready, _BasePlatform.UpdateCheckStatus.Error, _BasePlatform.UpdateCheckStatus.NotAvailable];

const UpdateCheckButton = () => {
  const [state, setState] = (0, _react.useState)(null);

  const onCheckForUpdateClick = () => {
    setState(null);

    _PlatformPeg.default.get().startUpdateCheck();
  };

  (0, _useDispatcher.useDispatcher)(_dispatcher.default, _ref => {
    let {
      action
    } = _ref,
        params = (0, _objectWithoutProperties2.default)(_ref, _excluded);

    if (action === _actions.Action.CheckUpdates) {
      setState(params);
    }
  });
  const busy = state && !doneStatuses.includes(state.status);
  let suffix;

  if (state) {
    suffix = /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_UpdateCheckButton_summary"
    }, getStatusText(state.status, state.detail), busy && /*#__PURE__*/_react.default.createElement(_InlineSpinner.default, null));
  }

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    onClick: onCheckForUpdateClick,
    kind: "primary",
    disabled: busy
  }, (0, _languageHandler._t)("Check for update")), suffix);
};

var _default = UpdateCheckButton;
exports.default = _default;
