// matrix-react-sdk/lib/utils/AutoDiscoveryUtils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ValidatedServerConfig = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _autodiscovery = require("matrix-js-sdk/src/autodiscovery");

var _languageHandler = require("../languageHandler");

var _TypeUtils = require("./TypeUtils");

var _SdkConfig = _interopRequireDefault(require("../SdkConfig"));

var _logger = require("matrix-js-sdk/src/logger");

/*
Copyright 2019-2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const LIVELINESS_DISCOVERY_ERRORS = [_autodiscovery.AutoDiscovery.ERROR_INVALID_HOMESERVER, _autodiscovery.AutoDiscovery.ERROR_INVALID_IDENTITY_SERVER];

class ValidatedServerConfig {
  constructor() {
    (0, _defineProperty2.default)(this, "hsUrl", void 0);
    (0, _defineProperty2.default)(this, "hsName", void 0);
    (0, _defineProperty2.default)(this, "hsNameIsDifferent", void 0);
    (0, _defineProperty2.default)(this, "isUrl", void 0);
    (0, _defineProperty2.default)(this, "isDefault", void 0);
    (0, _defineProperty2.default)(this, "isNameResolvable", void 0);
    (0, _defineProperty2.default)(this, "warning", void 0);
  }

}

exports.ValidatedServerConfig = ValidatedServerConfig;

class AutoDiscoveryUtils {
  /**
   * Checks if a given error or error message is considered an error
   * relating to the liveliness of the server. Must be an error returned
   * from this AutoDiscoveryUtils class.
   * @param {string | Error} error The error to check
   * @returns {boolean} True if the error is a liveliness error.
   */
  static isLivelinessError(error) {
    if (!error) return false;
    return !!LIVELINESS_DISCOVERY_ERRORS.find(e => typeof error === "string" ? e === error : e === error.message);
  }
  /**
   * Gets the common state for auth components (login, registration, forgot
   * password) for a given validation error.
   * @param {Error} err The error encountered.
   * @param {string} pageName The page for which the error should be customized to. See
   * implementation for known values.
   * @returns {*} The state for the component, given the error.
   */


  static authComponentStateForError(err, pageName = "login") {
    if (!err) {
      return {
        serverIsAlive: true,
        serverErrorIsFatal: false,
        serverDeadError: null
      };
    }

    let title = (0, _languageHandler._t)("Cannot reach homeserver");
    let body = (0, _languageHandler._t)("Ensure you have a stable internet connection, or get in touch with the server admin");

    if (!AutoDiscoveryUtils.isLivelinessError(err)) {
      const brand = _SdkConfig.default.get().brand;

      title = (0, _languageHandler._t)("Your %(brand)s is misconfigured", {
        brand
      });
      body = (0, _languageHandler._t)("Ask your %(brand)s admin to check <a>your config</a> for incorrect or duplicate entries.", {
        brand
      }, {
        a: sub => {
          return /*#__PURE__*/_react.default.createElement("a", {
            href: "https://github.com/vector-im/element-web/blob/master/docs/config.md",
            target: "_blank",
            rel: "noreferrer noopener"
          }, sub);
        }
      });
    }

    let isFatalError = true;
    const errorMessage = typeof err === "string" ? err : err.message;

    if (errorMessage === _autodiscovery.AutoDiscovery.ERROR_INVALID_IDENTITY_SERVER) {
      isFatalError = false;
      title = (0, _languageHandler._t)("Cannot reach identity server"); // It's annoying having a ladder for the third word in the same sentence, but our translations
      // don't make this easy to avoid.

      if (pageName === "register") {
        body = (0, _languageHandler._t)("You can register, but some features will be unavailable until the identity server is " + "back online. If you keep seeing this warning, check your configuration or contact a server " + "admin.");
      } else if (pageName === "reset_password") {
        body = (0, _languageHandler._t)("You can reset your password, but some features will be unavailable until the identity " + "server is back online. If you keep seeing this warning, check your configuration or contact " + "a server admin.");
      } else {
        body = (0, _languageHandler._t)("You can log in, but some features will be unavailable until the identity server is " + "back online. If you keep seeing this warning, check your configuration or contact a server " + "admin.");
      }
    }

    return {
      serverIsAlive: false,
      serverErrorIsFatal: isFatalError,
      serverDeadError: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("strong", null, title), /*#__PURE__*/_react.default.createElement("div", null, body))
    };
  }
  /**
   * Validates a server configuration, using a pair of URLs as input.
   * @param {string} homeserverUrl The homeserver URL.
   * @param {string} identityUrl The identity server URL.
   * @param {boolean} syntaxOnly If true, errors relating to liveliness of the servers will
   * not be raised.
   * @returns {Promise<ValidatedServerConfig>} Resolves to the validated configuration.
   */


  static async validateServerConfigWithStaticUrls(homeserverUrl, identityUrl, syntaxOnly = false) {
    if (!homeserverUrl) {
      throw (0, _languageHandler.newTranslatableError)((0, _languageHandler._td)("No homeserver URL provided"));
    }

    const wellknownConfig = {
      "m.homeserver": {
        base_url: homeserverUrl
      }
    };

    if (identityUrl) {
      wellknownConfig['m.identity_server'] = {
        base_url: identityUrl
      };
    }

    const result = await _autodiscovery.AutoDiscovery.fromDiscoveryConfig(wellknownConfig);
    const url = new URL(homeserverUrl);
    const serverName = url.hostname;
    return AutoDiscoveryUtils.buildValidatedConfigFromDiscovery(serverName, result, syntaxOnly, true);
  }
  /**
   * Validates a server configuration, using a homeserver domain name as input.
   * @param {string} serverName The homeserver domain name (eg: "matrix.org") to validate.
   * @returns {Promise<ValidatedServerConfig>} Resolves to the validated configuration.
   */


  static async validateServerName(serverName) {
    const result = await _autodiscovery.AutoDiscovery.findClientConfig(serverName);
    return AutoDiscoveryUtils.buildValidatedConfigFromDiscovery(serverName, result);
  }
  /**
   * Validates a server configuration, using a pre-calculated AutoDiscovery result as
   * input.
   * @param {string} serverName The domain name the AutoDiscovery result is for.
   * @param {*} discoveryResult The AutoDiscovery result.
   * @param {boolean} syntaxOnly If true, errors relating to liveliness of the servers will not be raised.
   * @param {boolean} isSynthetic If true, then the discoveryResult was synthesised locally.
   * @returns {Promise<ValidatedServerConfig>} Resolves to the validated configuration.
   */


  static buildValidatedConfigFromDiscovery(serverName, discoveryResult, syntaxOnly = false, isSynthetic = false) {
    if (!discoveryResult || !discoveryResult["m.homeserver"]) {
      // This shouldn't happen without major misconfiguration, so we'll log a bit of information
      // in the log so we can find this bit of codee but otherwise tell teh user "it broke".
      _logger.logger.error("Ended up in a state of not knowing which homeserver to connect to.");

      throw (0, _languageHandler.newTranslatableError)((0, _languageHandler._td)("Unexpected error resolving homeserver configuration"));
    }

    const hsResult = discoveryResult['m.homeserver'];
    const isResult = discoveryResult['m.identity_server'];

    const defaultConfig = _SdkConfig.default.get()["validated_server_config"]; // Validate the identity server first because an invalid identity server causes
    // an invalid homeserver, which may not be picked up correctly.
    // Note: In the cases where we rely on the default IS from the config (namely
    // lack of identity server provided by the discovery method), we intentionally do not
    // validate it. This has already been validated and this helps some off-the-grid usage
    // of Element.


    let preferredIdentityUrl = defaultConfig && defaultConfig['isUrl'];

    if (isResult && isResult.state === _autodiscovery.AutoDiscovery.SUCCESS) {
      preferredIdentityUrl = isResult["base_url"];
    } else if (isResult && isResult.state !== _autodiscovery.AutoDiscovery.PROMPT) {
      _logger.logger.error("Error determining preferred identity server URL:", isResult);

      if (isResult.state === _autodiscovery.AutoDiscovery.FAIL_ERROR) {
        if (_autodiscovery.AutoDiscovery.ALL_ERRORS.indexOf(isResult.error) !== -1) {
          throw (0, _languageHandler.newTranslatableError)(isResult.error);
        }

        throw (0, _languageHandler.newTranslatableError)((0, _languageHandler._td)("Unexpected error resolving identity server configuration"));
      } // else the error is not related to syntax - continue anyways.
      // rewrite homeserver error since we don't care about problems


      hsResult.error = _autodiscovery.AutoDiscovery.ERROR_INVALID_IDENTITY_SERVER; // Also use the user's supplied identity server if provided

      if (isResult["base_url"]) preferredIdentityUrl = isResult["base_url"];
    }

    if (hsResult.state !== _autodiscovery.AutoDiscovery.SUCCESS) {
      _logger.logger.error("Error processing homeserver config:", hsResult);

      if (!syntaxOnly || !AutoDiscoveryUtils.isLivelinessError(hsResult.error)) {
        if (_autodiscovery.AutoDiscovery.ALL_ERRORS.indexOf(hsResult.error) !== -1) {
          throw (0, _languageHandler.newTranslatableError)(hsResult.error);
        }

        throw (0, _languageHandler.newTranslatableError)((0, _languageHandler._td)("Unexpected error resolving homeserver configuration"));
      } // else the error is not related to syntax - continue anyways.

    }

    const preferredHomeserverUrl = hsResult["base_url"];
    let preferredHomeserverName = serverName ? serverName : hsResult["server_name"];
    const url = new URL(preferredHomeserverUrl);
    if (!preferredHomeserverName) preferredHomeserverName = url.hostname; // It should have been set by now, so check it

    if (!preferredHomeserverName) {
      _logger.logger.error("Failed to parse homeserver name from homeserver URL");

      throw (0, _languageHandler.newTranslatableError)((0, _languageHandler._td)("Unexpected error resolving homeserver configuration"));
    }

    return (0, _TypeUtils.makeType)(ValidatedServerConfig, {
      hsUrl: preferredHomeserverUrl,
      hsName: preferredHomeserverName,
      hsNameIsDifferent: url.hostname !== preferredHomeserverName,
      isUrl: preferredIdentityUrl,
      isDefault: false,
      warning: hsResult.error,
      isNameResolvable: !isSynthetic
    });
  }

}

exports.default = AutoDiscoveryUtils;
