// matrix-react-sdk/lib/stores/notifications/RoomNotificationStateStore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RoomNotificationStateStore = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _AsyncStoreWithClient = require("../AsyncStoreWithClient");

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _models = require("../room-list/models");

var _ListNotificationState = require("./ListNotificationState");

var _RoomNotificationState = require("./RoomNotificationState");

var _SummarizedNotificationState = require("./SummarizedNotificationState");

var _VisibilityProvider = require("../room-list/filters/VisibilityProvider");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class RoomNotificationStateStore extends _AsyncStoreWithClient.AsyncStoreWithClient {
  constructor() {
    super(_dispatcher.default, {});
    (0, _defineProperty2.default)(this, "roomMap", new Map());
    (0, _defineProperty2.default)(this, "listMap", new Map());
  }
  /**
   * Gets a snapshot of notification state for all visible rooms. The number of states recorded
   * on the SummarizedNotificationState is equivalent to rooms.
   */


  get globalState() {
    // If we're not ready yet, just return an empty state
    if (!this.matrixClient) return new _SummarizedNotificationState.SummarizedNotificationState(); // Only count visible rooms to not torment the user with notification counts in rooms they can't see.
    // This will include highlights from the previous version of the room internally

    const globalState = new _SummarizedNotificationState.SummarizedNotificationState();

    for (const room of this.matrixClient.getVisibleRooms()) {
      if (_VisibilityProvider.VisibilityProvider.instance.isRoomVisible(room)) {
        globalState.add(this.getRoomState(room));
      }
    }

    return globalState;
  }
  /**
   * Gets an instance of the list state class for the given tag.
   * @param tagId The tag to get the notification state for.
   * @returns The notification state for the tag.
   */


  getListState(tagId) {
    if (this.listMap.has(tagId)) {
      return this.listMap.get(tagId);
    } // TODO: Update if/when invites move out of the room list.


    const useTileCount = tagId === _models.DefaultTagID.Invite;

    const getRoomFn = room => {
      return this.getRoomState(room);
    };

    const state = new _ListNotificationState.ListNotificationState(useTileCount, tagId, getRoomFn);
    this.listMap.set(tagId, state);
    return state;
  }
  /**
   * Gets a copy of the notification state for a room. The consumer should not
   * attempt to destroy the returned state as it may be shared with other
   * consumers.
   * @param room The room to get the notification state for.
   * @returns The room's notification state.
   */


  getRoomState(room) {
    if (!this.roomMap.has(room)) {
      this.roomMap.set(room, new _RoomNotificationState.RoomNotificationState(room));
    }

    return this.roomMap.get(room);
  }

  static get instance() {
    return RoomNotificationStateStore.internalInstance;
  }

  async onNotReady() {
    for (const roomState of this.roomMap.values()) {
      roomState.destroy();
    }
  } // We don't need this, but our contract says we do.


  async onAction(payload) {
    return Promise.resolve();
  }

}

exports.RoomNotificationStateStore = RoomNotificationStateStore;
(0, _defineProperty2.default)(RoomNotificationStateStore, "internalInstance", new RoomNotificationStateStore());
