// matrix-react-sdk/lib/components/views/settings/tabs/room/SecurityRoomSettingsTab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _partials = require("matrix-js-sdk/src/@types/partials");

var _event = require("matrix-js-sdk/src/@types/event");

var _languageHandler = require("../../../../../languageHandler");

var _MatrixClientPeg = require("../../../../../MatrixClientPeg");

var _LabelledToggleSwitch = _interopRequireDefault(require("../../../elements/LabelledToggleSwitch"));

var _Modal = _interopRequireDefault(require("../../../../../Modal"));

var _QuestionDialog = _interopRequireDefault(require("../../../dialogs/QuestionDialog"));

var _StyledRadioGroup = _interopRequireDefault(require("../../../elements/StyledRadioGroup"));

var _SettingLevel = require("../../../../../settings/SettingLevel");

var _SettingsStore = _interopRequireDefault(require("../../../../../settings/SettingsStore"));

var _UIFeature = require("../../../../../settings/UIFeature");

var _replaceableComponent = require("../../../../../utils/replaceableComponent");

var _AccessibleButton = _interopRequireDefault(require("../../../elements/AccessibleButton"));

var _SettingsFlag = _interopRequireDefault(require("../../../elements/SettingsFlag"));

var _createRoom = _interopRequireDefault(require("../../../../../createRoom"));

var _CreateRoomDialog = _interopRequireDefault(require("../../../dialogs/CreateRoomDialog"));

var _JoinRuleSettings = _interopRequireDefault(require("../../JoinRuleSettings"));

var _ErrorDialog = _interopRequireDefault(require("../../../dialogs/ErrorDialog"));

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class;

let SecurityRoomSettingsTab = (_dec = (0, _replaceableComponent.replaceableComponent)("views.settings.tabs.room.SecurityRoomSettingsTab"), _dec(_class = class SecurityRoomSettingsTab extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onStateEvent", e => {
      const refreshWhenTypes = [_event.EventType.RoomJoinRules, _event.EventType.RoomGuestAccess, _event.EventType.RoomHistoryVisibility, _event.EventType.RoomEncryption];
      if (refreshWhenTypes.includes(e.getType())) this.forceUpdate();
    });
    (0, _defineProperty2.default)(this, "onEncryptionChange", async () => {
      var _MatrixClientPeg$get$;

      if (((_MatrixClientPeg$get$ = _MatrixClientPeg.MatrixClientPeg.get().getRoom(this.props.roomId)) === null || _MatrixClientPeg$get$ === void 0 ? void 0 : _MatrixClientPeg$get$.getJoinRule()) === _partials.JoinRule.Public) {
        const dialog = _Modal.default.createTrackedDialog('Confirm Public Encrypted Room', '', _QuestionDialog.default, {
          title: (0, _languageHandler._t)('Are you sure you want to add encryption to this public room?'),
          description: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, " ", (0, _languageHandler._t)("<b>It's not recommended to add encryption to public rooms.</b>" + "Anyone can find and join public rooms, so anyone can read messages in them. " + "You'll get none of the benefits of encryption, and you won't be able to turn it " + "off later. Encrypting messages in a public room will make receiving and sending " + "messages slower.", null, {
            "b": sub => /*#__PURE__*/_react.default.createElement("b", null, sub)
          }), " "), /*#__PURE__*/_react.default.createElement("p", null, " ", (0, _languageHandler._t)("To avoid these issues, create a <a>new encrypted room</a> for " + "the conversation you plan to have.", null, {
            "a": sub => /*#__PURE__*/_react.default.createElement("a", {
              className: "mx_linkButton",
              onClick: () => {
                dialog.close();
                this.createNewRoom(false, true);
              }
            }, " ", sub, " ")
          }), " "))
        });

        const {
          finished
        } = dialog;
        const [confirm] = await finished;
        if (!confirm) return;
      }

      _Modal.default.createTrackedDialog('Enable encryption', '', _QuestionDialog.default, {
        title: (0, _languageHandler._t)('Enable encryption?'),
        description: (0, _languageHandler._t)("Once enabled, encryption for a room cannot be disabled. Messages sent in an encrypted " + "room cannot be seen by the server, only by the participants of the room. Enabling encryption " + "may prevent many bots and bridges from working correctly. <a>Learn more about encryption.</a>", {}, {
          a: sub => /*#__PURE__*/_react.default.createElement("a", {
            href: "https://element.io/help#encryption",
            rel: "noreferrer noopener",
            target: "_blank"
          }, sub)
        }),
        onFinished: confirm => {
          if (!confirm) {
            this.setState({
              encrypted: false
            });
            return;
          }

          const beforeEncrypted = this.state.encrypted;
          this.setState({
            encrypted: true
          });

          _MatrixClientPeg.MatrixClientPeg.get().sendStateEvent(this.props.roomId, _event.EventType.RoomEncryption, {
            algorithm: "m.megolm.v1.aes-sha2"
          }).catch(e => {
            _logger.logger.error(e);

            this.setState({
              encrypted: beforeEncrypted
            });
          });
        }
      });
    });
    (0, _defineProperty2.default)(this, "onGuestAccessChange", allowed => {
      const guestAccess = allowed ? _partials.GuestAccess.CanJoin : _partials.GuestAccess.Forbidden;
      const beforeGuestAccess = this.state.guestAccess;
      if (beforeGuestAccess === guestAccess) return;
      this.setState({
        guestAccess
      });

      const client = _MatrixClientPeg.MatrixClientPeg.get();

      client.sendStateEvent(this.props.roomId, _event.EventType.RoomGuestAccess, {
        guest_access: guestAccess
      }, "").catch(e => {
        _logger.logger.error(e);

        this.setState({
          guestAccess: beforeGuestAccess
        });
      });
    });
    (0, _defineProperty2.default)(this, "createNewRoom", async (defaultPublic, defaultEncrypted) => {
      const modal = _Modal.default.createTrackedDialog("Create Room", "Create room after trying to make an E2EE room public", _CreateRoomDialog.default, {
        defaultPublic,
        defaultEncrypted
      });

      const [shouldCreate, opts] = await modal.finished;

      if (shouldCreate) {
        await (0, _createRoom.default)(opts);
      }

      return shouldCreate;
    });
    (0, _defineProperty2.default)(this, "onHistoryRadioToggle", history => {
      const beforeHistory = this.state.history;
      if (beforeHistory === history) return;
      this.setState({
        history: history
      });

      _MatrixClientPeg.MatrixClientPeg.get().sendStateEvent(this.props.roomId, _event.EventType.RoomHistoryVisibility, {
        history_visibility: history
      }, "").catch(e => {
        _logger.logger.error(e);

        this.setState({
          history: beforeHistory
        });
      });
    });
    (0, _defineProperty2.default)(this, "updateBlacklistDevicesFlag", checked => {
      _MatrixClientPeg.MatrixClientPeg.get().getRoom(this.props.roomId).setBlacklistUnverifiedDevices(checked);
    });
    (0, _defineProperty2.default)(this, "onJoinRuleChangeError", error => {
      _Modal.default.createTrackedDialog('Room not found', '', _ErrorDialog.default, {
        title: (0, _languageHandler._t)("Failed to update the join rules"),
        description: error.message ?? (0, _languageHandler._t)("Unknown failure")
      });
    });
    (0, _defineProperty2.default)(this, "onBeforeJoinRuleChange", async joinRule => {
      if (this.state.encrypted && joinRule === _partials.JoinRule.Public) {
        const dialog = _Modal.default.createTrackedDialog('Confirm Public Encrypted Room', '', _QuestionDialog.default, {
          title: (0, _languageHandler._t)("Are you sure you want to make this encrypted room public?"),
          description: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, " ", (0, _languageHandler._t)("<b>It's not recommended to make encrypted rooms public.</b> " + "It will mean anyone can find and join the room, so anyone can read messages. " + "You'll get none of the benefits of encryption. Encrypting messages in a public " + "room will make receiving and sending messages slower.", null, {
            "b": sub => /*#__PURE__*/_react.default.createElement("b", null, sub)
          }), " "), /*#__PURE__*/_react.default.createElement("p", null, " ", (0, _languageHandler._t)("To avoid these issues, create a <a>new public room</a> for the conversation " + "you plan to have.", null, {
            "a": sub => /*#__PURE__*/_react.default.createElement("a", {
              className: "mx_linkButton",
              onClick: () => {
                dialog.close();
                this.createNewRoom(true, false);
              }
            }, " ", sub, " ")
          }), " "))
        });

        const {
          finished
        } = dialog;
        const [confirm] = await finished;
        if (!confirm) return false;
      }

      return true;
    });
    (0, _defineProperty2.default)(this, "toggleAdvancedSection", () => {
      this.setState({
        showAdvancedSection: !this.state.showAdvancedSection
      });
    });
    this.state = {
      guestAccess: _partials.GuestAccess.Forbidden,
      history: _partials.HistoryVisibility.Shared,
      hasAliases: false,
      encrypted: false,
      showAdvancedSection: false
    };
  } // TODO: [REACT-WARNING] Move this to constructor


  UNSAFE_componentWillMount() {
    var _joinRuleEvent$getCon, _joinRuleEvent$getCon2;

    // eslint-disable-line
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    cli.on("RoomState.events", this.onStateEvent);
    const room = cli.getRoom(this.props.roomId);
    const state = room.currentState;
    const joinRuleEvent = state.getStateEvents(_event.EventType.RoomJoinRules, "");
    const joinRule = this.pullContentPropertyFromEvent(joinRuleEvent, 'join_rule', _partials.JoinRule.Invite);
    const restrictedAllowRoomIds = joinRule === _partials.JoinRule.Restricted ? joinRuleEvent === null || joinRuleEvent === void 0 ? void 0 : (_joinRuleEvent$getCon = joinRuleEvent.getContent().allow) === null || _joinRuleEvent$getCon === void 0 ? void 0 : (_joinRuleEvent$getCon2 = _joinRuleEvent$getCon.filter(a => a.type === _partials.RestrictedAllowType.RoomMembership)) === null || _joinRuleEvent$getCon2 === void 0 ? void 0 : _joinRuleEvent$getCon2.map(a => a.room_id) : undefined;
    const guestAccess = this.pullContentPropertyFromEvent(state.getStateEvents(_event.EventType.RoomGuestAccess, ""), 'guest_access', _partials.GuestAccess.Forbidden);
    const history = this.pullContentPropertyFromEvent(state.getStateEvents(_event.EventType.RoomHistoryVisibility, ""), 'history_visibility', _partials.HistoryVisibility.Shared);

    const encrypted = _MatrixClientPeg.MatrixClientPeg.get().isRoomEncrypted(this.props.roomId);

    this.setState({
      restrictedAllowRoomIds,
      guestAccess,
      history,
      encrypted
    });
    this.hasAliases().then(hasAliases => this.setState({
      hasAliases
    }));
  }

  pullContentPropertyFromEvent(event, key, defaultValue) {
    return (event === null || event === void 0 ? void 0 : event.getContent()[key]) || defaultValue;
  }

  componentWillUnmount() {
    _MatrixClientPeg.MatrixClientPeg.get().removeListener("RoomState.events", this.onStateEvent);
  }

  async hasAliases() {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    if (await cli.doesServerSupportUnstableFeature("org.matrix.msc2432")) {
      const response = await cli.unstableGetLocalAliases(this.props.roomId);
      const localAliases = response.aliases;
      return Array.isArray(localAliases) && localAliases.length !== 0;
    } else {
      const room = cli.getRoom(this.props.roomId);
      const aliasEvents = room.currentState.getStateEvents(_event.EventType.RoomAliases) || [];
      const hasAliases = !!aliasEvents.find(ev => (ev.getContent().aliases || []).length > 0);
      return hasAliases;
    }
  }

  renderJoinRule() {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const room = client.getRoom(this.props.roomId);
    let aliasWarning = null;

    if (room.getJoinRule() === _partials.JoinRule.Public && !this.state.hasAliases) {
      aliasWarning = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SecurityRoomSettingsTab_warning"
      }, /*#__PURE__*/_react.default.createElement("img", {
        src: require("../../../../../../res/img/warning.svg"),
        width: 15,
        height: 15
      }), /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("To link to this room, please add an address.")));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SecurityRoomSettingsTab_joinRule"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_subsectionText"
    }, /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Decide who can join %(roomName)s.", {
      roomName: room === null || room === void 0 ? void 0 : room.name
    }))), aliasWarning, /*#__PURE__*/_react.default.createElement(_JoinRuleSettings.default, {
      room: room,
      beforeChange: this.onBeforeJoinRuleChange,
      onError: this.onJoinRuleChangeError,
      closeSettingsFn: this.props.closeSettingsFn,
      promptUpgrade: true
    }));
  }

  renderHistory() {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const history = this.state.history;
    const state = client.getRoom(this.props.roomId).currentState;
    const canChangeHistory = state.mayClientSendStateEvent(_event.EventType.RoomHistoryVisibility, client);
    const options = [{
      value: _partials.HistoryVisibility.Shared,
      label: (0, _languageHandler._t)('Members only (since the point in time of selecting this option)')
    }, {
      value: _partials.HistoryVisibility.Invited,
      label: (0, _languageHandler._t)('Members only (since they were invited)')
    }, {
      value: _partials.HistoryVisibility.Joined,
      label: (0, _languageHandler._t)('Members only (since they joined)')
    }]; // World readable doesn't make sense for encrypted rooms

    if (!this.state.encrypted || history === _partials.HistoryVisibility.WorldReadable) {
      options.unshift({
        value: _partials.HistoryVisibility.WorldReadable,
        label: (0, _languageHandler._t)("Anyone")
      });
    }

    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)('Changes to who can read history will only apply to future messages in this room. ' + 'The visibility of existing history will be unchanged.')), /*#__PURE__*/_react.default.createElement(_StyledRadioGroup.default, {
      name: "historyVis",
      value: history,
      onChange: this.onHistoryRadioToggle,
      disabled: !canChangeHistory,
      definitions: options
    }));
  }

  renderAdvanced() {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const guestAccess = this.state.guestAccess;
    const state = client.getRoom(this.props.roomId).currentState;
    const canSetGuestAccess = state.mayClientSendStateEvent(_event.EventType.RoomGuestAccess, client);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_section"
    }, /*#__PURE__*/_react.default.createElement(_LabelledToggleSwitch.default, {
      value: guestAccess === _partials.GuestAccess.CanJoin,
      onChange: this.onGuestAccessChange,
      disabled: !canSetGuestAccess,
      label: (0, _languageHandler._t)("Enable guest access")
    }), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("People with supported clients will be able to join " + "the room without having a registered account.")));
  }

  render() {
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const room = client.getRoom(this.props.roomId);
    const isEncrypted = this.state.encrypted;
    const hasEncryptionPermission = room.currentState.mayClientSendStateEvent(_event.EventType.RoomEncryption, client);
    const canEnableEncryption = !isEncrypted && hasEncryptionPermission;
    let encryptionSettings = null;

    if (isEncrypted && _SettingsStore.default.isEnabled("blacklistUnverifiedDevices")) {
      encryptionSettings = /*#__PURE__*/_react.default.createElement(_SettingsFlag.default, {
        name: "blacklistUnverifiedDevices",
        level: _SettingLevel.SettingLevel.ROOM_DEVICE,
        onChange: this.updateBlacklistDevicesFlag,
        roomId: this.props.roomId
      });
    }

    let historySection = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subheading"
    }, (0, _languageHandler._t)("Who can read history?")), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_section mx_SettingsTab_subsectionText"
    }, this.renderHistory()));

    if (!_SettingsStore.default.getValue(_UIFeature.UIFeature.RoomHistorySettings)) {
      historySection = null;
    }

    let advanced;

    if (room.getJoinRule() === _partials.JoinRule.Public) {
      advanced = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
        onClick: this.toggleAdvancedSection,
        kind: "link",
        className: "mx_SettingsTab_showAdvanced"
      }, this.state.showAdvancedSection ? (0, _languageHandler._t)("Hide advanced") : (0, _languageHandler._t)("Show advanced")), this.state.showAdvancedSection && this.renderAdvanced());
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab mx_SecurityRoomSettingsTab"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_heading"
    }, (0, _languageHandler._t)("Security & Privacy")), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subheading"
    }, (0, _languageHandler._t)("Encryption")), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_section mx_SecurityRoomSettingsTab_encryptionSection"
    }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_subsectionText"
    }, /*#__PURE__*/_react.default.createElement("span", null, (0, _languageHandler._t)("Once enabled, encryption cannot be disabled."))), /*#__PURE__*/_react.default.createElement(_LabelledToggleSwitch.default, {
      value: isEncrypted,
      onChange: this.onEncryptionChange,
      label: (0, _languageHandler._t)("Encrypted"),
      disabled: !canEnableEncryption
    })), encryptionSettings), /*#__PURE__*/_react.default.createElement("span", {
      className: "mx_SettingsTab_subheading"
    }, (0, _languageHandler._t)("Access")), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SettingsTab_section mx_SettingsTab_subsectionText"
    }, this.renderJoinRule()), advanced, historySection);
  }

}) || _class);
exports.default = SecurityRoomSettingsTab;
