// matrix-react-sdk/lib/components/views/rooms/SimpleRoomHeader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

let SimpleRoomHeader = (
/*
 * A stripped-down room header used for things like the user settings
 * and room directory.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.SimpleRoomHeader"), _dec(_class = class SimpleRoomHeader extends _react.default.PureComponent {
  render() {
    let icon;

    if (this.props.icon) {
      icon = /*#__PURE__*/_react.default.createElement("img", {
        className: "mx_RoomHeader_icon",
        src: this.props.icon,
        width: "25",
        height: "25"
      });
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_RoomHeader mx_RoomHeader_wrapper"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_RoomHeader_simpleHeader"
    }, icon, this.props.title));
  }

}) || _class);
exports.default = SimpleRoomHeader;
