// matrix-react-sdk/lib/components/views/dialogs/AskInviteAnywayDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _SettingLevel = require("../../../settings/SettingLevel");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _dec, _class;

let AskInviteAnywayDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.AskInviteAnywayDialog"), _dec(_class = class AskInviteAnywayDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onInviteClicked", () => {
      this.props.onInviteAnyways();
      this.props.onFinished(true);
    });
    (0, _defineProperty2.default)(this, "onInviteNeverWarnClicked", () => {
      _SettingsStore.default.setValue("promptBeforeInviteUnknownUsers", null, _SettingLevel.SettingLevel.ACCOUNT, false);

      this.props.onInviteAnyways();
      this.props.onFinished(true);
    });
    (0, _defineProperty2.default)(this, "onGiveUpClicked", () => {
      this.props.onGiveUp();
      this.props.onFinished(false);
    });
  }

  render() {
    const errorList = this.props.unknownProfileUsers.map(address => /*#__PURE__*/_react.default.createElement("li", {
      key: address.userId
    }, address.userId, ": ", address.errorText));
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_RetryInvitesDialog",
      onFinished: this.onGiveUpClicked,
      title: (0, _languageHandler._t)('The following users may not exist'),
      contentId: "mx_Dialog_content"
    }, /*#__PURE__*/_react.default.createElement("div", {
      id: "mx_Dialog_content"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Unable to find profiles for the Matrix IDs listed below - " + "would you like to invite them anyway?")), /*#__PURE__*/_react.default.createElement("ul", null, errorList)), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_buttons"
    }, /*#__PURE__*/_react.default.createElement("button", {
      onClick: this.onGiveUpClicked
    }, (0, _languageHandler._t)('Close')), /*#__PURE__*/_react.default.createElement("button", {
      onClick: this.onInviteNeverWarnClicked
    }, (0, _languageHandler._t)('Invite anyway and never warn me again')), /*#__PURE__*/_react.default.createElement("button", {
      onClick: this.onInviteClicked,
      autoFocus: true
    }, (0, _languageHandler._t)('Invite anyway'))));
  }

}) || _class);
exports.default = AskInviteAnywayDialog;
