// matrix-react-sdk/lib/components/views/dialogs/BetaFeedbackDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _UserSettingsDialog = require("./UserSettingsDialog");

var _GenericFeatureFeedbackDialog = _interopRequireDefault(require("./GenericFeatureFeedbackDialog"));

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const BetaFeedbackDialog = ({
  featureId,
  onFinished
}) => {
  var _SettingsStore$getBet;

  const info = _SettingsStore.default.getBetaInfo(featureId);

  return /*#__PURE__*/_react.default.createElement(_GenericFeatureFeedbackDialog.default, {
    title: (0, _languageHandler._t)("%(featureName)s beta feedback", {
      featureName: info.title
    }),
    subheading: (0, _languageHandler._t)(info.feedbackSubheading),
    onFinished: onFinished,
    rageshakeLabel: info.feedbackLabel,
    rageshakeData: Object.fromEntries((((_SettingsStore$getBet = _SettingsStore.default.getBetaInfo(featureId)) === null || _SettingsStore$getBet === void 0 ? void 0 : _SettingsStore$getBet.extraSettings) || []).map(k => {
      return _SettingsStore.default.getValue(k);
    }))
  }, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    kind: "link",
    onClick: () => {
      onFinished(false);

      _dispatcher.default.dispatch({
        action: _actions.Action.ViewUserSettings,
        initialTabId: _UserSettingsDialog.UserTab.Labs
      });
    }
  }, (0, _languageHandler._t)("To leave the beta, visit your settings.")));
};

var _default = BetaFeedbackDialog;
exports.default = _default;
