// matrix-react-sdk/lib/components/views/dialogs/ConfirmAndWaitRedactDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ConfirmRedactDialog = _interopRequireDefault(require("./ConfirmRedactDialog"));

var _ErrorDialog = _interopRequireDefault(require("./ErrorDialog"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _dec, _class;

let ConfirmAndWaitRedactDialog = (
/*
 * A dialog for confirming a redaction.
 * Also shows a spinner (and possible error) while the redaction is ongoing,
 * and only closes the dialog when the redaction is done or failed.
 *
 * This is done to prevent the edit history dialog racing with the redaction:
 * if this dialog closes and the MessageEditHistoryDialog is shown again,
 * it will fetch the relations again, which will race with the ongoing /redact request.
 * which will cause the edit to appear unredacted.
 *
 * To avoid this, we keep the dialog open as long as /redact is in progress.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ConfirmAndWaitRedactDialog"), _dec(_class = class ConfirmAndWaitRedactDialog extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onParentFinished", async proceed => {
      if (proceed) {
        this.setState({
          isRedacting: true
        });

        try {
          await this.props.redact();
          this.props.onFinished(true);
        } catch (error) {
          const code = error.errcode || error.statusCode;

          if (typeof code !== "undefined") {
            this.setState({
              redactionErrorCode: code
            });
          } else {
            this.props.onFinished(true);
          }
        }
      } else {
        this.props.onFinished(false);
      }
    });
    this.state = {
      isRedacting: false,
      redactionErrorCode: null
    };
  }

  render() {
    if (this.state.isRedacting) {
      if (this.state.redactionErrorCode) {
        const code = this.state.redactionErrorCode;
        return /*#__PURE__*/_react.default.createElement(_ErrorDialog.default, {
          onFinished: this.props.onFinished,
          title: (0, _languageHandler._t)('Error'),
          description: (0, _languageHandler._t)('You cannot delete this message. (%(code)s)', {
            code
          })
        });
      } else {
        return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
          onFinished: this.props.onFinished,
          hasCancel: false,
          title: (0, _languageHandler._t)("Removing…")
        }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
      }
    } else {
      return /*#__PURE__*/_react.default.createElement(_ConfirmRedactDialog.default, {
        onFinished: this.onParentFinished
      });
    }
  }

}) || _class);
exports.default = ConfirmAndWaitRedactDialog;
