// matrix-react-sdk/lib/components/views/dialogs/ConfirmWipeDeviceDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _dec, _class;

let ConfirmWipeDeviceDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ConfirmWipeDeviceDialog"), _dec(_class = class ConfirmWipeDeviceDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onConfirm", () => {
      this.props.onFinished(true);
    });
    (0, _defineProperty2.default)(this, "onDecline", () => {
      this.props.onFinished(false);
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_ConfirmWipeDeviceDialog",
      hasCancel: true,
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Clear all data in this session?")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ConfirmWipeDeviceDialog_content"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Clearing all data from this session is permanent. Encrypted messages will be lost " + "unless their keys have been backed up."))), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("Clear all data"),
      onPrimaryButtonClick: this.onConfirm,
      primaryButtonClass: "danger",
      cancelButton: (0, _languageHandler._t)("Cancel"),
      onCancel: this.onDecline
    }));
  }

}) || _class);
exports.default = ConfirmWipeDeviceDialog;
