// matrix-react-sdk/lib/components/views/dialogs/ConfirmSpaceUserActionDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _ConfirmUserActionDialog = _interopRequireDefault(require("./ConfirmUserActionDialog"));

var _SpaceStore = _interopRequireDefault(require("../../../stores/spaces/SpaceStore"));

var _SpaceChildrenPicker = _interopRequireDefault(require("../spaces/SpaceChildrenPicker"));

const _excluded = ["space", "spaceChildFilter", "allLabel", "specificLabel", "noneLabel", "warningMessage", "onFinished"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const ConfirmSpaceUserActionDialog = _ref => {
  let {
    space,
    spaceChildFilter,
    allLabel,
    specificLabel,
    noneLabel,
    warningMessage,
    onFinished
  } = _ref,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  const spaceChildren = (0, _react.useMemo)(() => {
    const children = _SpaceStore.default.instance.getChildren(space.roomId);

    if (spaceChildFilter) {
      return children.filter(spaceChildFilter);
    }

    return children;
  }, [space.roomId, spaceChildFilter]);
  const [roomsToLeave, setRoomsToLeave] = (0, _react.useState)([]);
  const selectedRooms = (0, _react.useMemo)(() => new Set(roomsToLeave), [roomsToLeave]);
  let warning;

  if (warningMessage) {
    warning = /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ConfirmSpaceUserActionDialog_warning"
    }, warningMessage);
  }

  return /*#__PURE__*/_react.default.createElement(_ConfirmUserActionDialog.default, (0, _extends2.default)({}, props, {
    onFinished: (success, reason) => {
      onFinished(success, reason, roomsToLeave);
    },
    className: "mx_ConfirmSpaceUserActionDialog"
  }), warning, /*#__PURE__*/_react.default.createElement(_SpaceChildrenPicker.default, {
    space: space,
    spaceChildren: spaceChildren,
    selected: selectedRooms,
    allLabel: allLabel,
    specificLabel: specificLabel,
    noneLabel: noneLabel,
    onChange: setRoomsToLeave
  }));
};

var _default = ConfirmSpaceUserActionDialog;
exports.default = _default;
