// matrix-react-sdk/lib/components/views/dialogs/ConfirmRedactDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRedactEventDialog = createRedactEventDialog;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ErrorDialog = _interopRequireDefault(require("./ErrorDialog"));

var _TextInputDialog = _interopRequireDefault(require("./TextInputDialog"));

var _dec, _class;

let ConfirmRedactDialog = (
/*
 * A dialog for confirming a redaction.
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ConfirmRedactDialog"), _dec(_class = class ConfirmRedactDialog extends _react.default.Component {
  render() {
    return /*#__PURE__*/_react.default.createElement(_TextInputDialog.default, {
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Confirm Removal"),
      description: (0, _languageHandler._t)("Are you sure you wish to remove (delete) this event? " + "Note that if you delete a room name or topic change, it could undo the change."),
      placeholder: (0, _languageHandler._t)("Reason (optional)"),
      focus: true,
      button: (0, _languageHandler._t)("Remove")
    });
  }

}) || _class);
exports.default = ConfirmRedactDialog;

function createRedactEventDialog({
  mxEvent,
  onCloseDialog = () => {}
}) {
  _Modal.default.createTrackedDialog('Confirm Redact Dialog', '', ConfirmRedactDialog, {
    onFinished: async (proceed, reason) => {
      if (!proceed) return;

      const cli = _MatrixClientPeg.MatrixClientPeg.get();

      try {
        onCloseDialog === null || onCloseDialog === void 0 ? void 0 : onCloseDialog();
        await cli.redactEvent(mxEvent.getRoomId(), mxEvent.getId(), undefined, reason ? {
          reason
        } : {});
      } catch (e) {
        const code = e.errcode || e.statusCode; // only show the dialog if failing for something other than a network error
        // (e.g. no errcode or statusCode) as in that case the redactions end up in the
        // detached queue and we show the room status bar to allow retry

        if (typeof code !== "undefined") {
          // display error message stating you couldn't delete this.
          _Modal.default.createTrackedDialog('You cannot delete this message', '', _ErrorDialog.default, {
            title: (0, _languageHandler._t)('Error'),
            description: (0, _languageHandler._t)('You cannot delete this message. (%(code)s)', {
              code
            })
          });
        }
      }
    }
  }, 'mx_Dialog_confirmredact');
}
