// matrix-react-sdk/lib/components/views/dialogs/CryptoStoreTooNewDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _languageHandler = require("../../../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _QuestionDialog = _interopRequireDefault(require("./QuestionDialog"));

/*
Copyright 2018 New Vector Ltd
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const CryptoStoreTooNewDialog = props => {
  const brand = _SdkConfig.default.get().brand;

  const _onLogoutClicked = () => {
    _Modal.default.createTrackedDialog('Logout e2e db too new', '', _QuestionDialog.default, {
      title: (0, _languageHandler._t)("Sign out"),
      description: (0, _languageHandler._t)("To avoid losing your chat history, you must export your room keys " + "before logging out. You will need to go back to the newer version of " + "%(brand)s to do this", {
        brand
      }),
      button: (0, _languageHandler._t)("Sign out"),
      focus: false,
      onFinished: doLogout => {
        if (doLogout) {
          _dispatcher.default.dispatch({
            action: 'logout'
          });

          props.onFinished(true);
        }
      }
    });
  };

  const description = (0, _languageHandler._t)("You've previously used a newer version of %(brand)s with this session. " + "To use this version again with end to end encryption, you will " + "need to sign out and back in again.", {
    brand
  });
  return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
    className: "mx_CryptoStoreTooNewDialog",
    contentId: "mx_Dialog_content",
    title: (0, _languageHandler._t)("Incompatible Database"),
    hasCancel: false,
    onFinished: props.onFinished
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_Dialog_content",
    id: "mx_Dialog_content"
  }, description), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
    primaryButton: (0, _languageHandler._t)('Continue With Encryption Disabled'),
    hasCancel: false,
    onPrimaryButtonClick: props.onFinished
  }, /*#__PURE__*/_react.default.createElement("button", {
    onClick: _onLogoutClicked
  }, (0, _languageHandler._t)('Sign out'))));
};

var _default = CryptoStoreTooNewDialog;
exports.default = _default;
