// matrix-react-sdk/lib/components/views/dialogs/GenericFeatureFeedbackDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _QuestionDialog = _interopRequireDefault(require("./QuestionDialog"));

var _languageHandler = require("../../../languageHandler");

var _Field = _interopRequireDefault(require("../elements/Field"));

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _submitRageshake = require("../../../rageshake/submit-rageshake");

var _StyledCheckbox = _interopRequireDefault(require("../elements/StyledCheckbox"));

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _InfoDialog = _interopRequireDefault(require("./InfoDialog"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const GenericFeatureFeedbackDialog = ({
  title,
  subheading,
  children,
  rageshakeLabel,
  rageshakeData = {},
  onFinished
}) => {
  const [comment, setComment] = (0, _react.useState)("");
  const [canContact, setCanContact] = (0, _react.useState)(false);

  const sendFeedback = async ok => {
    if (!ok) return onFinished(false);
    (0, _submitRageshake.submitFeedback)(_SdkConfig.default.get().bug_report_endpoint_url, rageshakeLabel, comment, canContact, rageshakeData);
    onFinished(true);

    _Modal.default.createTrackedDialog("Feedback Sent", rageshakeLabel, _InfoDialog.default, {
      title,
      description: (0, _languageHandler._t)("Thank you for your feedback, we really appreciate it."),
      button: (0, _languageHandler._t)("Done"),
      hasCloseButton: false,
      fixedWidth: false
    });
  };

  return /*#__PURE__*/_react.default.createElement(_QuestionDialog.default, {
    className: "mx_GenericFeatureFeedbackDialog",
    hasCancelButton: true,
    title: title,
    description: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_GenericFeatureFeedbackDialog_subheading"
    }, subheading, "\xA0", (0, _languageHandler._t)("Your platform and username will be noted to help us use your feedback as much as we can."), children), /*#__PURE__*/_react.default.createElement(_Field.default, {
      id: "feedbackComment",
      label: (0, _languageHandler._t)("Feedback"),
      type: "text",
      autoComplete: "off",
      value: comment,
      element: "textarea",
      onChange: ev => {
        setComment(ev.target.value);
      },
      autoFocus: true
    }), /*#__PURE__*/_react.default.createElement(_StyledCheckbox.default, {
      checked: canContact,
      onChange: e => setCanContact(e.target.checked)
    }, (0, _languageHandler._t)("You may contact me if you have any follow up questions"))),
    button: (0, _languageHandler._t)("Send feedback"),
    buttonDisabled: !comment,
    onFinished: sendFeedback
  });
};

var _default = GenericFeatureFeedbackDialog;
exports.default = _default;
