// matrix-react-sdk/lib/components/views/dialogs/IntegrationsImpossibleDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _dec, _class;

let IntegrationsImpossibleDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.IntegrationsImpossibleDialog"), _dec(_class = class IntegrationsImpossibleDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onAcknowledgeClick", () => {
      this.props.onFinished();
    });
  }

  render() {
    const brand = _SdkConfig.default.get().brand;

    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_IntegrationsImpossibleDialog",
      hasCancel: false,
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)("Integrations not allowed")
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_IntegrationsImpossibleDialog_content"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Your %(brand)s doesn't allow you to use an integration manager to do this. " + "Please contact an admin.", {
      brand
    }))), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("OK"),
      onPrimaryButtonClick: this.onAcknowledgeClick,
      hasCancel: false
    }));
  }

}) || _class);
exports.default = IntegrationsImpossibleDialog;
