// matrix-react-sdk/lib/components/views/dialogs/LazyLoadingDisabledDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _QuestionDialog = _interopRequireDefault(require("./QuestionDialog"));

var _languageHandler = require("../../../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

/*
Copyright 2018 New Vector Ltd
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const LazyLoadingDisabledDialog = props => {
  const brand = _SdkConfig.default.get().brand;

  const description1 = (0, _languageHandler._t)("You've previously used %(brand)s on %(host)s with lazy loading of members enabled. " + "In this version lazy loading is disabled. " + "As the local cache is not compatible between these two settings, " + "%(brand)s needs to resync your account.", {
    brand,
    host: props.host
  });
  const description2 = (0, _languageHandler._t)("If the other version of %(brand)s is still open in another tab, " + "please close it as using %(brand)s on the same host with both " + "lazy loading enabled and disabled simultaneously will cause issues.", {
    brand
  });
  return /*#__PURE__*/_react.default.createElement(_QuestionDialog.default, {
    hasCancelButton: false,
    title: (0, _languageHandler._t)("Incompatible local cache"),
    description: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, description1), /*#__PURE__*/_react.default.createElement("p", null, description2)),
    button: (0, _languageHandler._t)("Clear cache and resync"),
    onFinished: props.onFinished
  });
};

var _default = LazyLoadingDisabledDialog;
exports.default = _default;
