// matrix-react-sdk/lib/components/views/dialogs/QuestionDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var sdk = _interopRequireWildcard(require("../../../index"));

var _languageHandler = require("../../../languageHandler");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 New Vector Ltd.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class QuestionDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onOk", () => {
      this.props.onFinished(true);
    });
    (0, _defineProperty2.default)(this, "onCancel", () => {
      this.props.onFinished(false);
    });
  }

  render() {
    // Converting these to imports breaks wrench tests
    const BaseDialog = sdk.getComponent('views.dialogs.BaseDialog');
    const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
    let primaryButtonClass = "";

    if (this.props.danger) {
      primaryButtonClass = "danger";
    }

    return /*#__PURE__*/_react.default.createElement(BaseDialog, {
      className: (0, _classnames.default)("mx_QuestionDialog", this.props.className),
      onFinished: this.props.onFinished,
      title: this.props.title,
      contentId: "mx_Dialog_content",
      headerImage: this.props.headerImage,
      hasCancel: this.props.hasCancelButton,
      fixedWidth: this.props.fixedWidth
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_content",
      id: "mx_Dialog_content"
    }, this.props.description), /*#__PURE__*/_react.default.createElement(DialogButtons, {
      primaryButton: this.props.button || (0, _languageHandler._t)('OK'),
      primaryButtonClass: primaryButtonClass,
      primaryDisabled: this.props.buttonDisabled,
      cancelButton: this.props.cancelButton,
      hasCancel: this.props.hasCancelButton && !this.props.quitOnly,
      onPrimaryButtonClick: this.onOk,
      focus: this.props.focus,
      onCancel: this.onCancel
    }, this.props.extraButtons));
  }

}

exports.default = QuestionDialog;
(0, _defineProperty2.default)(QuestionDialog, "defaultProps", {
  title: "",
  description: "",
  extraButtons: null,
  focus: true,
  hasCancelButton: true,
  danger: false,
  quitOnly: false
});
