// matrix-react-sdk/lib/components/views/dialogs/RegistrationEmailPromptDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _languageHandler = require("../../../languageHandler");

var _CountlyAnalytics = _interopRequireDefault(require("../../../CountlyAnalytics"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _EmailField = _interopRequireDefault(require("../auth/EmailField"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const RegistrationEmailPromptDialog = ({
  onFinished
}) => {
  const [email, setEmail] = (0, React.useState)("");
  const fieldRef = (0, React.useRef)();

  const onSubmit = async e => {
    e.preventDefault();

    if (email) {
      const valid = await fieldRef.current.validate({});

      if (!valid) {
        fieldRef.current.focus();
        fieldRef.current.validate({
          focused: true
        });
        return;
      }
    }

    onFinished(true, email);
  };

  return /*#__PURE__*/React.createElement(_BaseDialog.default, {
    title: (0, _languageHandler._t)("Continuing without email"),
    className: "mx_RegistrationEmailPromptDialog",
    contentId: "mx_RegistrationEmailPromptDialog",
    onFinished: () => onFinished(false),
    fixedWidth: false
  }, /*#__PURE__*/React.createElement("div", {
    className: "mx_Dialog_content",
    id: "mx_RegistrationEmailPromptDialog"
  }, /*#__PURE__*/React.createElement("p", null, (0, _languageHandler._t)("Just a heads up, if you don't add an email and forget your password, you could " + "<b>permanently lose access to your account</b>.", {}, {
    b: sub => /*#__PURE__*/React.createElement("b", null, sub)
  })), /*#__PURE__*/React.createElement("form", {
    onSubmit: onSubmit
  }, /*#__PURE__*/React.createElement(_EmailField.default, {
    fieldRef: fieldRef,
    autoFocus: true,
    label: (0, _languageHandler._t)("Email (optional)"),
    value: email,
    onChange: ev => {
      const target = ev.target;
      setEmail(target.value);
    },
    onFocus: () => _CountlyAnalytics.default.instance.track("onboarding_registration_email2_focus"),
    onBlur: () => _CountlyAnalytics.default.instance.track("onboarding_registration_email2_blur")
  }))), /*#__PURE__*/React.createElement(_DialogButtons.default, {
    primaryButton: (0, _languageHandler._t)("Continue"),
    onPrimaryButtonClick: onSubmit,
    hasCancel: false
  }));
};

var _default = RegistrationEmailPromptDialog;
exports.default = _default;
