// matrix-react-sdk/lib/components/views/dialogs/ServerPickerDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _autodiscovery = require("matrix-js-sdk/src/autodiscovery");

var _AutoDiscoveryUtils = _interopRequireDefault(require("../../../utils/AutoDiscoveryUtils"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _languageHandler = require("../../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _Field = _interopRequireDefault(require("../elements/Field"));

var _StyledRadioButton = _interopRequireDefault(require("../elements/StyledRadioButton"));

var _TextWithTooltip = _interopRequireDefault(require("../elements/TextWithTooltip"));

var _Validation = _interopRequireDefault(require("../elements/Validation"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let ServerPickerDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ServerPickerDialog"), _dec(_class = class ServerPickerDialog extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "defaultServer", void 0);
    (0, _defineProperty2.default)(this, "fieldRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "validatedConf", void 0);
    (0, _defineProperty2.default)(this, "onDefaultChosen", () => {
      this.setState({
        defaultChosen: true
      });
    });
    (0, _defineProperty2.default)(this, "onOtherChosen", () => {
      this.setState({
        defaultChosen: false
      });
    });
    (0, _defineProperty2.default)(this, "onHomeserverChange", ev => {
      this.setState({
        otherHomeserver: ev.target.value
      });
    });
    (0, _defineProperty2.default)(this, "validate", (0, _Validation.default)({
      deriveData: async ({
        value
      }) => {
        let hsUrl = value.trim(); // trim to account for random whitespace
        // if the URL has no protocol, try validate it as a serverName via well-known

        if (!hsUrl.includes("://")) {
          try {
            const discoveryResult = await _autodiscovery.AutoDiscovery.findClientConfig(hsUrl);
            this.validatedConf = _AutoDiscoveryUtils.default.buildValidatedConfigFromDiscovery(hsUrl, discoveryResult);
            return {}; // we have a validated config, we don't need to try the other paths
          } catch (e) {
            _logger.logger.error(`Attempted ${hsUrl} as a server_name but it failed`, e);
          }
        } // if we got to this stage then either the well-known failed or the URL had a protocol specified,
        // so validate statically only. If the URL has no protocol, default to https.


        if (!hsUrl.includes("://")) {
          hsUrl = "https://" + hsUrl;
        }

        try {
          this.validatedConf = await _AutoDiscoveryUtils.default.validateServerConfigWithStaticUrls(hsUrl);
          return {};
        } catch (e) {
          _logger.logger.error(e);

          const stateForError = _AutoDiscoveryUtils.default.authComponentStateForError(e);

          if (stateForError.serverErrorIsFatal) {
            let error = (0, _languageHandler._t)("Unable to validate homeserver");

            if (e.translatedMessage) {
              error = e.translatedMessage;
            }

            return {
              error
            };
          } // try to carry on anyway


          try {
            this.validatedConf = await _AutoDiscoveryUtils.default.validateServerConfigWithStaticUrls(hsUrl, null, true);
            return {};
          } catch (e) {
            _logger.logger.error(e);

            return {
              error: (0, _languageHandler._t)("Invalid URL")
            };
          }
        }
      },
      rules: [{
        key: "required",
        test: ({
          value,
          allowEmpty
        }) => allowEmpty || !!value,
        invalid: () => (0, _languageHandler._t)("Specify a homeserver")
      }, {
        key: "valid",
        test: async function ({
          value
        }, {
          error
        }) {
          if (!value) return true;
          return !error;
        },
        invalid: function ({
          error
        }) {
          return error;
        }
      }]
    }));
    (0, _defineProperty2.default)(this, "onHomeserverValidate", fieldState => this.validate(fieldState));
    (0, _defineProperty2.default)(this, "onSubmit", async ev => {
      ev.preventDefault();
      const valid = await this.fieldRef.current.validate({
        allowEmpty: false
      });

      if (!valid && !this.state.defaultChosen) {
        this.fieldRef.current.focus();
        this.fieldRef.current.validate({
          allowEmpty: false,
          focused: true
        });
        return;
      }

      this.props.onFinished(this.state.defaultChosen ? this.defaultServer : this.validatedConf);
    });

    const config = _SdkConfig.default.get();

    this.defaultServer = config["validated_server_config"];
    const {
      serverConfig
    } = this.props;
    let otherHomeserver = "";

    if (!serverConfig.isDefault) {
      if (serverConfig.isNameResolvable && serverConfig.hsName) {
        otherHomeserver = serverConfig.hsName;
      } else {
        otherHomeserver = serverConfig.hsUrl;
      }
    }

    this.state = {
      defaultChosen: serverConfig.isDefault,
      otherHomeserver
    };
  }

  render() {
    let text;

    if (this.defaultServer.hsName === "matrix.org") {
      text = (0, _languageHandler._t)("Matrix.org is the biggest public homeserver in the world, so it's a good place for many.");
    }

    let defaultServerName = this.defaultServer.hsName;

    if (this.defaultServer.hsNameIsDifferent) {
      defaultServerName = /*#__PURE__*/_react.default.createElement(_TextWithTooltip.default, {
        class: "mx_Login_underlinedServerName",
        tooltip: this.defaultServer.hsUrl
      }, this.defaultServer.hsName);
    }

    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      title: this.props.title || (0, _languageHandler._t)("Sign into your homeserver"),
      className: "mx_ServerPickerDialog",
      contentId: "mx_ServerPickerDialog",
      onFinished: this.props.onFinished,
      fixedWidth: false,
      hasCancel: true
    }, /*#__PURE__*/_react.default.createElement("form", {
      className: "mx_Dialog_content",
      id: "mx_ServerPickerDialog",
      onSubmit: this.onSubmit
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("We call the places where you can host your account 'homeservers'."), " ", text), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
      name: "defaultChosen",
      value: "true",
      checked: this.state.defaultChosen,
      onChange: this.onDefaultChosen
    }, defaultServerName), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
      name: "defaultChosen",
      value: "false",
      className: "mx_ServerPickerDialog_otherHomeserverRadio",
      checked: !this.state.defaultChosen,
      onChange: this.onOtherChosen,
      childrenInLabel: false
    }, /*#__PURE__*/_react.default.createElement(_Field.default, {
      type: "text",
      className: "mx_ServerPickerDialog_otherHomeserver",
      label: (0, _languageHandler._t)("Other homeserver"),
      onChange: this.onHomeserverChange,
      onFocus: this.onOtherChosen,
      ref: this.fieldRef,
      onValidate: this.onHomeserverValidate,
      value: this.state.otherHomeserver,
      validateOnChange: false,
      validateOnFocus: false,
      id: "mx_homeserverInput"
    })), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Use your preferred Matrix homeserver if you have one, or host your own.")), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      className: "mx_ServerPickerDialog_continue",
      kind: "primary",
      onClick: this.onSubmit
    }, (0, _languageHandler._t)("Continue")), /*#__PURE__*/_react.default.createElement("h4", null, (0, _languageHandler._t)("Learn more")), /*#__PURE__*/_react.default.createElement("a", {
      href: "https://matrix.org/faq/#what-is-a-homeserver%3F",
      target: "_blank",
      rel: "noreferrer noopener"
    }, (0, _languageHandler._t)("About homeservers"))));
  }

}) || _class);
exports.default = ServerPickerDialog;
