// matrix-react-sdk/lib/components/views/dialogs/ServerOfflineDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _languageHandler = require("../../../languageHandler");

var _EchoStore = require("../../../stores/local-echo/EchoStore");

var _DateUtils = require("../../../DateUtils");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _RoomEchoContext = require("../../../stores/local-echo/RoomEchoContext");

var _RoomAvatar = _interopRequireDefault(require("../avatars/RoomAvatar"));

var _EchoTransaction = require("../../../stores/local-echo/EchoTransaction");

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _AsyncStore = require("../../../stores/AsyncStore");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let ServerOfflineDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ServerOfflineDialog"), _dec(_class = class ServerOfflineDialog extends React.PureComponent {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onEchosUpdated", () => {
      this.forceUpdate(); // no state to worry about
    });
  }

  componentDidMount() {
    _EchoStore.EchoStore.instance.on(_AsyncStore.UPDATE_EVENT, this.onEchosUpdated);
  }

  componentWillUnmount() {
    _EchoStore.EchoStore.instance.off(_AsyncStore.UPDATE_EVENT, this.onEchosUpdated);
  }

  renderTimeline() {
    return _EchoStore.EchoStore.instance.contexts.map((c, i) => {
      if (!c.firstFailedTime) return null; // not useful

      if (!(c instanceof _RoomEchoContext.RoomEchoContext)) throw new Error("Cannot render unknown context: " + c);
      const header = /*#__PURE__*/React.createElement("div", {
        className: "mx_ServerOfflineDialog_content_context_timeline_header"
      }, /*#__PURE__*/React.createElement(_RoomAvatar.default, {
        width: 24,
        height: 24,
        room: c.room
      }), /*#__PURE__*/React.createElement("span", null, c.room.name));
      const entries = c.transactions.filter(t => t.status === _EchoTransaction.TransactionStatus.Error || t.didPreviouslyFail).map((t, j) => {
        let button = /*#__PURE__*/React.createElement(_Spinner.default, {
          w: 19,
          h: 19
        });

        if (t.status === _EchoTransaction.TransactionStatus.Error) {
          button = /*#__PURE__*/React.createElement(_AccessibleButton.default, {
            kind: "link",
            onClick: () => t.run()
          }, (0, _languageHandler._t)("Resend"));
        }

        return /*#__PURE__*/React.createElement("div", {
          className: "mx_ServerOfflineDialog_content_context_txn",
          key: `txn-${j}`
        }, /*#__PURE__*/React.createElement("span", {
          className: "mx_ServerOfflineDialog_content_context_txn_desc"
        }, t.auditName), button);
      });
      return /*#__PURE__*/React.createElement("div", {
        className: "mx_ServerOfflineDialog_content_context",
        key: `context-${i}`
      }, /*#__PURE__*/React.createElement("div", {
        className: "mx_ServerOfflineDialog_content_context_timestamp"
      }, (0, _DateUtils.formatTime)(c.firstFailedTime, _SettingsStore.default.getValue("showTwelveHourTimestamps"))), /*#__PURE__*/React.createElement("div", {
        className: "mx_ServerOfflineDialog_content_context_timeline"
      }, header, entries));
    });
  }

  render() {
    let timeline = this.renderTimeline().filter(c => !!c); // remove nulls for next check

    if (timeline.length === 0) {
      timeline = [/*#__PURE__*/React.createElement("div", {
        key: 1
      }, (0, _languageHandler._t)("You're all caught up."))];
    }

    const serverName = _MatrixClientPeg.MatrixClientPeg.getHomeserverName();

    return /*#__PURE__*/React.createElement(_BaseDialog.default, {
      title: (0, _languageHandler._t)("Server isn't responding"),
      className: "mx_ServerOfflineDialog",
      contentId: "mx_Dialog_content",
      onFinished: this.props.onFinished,
      hasCancel: true
    }, /*#__PURE__*/React.createElement("div", {
      className: "mx_ServerOfflineDialog_content"
    }, /*#__PURE__*/React.createElement("p", null, (0, _languageHandler._t)("Your server isn't responding to some of your requests. " + "Below are some of the most likely reasons.")), /*#__PURE__*/React.createElement("ul", null, /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("The server (%(serverName)s) took too long to respond.", {
      serverName
    })), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("Your firewall or anti-virus is blocking the request.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("A browser extension is preventing the request.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("The server is offline.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("The server has denied your request.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("Your area is experiencing difficulties connecting to the internet.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("A connection error occurred while trying to contact the server.")), /*#__PURE__*/React.createElement("li", null, (0, _languageHandler._t)("The server is not configured to indicate what the problem is (CORS)."))), /*#__PURE__*/React.createElement("hr", null), /*#__PURE__*/React.createElement("h2", null, (0, _languageHandler._t)("Recent changes that have not yet been received")), timeline));
  }

}) || _class);
exports.default = ServerOfflineDialog;
