// matrix-react-sdk/lib/components/views/dialogs/SeshatResetDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _dec, _class;

let SeshatResetDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.SeshatResetDialog"), _dec(_class = class SeshatResetDialog extends _react.default.PureComponent {
  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      hasCancel: true,
      onFinished: this.props.onFinished.bind(null, false),
      title: (0, _languageHandler._t)("Reset event store?")
    }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("You most likely do not want to reset your event index store"), /*#__PURE__*/_react.default.createElement("br", null), (0, _languageHandler._t)("If you do, please note that none of your messages will be deleted, " + "but the search experience might be degraded for a few moments " + "whilst the index is recreated"))), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("Reset event store"),
      onPrimaryButtonClick: this.props.onFinished.bind(null, true),
      primaryButtonClass: "danger",
      cancelButton: (0, _languageHandler._t)("Cancel"),
      onCancel: this.props.onFinished.bind(null, false)
    }));
  }

}) || _class);
exports.default = SeshatResetDialog;
