// matrix-react-sdk/lib/components/views/dialogs/StorageEvictedDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _BugReportDialog = _interopRequireDefault(require("./BugReportDialog"));

var _dec, _class;

let StorageEvictedDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.StorageEvictedDialog"), _dec(_class = class StorageEvictedDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "sendBugReport", ev => {
      ev.preventDefault();

      _Modal.default.createTrackedDialog('Storage evicted', 'Send Bug Report Dialog', _BugReportDialog.default, {});
    });
    (0, _defineProperty2.default)(this, "onSignOutClick", () => {
      this.props.onFinished(true);
    });
  }

  render() {
    let logRequest;

    if (_SdkConfig.default.get().bug_report_endpoint_url) {
      logRequest = (0, _languageHandler._t)("To help us prevent this in future, please <a>send us logs</a>.", {}, {
        a: text => /*#__PURE__*/_react.default.createElement("a", {
          href: "#",
          onClick: this.sendBugReport
        }, text)
      });
    }

    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_ErrorDialog",
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)('Missing session data'),
      contentId: "mx_Dialog_content",
      hasCancel: false
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_content",
      id: "mx_Dialog_content"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Some session data, including encrypted message keys, is " + "missing. Sign out and sign in to fix this, restoring keys " + "from backup.")), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Your browser likely removed this data when running low on " + "disk space."), " ", logRequest)), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("Sign out"),
      onPrimaryButtonClick: this.onSignOutClick,
      focus: true,
      hasCancel: false
    }));
  }

}) || _class);
exports.default = StorageEvictedDialog;
