// matrix-react-sdk/lib/components/views/dialogs/UntrustedDeviceDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _E2EIcon = _interopRequireWildcard(require("../rooms/E2EIcon"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2019, 2020, 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const UntrustedDeviceDialog = ({
  device,
  user,
  onFinished
}) => {
  let askToVerifyText;
  let newSessionText;

  if (_MatrixClientPeg.MatrixClientPeg.get().getUserId() === user.userId) {
    newSessionText = (0, _languageHandler._t)("You signed in to a new session without verifying it:");
    askToVerifyText = (0, _languageHandler._t)("Verify your other session using one of the options below.");
  } else {
    newSessionText = (0, _languageHandler._t)("%(name)s (%(userId)s) signed in to a new session without verifying it:", {
      name: user.displayName,
      userId: user.userId
    });
    askToVerifyText = (0, _languageHandler._t)("Ask this user to verify their session, or manually verify it below.");
  }

  return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
    onFinished: onFinished,
    className: "mx_UntrustedDeviceDialog",
    title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_E2EIcon.default, {
      status: _E2EIcon.E2EState.Warning,
      size: 24,
      hideTooltip: true
    }), (0, _languageHandler._t)("Not Trusted"))
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_Dialog_content",
    id: "mx_Dialog_content"
  }, /*#__PURE__*/_react.default.createElement("p", null, newSessionText), /*#__PURE__*/_react.default.createElement("p", null, device.getDisplayName(), " (", device.deviceId, ")"), /*#__PURE__*/_react.default.createElement("p", null, askToVerifyText)), /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_Dialog_buttons"
  }, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    element: "button",
    kind: "secondary",
    onClick: () => onFinished("legacy")
  }, (0, _languageHandler._t)("Manually Verify by Text")), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    element: "button",
    kind: "secondary",
    onClick: () => onFinished("sas")
  }, (0, _languageHandler._t)("Interactively verify by Emoji")), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    kind: "primary",
    onClick: () => onFinished(false)
  }, (0, _languageHandler._t)("Done"))));
};

var _default = UntrustedDeviceDialog;
exports.default = _default;
