// matrix-react-sdk/lib/components/views/dialogs/VerificationRequestDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _EncryptionPanel = _interopRequireDefault(require("../right_panel/EncryptionPanel"));

var _dec, _class;

let VerificationRequestDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.VerificationRequestDialog"), _dec(_class = class VerificationRequestDialog extends _react.default.Component {
  constructor(props) {
    super(props);
    this.state = {
      verificationRequest: this.props.verificationRequest
    };

    if (this.props.verificationRequestPromise) {
      this.props.verificationRequestPromise.then(r => {
        this.setState({
          verificationRequest: r
        });
      });
    }
  }

  render() {
    const request = this.state.verificationRequest;
    const otherUserId = request && request.otherUserId;

    const member = this.props.member || otherUserId && _MatrixClientPeg.MatrixClientPeg.get().getUser(otherUserId);

    const title = request && request.isSelfVerification ? (0, _languageHandler._t)("Verify other login") : (0, _languageHandler._t)("Verification Request");
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_InfoDialog",
      onFinished: this.props.onFinished,
      contentId: "mx_Dialog_content",
      title: title,
      hasCancel: true
    }, /*#__PURE__*/_react.default.createElement(_EncryptionPanel.default, {
      layout: "dialog",
      verificationRequest: this.props.verificationRequest,
      verificationRequestPromise: this.props.verificationRequestPromise,
      onClose: this.props.onFinished,
      member: member,
      isRoomEncrypted: false
    }));
  }

}) || _class);
exports.default = VerificationRequestDialog;
