// matrix-react-sdk/lib/components/structures/AutoHideScrollbar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

const _excluded = ["className", "onScroll", "onWheel", "style", "tabIndex", "wrappedRef", "children"];

class AutoHideScrollbar extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "containerRef", /*#__PURE__*/_react.default.createRef());
  }

  componentDidMount() {
    if (this.containerRef.current && this.props.onScroll) {
      // Using the passive option to not block the main thread
      // https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners
      this.containerRef.current.addEventListener("scroll", this.props.onScroll, {
        passive: true
      });
    }

    if (this.props.wrappedRef) {
      this.props.wrappedRef(this.containerRef.current);
    }
  }

  componentWillUnmount() {
    if (this.containerRef.current && this.props.onScroll) {
      this.containerRef.current.removeEventListener("scroll", this.props.onScroll);
    }
  }

  getScrollTop() {
    return this.containerRef.current.scrollTop;
  }

  render() {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const _this$props = this.props,
          {
      className,
      onScroll,
      onWheel,
      style,
      tabIndex,
      wrappedRef,
      children
    } = _this$props,
          otherProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
    return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, otherProps, {
      ref: this.containerRef,
      style: style,
      className: ["mx_AutoHideScrollbar", className].join(" "),
      onWheel: onWheel // Firefox sometimes makes this element focusable due to
      // overflow:scroll;, so force it out of tab order by default.
      ,
      tabIndex: tabIndex ?? -1
    }), children);
  }

}

exports.default = AutoHideScrollbar;
