// matrix-react-sdk/lib/components/structures/HomePage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _AutoHideScrollbar = _interopRequireDefault(require("./AutoHideScrollbar"));

var _pages = require("../../utils/pages");

var _languageHandler = require("../../languageHandler");

var _SdkConfig = _interopRequireDefault(require("../../SdkConfig"));

var sdk = _interopRequireWildcard(require("../../index"));

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _actions = require("../../dispatcher/actions");

var _BaseAvatar = _interopRequireDefault(require("../views/avatars/BaseAvatar"));

var _OwnProfileStore = require("../../stores/OwnProfileStore");

var _AccessibleButton = _interopRequireDefault(require("../views/elements/AccessibleButton"));

var _AsyncStore = require("../../stores/AsyncStore");

var _useEventEmitter = require("../../hooks/useEventEmitter");

var _MatrixClientContext = _interopRequireDefault(require("../../contexts/MatrixClientContext"));

var _MiniAvatarUploader = _interopRequireWildcard(require("../views/elements/MiniAvatarUploader"));

var _Analytics = _interopRequireDefault(require("../../Analytics"));

var _CountlyAnalytics = _interopRequireDefault(require("../../CountlyAnalytics"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const onClickSendDm = () => {
  _Analytics.default.trackEvent('home_page', 'button', 'dm');

  _CountlyAnalytics.default.instance.track("home_page_button", {
    button: "dm"
  });

  _dispatcher.default.dispatch({
    action: 'view_create_chat'
  });
};

const onClickExplore = () => {
  _Analytics.default.trackEvent('home_page', 'button', 'room_directory');

  _CountlyAnalytics.default.instance.track("home_page_button", {
    button: "room_directory"
  });

  _dispatcher.default.fire(_actions.Action.ViewRoomDirectory);
};

const onClickNewRoom = () => {
  _Analytics.default.trackEvent('home_page', 'button', 'create_room');

  _CountlyAnalytics.default.instance.track("home_page_button", {
    button: "create_room"
  });

  _dispatcher.default.dispatch({
    action: 'view_create_room'
  });
};

const getOwnProfile = userId => ({
  displayName: _OwnProfileStore.OwnProfileStore.instance.displayName || userId,
  avatarUrl: _OwnProfileStore.OwnProfileStore.instance.getHttpAvatarUrl(_MiniAvatarUploader.AVATAR_SIZE)
});

const UserWelcomeTop = () => {
  const cli = (0, React.useContext)(_MatrixClientContext.default);
  const userId = cli.getUserId();
  const [ownProfile, setOwnProfile] = (0, React.useState)(getOwnProfile(userId));
  (0, _useEventEmitter.useEventEmitter)(_OwnProfileStore.OwnProfileStore.instance, _AsyncStore.UPDATE_EVENT, () => {
    setOwnProfile(getOwnProfile(userId));
  });
  return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_MiniAvatarUploader.default, {
    hasAvatar: !!ownProfile.avatarUrl,
    hasAvatarLabel: (0, _languageHandler._t)("Great, that'll help people know it's you"),
    noAvatarLabel: (0, _languageHandler._t)("Add a photo so people know it's you."),
    setAvatarUrl: url => cli.setAvatarUrl(url)
  }, /*#__PURE__*/React.createElement(_BaseAvatar.default, {
    idName: userId,
    name: ownProfile.displayName,
    url: ownProfile.avatarUrl,
    width: _MiniAvatarUploader.AVATAR_SIZE,
    height: _MiniAvatarUploader.AVATAR_SIZE,
    resizeMethod: "crop"
  })), /*#__PURE__*/React.createElement("h1", null, (0, _languageHandler._t)("Welcome %(name)s", {
    name: ownProfile.displayName
  })), /*#__PURE__*/React.createElement("h4", null, (0, _languageHandler._t)("Now, let's help you get started")));
};

const HomePage = ({
  justRegistered = false
}) => {
  const config = _SdkConfig.default.get();

  const pageUrl = (0, _pages.getHomePageUrl)(config);

  if (pageUrl) {
    // FIXME: Using an import will result in wrench-element-tests failures
    const EmbeddedPage = sdk.getComponent('structures.EmbeddedPage');
    return /*#__PURE__*/React.createElement(EmbeddedPage, {
      className: "mx_HomePage",
      url: pageUrl,
      scrollbar: true
    });
  }

  let introSection;

  if (justRegistered) {
    introSection = /*#__PURE__*/React.createElement(UserWelcomeTop, null);
  } else {
    const brandingConfig = config.branding;
    let logoUrl = "themes/element/img/logos/element-logo.svg";

    if (brandingConfig && brandingConfig.authHeaderLogoUrl) {
      logoUrl = brandingConfig.authHeaderLogoUrl;
    }

    introSection = /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement("img", {
      src: logoUrl,
      alt: config.brand
    }), /*#__PURE__*/React.createElement("h1", null, (0, _languageHandler._t)("Welcome to %(appName)s", {
      appName: config.brand
    })), /*#__PURE__*/React.createElement("h4", null, (0, _languageHandler._t)("Own your conversations.")));
  }

  return /*#__PURE__*/React.createElement(_AutoHideScrollbar.default, {
    className: "mx_HomePage mx_HomePage_default"
  }, /*#__PURE__*/React.createElement("div", {
    className: "mx_HomePage_default_wrapper"
  }, introSection, /*#__PURE__*/React.createElement("div", {
    className: "mx_HomePage_default_buttons"
  }, /*#__PURE__*/React.createElement(_AccessibleButton.default, {
    onClick: onClickSendDm,
    className: "mx_HomePage_button_sendDm"
  }, (0, _languageHandler._t)("Send a Direct Message")), /*#__PURE__*/React.createElement(_AccessibleButton.default, {
    onClick: onClickExplore,
    className: "mx_HomePage_button_explore"
  }, (0, _languageHandler._t)("Explore Public Rooms")), /*#__PURE__*/React.createElement(_AccessibleButton.default, {
    onClick: onClickNewRoom,
    className: "mx_HomePage_button_createGroup"
  }, (0, _languageHandler._t)("Create a Group Chat")))));
};

var _default = HomePage;
exports.default = _default;
