// matrix-react-sdk/lib/components/structures/MainSplit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _reResizable = require("re-resizable");

var _replaceableComponent = require("../../utils/replaceableComponent");

var _dec, _class;

let MainSplit = (_dec = (0, _replaceableComponent.replaceableComponent)("structures.MainSplit"), _dec(_class = class MainSplit extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onResizeStart", () => {
      this.props.resizeNotifier.startResizing();
    });
    (0, _defineProperty2.default)(this, "onResize", () => {
      this.props.resizeNotifier.notifyRightHandleResized();
    });
    (0, _defineProperty2.default)(this, "onResizeStop", (event, direction, elementRef, delta) => {
      this.props.resizeNotifier.stopResizing();
      window.localStorage.setItem("mx_rhs_size", (this.loadSidePanelSize().width + delta.width).toString());
    });
  }

  loadSidePanelSize() {
    let rhsSize = parseInt(window.localStorage.getItem("mx_rhs_size"), 10);

    if (isNaN(rhsSize)) {
      rhsSize = 350;
    }

    return {
      height: "100%",
      width: rhsSize
    };
  }

  render() {
    const bodyView = _react.default.Children.only(this.props.children);

    const panelView = this.props.panel;
    const hasResizer = !this.props.collapsedRhs && panelView;
    let children;

    if (hasResizer) {
      children = /*#__PURE__*/_react.default.createElement(_reResizable.Resizable, {
        defaultSize: this.loadSidePanelSize(),
        minWidth: 264,
        maxWidth: "50%",
        enable: {
          top: false,
          right: false,
          bottom: false,
          left: true,
          topRight: false,
          bottomRight: false,
          bottomLeft: false,
          topLeft: false
        },
        onResizeStart: this.onResizeStart,
        onResize: this.onResize,
        onResizeStop: this.onResizeStop,
        className: "mx_RightPanel_ResizeWrapper",
        handleClasses: {
          left: "mx_ResizeHandle_horizontal"
        }
      }, panelView);
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_MainSplit"
    }, bodyView, children);
  }

}) || _class);
exports.default = MainSplit;
