// matrix-react-sdk/lib/components/structures/MatrixChat.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.Views = void 0;
exports.isLoggedIn = isLoggedIn;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _matrix = require("matrix-js-sdk/src/matrix");

var _errors = require("matrix-js-sdk/src/errors");

var _roomMember = require("matrix-js-sdk/src/models/room-member");

var _utils = require("matrix-js-sdk/src/utils");

require("focus-visible");

require("what-input");

var _Analytics = _interopRequireDefault(require("../../Analytics"));

var _CountlyAnalytics = _interopRequireDefault(require("../../CountlyAnalytics"));

var _DecryptionFailureTracker = require("../../DecryptionFailureTracker");

var _MatrixClientPeg = require("../../MatrixClientPeg");

var _PlatformPeg = _interopRequireDefault(require("../../PlatformPeg"));

var _SdkConfig = _interopRequireDefault(require("../../SdkConfig"));

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _Notifier = _interopRequireDefault(require("../../Notifier"));

var _Modal = _interopRequireDefault(require("../../Modal"));

var _RoomInvite = require("../../RoomInvite");

var Rooms = _interopRequireWildcard(require("../../Rooms"));

var _linkifyMatrix = _interopRequireDefault(require("../../linkify-matrix"));

var Lifecycle = _interopRequireWildcard(require("../../Lifecycle"));

require("../../stores/LifecycleStore");

var _PageTypes = _interopRequireDefault(require("../../PageTypes"));

var _createRoom = _interopRequireDefault(require("../../createRoom"));

var _languageHandler = require("../../languageHandler");

var _SettingsStore = _interopRequireDefault(require("../../settings/SettingsStore"));

var _ThemeController = _interopRequireDefault(require("../../settings/controllers/ThemeController"));

var _Registration = require("../../Registration");

var _ErrorUtils = require("../../utils/ErrorUtils");

var _ResizeNotifier = _interopRequireDefault(require("../../utils/ResizeNotifier"));

var _AutoDiscoveryUtils = _interopRequireDefault(require("../../utils/AutoDiscoveryUtils"));

var _DMRoomMap = _interopRequireDefault(require("../../utils/DMRoomMap"));

var _ThemeWatcher = _interopRequireDefault(require("../../settings/watchers/ThemeWatcher"));

var _FontWatcher = require("../../settings/watchers/FontWatcher");

var _RoomAliasCache = require("../../RoomAliasCache");

var _ToastStore = _interopRequireDefault(require("../../stores/ToastStore"));

var StorageManager = _interopRequireWildcard(require("../../utils/StorageManager"));

var _actions = require("../../dispatcher/actions");

var _AnalyticsToast = require("../../toasts/AnalyticsToast");

var _DesktopNotificationsToast = require("../../toasts/DesktopNotificationsToast");

var _ErrorDialog = _interopRequireDefault(require("../views/dialogs/ErrorDialog"));

var _RoomNotificationStateStore = require("../../stores/notifications/RoomNotificationStateStore");

var _SettingLevel = require("../../settings/SettingLevel");

var _membership = require("../../utils/membership");

var _CreateCommunityPrototypeDialog = _interopRequireDefault(require("../views/dialogs/CreateCommunityPrototypeDialog"));

var _ThreepidInviteStore = _interopRequireDefault(require("../../stores/ThreepidInviteStore"));

var _UIFeature = require("../../settings/UIFeature");

var _CommunityPrototypeStore = require("../../stores/CommunityPrototypeStore");

var _DialPadModal = _interopRequireDefault(require("../views/voip/DialPadModal"));

var _MobileGuideToast = require("../../toasts/MobileGuideToast");

var _pages = require("../../utils/pages");

var _SpaceStore = _interopRequireDefault(require("../../stores/spaces/SpaceStore"));

var _replaceableComponent = require("../../utils/replaceableComponent");

var _RoomListStore = _interopRequireDefault(require("../../stores/room-list/RoomListStore"));

var _models = require("../../stores/room-list/models");

var _Security = _interopRequireDefault(require("../../customisations/Security"));

var _Spinner = _interopRequireDefault(require("../views/elements/Spinner"));

var _QuestionDialog = _interopRequireDefault(require("../views/dialogs/QuestionDialog"));

var _UserSettingsDialog = _interopRequireDefault(require("../views/dialogs/UserSettingsDialog"));

var _CreateGroupDialog = _interopRequireDefault(require("../views/dialogs/CreateGroupDialog"));

var _CreateRoomDialog = _interopRequireDefault(require("../views/dialogs/CreateRoomDialog"));

var _RoomDirectory = _interopRequireDefault(require("./RoomDirectory"));

var _KeySignatureUploadFailedDialog = _interopRequireDefault(require("../views/dialogs/KeySignatureUploadFailedDialog"));

var _IncomingSasDialog = _interopRequireDefault(require("../views/dialogs/IncomingSasDialog"));

var _CompleteSecurity = _interopRequireDefault(require("./auth/CompleteSecurity"));

var _LoggedInView = _interopRequireDefault(require("./LoggedInView"));

var _Welcome = _interopRequireDefault(require("../views/auth/Welcome"));

var _ForgotPassword = _interopRequireDefault(require("./auth/ForgotPassword"));

var _E2eSetup = _interopRequireDefault(require("./auth/E2eSetup"));

var _Registration2 = _interopRequireDefault(require("./auth/Registration"));

var _Login = _interopRequireDefault(require("./auth/Login"));

var _ErrorBoundary = _interopRequireDefault(require("../views/elements/ErrorBoundary"));

var _VerificationRequestToast = _interopRequireDefault(require("../views/toasts/VerificationRequestToast"));

var _performance = _interopRequireWildcard(require("../../performance"));

var _UIStore = _interopRequireWildcard(require("../../stores/UIStore"));

var _SoftLogout = _interopRequireDefault(require("./auth/SoftLogout"));

var _Permalinks = require("../../utils/permalinks/Permalinks");

var _strings = require("../../utils/strings");

var _PosthogAnalytics = require("../../PosthogAnalytics");

var _sentry = require("../../sentry");

var _logger = require("matrix-js-sdk/src/logger");

var _space = require("../../utils/space");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/** constants for MatrixChat.state.view */
let Views;
exports.Views = Views;

(function (Views) {
  Views[Views["LOADING"] = 0] = "LOADING";
  Views[Views["WELCOME"] = 1] = "WELCOME";
  Views[Views["LOGIN"] = 2] = "LOGIN";
  Views[Views["REGISTER"] = 3] = "REGISTER";
  Views[Views["FORGOT_PASSWORD"] = 4] = "FORGOT_PASSWORD";
  Views[Views["COMPLETE_SECURITY"] = 5] = "COMPLETE_SECURITY";
  Views[Views["E2E_SETUP"] = 6] = "E2E_SETUP";
  Views[Views["LOGGED_IN"] = 7] = "LOGGED_IN";
  Views[Views["SOFT_LOGOUT"] = 8] = "SOFT_LOGOUT";
})(Views || (exports.Views = Views = {}));

const AUTH_SCREENS = ["register", "login", "forgot_password", "start_sso", "start_cas", "welcome"]; // Actions that are redirected through the onboarding process prior to being
// re-dispatched. NOTE: some actions are non-trivial and would require
// re-factoring to be included in this list in future.

const ONBOARDING_FLOW_STARTERS = [_actions.Action.ViewUserSettings, 'view_create_chat', 'view_create_room', 'view_create_group'];
let MatrixChat = (_dec = (0, _replaceableComponent.replaceableComponent)("structures.MatrixChat"), _dec(_class = (_temp = _class2 = class MatrixChat extends _react.default.PureComponent {
  constructor(props, context) {
    super(props, context);
    (0, _defineProperty2.default)(this, "firstSyncComplete", void 0);
    (0, _defineProperty2.default)(this, "firstSyncPromise", void 0);
    (0, _defineProperty2.default)(this, "screenAfterLogin", void 0);
    (0, _defineProperty2.default)(this, "pageChanging", void 0);
    (0, _defineProperty2.default)(this, "tokenLogin", void 0);
    (0, _defineProperty2.default)(this, "accountPassword", void 0);
    (0, _defineProperty2.default)(this, "accountPasswordTimer", void 0);
    (0, _defineProperty2.default)(this, "focusComposer", void 0);
    (0, _defineProperty2.default)(this, "subTitleStatus", void 0);
    (0, _defineProperty2.default)(this, "prevWindowWidth", void 0);
    (0, _defineProperty2.default)(this, "loggedInView", void 0);
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "themeWatcher", void 0);
    (0, _defineProperty2.default)(this, "fontWatcher", void 0);
    (0, _defineProperty2.default)(this, "onAction", payload => {
      // console.log(`MatrixClientPeg.onAction: ${payload.action}`);
      // Start the onboarding process for certain actions
      if (_MatrixClientPeg.MatrixClientPeg.get() && _MatrixClientPeg.MatrixClientPeg.get().isGuest() && ONBOARDING_FLOW_STARTERS.includes(payload.action)) {
        // This will cause `payload` to be dispatched later, once a
        // sync has reached the "prepared" state. Setting a matrix ID
        // will cause a full login and sync and finally the deferred
        // action will be dispatched.
        _dispatcher.default.dispatch({
          action: 'do_after_sync_prepared',
          deferred_action: payload
        });

        _dispatcher.default.dispatch({
          action: 'require_registration'
        });

        return;
      }

      switch (payload.action) {
        case 'MatrixActions.accountData':
          // XXX: This is a collection of several hacks to solve a minor problem. We want to
          // update our local state when the identity server changes, but don't want to put that in
          // the js-sdk as we'd be then dictating how all consumers need to behave. However,
          // this component is already bloated and we probably don't want this tiny logic in
          // here, but there's no better place in the react-sdk for it. Additionally, we're
          // abusing the MatrixActionCreator stuff to avoid errors on dispatches.
          if (payload.event_type === 'm.identity_server') {
            const fullUrl = payload.event_content ? payload.event_content['base_url'] : null;

            if (!fullUrl) {
              _MatrixClientPeg.MatrixClientPeg.get().setIdentityServerUrl(null);

              localStorage.removeItem("mx_is_access_token");
              localStorage.removeItem("mx_is_url");
            } else {
              _MatrixClientPeg.MatrixClientPeg.get().setIdentityServerUrl(fullUrl);

              localStorage.removeItem("mx_is_access_token"); // clear token

              localStorage.setItem("mx_is_url", fullUrl); // XXX: Do we still need this?
            } // redispatch the change with a more specific action


            _dispatcher.default.dispatch({
              action: 'id_server_changed'
            });
          }

          break;

        case 'logout':
          _dispatcher.default.dispatch({
            action: "hangup_all"
          });

          Lifecycle.logout();
          break;

        case 'require_registration':
          (0, _Registration.startAnyRegistrationFlow)(payload);
          break;

        case 'start_registration':
          if (Lifecycle.isSoftLogout()) {
            this.onSoftLogout();
            break;
          } // This starts the full registration flow


          if (payload.screenAfterLogin) {
            this.screenAfterLogin = payload.screenAfterLogin;
          }

          this.startRegistration(payload.params || {});
          break;

        case 'start_login':
          if (Lifecycle.isSoftLogout()) {
            this.onSoftLogout();
            break;
          }

          if (payload.screenAfterLogin) {
            this.screenAfterLogin = payload.screenAfterLogin;
          }

          this.viewLogin();
          break;

        case 'start_password_recovery':
          this.setStateForNewView({
            view: Views.FORGOT_PASSWORD
          });
          this.notifyNewScreen('forgot_password');
          break;

        case 'start_chat':
          (0, _createRoom.default)({
            dmUserId: payload.user_id
          });
          break;

        case 'leave_room':
          this.leaveRoom(payload.room_id);
          break;

        case 'forget_room':
          this.forgetRoom(payload.room_id);
          break;

        case 'copy_room':
          this.copyRoom(payload.room_id);
          break;

        case 'reject_invite':
          _Modal.default.createTrackedDialog('Reject invitation', '', _QuestionDialog.default, {
            title: (0, _languageHandler._t)('Reject invitation'),
            description: (0, _languageHandler._t)('Are you sure you want to reject the invitation?'),
            onFinished: confirm => {
              if (confirm) {
                // FIXME: controller shouldn't be loading a view :(
                const modal = _Modal.default.createDialog(_Spinner.default, null, 'mx_Dialog_spinner');

                _MatrixClientPeg.MatrixClientPeg.get().leave(payload.room_id).then(() => {
                  modal.close();

                  if (this.state.currentRoomId === payload.room_id) {
                    _dispatcher.default.dispatch({
                      action: 'view_home_page'
                    });
                  }
                }, err => {
                  modal.close();

                  _Modal.default.createTrackedDialog('Failed to reject invitation', '', _ErrorDialog.default, {
                    title: (0, _languageHandler._t)('Failed to reject invitation'),
                    description: err.toString()
                  });
                });
              }
            }
          });

          break;

        case 'view_user_info':
          this.viewUser(payload.userId, payload.subAction);
          break;

        case _actions.Action.ViewRoom:
          {
            // Takes either a room ID or room alias: if switching to a room the client is already
            // known to be in (eg. user clicks on a room in the recents panel), supply the ID
            // If the user is clicking on a room in the context of the alias being presented
            // to them, supply the room alias. If both are supplied, the room ID will be ignored.
            const promise = this.viewRoom(payload);

            if (payload.deferred_action) {
              promise.then(() => {
                _dispatcher.default.dispatch(payload.deferred_action);
              });
            }

            break;
          }

        case _actions.Action.ViewUserSettings:
          {
            const tabPayload = payload;

            _Modal.default.createTrackedDialog('User settings', '', _UserSettingsDialog.default, {
              initialTabId: tabPayload.initialTabId
            },
            /*className=*/
            null,
            /*isPriority=*/
            false,
            /*isStatic=*/
            true); // View the welcome or home page if we need something to look at


            this.viewSomethingBehindModal();
            break;
          }

        case 'view_create_room':
          this.createRoom(payload.public, payload.defaultName);
          break;

        case 'view_create_group':
          {
            const prototype = _SettingsStore.default.getValue("feature_communities_v2_prototypes");

            _Modal.default.createTrackedDialog('Create Community', '', prototype ? _CreateCommunityPrototypeDialog.default : _CreateGroupDialog.default);

            break;
          }

        case _actions.Action.ViewRoomDirectory:
          {
            if (_SpaceStore.default.instance.activeSpace[0] === "!") {
              _dispatcher.default.dispatch({
                action: "view_room",
                room_id: _SpaceStore.default.instance.activeSpace
              });
            } else {
              _Modal.default.createTrackedDialog('Room directory', '', _RoomDirectory.default, {
                initialText: payload.initialText
              }, 'mx_RoomDirectory_dialogWrapper', false, true);
            } // View the welcome or home page if we need something to look at


            this.viewSomethingBehindModal();
            break;
          }

        case 'view_my_groups':
          this.setPage(_PageTypes.default.MyGroups);
          this.notifyNewScreen('groups');
          break;

        case 'view_group':
          this.viewGroup(payload);
          break;

        case 'view_welcome_page':
          this.viewWelcome();
          break;

        case 'view_home_page':
          this.viewHome(payload.justRegistered);
          break;

        case 'view_start_chat_or_reuse':
          this.chatCreateOrReuse(payload.user_id);
          break;

        case 'view_create_chat':
          (0, _RoomInvite.showStartChatInviteDialog)(payload.initialText || "");
          break;

        case 'view_invite':
          {
            const room = _MatrixClientPeg.MatrixClientPeg.get().getRoom(payload.roomId);

            if (room !== null && room !== void 0 && room.isSpaceRoom()) {
              (0, _space.showSpaceInvite)(room);
            } else {
              (0, _RoomInvite.showRoomInviteDialog)(payload.roomId);
            }

            break;
          }

        case 'view_last_screen':
          // This function does what we want, despite the name. The idea is that it shows
          // the last room we were looking at or some reasonable default/guess. We don't
          // have to worry about email invites or similar being re-triggered because the
          // function will have cleared that state and not execute that path.
          this.showScreenAfterLogin();
          break;

        case 'toggle_my_groups':
          // persist that the user has interacted with this, use it to dismiss the beta dot
          localStorage.setItem("mx_seenSpacesBeta", "1"); // We just dispatch the page change rather than have to worry about
          // what the logic is for each of these branches.

          if (this.state.page_type === _PageTypes.default.MyGroups) {
            _dispatcher.default.dispatch({
              action: 'view_last_screen'
            });
          } else {
            _dispatcher.default.dispatch({
              action: 'view_my_groups'
            });
          }

          break;

        case 'hide_left_panel':
          this.setState({
            collapseLhs: true
          }, () => {
            this.state.resizeNotifier.notifyLeftHandleResized();
          });
          break;

        case 'focus_room_filter': // for CtrlOrCmd+K to work by expanding the left panel first

        case 'show_left_panel':
          this.setState({
            collapseLhs: false
          }, () => {
            this.state.resizeNotifier.notifyLeftHandleResized();
          });
          break;

        case _actions.Action.OpenDialPad:
          _Modal.default.createTrackedDialog('Dial pad', '', _DialPadModal.default, {}, "mx_Dialog_dialPadWrapper");

          break;

        case 'on_logged_in':
          if ( // Skip this handling for token login as that always calls onLoggedIn itself
          !this.tokenLogin && !Lifecycle.isSoftLogout() && this.state.view !== Views.LOGIN && this.state.view !== Views.REGISTER && this.state.view !== Views.COMPLETE_SECURITY && this.state.view !== Views.E2E_SETUP) {
            this.onLoggedIn();
          }

          break;

        case 'on_client_not_viable':
          this.onSoftLogout();
          break;

        case 'on_logged_out':
          this.onLoggedOut();
          break;

        case 'will_start_client':
          this.setState({
            ready: false
          }, () => {
            // if the client is about to start, we are, by definition, not ready.
            // Set ready to false now, then it'll be set to true when the sync
            // listener we set below fires.
            this.onWillStartClient();
          });
          break;

        case 'client_started':
          this.onClientStarted();
          break;

        case 'send_event':
          this.onSendEvent(payload.room_id, payload.event);
          break;

        case 'aria_hide_main_app':
          this.setState({
            hideToSRUsers: true
          });
          break;

        case 'aria_unhide_main_app':
          this.setState({
            hideToSRUsers: false
          });
          break;

        case 'accept_cookies':
          _SettingsStore.default.setValue("analyticsOptIn", null, _SettingLevel.SettingLevel.DEVICE, true);

          _SettingsStore.default.setValue("showCookieBar", null, _SettingLevel.SettingLevel.DEVICE, false);

          (0, _AnalyticsToast.hideToast)();

          if (_Analytics.default.canEnable()) {
            _Analytics.default.enable();
          }

          if (_CountlyAnalytics.default.instance.canEnable()) {
            _CountlyAnalytics.default.instance.enable(
            /* anonymous = */
            false);
          }

          break;

        case 'reject_cookies':
          _SettingsStore.default.setValue("analyticsOptIn", null, _SettingLevel.SettingLevel.DEVICE, false);

          _SettingsStore.default.setValue("showCookieBar", null, _SettingLevel.SettingLevel.DEVICE, false);

          (0, _AnalyticsToast.hideToast)();
          break;
      }
    });
    (0, _defineProperty2.default)(this, "handleResize", () => {
      const LHS_THRESHOLD = 1000;
      const width = _UIStore.default.instance.windowWidth;

      if (this.prevWindowWidth < LHS_THRESHOLD && width >= LHS_THRESHOLD) {
        _dispatcher.default.dispatch({
          action: 'show_left_panel'
        });
      }

      if (this.prevWindowWidth >= LHS_THRESHOLD && width < LHS_THRESHOLD) {
        _dispatcher.default.dispatch({
          action: 'hide_left_panel'
        });
      }

      this.prevWindowWidth = width;
      this.state.resizeNotifier.notifyWindowResized();
    });
    (0, _defineProperty2.default)(this, "onRegisterClick", () => {
      this.showScreen("register");
    });
    (0, _defineProperty2.default)(this, "onLoginClick", () => {
      this.showScreen("login");
    });
    (0, _defineProperty2.default)(this, "onForgotPasswordClick", () => {
      this.showScreen("forgot_password");
    });
    (0, _defineProperty2.default)(this, "onRegisterFlowComplete", (credentials, password) => {
      return this.onUserCompletedLoginFlow(credentials, password);
    });
    (0, _defineProperty2.default)(this, "onServerConfigChange", serverConfig => {
      this.setState({
        serverConfig
      });
    });
    (0, _defineProperty2.default)(this, "makeRegistrationUrl", params => {
      if (this.props.startingFragmentQueryParams.referrer) {
        params.referrer = this.props.startingFragmentQueryParams.referrer;
      }

      return this.props.makeRegistrationUrl(params);
    });
    (0, _defineProperty2.default)(this, "onUserCompletedLoginFlow", async (credentials, password) => {
      this.accountPassword = password; // self-destruct the password after 5mins

      if (this.accountPasswordTimer !== null) clearTimeout(this.accountPasswordTimer);
      this.accountPasswordTimer = setTimeout(() => {
        this.accountPassword = null;
        this.accountPasswordTimer = null;
      }, 60 * 5 * 1000); // Create and start the client

      await Lifecycle.setLoggedIn(credentials);
      await this.postLoginSetup();

      _performance.default.instance.stop(_performance.PerformanceEntryNames.LOGIN);

      _performance.default.instance.stop(_performance.PerformanceEntryNames.REGISTER);
    });
    (0, _defineProperty2.default)(this, "onCompleteSecurityE2eSetupFinished", () => {
      this.onLoggedIn();
    });
    this.state = {
      view: Views.LOADING,
      collapseLhs: false,
      hideToSRUsers: false,
      syncError: null,
      // If the current syncing status is ERROR, the error object, otherwise null.
      resizeNotifier: new _ResizeNotifier.default(),
      ready: false
    };
    this.loggedInView = /*#__PURE__*/(0, _react.createRef)();

    _SdkConfig.default.put(this.props.config); // Used by _viewRoom before getting state from sync


    this.firstSyncComplete = false;
    this.firstSyncPromise = (0, _utils.defer)();

    if (this.props.config.sync_timeline_limit) {
      _MatrixClientPeg.MatrixClientPeg.opts.initialSyncLimit = this.props.config.sync_timeline_limit;
    } // a thing to call showScreen with once login completes.  this is kept
    // outside this.state because updating it should never trigger a
    // rerender.


    this.screenAfterLogin = this.props.initialScreenAfterLogin;

    if (this.screenAfterLogin) {
      const params = this.screenAfterLogin.params || {};

      if (this.screenAfterLogin.screen.startsWith("room/") && params['signurl'] && params['email']) {
        // probably a threepid invite - try to store it
        const roomId = this.screenAfterLogin.screen.substring("room/".length);

        _ThreepidInviteStore.default.instance.storeInvite(roomId, params);
      }
    }

    this.prevWindowWidth = _UIStore.default.instance.windowWidth || 1000;

    _UIStore.default.instance.on(_UIStore.UI_EVENTS.Resize, this.handleResize);

    this.pageChanging = false; // For PersistentElement

    this.state.resizeNotifier.on("middlePanelResized", this.dispatchTimelineResize); // Force users to go through the soft logout page if they're soft logged out

    if (Lifecycle.isSoftLogout()) {
      // When the session loads it'll be detected as soft logged out and a dispatch
      // will be sent out to say that, triggering this MatrixChat to show the soft
      // logout page.
      Lifecycle.loadSession();
    }

    this.accountPassword = null;
    this.accountPasswordTimer = null;
    this.dispatcherRef = _dispatcher.default.register(this.onAction);
    this.themeWatcher = new _ThemeWatcher.default();
    this.fontWatcher = new _FontWatcher.FontWatcher();
    this.themeWatcher.start();
    this.fontWatcher.start();
    this.focusComposer = false; // object field used for tracking the status info appended to the title tag.
    // we don't do it as react state as i'm scared about triggering needless react refreshes.

    this.subTitleStatus = ''; // this can technically be done anywhere but doing this here keeps all
    // the routing url path logic together.

    if (this.onAliasClick) {
      _linkifyMatrix.default.onAliasClick = this.onAliasClick;
    }

    if (this.onUserClick) {
      _linkifyMatrix.default.onUserClick = this.onUserClick;
    }

    if (this.onGroupClick) {
      _linkifyMatrix.default.onGroupClick = this.onGroupClick;
    } // the first thing to do is to try the token params in the query-string
    // if the session isn't soft logged out (ie: is a clean session being logged in)


    if (!Lifecycle.isSoftLogout()) {
      Lifecycle.attemptTokenLogin(this.props.realQueryParams, this.props.defaultDeviceDisplayName, this.getFragmentAfterLogin()).then(async loggedIn => {
        var _this$props$realQuery;

        if ((_this$props$realQuery = this.props.realQueryParams) !== null && _this$props$realQuery !== void 0 && _this$props$realQuery.loginToken) {
          // remove the loginToken from the URL regardless
          this.props.onTokenLoginCompleted();
        }

        if (loggedIn) {
          this.tokenLogin = true; // Create and start the client

          await Lifecycle.restoreFromLocalStorage({
            ignoreGuest: true
          });
          return this.postLoginSetup();
        } // if the user has followed a login or register link, don't reanimate
        // the old creds, but rather go straight to the relevant page


        const firstScreen = this.screenAfterLogin ? this.screenAfterLogin.screen : null;

        if (firstScreen === 'login' || firstScreen === 'register' || firstScreen === 'forgot_password') {
          this.showScreenAfterLogin();
          return;
        }

        return this.loadSession();
      });
    }

    if (_SettingsStore.default.getValue("analyticsOptIn")) {
      _Analytics.default.enable();
    }

    _PosthogAnalytics.PosthogAnalytics.instance.updateAnonymityFromSettings();

    _PosthogAnalytics.PosthogAnalytics.instance.updatePlatformSuperProperties();

    _CountlyAnalytics.default.instance.enable(
    /* anonymous = */
    true);

    (0, _sentry.initSentry)(_SdkConfig.default.get()["sentry"]);
  }

  async postLoginSetup() {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    const cryptoEnabled = cli.isCryptoEnabled();

    if (!cryptoEnabled) {
      this.onLoggedIn();
    }

    const promisesList = [this.firstSyncPromise.promise];

    if (cryptoEnabled) {
      // wait for the client to finish downloading cross-signing keys for us so we
      // know whether or not we have keys set up on this account
      promisesList.push(cli.downloadKeys([cli.getUserId()]));
    } // Now update the state to say we're waiting for the first sync to complete rather
    // than for the login to finish.


    this.setState({
      pendingInitialSync: true
    });
    await Promise.all(promisesList);

    if (!cryptoEnabled) {
      this.setState({
        pendingInitialSync: false
      });
      return;
    }

    const crossSigningIsSetUp = cli.getStoredCrossSigningForUser(cli.getUserId());

    if (crossSigningIsSetUp) {
      if (_Security.default.SHOW_ENCRYPTION_SETUP_UI === false) {
        this.onLoggedIn();
      } else {
        this.setStateForNewView({
          view: Views.COMPLETE_SECURITY
        });
      }
    } else if (await cli.doesServerSupportUnstableFeature("org.matrix.e2e_cross_signing")) {
      this.setStateForNewView({
        view: Views.E2E_SETUP
      });
    } else {
      this.onLoggedIn();
    }

    this.setState({
      pendingInitialSync: false
    });
  } // TODO: [REACT-WARNING] Replace with appropriate lifecycle stage
  // eslint-disable-next-line


  UNSAFE_componentWillUpdate(props, state) {
    if (this.shouldTrackPageChange(this.state, state)) {
      this.startPageChangeTimer();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.shouldTrackPageChange(prevState, this.state)) {
      const durationMs = this.stopPageChangeTimer();

      _Analytics.default.trackPageChange(durationMs);

      _CountlyAnalytics.default.instance.trackPageChange(durationMs);

      _PosthogAnalytics.PosthogAnalytics.instance.trackPageView(durationMs);
    }

    if (this.focusComposer) {
      _dispatcher.default.fire(_actions.Action.FocusSendMessageComposer);

      this.focusComposer = false;
    }
  }

  componentWillUnmount() {
    Lifecycle.stopMatrixClient();

    _dispatcher.default.unregister(this.dispatcherRef);

    this.themeWatcher.stop();
    this.fontWatcher.stop();

    _UIStore.default.destroy();

    this.state.resizeNotifier.removeListener("middlePanelResized", this.dispatchTimelineResize);
    if (this.accountPasswordTimer !== null) clearTimeout(this.accountPasswordTimer);
  }

  getFallbackHsUrl() {
    if (this.props.serverConfig && this.props.serverConfig.isDefault) {
      return this.props.config.fallback_hs_url;
    } else {
      return null;
    }
  }

  getServerProperties() {
    let props = this.state.serverConfig;
    if (!props) props = this.props.serverConfig; // for unit tests

    if (!props) props = _SdkConfig.default.get()["validated_server_config"];
    return {
      serverConfig: props
    };
  }

  loadSession() {
    // the extra Promise.resolve() ensures that synchronous exceptions hit the same codepath as
    // asynchronous ones.
    return Promise.resolve().then(() => {
      return Lifecycle.loadSession({
        fragmentQueryParams: this.props.startingFragmentQueryParams,
        enableGuest: this.props.enableGuest,
        guestHsUrl: this.getServerProperties().serverConfig.hsUrl,
        guestIsUrl: this.getServerProperties().serverConfig.isUrl,
        defaultDeviceDisplayName: this.props.defaultDeviceDisplayName
      });
    }).then(loadedSession => {
      if (!loadedSession) {
        // fall back to showing the welcome screen... unless we have a 3pid invite pending
        if (_ThreepidInviteStore.default.instance.pickBestInvite()) {
          _dispatcher.default.dispatch({
            action: 'start_registration'
          });
        } else {
          _dispatcher.default.dispatch({
            action: "view_welcome_page"
          });
        }
      } else if (_SettingsStore.default.getValue("analyticsOptIn")) {
        _CountlyAnalytics.default.instance.enable(
        /* anonymous = */
        false);
      }
    }); // Note we don't catch errors from this: we catch everything within
    // loadSession as there's logic there to ask the user if they want
    // to try logging out.
  }

  startPageChangeTimer() {
    _performance.default.instance.start(_performance.PerformanceEntryNames.PAGE_CHANGE);
  }

  stopPageChangeTimer() {
    const perfMonitor = _performance.default.instance;
    perfMonitor.stop(_performance.PerformanceEntryNames.PAGE_CHANGE);
    const entries = perfMonitor.getEntries({
      name: _performance.PerformanceEntryNames.PAGE_CHANGE
    });
    const measurement = entries.pop();
    return measurement ? measurement.duration : null;
  }

  shouldTrackPageChange(prevState, state) {
    return prevState.currentRoomId !== state.currentRoomId || prevState.view !== state.view || prevState.page_type !== state.page_type;
  }

  setStateForNewView(state) {
    if (state.view === undefined) {
      throw new Error("setStateForNewView with no view!");
    }

    const newState = {
      currentUserId: null,
      justRegistered: false
    };
    Object.assign(newState, state);
    this.setState(newState);
  }

  setPage(pageType) {
    this.setState({
      page_type: pageType
    });
  }

  async startRegistration(params) {
    const newState = {
      view: Views.REGISTER
    }; // Only honour params if they are all present, otherwise we reset
    // HS and IS URLs when switching to registration.

    if (params.client_secret && params.session_id && params.hs_url && params.is_url && params.sid) {
      newState.serverConfig = await _AutoDiscoveryUtils.default.validateServerConfigWithStaticUrls(params.hs_url, params.is_url); // If the hs url matches then take the hs name we know locally as it is likely prettier

      const defaultConfig = _SdkConfig.default.get()["validated_server_config"];

      if (defaultConfig && defaultConfig.hsUrl === newState.serverConfig.hsUrl) {
        newState.serverConfig.hsName = defaultConfig.hsName;
        newState.serverConfig.hsNameIsDifferent = defaultConfig.hsNameIsDifferent;
        newState.serverConfig.isDefault = defaultConfig.isDefault;
        newState.serverConfig.isNameResolvable = defaultConfig.isNameResolvable;
      }

      newState.register_client_secret = params.client_secret;
      newState.register_session_id = params.session_id;
      newState.register_id_sid = params.sid;
    }

    this.setStateForNewView(newState);
    _ThemeController.default.isLogin = true;
    this.themeWatcher.recheck();
    this.notifyNewScreen('register');
  } // switch view to the given room
  //
  // @param {Object} roomInfo Object containing data about the room to be joined
  // @param {string=} roomInfo.room_id ID of the room to join. One of room_id or room_alias must be given.
  // @param {string=} roomInfo.room_alias Alias of the room to join. One of room_id or room_alias must be given.
  // @param {boolean=} roomInfo.auto_join If true, automatically attempt to join the room if not already a member.
  // @param {string=} roomInfo.event_id ID of the event in this room to show: this will cause a switch to the
  //                                    context of that particular event.
  // @param {boolean=} roomInfo.highlighted If true, add event_id to the hash of the URL
  //                                        and alter the EventTile to appear highlighted.
  // @param {Object=} roomInfo.threepid_invite Object containing data about the third party
  //                                           we received to join the room, if any.
  // @param {Object=} roomInfo.oob_data Object of additional data about the room
  //                               that has been passed out-of-band (eg.
  //                               room name and avatar from an invite email)


  viewRoom(roomInfo) {
    this.focusComposer = true;

    if (roomInfo.room_alias) {
      _logger.logger.log(`Switching to room alias ${roomInfo.room_alias} at event ` + roomInfo.event_id);
    } else {
      _logger.logger.log(`Switching to room id ${roomInfo.room_id} at event ` + roomInfo.event_id);
    } // Wait for the first sync to complete so that if a room does have an alias,
    // it would have been retrieved.


    let waitFor = Promise.resolve(null);

    if (!this.firstSyncComplete) {
      if (!this.firstSyncPromise) {
        _logger.logger.warn('Cannot view a room before first sync. room_id:', roomInfo.room_id);

        return;
      }

      waitFor = this.firstSyncPromise.promise;
    }

    return waitFor.then(() => {
      let presentedId = roomInfo.room_alias || roomInfo.room_id;

      const room = _MatrixClientPeg.MatrixClientPeg.get().getRoom(roomInfo.room_id);

      if (room) {
        // Not all timeline events are decrypted ahead of time anymore
        // Only the critical ones for a typical UI are
        // This will start the decryption process for all events when a
        // user views a room
        room.decryptAllEvents();
        const theAlias = Rooms.getDisplayAliasForRoom(room);

        if (theAlias) {
          presentedId = theAlias; // Store display alias of the presented room in cache to speed future
          // navigation.

          (0, _RoomAliasCache.storeRoomAliasInCache)(theAlias, room.roomId);
        } // Store this as the ID of the last room accessed. This is so that we can
        // persist which room is being stored across refreshes and browser quits.


        if (localStorage) {
          localStorage.setItem('mx_last_room_id', room.roomId);
        }
      } // If we are redirecting to a Room Alias and it is for the room we already showing then replace history item


      const replaceLast = presentedId[0] === "#" && roomInfo.room_id === this.state.currentRoomId;

      if (roomInfo.event_id && roomInfo.highlighted) {
        presentedId += "/" + roomInfo.event_id;
      }

      this.setState({
        view: Views.LOGGED_IN,
        currentRoomId: roomInfo.room_id || null,
        page_type: _PageTypes.default.RoomView,
        threepidInvite: roomInfo.threepid_invite,
        roomOobData: roomInfo.oob_data,
        forceTimeline: roomInfo.forceTimeline,
        ready: true,
        roomJustCreatedOpts: roomInfo.justCreatedOpts
      }, () => {
        this.notifyNewScreen('room/' + presentedId, replaceLast);
      });
    });
  }

  async viewGroup(payload) {
    const groupId = payload.group_id; // Wait for the first sync to complete

    if (!this.firstSyncComplete) {
      if (!this.firstSyncPromise) {
        _logger.logger.warn('Cannot view a group before first sync. group_id:', groupId);

        return;
      }

      await this.firstSyncPromise.promise;
    }

    this.setState({
      view: Views.LOGGED_IN,
      currentGroupId: groupId,
      currentGroupIsNew: payload.group_is_new
    });
    this.setPage(_PageTypes.default.GroupView);
    this.notifyNewScreen('group/' + groupId);
  }

  viewSomethingBehindModal() {
    if (this.state.view !== Views.LOGGED_IN) {
      this.viewWelcome();
      return;
    }

    if (!this.state.currentGroupId && !this.state.currentRoomId) {
      this.viewHome();
    }
  }

  viewWelcome() {
    if ((0, _pages.shouldUseLoginForWelcome)(_SdkConfig.default.get())) {
      return this.viewLogin();
    }

    this.setStateForNewView({
      view: Views.WELCOME
    });
    this.notifyNewScreen('welcome');
    _ThemeController.default.isLogin = true;
    this.themeWatcher.recheck();
  }

  viewLogin(otherState) {
    this.setStateForNewView(_objectSpread({
      view: Views.LOGIN
    }, otherState));
    this.notifyNewScreen('login');
    _ThemeController.default.isLogin = true;
    this.themeWatcher.recheck();
  }

  viewHome(justRegistered = false) {
    // The home page requires the "logged in" view, so we'll set that.
    this.setStateForNewView({
      view: Views.LOGGED_IN,
      justRegistered,
      currentRoomId: null
    });
    this.setPage(_PageTypes.default.HomePage);
    this.notifyNewScreen('home');
    _ThemeController.default.isLogin = false;
    this.themeWatcher.recheck();
  }

  viewUser(userId, subAction) {
    // Wait for the first sync so that `getRoom` gives us a room object if it's
    // in the sync response
    const waitForSync = this.firstSyncPromise ? this.firstSyncPromise.promise : Promise.resolve();
    waitForSync.then(() => {
      if (subAction === 'chat') {
        this.chatCreateOrReuse(userId);
        return;
      }

      this.notifyNewScreen('user/' + userId);
      this.setState({
        currentUserId: userId
      });
      this.setPage(_PageTypes.default.UserView);
    });
  }

  async createRoom(defaultPublic = false, defaultName) {
    const communityId = _CommunityPrototypeStore.CommunityPrototypeStore.instance.getSelectedCommunityId();

    if (communityId) {
      // double check the user will have permission to associate this room with the community
      if (!_CommunityPrototypeStore.CommunityPrototypeStore.instance.isAdminOf(communityId)) {
        _Modal.default.createTrackedDialog('Pre-failure to create room', '', _ErrorDialog.default, {
          title: (0, _languageHandler._t)("Cannot create rooms in this community"),
          description: (0, _languageHandler._t)("You do not have permission to create rooms in this community.")
        });

        return;
      }
    }

    const modal = _Modal.default.createTrackedDialog('Create Room', '', _CreateRoomDialog.default, {
      defaultPublic,
      defaultName
    });

    const [shouldCreate, opts] = await modal.finished;

    if (shouldCreate) {
      (0, _createRoom.default)(opts);
    }
  }

  chatCreateOrReuse(userId) {
    // Use a deferred action to reshow the dialog once the user has registered
    if (_MatrixClientPeg.MatrixClientPeg.get().isGuest()) {
      // No point in making 2 DMs with welcome bot. This assumes view_set_mxid will
      // result in a new DM with the welcome user.
      if (userId !== this.props.config.welcomeUserId) {
        _dispatcher.default.dispatch({
          action: 'do_after_sync_prepared',
          deferred_action: {
            action: 'view_start_chat_or_reuse',
            user_id: userId
          }
        });
      }

      _dispatcher.default.dispatch({
        action: 'require_registration',
        // If the set_mxid dialog is cancelled, view /welcome because if the
        // browser was pointing at /user/@someone:domain?action=chat, the URL
        // needs to be reset so that they can revisit /user/.. // (and trigger
        // `_chatCreateOrReuse` again)
        go_welcome_on_cancel: true,
        screen_after: {
          screen: `user/${this.props.config.welcomeUserId}`,
          params: {
            action: 'chat'
          }
        }
      });

      return;
    } // TODO: Immutable DMs replaces this


    const client = _MatrixClientPeg.MatrixClientPeg.get();

    const dmRoomMap = new _DMRoomMap.default(client);
    const dmRooms = dmRoomMap.getDMRoomsForUserId(userId);

    if (dmRooms.length > 0) {
      _dispatcher.default.dispatch({
        action: _actions.Action.ViewRoom,
        room_id: dmRooms[0]
      });
    } else {
      _dispatcher.default.dispatch({
        action: 'start_chat',
        user_id: userId
      });
    }
  }

  leaveRoomWarnings(roomId) {
    const roomToLeave = _MatrixClientPeg.MatrixClientPeg.get().getRoom(roomId);

    const isSpace = _SpaceStore.default.spacesEnabled && (roomToLeave === null || roomToLeave === void 0 ? void 0 : roomToLeave.isSpaceRoom()); // Show a warning if there are additional complications.

    const warnings = [];
    const memberCount = roomToLeave.currentState.getJoinedMemberCount();

    if (memberCount === 1) {
      warnings.push( /*#__PURE__*/_react.default.createElement("span", {
        className: "warning",
        key: "only_member_warning"
      }, ' '
      /* Whitespace, otherwise the sentences get smashed together */
      , (0, _languageHandler._t)("You are the only person here. " + "If you leave, no one will be able to join in the future, including you.")));
      return warnings;
    }

    const joinRules = roomToLeave.currentState.getStateEvents('m.room.join_rules', '');

    if (joinRules) {
      const rule = joinRules.getContent().join_rule;

      if (rule !== "public") {
        warnings.push( /*#__PURE__*/_react.default.createElement("span", {
          className: "warning",
          key: "non_public_warning"
        }, ' '
        /* Whitespace, otherwise the sentences get smashed together */
        , isSpace ? (0, _languageHandler._t)("This space is not public. You will not be able to rejoin without an invite.") : (0, _languageHandler._t)("This room is not public. You will not be able to rejoin without an invite.")));
      }
    }

    return warnings;
  }

  leaveRoom(roomId) {
    const roomToLeave = _MatrixClientPeg.MatrixClientPeg.get().getRoom(roomId);

    const warnings = this.leaveRoomWarnings(roomId);
    const isSpace = _SpaceStore.default.spacesEnabled && (roomToLeave === null || roomToLeave === void 0 ? void 0 : roomToLeave.isSpaceRoom());

    _Modal.default.createTrackedDialog(isSpace ? "Leave space" : "Leave room", '', _QuestionDialog.default, {
      title: isSpace ? (0, _languageHandler._t)("Leave space") : (0, _languageHandler._t)("Leave room"),
      description: /*#__PURE__*/_react.default.createElement("span", null, isSpace ? (0, _languageHandler._t)("Are you sure you want to leave the space '%(spaceName)s'?", {
        spaceName: roomToLeave.name
      }) : (0, _languageHandler._t)("Are you sure you want to leave the room '%(roomName)s'?", {
        roomName: roomToLeave.name
      }), warnings),
      button: (0, _languageHandler._t)("Leave"),
      onFinished: shouldLeave => {
        if (shouldLeave) {
          const d = (0, _membership.leaveRoomBehaviour)(roomId); // FIXME: controller shouldn't be loading a view :(

          const modal = _Modal.default.createDialog(_Spinner.default, null, 'mx_Dialog_spinner');

          d.finally(() => modal.close());

          _dispatcher.default.dispatch({
            action: "after_leave_room",
            room_id: roomId
          });
        }
      }
    });
  }

  forgetRoom(roomId) {
    const room = _MatrixClientPeg.MatrixClientPeg.get().getRoom(roomId);

    _MatrixClientPeg.MatrixClientPeg.get().forget(roomId).then(() => {
      // Switch to home page if we're currently viewing the forgotten room
      if (this.state.currentRoomId === roomId) {
        _dispatcher.default.dispatch({
          action: "view_home_page"
        });
      } // We have to manually update the room list because the forgotten room will not
      // be notified to us, therefore the room list will have no other way of knowing
      // the room is forgotten.


      _RoomListStore.default.instance.manualRoomUpdate(room, _models.RoomUpdateCause.RoomRemoved);
    }).catch(err => {
      const errCode = err.errcode || (0, _languageHandler._td)("unknown error code");

      _Modal.default.createTrackedDialog("Failed to forget room", '', _ErrorDialog.default, {
        title: (0, _languageHandler._t)("Failed to forget room %(errCode)s", {
          errCode
        }),
        description: err && err.message ? err.message : (0, _languageHandler._t)("Operation failed")
      });
    });
  }

  async copyRoom(roomId) {
    const roomLink = (0, _Permalinks.makeRoomPermalink)(roomId);
    const success = await (0, _strings.copyPlaintext)(roomLink);

    if (!success) {
      _Modal.default.createTrackedDialog("Unable to copy room link", "", _ErrorDialog.default, {
        title: (0, _languageHandler._t)("Unable to copy room link"),
        description: (0, _languageHandler._t)("Unable to copy a link to the room to the clipboard.")
      });
    }
  }
  /**
   * Starts a chat with the welcome user, if the user doesn't already have one
   * @returns {string} The room ID of the new room, or null if no room was created
   */


  async startWelcomeUserChat() {
    // We can end up with multiple tabs post-registration where the user
    // might then end up with a session and we don't want them all making
    // a chat with the welcome user: try to de-dupe.
    // We need to wait for the first sync to complete for this to
    // work though.
    let waitFor;

    if (!this.firstSyncComplete) {
      waitFor = this.firstSyncPromise.promise;
    } else {
      waitFor = Promise.resolve();
    }

    await waitFor;

    const welcomeUserRooms = _DMRoomMap.default.shared().getDMRoomsForUserId(this.props.config.welcomeUserId);

    if (welcomeUserRooms.length === 0) {
      const roomId = await (0, _createRoom.default)({
        dmUserId: this.props.config.welcomeUserId,
        // Only view the welcome user if we're NOT looking at a room
        andView: !this.state.currentRoomId,
        spinner: false // we're already showing one: we don't need another one

      }); // This is a bit of a hack, but since the deduplication relies
      // on m.direct being up to date, we need to force a sync
      // of the database, otherwise if the user goes to the other
      // tab before the next save happens (a few minutes), the
      // saved sync will be restored from the db and this code will
      // run without the update to m.direct, making another welcome
      // user room (it doesn't wait for new data from the server, just
      // the saved sync to be loaded).

      const saveWelcomeUser = ev => {
        if (ev.getType() === 'm.direct' && ev.getContent() && ev.getContent()[this.props.config.welcomeUserId]) {
          _MatrixClientPeg.MatrixClientPeg.get().store.save(true);

          _MatrixClientPeg.MatrixClientPeg.get().removeListener("accountData", saveWelcomeUser);
        }
      };

      _MatrixClientPeg.MatrixClientPeg.get().on("accountData", saveWelcomeUser);

      return roomId;
    }

    return null;
  }
  /**
   * Called when a new logged in session has started
   */


  async onLoggedIn() {
    _ThemeController.default.isLogin = false;
    this.themeWatcher.recheck();
    this.setStateForNewView({
      view: Views.LOGGED_IN
    }); // If a specific screen is set to be shown after login, show that above
    // all else, as it probably means the user clicked on something already.

    if (this.screenAfterLogin && this.screenAfterLogin.screen) {
      this.showScreen(this.screenAfterLogin.screen, this.screenAfterLogin.params);
      this.screenAfterLogin = null;
    } else if (_MatrixClientPeg.MatrixClientPeg.currentUserIsJustRegistered()) {
      _MatrixClientPeg.MatrixClientPeg.setJustRegisteredUserId(null);

      if (this.props.config.welcomeUserId && (0, _languageHandler.getCurrentLanguage)().startsWith("en")) {
        const welcomeUserRoom = await this.startWelcomeUserChat();

        if (welcomeUserRoom === null) {
          // We didn't redirect to the welcome user room, so show
          // the homepage.
          _dispatcher.default.dispatch({
            action: 'view_home_page',
            justRegistered: true
          });
        }
      } else if (_ThreepidInviteStore.default.instance.pickBestInvite()) {
        // The user has a 3pid invite pending - show them that
        const threepidInvite = _ThreepidInviteStore.default.instance.pickBestInvite(); // HACK: This is a pretty brutal way of threading the invite back through
        // our systems, but it's the safest we have for now.


        const params = _ThreepidInviteStore.default.instance.translateToWireFormat(threepidInvite);

        this.showScreen(`room/${threepidInvite.roomId}`, params);
      } else {
        // The user has just logged in after registering,
        // so show the homepage.
        _dispatcher.default.dispatch({
          action: 'view_home_page',
          justRegistered: true
        });
      }
    } else {
      this.showScreenAfterLogin();
    }

    StorageManager.tryPersistStorage(); // defer the following actions by 30 seconds to not throw them at the user immediately

    await (0, _utils.sleep)(30);

    if (_SettingsStore.default.getValue("showCookieBar") && (_Analytics.default.canEnable() || _CountlyAnalytics.default.instance.canEnable())) {
      var _this$props$config$pi;

      (0, _AnalyticsToast.showToast)((_this$props$config$pi = this.props.config.piwik) === null || _this$props$config$pi === void 0 ? void 0 : _this$props$config$pi.policyUrl);
    }

    if (_SdkConfig.default.get().mobileGuideToast) {
      // The toast contains further logic to detect mobile platforms,
      // check if it has been dismissed before, etc.
      (0, _MobileGuideToast.showToast)();
    }
  }

  showScreenAfterLogin() {
    // If screenAfterLogin is set, use that, then null it so that a second login will
    // result in view_home_page, _user_settings or _room_directory
    if (this.screenAfterLogin && this.screenAfterLogin.screen) {
      this.showScreen(this.screenAfterLogin.screen, this.screenAfterLogin.params);
      this.screenAfterLogin = null;
    } else if (localStorage && localStorage.getItem('mx_last_room_id')) {
      // Before defaulting to directory, show the last viewed room
      this.viewLastRoom();
    } else {
      if (_MatrixClientPeg.MatrixClientPeg.get().isGuest()) {
        _dispatcher.default.dispatch({
          action: 'view_welcome_page'
        });
      } else {
        _dispatcher.default.dispatch({
          action: 'view_home_page'
        });
      }
    }
  }

  viewLastRoom() {
    _dispatcher.default.dispatch({
      action: _actions.Action.ViewRoom,
      room_id: localStorage.getItem('mx_last_room_id')
    });
  }
  /**
   * Called when the session is logged out
   */


  onLoggedOut() {
    this.viewLogin({
      ready: false,
      collapseLhs: false,
      currentRoomId: null
    });
    this.subTitleStatus = '';
    this.setPageSubtitle();
  }
  /**
   * Called when the session is softly logged out
   */


  onSoftLogout() {
    this.notifyNewScreen('soft_logout');
    this.setStateForNewView({
      view: Views.SOFT_LOGOUT,
      ready: false,
      collapseLhs: false,
      currentRoomId: null
    });
    this.subTitleStatus = '';
    this.setPageSubtitle();
  }
  /**
   * Called just before the matrix client is started
   * (useful for setting listeners)
   */


  onWillStartClient() {
    // reset the 'have completed first sync' flag,
    // since we're about to start the client and therefore about
    // to do the first sync
    this.firstSyncComplete = false;
    this.firstSyncPromise = (0, _utils.defer)();

    const cli = _MatrixClientPeg.MatrixClientPeg.get(); // Allow the JS SDK to reap timeline events. This reduces the amount of
    // memory consumed as the JS SDK stores multiple distinct copies of room
    // state (each of which can be 10s of MBs) for each DISJOINT timeline. This is
    // particularly noticeable when there are lots of 'limited' /sync responses
    // such as when laptops unsleep.
    // https://github.com/vector-im/element-web/issues/3307#issuecomment-282895568


    cli.setCanResetTimelineCallback(roomId => {
      _logger.logger.log("Request to reset timeline in room ", roomId, " viewing:", this.state.currentRoomId);

      if (roomId !== this.state.currentRoomId) {
        // It is safe to remove events from rooms we are not viewing.
        return true;
      } // We are viewing the room which we want to reset. It is only safe to do
      // this if we are not scrolled up in the view. To find out, delegate to
      // the timeline panel. If the timeline panel doesn't exist, then we assume
      // it is safe to reset the timeline.


      if (!this.loggedInView.current) {
        return true;
      }

      return this.loggedInView.current.canResetTimelineInRoom(roomId);
    });
    cli.on('sync', (state, prevState, data) => {
      // LifecycleStore and others cannot directly subscribe to matrix client for
      // events because flux only allows store state changes during flux dispatches.
      // So dispatch directly from here. Ideally we'd use a SyncStateStore that
      // would do this dispatch and expose the sync state itself (by listening to
      // its own dispatch).
      _dispatcher.default.dispatch({
        action: 'sync_state',
        prevState,
        state
      });

      if (state === "ERROR" || state === "RECONNECTING") {
        if (data.error instanceof _errors.InvalidStoreError) {
          Lifecycle.handleInvalidStoreError(data.error);
        }

        this.setState({
          syncError: data.error || true
        });
      } else if (this.state.syncError) {
        this.setState({
          syncError: null
        });
      }

      this.updateStatusIndicator(state, prevState);

      if (state === "SYNCING" && prevState === "SYNCING") {
        return;
      }

      _logger.logger.info("MatrixClient sync state => %s", state);

      if (state !== "PREPARED") {
        return;
      }

      this.firstSyncComplete = true;
      this.firstSyncPromise.resolve();

      if (_Notifier.default.shouldShowPrompt() && !_MatrixClientPeg.MatrixClientPeg.userRegisteredWithinLastHours(24)) {
        (0, _DesktopNotificationsToast.showToast)(false);
      }

      _dispatcher.default.fire(_actions.Action.FocusSendMessageComposer);

      this.setState({
        ready: true
      });
    });
    cli.on('Session.logged_out', function (errObj) {
      if (Lifecycle.isLoggingOut()) return; // A modal might have been open when we were logged out by the server

      _Modal.default.closeCurrentModal('Session.logged_out');

      if (errObj.httpStatus === 401 && errObj.data && errObj.data['soft_logout']) {
        _logger.logger.warn("Soft logout issued by server - avoiding data deletion");

        Lifecycle.softLogout();
        return;
      }

      _Modal.default.createTrackedDialog('Signed out', '', _ErrorDialog.default, {
        title: (0, _languageHandler._t)('Signed Out'),
        description: (0, _languageHandler._t)('For security, this session has been signed out. Please sign in again.')
      });

      _dispatcher.default.dispatch({
        action: 'logout'
      });
    });
    cli.on('no_consent', function (message, consentUri) {
      _Modal.default.createTrackedDialog('No Consent Dialog', '', _QuestionDialog.default, {
        title: (0, _languageHandler._t)('Terms and Conditions'),
        description: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, " ", (0, _languageHandler._t)('To continue using the %(homeserverDomain)s homeserver ' + 'you must review and agree to our terms and conditions.', {
          homeserverDomain: cli.getDomain()
        }))),
        button: (0, _languageHandler._t)('Review terms and conditions'),
        cancelButton: (0, _languageHandler._t)('Dismiss'),
        onFinished: confirmed => {
          if (confirmed) {
            const wnd = window.open(consentUri, '_blank');
            wnd.opener = null;
          }
        }
      }, null, true);
    });
    const dft = new _DecryptionFailureTracker.DecryptionFailureTracker((total, errorCode) => {
      _Analytics.default.trackEvent('E2E', 'Decryption failure', errorCode, String(total));

      _CountlyAnalytics.default.instance.track("decryption_failure", {
        errorCode
      }, null, {
        sum: total
      });
    }, errorCode => {
      // Map JS-SDK error codes to tracker codes for aggregation
      switch (errorCode) {
        case 'MEGOLM_UNKNOWN_INBOUND_SESSION_ID':
          return 'olm_keys_not_sent_error';

        case 'OLM_UNKNOWN_MESSAGE_INDEX':
          return 'olm_index_error';

        case undefined:
          return 'unexpected_error';

        default:
          return 'unspecified_error';
      }
    }); // Shelved for later date when we have time to think about persisting history of
    // tracked events across sessions.
    // dft.loadTrackedEventHashMap();

    dft.start(); // When logging out, stop tracking failures and destroy state

    cli.on("Session.logged_out", () => dft.stop());
    cli.on("Event.decrypted", (e, err) => dft.eventDecrypted(e, err));
    cli.on("Room", room => {
      if (_MatrixClientPeg.MatrixClientPeg.get().isCryptoEnabled()) {
        const blacklistEnabled = _SettingsStore.default.getValueAt(_SettingLevel.SettingLevel.ROOM_DEVICE, "blacklistUnverifiedDevices", room.roomId,
        /*explicit=*/
        true);

        room.setBlacklistUnverifiedDevices(blacklistEnabled);
      }
    });
    cli.on("crypto.warning", type => {
      switch (type) {
        case 'CRYPTO_WARNING_OLD_VERSION_DETECTED':
          _Modal.default.createTrackedDialog('Crypto migrated', '', _ErrorDialog.default, {
            title: (0, _languageHandler._t)('Old cryptography data detected'),
            description: (0, _languageHandler._t)("Data from an older version of %(brand)s has been detected. " + "This will have caused end-to-end cryptography to malfunction " + "in the older version. End-to-end encrypted messages exchanged " + "recently whilst using the older version may not be decryptable " + "in this version. This may also cause messages exchanged with this " + "version to fail. If you experience problems, log out and back in " + "again. To retain message history, export and re-import your keys.", {
              brand: _SdkConfig.default.get().brand
            })
          });

          break;
      }
    });
    cli.on("crypto.keyBackupFailed", async errcode => {
      let haveNewVersion;
      let newVersionInfo; // if key backup is still enabled, there must be a new backup in place

      if (_MatrixClientPeg.MatrixClientPeg.get().getKeyBackupEnabled()) {
        haveNewVersion = true;
      } else {
        // otherwise check the server to see if there's a new one
        try {
          newVersionInfo = await _MatrixClientPeg.MatrixClientPeg.get().getKeyBackupVersion();
          if (newVersionInfo !== null) haveNewVersion = true;
        } catch (e) {
          _logger.logger.error("Saw key backup error but failed to check backup version!", e);

          return;
        }
      }

      if (haveNewVersion) {
        _Modal.default.createTrackedDialogAsync('New Recovery Method', 'New Recovery Method', Promise.resolve().then(() => _interopRequireWildcard(require('../../async-components/views/dialogs/security/NewRecoveryMethodDialog'))), {
          newVersionInfo
        });
      } else {
        _Modal.default.createTrackedDialogAsync('Recovery Method Removed', 'Recovery Method Removed', Promise.resolve().then(() => _interopRequireWildcard(require('../../async-components/views/dialogs/security/RecoveryMethodRemovedDialog'))));
      }
    });
    cli.on("crypto.keySignatureUploadFailure", (failures, source, continuation) => {
      _Modal.default.createTrackedDialog('Failed to upload key signatures', 'Failed to upload key signatures', _KeySignatureUploadFailedDialog.default, {
        failures,
        source,
        continuation
      });
    });
    cli.on("crypto.verification.request", request => {
      if (request.verifier) {
        _Modal.default.createTrackedDialog('Incoming Verification', '', _IncomingSasDialog.default, {
          verifier: request.verifier
        }, null,
        /* priority = */
        false,
        /* static = */
        true);
      } else if (request.pending) {
        _ToastStore.default.sharedInstance().addOrReplaceToast({
          key: 'verifreq_' + request.channel.transactionId,
          title: (0, _languageHandler._t)("Verification requested"),
          icon: "verification",
          props: {
            request
          },
          component: _VerificationRequestToast.default,
          priority: 90
        });
      }
    });
  }
  /**
   * Called shortly after the matrix client has started. Useful for
   * setting up anything that requires the client to be started.
   * @private
   */


  onClientStarted() {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    if (cli.isCryptoEnabled()) {
      const blacklistEnabled = _SettingsStore.default.getValueAt(_SettingLevel.SettingLevel.DEVICE, "blacklistUnverifiedDevices");

      cli.setGlobalBlacklistUnverifiedDevices(blacklistEnabled); // With cross-signing enabled, we send to unknown devices
      // without prompting. Any bad-device status the user should
      // be aware of will be signalled through the room shield
      // changing colour. More advanced behaviour will come once
      // we implement more settings.

      cli.setGlobalErrorOnUnknownDevices(false);
    }
  }

  showScreen(screen, params) {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    const isLoggedOutOrGuest = !cli || cli.isGuest();

    if (!isLoggedOutOrGuest && AUTH_SCREENS.includes(screen)) {
      // user is logged in and landing on an auth page which will uproot their session, redirect them home instead
      _dispatcher.default.dispatch({
        action: "view_home_page"
      });

      return;
    }

    if (screen === 'register') {
      _dispatcher.default.dispatch({
        action: 'start_registration',
        params: params
      });

      _performance.default.instance.start(_performance.PerformanceEntryNames.REGISTER);
    } else if (screen === 'login') {
      _dispatcher.default.dispatch({
        action: 'start_login',
        params: params
      });

      _performance.default.instance.start(_performance.PerformanceEntryNames.LOGIN);
    } else if (screen === 'forgot_password') {
      _dispatcher.default.dispatch({
        action: 'start_password_recovery',
        params: params
      });
    } else if (screen === 'soft_logout') {
      if (cli.getUserId() && !Lifecycle.isSoftLogout()) {
        // Logged in - visit a room
        this.viewLastRoom();
      } else {
        // Ultimately triggers soft_logout if needed
        _dispatcher.default.dispatch({
          action: 'start_login',
          params: params
        });
      }
    } else if (screen === 'new') {
      _dispatcher.default.dispatch({
        action: 'view_create_room'
      });
    } else if (screen === 'settings') {
      _dispatcher.default.fire(_actions.Action.ViewUserSettings);
    } else if (screen === 'welcome') {
      _dispatcher.default.dispatch({
        action: 'view_welcome_page'
      });
    } else if (screen === 'home') {
      _dispatcher.default.dispatch({
        action: 'view_home_page'
      });
    } else if (screen === 'start') {
      this.showScreen('home');

      _dispatcher.default.dispatch({
        action: 'require_registration'
      });
    } else if (screen === 'directory') {
      if (this.state.view === Views.WELCOME) {
        _CountlyAnalytics.default.instance.track("onboarding_room_directory");
      }

      _dispatcher.default.fire(_actions.Action.ViewRoomDirectory);
    } else if (screen === "start_sso" || screen === "start_cas") {
      // TODO if logged in, skip SSO
      let cli = _MatrixClientPeg.MatrixClientPeg.get();

      if (!cli) {
        const {
          hsUrl,
          isUrl
        } = this.props.serverConfig;
        cli = (0, _matrix.createClient)({
          baseUrl: hsUrl,
          idBaseUrl: isUrl
        });
      }

      const type = screen === "start_sso" ? "sso" : "cas";

      _PlatformPeg.default.get().startSingleSignOn(cli, type, this.getFragmentAfterLogin());
    } else if (screen === 'groups') {
      if (_SpaceStore.default.spacesEnabled) {
        _dispatcher.default.dispatch({
          action: "view_home_page"
        });

        return;
      }

      _dispatcher.default.dispatch({
        action: 'view_my_groups'
      });
    } else if (screen.indexOf('room/') === 0) {
      var _threepidInvite, _threepidInvite2, _threepidInvite3;

      // Rooms can have the following formats:
      // #room_alias:domain or !opaque_id:domain
      const room = screen.substring(5);
      const domainOffset = room.indexOf(':') + 1; // 0 in case room does not contain a :

      let eventOffset = room.length; // room aliases can contain slashes only look for slash after domain

      if (room.substring(domainOffset).indexOf('/') > -1) {
        eventOffset = domainOffset + room.substring(domainOffset).indexOf('/');
      }

      const roomString = room.substring(0, eventOffset);
      let eventId = room.substring(eventOffset + 1); // empty string if no event id given
      // Previously we pulled the eventID from the segments in such a way
      // where if there was no eventId then we'd get undefined. However, we
      // now do a splice and join to handle v3 event IDs which results in
      // an empty string. To maintain our potential contract with the rest
      // of the app, we coerce the eventId to be undefined where applicable.

      if (!eventId) eventId = undefined; // TODO: Handle encoded room/event IDs: https://github.com/vector-im/element-web/issues/9149

      let threepidInvite; // if we landed here from a 3PID invite, persist it

      if (params.signurl && params.email) {
        threepidInvite = _ThreepidInviteStore.default.instance.storeInvite(roomString, params);
      } // otherwise check that this room doesn't already have a known invite


      if (!threepidInvite) {
        const invites = _ThreepidInviteStore.default.instance.getInvites();

        threepidInvite = invites.find(invite => invite.roomId === roomString);
      } // on our URLs there might be a ?via=matrix.org or similar to help
      // joins to the room succeed. We'll pass these through as an array
      // to other levels. If there's just one ?via= then params.via is a
      // single string. If someone does something like ?via=one.com&via=two.com
      // then params.via is an array of strings.


      let via = [];

      if (params.via) {
        if (typeof params.via === 'string') via = [params.via];else via = params.via;
      }

      const payload = {
        action: _actions.Action.ViewRoom,
        event_id: eventId,
        via_servers: via,
        // If an event ID is given in the URL hash, notify RoomViewStore to mark
        // it as highlighted, which will propagate to RoomView and highlight the
        // associated EventTile.
        highlighted: Boolean(eventId),
        threepid_invite: threepidInvite,
        // TODO: Replace oob_data with the threepidInvite (which has the same info).
        // This isn't done yet because it's threaded through so many more places.
        // See https://github.com/vector-im/element-web/issues/15157
        oob_data: {
          name: (_threepidInvite = threepidInvite) === null || _threepidInvite === void 0 ? void 0 : _threepidInvite.roomName,
          avatarUrl: (_threepidInvite2 = threepidInvite) === null || _threepidInvite2 === void 0 ? void 0 : _threepidInvite2.roomAvatarUrl,
          inviterName: (_threepidInvite3 = threepidInvite) === null || _threepidInvite3 === void 0 ? void 0 : _threepidInvite3.inviterName
        },
        room_alias: undefined,
        room_id: undefined
      };

      if (roomString[0] === '#') {
        payload.room_alias = roomString;
      } else {
        payload.room_id = roomString;
      }

      _dispatcher.default.dispatch(payload);
    } else if (screen.indexOf('user/') === 0) {
      const userId = screen.substring(5);

      _dispatcher.default.dispatch({
        action: 'view_user_info',
        userId: userId,
        subAction: params.action
      });
    } else if (screen.indexOf('group/') === 0) {
      const groupId = screen.substring(6); // TODO: Check valid group ID

      _dispatcher.default.dispatch({
        action: 'view_group',
        group_id: groupId
      });
    } else {
      _logger.logger.info("Ignoring showScreen for '%s'", screen);
    }
  }

  notifyNewScreen(screen, replaceLast = false) {
    if (this.props.onNewScreen) {
      this.props.onNewScreen(screen, replaceLast);
    }

    this.setPageSubtitle();
  }

  onAliasClick(event, alias) {
    event.preventDefault();

    _dispatcher.default.dispatch({
      action: _actions.Action.ViewRoom,
      room_alias: alias
    });
  }

  onUserClick(event, userId) {
    event.preventDefault();
    const member = new _roomMember.RoomMember(null, userId);

    if (!member) {
      return;
    }

    _dispatcher.default.dispatch({
      action: _actions.Action.ViewUser,
      member: member
    });
  }

  onGroupClick(event, groupId) {
    event.preventDefault();

    _dispatcher.default.dispatch({
      action: 'view_group',
      group_id: groupId
    });
  }

  onLogoutClick(event) {
    _dispatcher.default.dispatch({
      action: 'logout'
    });

    event.stopPropagation();
    event.preventDefault();
  }

  dispatchTimelineResize() {
    _dispatcher.default.dispatch({
      action: 'timeline_resize'
    });
  }

  // returns a promise which resolves to the new MatrixClient
  onRegistered(credentials) {
    return Lifecycle.setLoggedIn(credentials);
  }

  onSendEvent(roomId, event) {
    const cli = _MatrixClientPeg.MatrixClientPeg.get();

    if (!cli) return;
    cli.sendEvent(roomId, event.getType(), event.getContent()).then(() => {
      _dispatcher.default.dispatch({
        action: 'message_sent'
      });
    });
  }

  setPageSubtitle(subtitle = '') {
    if (this.state.currentRoomId) {
      const client = _MatrixClientPeg.MatrixClientPeg.get();

      const room = client && client.getRoom(this.state.currentRoomId);

      if (room) {
        subtitle = `${this.subTitleStatus} | ${room.name} ${subtitle}`;
      }
    } else {
      subtitle = `${this.subTitleStatus} ${subtitle}`;
    }

    const title = `${_SdkConfig.default.get().brand} ${subtitle}`;

    if (document.title !== title) {
      document.title = title;
    }
  }

  updateStatusIndicator(state, prevState) {
    const notificationState = _RoomNotificationStateStore.RoomNotificationStateStore.instance.globalState;
    const numUnreadRooms = notificationState.numUnreadStates; // we know that states === rooms here

    if (_PlatformPeg.default.get()) {
      _PlatformPeg.default.get().setErrorStatus(state === 'ERROR');

      _PlatformPeg.default.get().setNotificationCount(numUnreadRooms);
    }

    this.subTitleStatus = '';

    if (state === "ERROR") {
      this.subTitleStatus += `[${(0, _languageHandler._t)("Offline")}] `;
    }

    if (numUnreadRooms > 0) {
      this.subTitleStatus += `[${numUnreadRooms}]`;
    }

    this.setPageSubtitle();
  }

  onCloseAllSettings() {
    _dispatcher.default.dispatch({
      action: 'close_settings'
    });
  }

  getFragmentAfterLogin() {
    let fragmentAfterLogin = "";
    const initialScreenAfterLogin = this.props.initialScreenAfterLogin;

    if (initialScreenAfterLogin && // XXX: workaround for https://github.com/vector-im/element-web/issues/11643 causing a login-loop
    !["welcome", "login", "register", "start_sso", "start_cas"].includes(initialScreenAfterLogin.screen)) {
      fragmentAfterLogin = `/${initialScreenAfterLogin.screen}`;
    }

    return fragmentAfterLogin;
  }

  render() {
    const fragmentAfterLogin = this.getFragmentAfterLogin();
    let view = null;

    if (this.state.view === Views.LOADING) {
      view = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_MatrixChat_splash"
      }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
    } else if (this.state.view === Views.COMPLETE_SECURITY) {
      view = /*#__PURE__*/_react.default.createElement(_CompleteSecurity.default, {
        onFinished: this.onCompleteSecurityE2eSetupFinished
      });
    } else if (this.state.view === Views.E2E_SETUP) {
      view = /*#__PURE__*/_react.default.createElement(_E2eSetup.default, {
        onFinished: this.onCompleteSecurityE2eSetupFinished,
        accountPassword: this.accountPassword,
        tokenLogin: !!this.tokenLogin
      });
    } else if (this.state.view === Views.LOGGED_IN) {
      // store errors stop the client syncing and require user intervention, so we'll
      // be showing a dialog. Don't show anything else.
      const isStoreError = this.state.syncError && this.state.syncError instanceof _errors.InvalidStoreError; // `ready` and `view==LOGGED_IN` may be set before `page_type` (because the
      // latter is set via the dispatcher). If we don't yet have a `page_type`,
      // keep showing the spinner for now.

      if (this.state.ready && this.state.page_type && !isStoreError) {
        /* for now, we stuff the entirety of our props and state into the LoggedInView.
         * we should go through and figure out what we actually need to pass down, as well
         * as using something like redux to avoid having a billion bits of state kicking around.
         */
        view = /*#__PURE__*/_react.default.createElement(_LoggedInView.default, (0, _extends2.default)({}, this.props, this.state, {
          ref: this.loggedInView,
          matrixClient: _MatrixClientPeg.MatrixClientPeg.get(),
          onRegistered: this.onRegistered,
          currentRoomId: this.state.currentRoomId
        }));
      } else {
        // we think we are logged in, but are still waiting for the /sync to complete
        let errorBox;

        if (this.state.syncError && !isStoreError) {
          errorBox = /*#__PURE__*/_react.default.createElement("div", {
            className: "mx_MatrixChat_syncError"
          }, (0, _ErrorUtils.messageForSyncError)(this.state.syncError));
        }

        view = /*#__PURE__*/_react.default.createElement("div", {
          className: "mx_MatrixChat_splash"
        }, errorBox, /*#__PURE__*/_react.default.createElement(_Spinner.default, null), /*#__PURE__*/_react.default.createElement("a", {
          href: "#",
          className: "mx_MatrixChat_splashButtons",
          onClick: this.onLogoutClick
        }, (0, _languageHandler._t)('Logout')));
      }
    } else if (this.state.view === Views.WELCOME) {
      view = /*#__PURE__*/_react.default.createElement(_Welcome.default, null);
    } else if (this.state.view === Views.REGISTER && _SettingsStore.default.getValue(_UIFeature.UIFeature.Registration)) {
      var _ThreepidInviteStore$;

      const email = (_ThreepidInviteStore$ = _ThreepidInviteStore.default.instance.pickBestInvite()) === null || _ThreepidInviteStore$ === void 0 ? void 0 : _ThreepidInviteStore$.toEmail;
      view = /*#__PURE__*/_react.default.createElement(_Registration2.default, (0, _extends2.default)({
        clientSecret: this.state.register_client_secret,
        sessionId: this.state.register_session_id,
        idSid: this.state.register_id_sid,
        email: email,
        brand: this.props.config.brand,
        makeRegistrationUrl: this.makeRegistrationUrl,
        onLoggedIn: this.onRegisterFlowComplete,
        onLoginClick: this.onLoginClick,
        onServerConfigChange: this.onServerConfigChange,
        defaultDeviceDisplayName: this.props.defaultDeviceDisplayName,
        fragmentAfterLogin: fragmentAfterLogin
      }, this.getServerProperties()));
    } else if (this.state.view === Views.FORGOT_PASSWORD && _SettingsStore.default.getValue(_UIFeature.UIFeature.PasswordReset)) {
      view = /*#__PURE__*/_react.default.createElement(_ForgotPassword.default, (0, _extends2.default)({
        onComplete: this.onLoginClick,
        onLoginClick: this.onLoginClick,
        onServerConfigChange: this.onServerConfigChange
      }, this.getServerProperties()));
    } else if (this.state.view === Views.LOGIN) {
      const showPasswordReset = _SettingsStore.default.getValue(_UIFeature.UIFeature.PasswordReset);

      view = /*#__PURE__*/_react.default.createElement(_Login.default, (0, _extends2.default)({
        isSyncing: this.state.pendingInitialSync,
        onLoggedIn: this.onUserCompletedLoginFlow,
        onRegisterClick: this.onRegisterClick,
        fallbackHsUrl: this.getFallbackHsUrl(),
        defaultDeviceDisplayName: this.props.defaultDeviceDisplayName,
        onForgotPasswordClick: showPasswordReset ? this.onForgotPasswordClick : undefined,
        onServerConfigChange: this.onServerConfigChange,
        fragmentAfterLogin: fragmentAfterLogin,
        defaultUsername: this.props.startingFragmentQueryParams.defaultUsername
      }, this.getServerProperties()));
    } else if (this.state.view === Views.SOFT_LOGOUT) {
      view = /*#__PURE__*/_react.default.createElement(_SoftLogout.default, {
        realQueryParams: this.props.realQueryParams,
        onTokenLoginCompleted: this.props.onTokenLoginCompleted,
        fragmentAfterLogin: fragmentAfterLogin
      });
    } else {
      _logger.logger.error(`Unknown view ${this.state.view}`);
    }

    return /*#__PURE__*/_react.default.createElement(_ErrorBoundary.default, null, view);
  }

}, (0, _defineProperty2.default)(_class2, "displayName", "MatrixChat"), (0, _defineProperty2.default)(_class2, "defaultProps", {
  realQueryParams: {},
  startingFragmentQueryParams: {},
  config: {},
  onTokenLoginCompleted: () => {}
}), _temp)) || _class);
exports.default = MatrixChat;

function isLoggedIn() {
  // JRS: Maybe we should move the step that writes this to the window out of
  // `element-web` and into this file? Better yet, we should probably create a
  // store to hold this state.
  // See also https://github.com/vector-im/element-web/issues/15034.
  const app = window.matrixChat;
  return app && app.state.view === Views.LOGGED_IN;
}
