// matrix-react-sdk/lib/components/structures/RightPanel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _GroupStore = _interopRequireDefault(require("../../stores/GroupStore"));

var _RightPanelStorePhases = require("../../stores/RightPanelStorePhases");

var _RightPanelStore = _interopRequireDefault(require("../../stores/RightPanelStore"));

var _MatrixClientContext = _interopRequireDefault(require("../../contexts/MatrixClientContext"));

var _actions = require("../../dispatcher/actions");

var _RoomSummaryCard = _interopRequireDefault(require("../views/right_panel/RoomSummaryCard"));

var _WidgetCard = _interopRequireDefault(require("../views/right_panel/WidgetCard"));

var _replaceableComponent = require("../../utils/replaceableComponent");

var _SettingsStore = _interopRequireDefault(require("../../settings/SettingsStore"));

var _MemberList = _interopRequireDefault(require("../views/rooms/MemberList"));

var _GroupMemberList = _interopRequireDefault(require("../views/groups/GroupMemberList"));

var _GroupRoomList = _interopRequireDefault(require("../views/groups/GroupRoomList"));

var _GroupRoomInfo = _interopRequireDefault(require("../views/groups/GroupRoomInfo"));

var _UserInfo = _interopRequireDefault(require("../views/right_panel/UserInfo"));

var _ThirdPartyMemberInfo = _interopRequireDefault(require("../views/rooms/ThirdPartyMemberInfo"));

var _FilePanel = _interopRequireDefault(require("./FilePanel"));

var _ThreadView = _interopRequireDefault(require("./ThreadView"));

var _ThreadPanel = _interopRequireDefault(require("./ThreadPanel"));

var _NotificationPanel = _interopRequireDefault(require("./NotificationPanel"));

var _PinnedMessagesCard = _interopRequireDefault(require("../views/right_panel/PinnedMessagesCard"));

var _lodash = require("lodash");

var _SpaceStore = _interopRequireDefault(require("../../stores/spaces/SpaceStore"));

var _threads = require("../../dispatcher/dispatch-actions/threads");

var _TimelineCard = _interopRequireDefault(require("../views/right_panel/TimelineCard"));

var _dec, _class, _class2, _temp;

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

let RightPanel = (_dec = (0, _replaceableComponent.replaceableComponent)("structures.RightPanel"), _dec(_class = (_temp = _class2 = class RightPanel extends _react.default.Component {
  constructor(props, context) {
    super(props, context);
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "delayedUpdate", (0, _lodash.throttle)(() => {
      this.forceUpdate();
    }, 500, {
      leading: true,
      trailing: true
    }));
    (0, _defineProperty2.default)(this, "onGroupStoreUpdated", () => {
      this.setState({
        isUserPrivilegedInGroup: _GroupStore.default.isUserPrivileged(this.props.groupId)
      });
    });
    (0, _defineProperty2.default)(this, "onRoomStateMember", (ev, state, member) => {
      if (!this.props.room || member.roomId !== this.props.room.roomId) {
        return;
      } // redraw the badge on the membership list


      if (this.state.phase === _RightPanelStorePhases.RightPanelPhases.RoomMemberList && member.roomId === this.props.room.roomId) {
        this.delayedUpdate();
      } else if (this.state.phase === _RightPanelStorePhases.RightPanelPhases.RoomMemberInfo && member.roomId === this.props.room.roomId && member.userId === this.state.member.userId) {
        // refresh the member info (e.g. new power level)
        this.delayedUpdate();
      }
    });
    (0, _defineProperty2.default)(this, "onAction", payload => {
      const isChangingRoom = payload.action === _actions.Action.ViewRoom && payload.room_id !== this.props.room.roomId;
      const isViewingThread = this.state.phase === _RightPanelStorePhases.RightPanelPhases.ThreadView;

      if (isChangingRoom && isViewingThread) {
        (0, _threads.dispatchShowThreadsPanelEvent)();
      }

      if (payload.action === _actions.Action.AfterRightPanelPhaseChange) {
        this.setState({
          phase: payload.phase,
          groupRoomId: payload.groupRoomId,
          groupId: payload.groupId,
          member: payload.member,
          event: payload.event,
          initialEvent: payload.initialEvent,
          initialEventHighlighted: payload.highlighted,
          verificationRequest: payload.verificationRequest,
          verificationRequestPromise: payload.verificationRequestPromise,
          widgetId: payload.widgetId,
          space: payload.space
        });
      }
    });
    (0, _defineProperty2.default)(this, "onClose", () => {
      // XXX: There are three different ways of 'closing' this panel depending on what state
      // things are in... this knows far more than it should do about the state of the rest
      // of the app and is generally a bit silly.
      if (this.props.member) {
        // If we have a user prop then we're displaying a user from the 'user' page type
        // in LoggedInView, so need to change the page type to close the panel (we switch
        // to the home page which is not obviously the correct thing to do, but I'm not sure
        // anything else is - we could hide the close button altogether?)
        _dispatcher.default.dispatch({
          action: "view_home_page"
        });
      } else if (this.state.phase === _RightPanelStorePhases.RightPanelPhases.EncryptionPanel && this.state.verificationRequest && this.state.verificationRequest.pending) {
        // When the user clicks close on the encryption panel cancel the pending request first if any
        this.state.verificationRequest.cancel();
      } else {
        // the RightPanelStore has no way of knowing which mode room/group it is in, so we handle closing here
        _dispatcher.default.dispatch({
          action: _actions.Action.ToggleRightPanel,
          type: this.props.groupId ? "group" : "room"
        });
      }
    });
    (0, _defineProperty2.default)(this, "onSearchQueryChanged", searchQuery => {
      this.setState({
        searchQuery
      });
    });
    this.state = _objectSpread(_objectSpread({}, _RightPanelStore.default.getSharedInstance().roomPanelPhaseParams), {}, {
      phase: this.getPhaseFromProps(),
      isUserPrivilegedInGroup: null,
      member: this.getUserForPanel(),
      searchQuery: ""
    });
  }

  // Helper function to split out the logic for getPhaseFromProps() and the constructor
  // as both are called at the same time in the constructor.
  getUserForPanel() {
    if (this.state && this.state.member) return this.state.member;

    const lastParams = _RightPanelStore.default.getSharedInstance().roomPanelPhaseParams;

    return this.props.member || lastParams['member'];
  } // gets the current phase from the props and also maybe the store


  getPhaseFromProps() {
    var _this$props$room;

    const rps = _RightPanelStore.default.getSharedInstance();

    const userForPanel = this.getUserForPanel();

    if (this.props.groupId) {
      if (!_RightPanelStorePhases.RIGHT_PANEL_PHASES_NO_ARGS.includes(rps.groupPanelPhase)) {
        _dispatcher.default.dispatch({
          action: _actions.Action.SetRightPanelPhase,
          phase: _RightPanelStorePhases.RightPanelPhases.GroupMemberList
        });

        return _RightPanelStorePhases.RightPanelPhases.GroupMemberList;
      }

      return rps.groupPanelPhase;
    } else if (_SpaceStore.default.spacesEnabled && (_this$props$room = this.props.room) !== null && _this$props$room !== void 0 && _this$props$room.isSpaceRoom() && !_RightPanelStorePhases.RIGHT_PANEL_SPACE_PHASES.includes(rps.roomPanelPhase)) {
      return _RightPanelStorePhases.RightPanelPhases.SpaceMemberList;
    } else if (userForPanel) {
      // XXX FIXME AAAAAARGH: What is going on with this class!? It takes some of its state
      // from its props and some from a store, except if the contents of the store changes
      // while it's mounted in which case it replaces all of its state with that of the store,
      // except it uses a dispatch instead of a normal store listener?
      // Unfortunately rewriting this would almost certainly break showing the right panel
      // in some of the many cases, and I don't have time to re-architect it and test all
      // the flows now, so adding yet another special case so if the store thinks there is
      // a verification going on for the member we're displaying, we show that, otherwise
      // we race if a verification is started while the panel isn't displayed because we're
      // not mounted in time to get the dispatch.
      // Until then, let this code serve as a warning from history.
      if (rps.roomPanelPhaseParams.member && userForPanel.userId === rps.roomPanelPhaseParams.member.userId && rps.roomPanelPhaseParams.verificationRequest) {
        return rps.roomPanelPhase;
      }

      return _RightPanelStorePhases.RightPanelPhases.RoomMemberInfo;
    }

    return rps.roomPanelPhase;
  }

  componentDidMount() {
    this.dispatcherRef = _dispatcher.default.register(this.onAction);
    const cli = this.context;
    cli.on("RoomState.members", this.onRoomStateMember);
    this.initGroupStore(this.props.groupId);
  }

  componentWillUnmount() {
    _dispatcher.default.unregister(this.dispatcherRef);

    if (this.context) {
      this.context.removeListener("RoomState.members", this.onRoomStateMember);
    }

    this.unregisterGroupStore();
  } // TODO: [REACT-WARNING] Replace with appropriate lifecycle event


  UNSAFE_componentWillReceiveProps(newProps) {
    // eslint-disable-line
    if (newProps.groupId !== this.props.groupId) {
      this.unregisterGroupStore();
      this.initGroupStore(newProps.groupId);
    }
  }

  initGroupStore(groupId) {
    if (!groupId) return;

    _GroupStore.default.registerListener(groupId, this.onGroupStoreUpdated);
  }

  unregisterGroupStore() {
    _GroupStore.default.unregisterListener(this.onGroupStoreUpdated);
  }

  render() {
    let panel = /*#__PURE__*/_react.default.createElement("div", null);

    const roomId = this.props.room ? this.props.room.roomId : undefined;

    switch (this.state.phase) {
      case _RightPanelStorePhases.RightPanelPhases.RoomMemberList:
        if (roomId) {
          panel = /*#__PURE__*/_react.default.createElement(_MemberList.default, {
            roomId: roomId,
            key: roomId,
            onClose: this.onClose,
            searchQuery: this.state.searchQuery,
            onSearchQueryChanged: this.onSearchQueryChanged
          });
        }

        break;

      case _RightPanelStorePhases.RightPanelPhases.SpaceMemberList:
        panel = /*#__PURE__*/_react.default.createElement(_MemberList.default, {
          roomId: this.state.space ? this.state.space.roomId : roomId,
          key: this.state.space ? this.state.space.roomId : roomId,
          onClose: this.onClose,
          searchQuery: this.state.searchQuery,
          onSearchQueryChanged: this.onSearchQueryChanged
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.GroupMemberList:
        if (this.props.groupId) {
          panel = /*#__PURE__*/_react.default.createElement(_GroupMemberList.default, {
            groupId: this.props.groupId,
            key: this.props.groupId
          });
        }

        break;

      case _RightPanelStorePhases.RightPanelPhases.GroupRoomList:
        panel = /*#__PURE__*/_react.default.createElement(_GroupRoomList.default, {
          groupId: this.props.groupId,
          key: this.props.groupId
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.RoomMemberInfo:
      case _RightPanelStorePhases.RightPanelPhases.SpaceMemberInfo:
      case _RightPanelStorePhases.RightPanelPhases.EncryptionPanel:
        panel = /*#__PURE__*/_react.default.createElement(_UserInfo.default, {
          user: this.state.member,
          room: this.context.getRoom(this.state.member.roomId) ?? this.props.room,
          key: roomId || this.state.member.userId,
          onClose: this.onClose,
          phase: this.state.phase,
          verificationRequest: this.state.verificationRequest,
          verificationRequestPromise: this.state.verificationRequestPromise
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.Room3pidMemberInfo:
      case _RightPanelStorePhases.RightPanelPhases.Space3pidMemberInfo:
        panel = /*#__PURE__*/_react.default.createElement(_ThirdPartyMemberInfo.default, {
          event: this.state.event,
          key: roomId
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.GroupMemberInfo:
        panel = /*#__PURE__*/_react.default.createElement(_UserInfo.default, {
          user: this.state.member,
          groupId: this.props.groupId,
          key: this.state.member.userId,
          phase: this.state.phase,
          onClose: this.onClose
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.GroupRoomInfo:
        panel = /*#__PURE__*/_react.default.createElement(_GroupRoomInfo.default, {
          groupRoomId: this.state.groupRoomId,
          groupId: this.props.groupId,
          key: this.state.groupRoomId
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.NotificationPanel:
        panel = /*#__PURE__*/_react.default.createElement(_NotificationPanel.default, {
          onClose: this.onClose
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.PinnedMessages:
        if (_SettingsStore.default.getValue("feature_pinning")) {
          panel = /*#__PURE__*/_react.default.createElement(_PinnedMessagesCard.default, {
            room: this.props.room,
            onClose: this.onClose
          });
        }

        break;

      case _RightPanelStorePhases.RightPanelPhases.Timeline:
        if (!_SettingsStore.default.getValue("feature_maximised_widgets")) break;
        panel = /*#__PURE__*/_react.default.createElement(_TimelineCard.default, {
          room: this.props.room,
          resizeNotifier: this.props.resizeNotifier,
          onClose: this.onClose
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.FilePanel:
        panel = /*#__PURE__*/_react.default.createElement(_FilePanel.default, {
          roomId: roomId,
          resizeNotifier: this.props.resizeNotifier,
          onClose: this.onClose
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.ThreadView:
        panel = /*#__PURE__*/_react.default.createElement(_ThreadView.default, {
          room: this.props.room,
          resizeNotifier: this.props.resizeNotifier,
          onClose: this.onClose,
          mxEvent: this.state.event,
          initialEvent: this.state.initialEvent,
          initialEventHighlighted: this.state.initialEventHighlighted,
          permalinkCreator: this.props.permalinkCreator,
          e2eStatus: this.props.e2eStatus
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.ThreadPanel:
        panel = /*#__PURE__*/_react.default.createElement(_ThreadPanel.default, {
          roomId: roomId,
          resizeNotifier: this.props.resizeNotifier,
          onClose: this.onClose,
          permalinkCreator: this.props.permalinkCreator
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.RoomSummary:
        panel = /*#__PURE__*/_react.default.createElement(_RoomSummaryCard.default, {
          room: this.props.room,
          onClose: this.onClose
        });
        break;

      case _RightPanelStorePhases.RightPanelPhases.Widget:
        panel = /*#__PURE__*/_react.default.createElement(_WidgetCard.default, {
          room: this.props.room,
          widgetId: this.state.widgetId,
          onClose: this.onClose
        });
        break;
    }

    return /*#__PURE__*/_react.default.createElement("aside", {
      className: "mx_RightPanel dark-panel",
      id: "mx_RightPanel"
    }, panel);
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _MatrixClientContext.default), _temp)) || _class);
exports.default = RightPanel;
