// matrix-react-sdk/lib/components/structures/SpaceHierarchy.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useRoomHierarchy = exports.showRoom = exports.joinRoom = exports.default = exports.HierarchyLevel = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _roomHierarchy = require("matrix-js-sdk/src/room-hierarchy");

var _event = require("matrix-js-sdk/src/@types/event");

var _classnames = _interopRequireDefault(require("classnames"));

var _lodash = require("lodash");

var _partials = require("matrix-js-sdk/src/@types/partials");

var _dispatcher = _interopRequireDefault(require("../../dispatcher/dispatcher"));

var _languageHandler = require("../../languageHandler");

var _AccessibleButton = _interopRequireDefault(require("../views/elements/AccessibleButton"));

var _Spinner = _interopRequireDefault(require("../views/elements/Spinner"));

var _SearchBox = _interopRequireDefault(require("./SearchBox"));

var _RoomAvatar = _interopRequireDefault(require("../views/avatars/RoomAvatar"));

var _StyledCheckbox = _interopRequireDefault(require("../views/elements/StyledCheckbox"));

var _BaseAvatar = _interopRequireDefault(require("../views/avatars/BaseAvatar"));

var _Media = require("../../customisations/Media");

var _InfoTooltip = _interopRequireDefault(require("../views/elements/InfoTooltip"));

var _TextWithTooltip = _interopRequireDefault(require("../views/elements/TextWithTooltip"));

var _useStateToggle = require("../../hooks/useStateToggle");

var _SpaceStore = require("../../stores/spaces/SpaceStore");

var _AccessibleTooltipButton = _interopRequireDefault(require("../views/elements/AccessibleTooltipButton"));

var _HtmlUtils = require("../../HtmlUtils");

var _useDispatcher = require("../../hooks/useDispatcher");

var _actions = require("../../dispatcher/actions");

var _Keyboard = require("../../Keyboard");

var _RovingTabIndex = require("../../accessibility/RovingTabIndex");

var _RoomDirectory = require("./RoomDirectory");

var _MatrixClientContext = _interopRequireDefault(require("../../contexts/MatrixClientContext"));

var _useEventEmitter = require("../../hooks/useEventEmitter");

var _RoomUpgrade = require("../../utils/RoomUpgrade");

var _RoomViewStore = _interopRequireDefault(require("../../stores/RoomViewStore"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

const Tile = ({
  room,
  suggested,
  selected,
  hasPermissions,
  onToggleClick,
  onViewRoomClick,
  onJoinRoomClick,
  numChildRooms,
  children
}) => {
  var _room$aliases, _joinedRoom$currentSt, _joinedRoom$currentSt2, _joinedRoom$currentSt3;

  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const [joinedRoom, setJoinedRoom] = (0, _react.useState)(() => {
    const cliRoom = cli.getRoom(room.room_id);
    return (cliRoom === null || cliRoom === void 0 ? void 0 : cliRoom.getMyMembership()) === "join" ? cliRoom : null;
  });
  const joinedRoomName = (0, _useEventEmitter.useEventEmitterState)(joinedRoom, "Room.name", room => room === null || room === void 0 ? void 0 : room.name);
  const name = joinedRoomName || room.name || room.canonical_alias || ((_room$aliases = room.aliases) === null || _room$aliases === void 0 ? void 0 : _room$aliases[0]) || (room.room_type === _event.RoomType.Space ? (0, _languageHandler._t)("Unnamed Space") : (0, _languageHandler._t)("Unnamed Room"));
  const [showChildren, toggleShowChildren] = (0, _useStateToggle.useStateToggle)(true);
  const [onFocus, isActive, ref] = (0, _RovingTabIndex.useRovingTabIndex)();
  const [busy, setBusy] = (0, _react.useState)(false);

  const onPreviewClick = ev => {
    ev.preventDefault();
    ev.stopPropagation();
    onViewRoomClick();
  };

  const onJoinClick = async ev => {
    setBusy(true);
    ev.preventDefault();
    ev.stopPropagation();
    onJoinRoomClick().then(() => (0, _RoomUpgrade.awaitRoomDownSync)(cli, room.room_id)).then(setJoinedRoom).finally(() => {
      setBusy(false);
    });
  };

  let button;

  if (busy) {
    button = /*#__PURE__*/_react.default.createElement(_AccessibleTooltipButton.default, {
      disabled: true,
      onClick: onJoinClick,
      kind: "primary_outline",
      onFocus: onFocus,
      tabIndex: isActive ? 0 : -1,
      title: (0, _languageHandler._t)("Joining")
    }, /*#__PURE__*/_react.default.createElement(_Spinner.default, {
      w: 24,
      h: 24
    }));
  } else if (joinedRoom) {
    button = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: onPreviewClick,
      kind: "primary_outline",
      onFocus: onFocus,
      tabIndex: isActive ? 0 : -1
    }, (0, _languageHandler._t)("View"));
  } else {
    button = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: onJoinClick,
      kind: "primary",
      onFocus: onFocus,
      tabIndex: isActive ? 0 : -1
    }, (0, _languageHandler._t)("Join"));
  }

  let checkbox;

  if (onToggleClick) {
    if (hasPermissions) {
      checkbox = /*#__PURE__*/_react.default.createElement(_StyledCheckbox.default, {
        checked: !!selected,
        onChange: onToggleClick,
        tabIndex: isActive ? 0 : -1
      });
    } else {
      checkbox = /*#__PURE__*/_react.default.createElement(_TextWithTooltip.default, {
        tooltip: (0, _languageHandler._t)("You don't have permission"),
        onClick: ev => {
          ev.stopPropagation();
        }
      }, /*#__PURE__*/_react.default.createElement(_StyledCheckbox.default, {
        disabled: true,
        tabIndex: isActive ? 0 : -1
      }));
    }
  }

  let avatar;

  if (joinedRoom) {
    avatar = /*#__PURE__*/_react.default.createElement(_RoomAvatar.default, {
      room: joinedRoom,
      width: 20,
      height: 20
    });
  } else {
    avatar = /*#__PURE__*/_react.default.createElement(_BaseAvatar.default, {
      name: name,
      idName: room.room_id,
      url: room.avatar_url ? (0, _Media.mediaFromMxc)(room.avatar_url).getSquareThumbnailHttp(20) : null,
      width: 20,
      height: 20
    });
  }

  let description = (0, _languageHandler._t)("%(count)s members", {
    count: room.num_joined_members
  });

  if (numChildRooms !== undefined) {
    description += " · " + (0, _languageHandler._t)("%(count)s rooms", {
      count: numChildRooms
    });
  }

  const topic = (joinedRoom === null || joinedRoom === void 0 ? void 0 : (_joinedRoom$currentSt = joinedRoom.currentState) === null || _joinedRoom$currentSt === void 0 ? void 0 : (_joinedRoom$currentSt2 = _joinedRoom$currentSt.getStateEvents(_event.EventType.RoomTopic, "")) === null || _joinedRoom$currentSt2 === void 0 ? void 0 : (_joinedRoom$currentSt3 = _joinedRoom$currentSt2.getContent()) === null || _joinedRoom$currentSt3 === void 0 ? void 0 : _joinedRoom$currentSt3.topic) || room.topic;

  if (topic) {
    description += " · " + topic;
  }

  let joinedSection;

  if (joinedRoom) {
    joinedSection = /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_SpaceHierarchy_roomTile_joined"
    }, (0, _languageHandler._t)("Joined"));
  }

  let suggestedSection;

  if (suggested && (!joinedRoom || hasPermissions)) {
    suggestedSection = /*#__PURE__*/_react.default.createElement(_InfoTooltip.default, {
      tooltip: (0, _languageHandler._t)("This room is suggested as a good one to join")
    }, (0, _languageHandler._t)("Suggested"));
  }

  const content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, avatar, /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceHierarchy_roomTile_name"
  }, name, joinedSection, suggestedSection), /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceHierarchy_roomTile_info",
    ref: e => e && (0, _HtmlUtils.linkifyElement)(e),
    onClick: ev => {
      // prevent clicks on links from bubbling up to the room tile
      if (ev.target.tagName === "A") {
        ev.stopPropagation();
      }
    }
  }, description), /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpaceHierarchy_actions"
  }, button, checkbox));

  let childToggle;
  let childSection;
  let onKeyDown;

  if (children) {
    // the chevron is purposefully a div rather than a button as it should be ignored for a11y
    childToggle = /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)("mx_SpaceHierarchy_subspace_toggle", {
        mx_SpaceHierarchy_subspace_toggle_shown: showChildren
      }),
      onClick: ev => {
        ev.stopPropagation();
        toggleShowChildren();
      }
    });

    if (showChildren) {
      const onChildrenKeyDown = e => {
        if (e.key === _Keyboard.Key.ARROW_LEFT) {
          var _ref$current;

          e.preventDefault();
          e.stopPropagation();
          (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.focus();
        }
      };

      childSection = /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SpaceHierarchy_subspace_children",
        onKeyDown: onChildrenKeyDown,
        role: "group"
      }, children);
    }

    onKeyDown = e => {
      let handled = false;

      switch (e.key) {
        case _Keyboard.Key.ARROW_LEFT:
          if (showChildren) {
            handled = true;
            toggleShowChildren();
          }

          break;

        case _Keyboard.Key.ARROW_RIGHT:
          handled = true;

          if (showChildren) {
            var _ref$current2, _childSection$querySe;

            const childSection = (_ref$current2 = ref.current) === null || _ref$current2 === void 0 ? void 0 : _ref$current2.nextElementSibling;
            childSection === null || childSection === void 0 ? void 0 : (_childSection$querySe = childSection.querySelector(".mx_SpaceHierarchy_roomTile")) === null || _childSection$querySe === void 0 ? void 0 : _childSection$querySe.focus();
          } else {
            toggleShowChildren();
          }

          break;
      }

      if (handled) {
        e.preventDefault();
        e.stopPropagation();
      }
    };
  }

  return /*#__PURE__*/_react.default.createElement("li", {
    className: "mx_SpaceHierarchy_roomTileWrapper",
    role: "treeitem",
    "aria-expanded": children ? showChildren : undefined
  }, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    className: (0, _classnames.default)("mx_SpaceHierarchy_roomTile", {
      mx_SpaceHierarchy_subspace: room.room_type === _event.RoomType.Space,
      mx_SpaceHierarchy_joining: busy
    }),
    onClick: hasPermissions && onToggleClick ? onToggleClick : onPreviewClick,
    onKeyDown: onKeyDown,
    inputRef: ref,
    onFocus: onFocus,
    tabIndex: isActive ? 0 : -1
  }, content, childToggle), childSection);
};

const showRoom = (cli, hierarchy, roomId, roomType) => {
  const room = hierarchy.roomMap.get(roomId); // Don't let the user view a room they won't be able to either peek or join:
  // fail earlier so they don't have to click back to the directory.

  if (cli.isGuest()) {
    if (!room.world_readable && !room.guest_can_join) {
      _dispatcher.default.dispatch({
        action: "require_registration"
      });

      return;
    }
  }

  const roomAlias = (0, _RoomDirectory.getDisplayAliasForRoom)(room) || undefined;

  _dispatcher.default.dispatch({
    action: "view_room",
    should_peek: true,
    _type: "room_directory",
    // instrumentation
    room_alias: roomAlias,
    room_id: room.room_id,
    via_servers: Array.from(hierarchy.viaMap.get(roomId) || []),
    oob_data: {
      avatarUrl: room.avatar_url,
      // XXX: This logic is duplicated from the JS SDK which would normally decide what the name is.
      name: room.name || roomAlias || (0, _languageHandler._t)("Unnamed room"),
      roomType
    }
  });
};

exports.showRoom = showRoom;

const joinRoom = (cli, hierarchy, roomId) => {
  // Don't let the user view a room they won't be able to either peek or join:
  // fail earlier so they don't have to click back to the directory.
  if (cli.isGuest()) {
    _dispatcher.default.dispatch({
      action: "require_registration"
    });

    return;
  }

  const prom = cli.joinRoom(roomId, {
    viaServers: Array.from(hierarchy.viaMap.get(roomId) || [])
  });
  prom.catch(err => {
    _RoomViewStore.default.showJoinRoomError(err, roomId);
  });
  return prom;
};

exports.joinRoom = joinRoom;

const toLocalRoom = (cli, room) => {
  const history = cli.getRoomUpgradeHistory(room.room_id, true);
  const cliRoom = history[history.length - 1];

  if (cliRoom) {
    var _cliRoom$currentState, _cliRoom$currentState2, _cliRoom$currentState3;

    return _objectSpread(_objectSpread({}, room), {}, {
      room_id: cliRoom.roomId,
      room_type: cliRoom.getType(),
      name: cliRoom.name,
      topic: (_cliRoom$currentState = cliRoom.currentState.getStateEvents(_event.EventType.RoomTopic, "")) === null || _cliRoom$currentState === void 0 ? void 0 : _cliRoom$currentState.getContent().topic,
      avatar_url: cliRoom.getMxcAvatarUrl(),
      canonical_alias: cliRoom.getCanonicalAlias(),
      aliases: cliRoom.getAltAliases(),
      world_readable: ((_cliRoom$currentState2 = cliRoom.currentState.getStateEvents(_event.EventType.RoomHistoryVisibility, "")) === null || _cliRoom$currentState2 === void 0 ? void 0 : _cliRoom$currentState2.getContent().history_visibility) === _partials.HistoryVisibility.WorldReadable,
      guest_can_join: ((_cliRoom$currentState3 = cliRoom.currentState.getStateEvents(_event.EventType.RoomGuestAccess, "")) === null || _cliRoom$currentState3 === void 0 ? void 0 : _cliRoom$currentState3.getContent().guest_access) === _partials.GuestAccess.CanJoin,
      num_joined_members: cliRoom.getJoinedMemberCount()
    });
  }

  return room;
};

const HierarchyLevel = ({
  root,
  roomSet,
  hierarchy,
  parents,
  selectedMap,
  onViewRoomClick,
  onJoinRoomClick,
  onToggleClick
}) => {
  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const space = cli.getRoom(root.room_id);
  const hasPermissions = space === null || space === void 0 ? void 0 : space.currentState.maySendStateEvent(_event.EventType.SpaceChild, cli.getUserId());
  const sortedChildren = (0, _lodash.sortBy)(root.children_state, ev => {
    return (0, _SpaceStore.getChildOrder)(ev.content.order, ev.origin_server_ts, ev.state_key);
  });
  const [subspaces, childRooms] = sortedChildren.reduce((result, ev) => {
    const room = hierarchy.roomMap.get(ev.state_key);

    if (room && roomSet.has(room)) {
      result[room.room_type === _event.RoomType.Space ? 0 : 1].push(toLocalRoom(cli, room));
    }

    return result;
  }, [[], []]);
  const newParents = new Set(parents).add(root.room_id);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, (0, _lodash.uniqBy)(childRooms, "room_id").map(room => {
    var _selectedMap$get;

    return /*#__PURE__*/_react.default.createElement(Tile, {
      key: room.room_id,
      room: room,
      suggested: hierarchy.isSuggested(root.room_id, room.room_id),
      selected: selectedMap === null || selectedMap === void 0 ? void 0 : (_selectedMap$get = selectedMap.get(root.room_id)) === null || _selectedMap$get === void 0 ? void 0 : _selectedMap$get.has(room.room_id),
      onViewRoomClick: () => onViewRoomClick(room.room_id, room.room_type),
      onJoinRoomClick: () => onJoinRoomClick(room.room_id),
      hasPermissions: hasPermissions,
      onToggleClick: onToggleClick ? () => onToggleClick(root.room_id, room.room_id) : undefined
    });
  }), subspaces.filter(room => !newParents.has(room.room_id)).map(space => {
    var _selectedMap$get2;

    return /*#__PURE__*/_react.default.createElement(Tile, {
      key: space.room_id,
      room: space,
      numChildRooms: space.children_state.filter(ev => {
        const room = hierarchy.roomMap.get(ev.state_key);
        return room && roomSet.has(room) && !room.room_type;
      }).length,
      suggested: hierarchy.isSuggested(root.room_id, space.room_id),
      selected: selectedMap === null || selectedMap === void 0 ? void 0 : (_selectedMap$get2 = selectedMap.get(root.room_id)) === null || _selectedMap$get2 === void 0 ? void 0 : _selectedMap$get2.has(space.room_id),
      onViewRoomClick: () => onViewRoomClick(space.room_id, _event.RoomType.Space),
      onJoinRoomClick: () => onJoinRoomClick(space.room_id),
      hasPermissions: hasPermissions,
      onToggleClick: onToggleClick ? () => onToggleClick(root.room_id, space.room_id) : undefined
    }, /*#__PURE__*/_react.default.createElement(HierarchyLevel, {
      root: space,
      roomSet: roomSet,
      hierarchy: hierarchy,
      parents: newParents,
      selectedMap: selectedMap,
      onViewRoomClick: onViewRoomClick,
      onJoinRoomClick: onJoinRoomClick,
      onToggleClick: onToggleClick
    }));
  }));
};

exports.HierarchyLevel = HierarchyLevel;
const INITIAL_PAGE_SIZE = 20;

const useRoomHierarchy = space => {
  const [rooms, setRooms] = (0, _react.useState)([]);
  const [hierarchy, setHierarchy] = (0, _react.useState)();
  const resetHierarchy = (0, _react.useCallback)(() => {
    const hierarchy = new _roomHierarchy.RoomHierarchy(space, INITIAL_PAGE_SIZE);
    hierarchy.load().then(() => {
      if (space !== hierarchy.root) return; // discard stale results

      setRooms(hierarchy.rooms);
    });
    setHierarchy(hierarchy);
  }, [space]);
  (0, _react.useEffect)(resetHierarchy, [resetHierarchy]);
  (0, _useDispatcher.useDispatcher)(_dispatcher.default, payload => {
    if (payload.action === _actions.Action.UpdateSpaceHierarchy) {
      setRooms([]); // TODO

      resetHierarchy();
    }
  });
  const loadMore = (0, _react.useCallback)(async pageSize => {
    if (hierarchy.loading || !hierarchy.canLoadMore || hierarchy.noSupport) return;
    await hierarchy.load(pageSize);
    setRooms(hierarchy.rooms);
  }, [hierarchy]);
  const loading = (hierarchy === null || hierarchy === void 0 ? void 0 : hierarchy.loading) ?? true;
  return {
    loading,
    rooms,
    hierarchy,
    loadMore
  };
};

exports.useRoomHierarchy = useRoomHierarchy;

const useIntersectionObserver = callback => {
  const handleObserver = entries => {
    const target = entries[0];

    if (target.isIntersecting) {
      callback();
    }
  };

  const observerRef = (0, _react.useRef)();
  return element => {
    if (observerRef.current) {
      observerRef.current.disconnect();
    } else if (element) {
      observerRef.current = new IntersectionObserver(handleObserver, {
        root: element.parentElement,
        rootMargin: "0px 0px 600px 0px"
      });
    }

    if (observerRef.current && element) {
      observerRef.current.observe(element);
    }
  };
};

const ManageButtons = ({
  hierarchy,
  selected,
  setSelected,
  setError
}) => {
  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const [removing, setRemoving] = (0, _react.useState)(false);
  const [saving, setSaving] = (0, _react.useState)(false);
  const selectedRelations = Array.from(selected.keys()).flatMap(parentId => {
    return [...selected.get(parentId).values()].map(childId => [parentId, childId]);
  });
  const selectionAllSuggested = selectedRelations.every(([parentId, childId]) => {
    return hierarchy.isSuggested(parentId, childId);
  });
  const disabled = !selectedRelations.length || removing || saving;
  let Button = _AccessibleButton.default;
  let props = {};

  if (!selectedRelations.length) {
    Button = _AccessibleTooltipButton.default;
    props = {
      tooltip: (0, _languageHandler._t)("Select a room below first"),
      yOffset: -40
    };
  }

  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(Button, (0, _extends2.default)({}, props, {
    onClick: async () => {
      setRemoving(true);

      try {
        const userId = cli.getUserId();

        for (const [parentId, childId] of selectedRelations) {
          await cli.sendStateEvent(parentId, _event.EventType.SpaceChild, {}, childId); // remove the child->parent relation too, if we have permission to.

          const childRoom = cli.getRoom(childId);
          const parentRelation = childRoom === null || childRoom === void 0 ? void 0 : childRoom.currentState.getStateEvents(_event.EventType.SpaceParent, parentId);

          if (childRoom !== null && childRoom !== void 0 && childRoom.currentState.maySendStateEvent(_event.EventType.SpaceParent, userId) && Array.isArray(parentRelation === null || parentRelation === void 0 ? void 0 : parentRelation.getContent().via)) {
            await cli.sendStateEvent(childId, _event.EventType.SpaceParent, {}, parentId);
          }

          hierarchy.removeRelation(parentId, childId);
        }
      } catch (e) {
        setError((0, _languageHandler._t)("Failed to remove some rooms. Try again later"));
      }

      setRemoving(false);
      setSelected(new Map());
    },
    kind: "danger_outline",
    disabled: disabled
  }), removing ? (0, _languageHandler._t)("Removing...") : (0, _languageHandler._t)("Remove")), /*#__PURE__*/_react.default.createElement(Button, (0, _extends2.default)({}, props, {
    onClick: async () => {
      setSaving(true);

      try {
        for (const [parentId, childId] of selectedRelations) {
          var _hierarchy$getRelatio;

          const suggested = !selectionAllSuggested;
          const existingContent = (_hierarchy$getRelatio = hierarchy.getRelation(parentId, childId)) === null || _hierarchy$getRelatio === void 0 ? void 0 : _hierarchy$getRelatio.content;
          if (!existingContent || existingContent.suggested === suggested) continue;

          const content = _objectSpread(_objectSpread({}, existingContent), {}, {
            suggested: !selectionAllSuggested
          });

          await cli.sendStateEvent(parentId, _event.EventType.SpaceChild, content, childId); // mutate the local state to save us having to refetch the world

          existingContent.suggested = content.suggested;
        }
      } catch (e) {
        setError("Failed to update some suggestions. Try again later");
      }

      setSaving(false);
      setSelected(new Map());
    },
    kind: "primary_outline",
    disabled: disabled
  }), saving ? (0, _languageHandler._t)("Saving...") : selectionAllSuggested ? (0, _languageHandler._t)("Mark as not suggested") : (0, _languageHandler._t)("Mark as suggested")));
};

const SpaceHierarchy = ({
  space,
  initialText = "",
  showRoom,
  additionalButtons
}) => {
  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const [query, setQuery] = (0, _react.useState)(initialText);
  const [selected, setSelected] = (0, _react.useState)(new Map()); // Map<parentId, Set<childId>>

  const {
    loading,
    rooms,
    hierarchy,
    loadMore
  } = useRoomHierarchy(space);
  const filteredRoomSet = (0, _react.useMemo)(() => {
    if (!(rooms !== null && rooms !== void 0 && rooms.length)) return new Set();
    const lcQuery = query.toLowerCase().trim();
    if (!lcQuery) return new Set(rooms);
    const directMatches = rooms.filter(r => {
      var _r$name, _r$topic;

      return ((_r$name = r.name) === null || _r$name === void 0 ? void 0 : _r$name.toLowerCase().includes(lcQuery)) || ((_r$topic = r.topic) === null || _r$topic === void 0 ? void 0 : _r$topic.toLowerCase().includes(lcQuery));
    }); // Walk back up the tree to find all parents of the direct matches to show their place in the hierarchy

    const visited = new Set();
    const queue = [...directMatches.map(r => r.room_id)];

    while (queue.length) {
      var _hierarchy$backRefs$g;

      const roomId = queue.pop();
      visited.add(roomId);
      (_hierarchy$backRefs$g = hierarchy.backRefs.get(roomId)) === null || _hierarchy$backRefs$g === void 0 ? void 0 : _hierarchy$backRefs$g.forEach(parentId => {
        if (!visited.has(parentId)) {
          queue.push(parentId);
        }
      });
    }

    return new Set(rooms.filter(r => visited.has(r.room_id)));
  }, [rooms, hierarchy, query]);
  const [error, setError] = (0, _react.useState)("");
  const loaderRef = useIntersectionObserver(loadMore);

  if (!loading && hierarchy.noSupport) {
    return /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Your server does not support showing space hierarchies."));
  }

  const onKeyDown = (ev, state) => {
    if (ev.key === _Keyboard.Key.ARROW_DOWN && ev.currentTarget.classList.contains("mx_SpaceHierarchy_search")) {
      var _state$refs$, _state$refs$$current;

      (_state$refs$ = state.refs[0]) === null || _state$refs$ === void 0 ? void 0 : (_state$refs$$current = _state$refs$.current) === null || _state$refs$$current === void 0 ? void 0 : _state$refs$$current.focus();
    }
  };

  const onToggleClick = (parentId, childId) => {
    setError("");

    if (!selected.has(parentId)) {
      setSelected(new Map(selected.set(parentId, new Set([childId]))));
      return;
    }

    const parentSet = selected.get(parentId);

    if (!parentSet.has(childId)) {
      setSelected(new Map(selected.set(parentId, new Set([...parentSet, childId]))));
      return;
    }

    parentSet.delete(childId);
    setSelected(new Map(selected.set(parentId, new Set(parentSet))));
  };

  return /*#__PURE__*/_react.default.createElement(_RovingTabIndex.RovingTabIndexProvider, {
    onKeyDown: onKeyDown,
    handleHomeEnd: true,
    handleUpDown: true
  }, ({
    onKeyDownHandler
  }) => {
    let content;

    if (loading && !rooms.length) {
      content = /*#__PURE__*/_react.default.createElement(_Spinner.default, null);
    } else {
      const hasPermissions = (space === null || space === void 0 ? void 0 : space.getMyMembership()) === "join" && space.currentState.maySendStateEvent(_event.EventType.SpaceChild, cli.getUserId());
      let results;

      if (filteredRoomSet.size) {
        results = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(HierarchyLevel, {
          root: hierarchy.roomMap.get(space.roomId),
          roomSet: filteredRoomSet,
          hierarchy: hierarchy,
          parents: new Set(),
          selectedMap: selected,
          onToggleClick: hasPermissions ? onToggleClick : undefined,
          onViewRoomClick: (roomId, roomType) => showRoom(cli, hierarchy, roomId, roomType),
          onJoinRoomClick: roomId => joinRoom(cli, hierarchy, roomId)
        }));
      } else if (!hierarchy.canLoadMore) {
        results = /*#__PURE__*/_react.default.createElement("div", {
          className: "mx_SpaceHierarchy_noResults"
        }, /*#__PURE__*/_react.default.createElement("h3", null, (0, _languageHandler._t)("No results found")), /*#__PURE__*/_react.default.createElement("div", null, (0, _languageHandler._t)("You may want to try a different search or check for typos.")));
      }

      let loader;

      if (hierarchy.canLoadMore) {
        loader = /*#__PURE__*/_react.default.createElement("div", {
          ref: loaderRef
        }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
      }

      content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SpaceHierarchy_listHeader"
      }, /*#__PURE__*/_react.default.createElement("h4", null, query.trim() ? (0, _languageHandler._t)("Results") : (0, _languageHandler._t)("Rooms and spaces")), /*#__PURE__*/_react.default.createElement("span", null, additionalButtons, hasPermissions && /*#__PURE__*/_react.default.createElement(ManageButtons, {
        hierarchy: hierarchy,
        selected: selected,
        setSelected: setSelected,
        setError: setError
      }))), error && /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_SpaceHierarchy_error"
      }, error), /*#__PURE__*/_react.default.createElement("ul", {
        className: "mx_SpaceHierarchy_list",
        onKeyDown: onKeyDownHandler,
        role: "tree",
        "aria-label": (0, _languageHandler._t)("Space")
      }, results), loader);
    }

    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_SearchBox.default, {
      className: "mx_SpaceHierarchy_search mx_textinput_icon mx_textinput_search",
      placeholder: (0, _languageHandler._t)("Search names and descriptions"),
      onSearch: setQuery,
      autoFocus: true,
      initialValue: initialText,
      onKeyDown: onKeyDownHandler
    }), content);
  });
};

var _default = SpaceHierarchy;
exports.default = _default;
