// matrix-react-sdk/lib/components/views/rooms/AppsDrawer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _reResizable = require("re-resizable");

var _AppTile = _interopRequireDefault(require("../elements/AppTile"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var ScalarMessaging = _interopRequireWildcard(require("../../../ScalarMessaging"));

var _WidgetUtils = _interopRequireDefault(require("../../../utils/WidgetUtils"));

var _WidgetEchoStore = _interopRequireDefault(require("../../../stores/WidgetEchoStore"));

var _ResizeHandle = _interopRequireDefault(require("../elements/ResizeHandle"));

var _resizer = _interopRequireDefault(require("../../../resizer/resizer"));

var _percentage = _interopRequireDefault(require("../../../resizer/distributors/percentage"));

var _WidgetLayoutStore = require("../../../stores/widgets/WidgetLayoutStore");

var _numbers = require("../../../utils/numbers");

var _useStateCallback = require("../../../hooks/useStateCallback");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _UIStore = _interopRequireDefault(require("../../../stores/UIStore"));

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let AppsDrawer = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.AppsDrawer"), _dec(_class = (_temp = _class2 = class AppsDrawer extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "resizeContainer", void 0);
    (0, _defineProperty2.default)(this, "resizer", void 0);
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "onIsResizing", resizing => {
      // This one is the vertical, ie. change height of apps drawer
      this.setState({
        resizingVertical: resizing
      });

      if (!resizing) {
        this.relaxResizer();
      }
    });
    (0, _defineProperty2.default)(this, "collectResizer", ref => {
      if (this.resizeContainer) {
        this.resizer.detach();
      }

      if (ref) {
        this.resizer.container = ref;
        this.resizer.attach();
      }

      this.resizeContainer = ref;
      this.loadResizerPreferences();
    });
    (0, _defineProperty2.default)(this, "getAppsHash", apps => apps.map(app => app.id).join("~"));
    (0, _defineProperty2.default)(this, "relaxResizer", () => {
      const distributors = this.resizer.getDistributors(); // relax all items if they had any overconstrained flexboxes

      distributors.forEach(d => d.start());
      distributors.forEach(d => d.finish());
    });
    (0, _defineProperty2.default)(this, "loadResizerPreferences", () => {
      const distributions = _WidgetLayoutStore.WidgetLayoutStore.instance.getResizerDistributions(this.props.room, _WidgetLayoutStore.Container.Top);

      if (this.state.apps && this.topApps().length - 1 === distributions.length) {
        distributions.forEach((size, i) => {
          const distributor = this.resizer.forHandleAt(i);

          if (distributor) {
            distributor.size = size;
            distributor.finish();
          }
        });
      } else if (this.state.apps) {
        const distributors = this.resizer.getDistributors();
        distributors.forEach(d => d.item.clearSize());
        distributors.forEach(d => d.start());
        distributors.forEach(d => d.finish());
      }
    });
    (0, _defineProperty2.default)(this, "onAction", action => {
      const hideWidgetKey = this.props.room.roomId + '_hide_widget_drawer';

      switch (action.action) {
        case 'appsDrawer':
          // Note: these booleans are awkward because localstorage is fundamentally
          // string-based. We also do exact equality on the strings later on.
          if (action.show) {
            localStorage.setItem(hideWidgetKey, "false");
          } else {
            // Store hidden state of widget
            // Don't show if previously hidden
            localStorage.setItem(hideWidgetKey, "true");
          }

          break;
      }
    });
    (0, _defineProperty2.default)(this, "getApps", () => {
      // @ts-ignore
      const appsDict = {};
      appsDict[_WidgetLayoutStore.Container.Top] = _WidgetLayoutStore.WidgetLayoutStore.instance.getContainerWidgets(this.props.room, _WidgetLayoutStore.Container.Top);
      appsDict[_WidgetLayoutStore.Container.Center] = _WidgetLayoutStore.WidgetLayoutStore.instance.getContainerWidgets(this.props.room, _WidgetLayoutStore.Container.Center);
      return appsDict;
    });
    (0, _defineProperty2.default)(this, "topApps", () => this.state.apps[_WidgetLayoutStore.Container.Top]);
    (0, _defineProperty2.default)(this, "centerApps", () => this.state.apps[_WidgetLayoutStore.Container.Center]);
    (0, _defineProperty2.default)(this, "updateApps", () => {
      this.setState({
        apps: this.getApps()
      });
    });
    this.state = {
      apps: this.getApps(),
      resizingVertical: false,
      resizingHorizontal: false,
      resizing: false
    };
    this.resizer = this.createResizer();
    this.props.resizeNotifier.on("isResizing", this.onIsResizing);
  }

  componentDidMount() {
    ScalarMessaging.startListening();

    _WidgetLayoutStore.WidgetLayoutStore.instance.on(_WidgetLayoutStore.WidgetLayoutStore.emissionForRoom(this.props.room), this.updateApps);

    this.dispatcherRef = _dispatcher.default.register(this.onAction);
  }

  componentWillUnmount() {
    ScalarMessaging.stopListening();

    _WidgetLayoutStore.WidgetLayoutStore.instance.off(_WidgetLayoutStore.WidgetLayoutStore.emissionForRoom(this.props.room), this.updateApps);

    if (this.dispatcherRef) _dispatcher.default.unregister(this.dispatcherRef);

    if (this.resizeContainer) {
      this.resizer.detach();
    }

    this.props.resizeNotifier.off("isResizing", this.onIsResizing);
  }

  createResizer() {
    // This is the horizontal one, changing the distribution of the width between the app tiles
    // (ie. a vertical resize handle because, the handle itself is vertical...)
    const classNames = {
      handle: "mx_ResizeHandle",
      vertical: "mx_ResizeHandle_vertical",
      reverse: "mx_ResizeHandle_reverse"
    };
    const collapseConfig = {
      onResizeStart: () => {
        this.resizeContainer.classList.add("mx_AppsDrawer_resizing");
        this.setState({
          resizingHorizontal: true
        });
      },
      onResizeStop: () => {
        this.resizeContainer.classList.remove("mx_AppsDrawer_resizing");

        _WidgetLayoutStore.WidgetLayoutStore.instance.setResizerDistributions(this.props.room, _WidgetLayoutStore.Container.Top, this.topApps().slice(1).map((_, i) => this.resizer.forHandleAt(i).size));

        this.setState({
          resizingHorizontal: false
        });
      }
    }; // pass a truthy container for now, we won't call attach until we update it

    const resizer = new _resizer.default(null, _percentage.default, collapseConfig);
    resizer.setClassNames(classNames);
    return resizer;
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevProps.userId !== this.props.userId || prevProps.room !== this.props.room) {
      // Room has changed, update apps
      this.updateApps();
    } else if (this.getAppsHash(this.topApps()) !== this.getAppsHash(prevState.apps[_WidgetLayoutStore.Container.Top])) {
      this.loadResizerPreferences();
    }
  }

  isResizing() {
    return this.state.resizingVertical || this.state.resizingHorizontal;
  }

  render() {
    if (!this.props.showApps) return /*#__PURE__*/_react.default.createElement("div", null);
    const widgetIsMaxmised = this.centerApps().length > 0;
    const appsToDisplay = widgetIsMaxmised ? this.centerApps() : this.topApps();
    const apps = appsToDisplay.map((app, index, arr) => {
      return /*#__PURE__*/_react.default.createElement(_AppTile.default, {
        key: app.id,
        app: app,
        fullWidth: arr.length < 2,
        room: this.props.room,
        userId: this.props.userId,
        creatorUserId: app.creatorUserId,
        widgetPageTitle: _WidgetUtils.default.getWidgetDataTitle(app),
        waitForIframeLoad: app.waitForIframeLoad,
        pointerEvents: this.isResizing() ? 'none' : undefined
      });
    });

    if (apps.length === 0) {
      return /*#__PURE__*/_react.default.createElement("div", null);
    }

    let spinner;

    if (apps.length === 0 && _WidgetEchoStore.default.roomHasPendingWidgets(this.props.room.roomId, _WidgetUtils.default.getRoomWidgets(this.props.room))) {
      spinner = /*#__PURE__*/_react.default.createElement(_Spinner.default, null);
    }

    const classes = (0, _classnames.default)({
      mx_AppsDrawer: true,
      mx_AppsDrawer_maximise: widgetIsMaxmised,
      mx_AppsDrawer_fullWidth: apps.length < 2,
      mx_AppsDrawer_resizing: this.state.resizing,
      mx_AppsDrawer_2apps: apps.length === 2,
      mx_AppsDrawer_3apps: apps.length === 3
    });

    const appConatiners = /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_AppsContainer",
      ref: this.collectResizer
    }, apps.map((app, i) => {
      if (i < 1) return app;
      return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
        key: app.key
      }, /*#__PURE__*/_react.default.createElement(_ResizeHandle.default, {
        reverse: i > apps.length / 2
      }), app);
    }));

    let drawer;

    if (widgetIsMaxmised) {
      drawer = appConatiners;
    } else {
      drawer = /*#__PURE__*/_react.default.createElement(PersistentVResizer, {
        room: this.props.room,
        minHeight: 100,
        maxHeight: this.props.maxHeight || !widgetIsMaxmised ? this.props.maxHeight - 50 : undefined,
        handleClass: "mx_AppsContainer_resizerHandle",
        handleWrapperClass: "mx_AppsContainer_resizerHandleContainer",
        className: "mx_AppsContainer_resizer",
        resizeNotifier: this.props.resizeNotifier
      }, appConatiners);
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: classes
    }, drawer, spinner);
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  showApps: true
}), _temp)) || _class);
exports.default = AppsDrawer;

const PersistentVResizer = ({
  room,
  minHeight,
  maxHeight,
  className,
  handleWrapperClass,
  handleClass,
  resizeNotifier,
  children
}) => {
  let defaultHeight = _WidgetLayoutStore.WidgetLayoutStore.instance.getContainerHeight(room, _WidgetLayoutStore.Container.Top); // Arbitrary defaults to avoid NaN problems. 100 px or 3/4 of the visible window.


  if (!minHeight) minHeight = 100;
  if (!maxHeight) maxHeight = _UIStore.default.instance.windowHeight / 4 * 3; // Convert from percentage to height. Note that the default height is 280px.

  if (defaultHeight) {
    defaultHeight = (0, _numbers.clamp)(defaultHeight, 0, 100);
    defaultHeight = (0, _numbers.percentageWithin)(defaultHeight / 100, minHeight, maxHeight);
  } else {
    defaultHeight = 280;
  }

  const [height, setHeight] = (0, _useStateCallback.useStateCallback)(defaultHeight, newHeight => {
    newHeight = (0, _numbers.percentageOf)(newHeight, minHeight, maxHeight) * 100;

    _WidgetLayoutStore.WidgetLayoutStore.instance.setContainerHeight(room, _WidgetLayoutStore.Container.Top, newHeight);
  });
  return /*#__PURE__*/_react.default.createElement(_reResizable.Resizable, {
    size: {
      height: Math.min(height, maxHeight),
      width: undefined
    },
    minHeight: minHeight,
    maxHeight: maxHeight,
    onResizeStart: () => {
      resizeNotifier.startResizing();
    },
    onResize: () => {
      resizeNotifier.notifyTimelineHeightChanged();
    },
    onResizeStop: (e, dir, ref, d) => {
      setHeight(height + d.height);
      resizeNotifier.stopResizing();
    },
    handleWrapperClass: handleWrapperClass,
    handleClasses: {
      bottom: handleClass
    },
    className: className,
    enable: {
      bottom: true
    }
  }, children);
};
