// matrix-react-sdk/lib/components/views/rooms/EditMessageComposer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _event = require("matrix-js-sdk/src/models/event");

var _languageHandler = require("../../../languageHandler");

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _model = _interopRequireDefault(require("../../../editor/model"));

var _dom = require("../../../editor/dom");

var _serialize = require("../../../editor/serialize");

var _EventUtils = require("../../../utils/EventUtils");

var _deserialize = require("../../../editor/deserialize");

var _parts = require("../../../editor/parts");

var _BasicMessageComposer = _interopRequireWildcard(require("./BasicMessageComposer"));

var _SlashCommands = require("../../../SlashCommands");

var _actions = require("../../../dispatcher/actions");

var _CountlyAnalytics = _interopRequireDefault(require("../../../CountlyAnalytics"));

var _KeyBindingsManager = require("../../../KeyBindingsManager");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _SendHistoryManager = _interopRequireDefault(require("../../../SendHistoryManager"));

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _event2 = require("matrix-js-sdk/src/@types/event");

var _ErrorDialog = _interopRequireDefault(require("../dialogs/ErrorDialog"));

var _QuestionDialog = _interopRequireDefault(require("../dialogs/QuestionDialog"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _ConfirmRedactDialog = require("../dialogs/ConfirmRedactDialog");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _logger = require("matrix-js-sdk/src/logger");

var _MatrixClientContext = require("../../../contexts/MatrixClientContext");

var _RoomContext = _interopRequireDefault(require("../../../contexts/RoomContext"));

var _ComposerInsertPayload = require("../../../dispatcher/payloads/ComposerInsertPayload");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function getHtmlReplyFallback(mxEvent) {
  const html = mxEvent.getContent().formatted_body;

  if (!html) {
    return "";
  }

  const rootNode = new DOMParser().parseFromString(html, "text/html").body;
  const mxReply = rootNode.querySelector("mx-reply");
  return mxReply && mxReply.outerHTML || "";
}

function getTextReplyFallback(mxEvent) {
  const body = mxEvent.getContent().body;
  const lines = body.split("\n").map(l => l.trim());

  if (lines.length > 2 && lines[0].startsWith("> ") && lines[1].length === 0) {
    return `${lines[0]}\n\n`;
  }

  return "";
}

function createEditContent(model, editedEvent) {
  const isEmote = (0, _serialize.containsEmote)(model);

  if (isEmote) {
    model = (0, _serialize.stripEmoteCommand)(model);
  }

  const isReply = !!editedEvent.replyEventId;
  let plainPrefix = "";
  let htmlPrefix = "";

  if (isReply) {
    plainPrefix = getTextReplyFallback(editedEvent);
    htmlPrefix = getHtmlReplyFallback(editedEvent);
  }

  const body = (0, _serialize.textSerialize)(model);
  const newContent = {
    "msgtype": isEmote ? _event2.MsgType.Emote : _event2.MsgType.Text,
    "body": body
  };
  const contentBody = {
    msgtype: newContent.msgtype,
    body: `${plainPrefix} * ${body}`
  };
  const formattedBody = (0, _serialize.htmlSerializeIfNeeded)(model, {
    forceHTML: isReply
  });

  if (formattedBody) {
    newContent.format = "org.matrix.custom.html";
    newContent.formatted_body = formattedBody;
    contentBody.format = newContent.format;
    contentBody.formatted_body = `${htmlPrefix} * ${formattedBody}`;
  }

  const relation = {
    "m.new_content": newContent,
    "m.relates_to": {
      "rel_type": "m.replace",
      "event_id": editedEvent.getId()
    }
  };
  return Object.assign(relation, contentBody);
}

let EditMessageComposer = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.EditMessageComposer"), _dec(_class = (_temp = _class2 = class EditMessageComposer extends _react.default.Component {
  constructor(props, context) {
    super(props);
    (0, _defineProperty2.default)(this, "context", void 0);
    (0, _defineProperty2.default)(this, "editorRef", /*#__PURE__*/(0, _react.createRef)());
    (0, _defineProperty2.default)(this, "dispatcherRef", void 0);
    (0, _defineProperty2.default)(this, "model", null);
    (0, _defineProperty2.default)(this, "onKeyDown", event => {
      var _this$editorRef$curre;

      // ignore any keypress while doing IME compositions
      if ((_this$editorRef$curre = this.editorRef.current) !== null && _this$editorRef$curre !== void 0 && _this$editorRef$curre.isComposing(event)) {
        return;
      }

      const action = (0, _KeyBindingsManager.getKeyBindingsManager)().getMessageComposerAction(event);

      switch (action) {
        case _KeyBindingsManager.MessageComposerAction.Send:
          this.sendEdit();
          event.preventDefault();
          break;

        case _KeyBindingsManager.MessageComposerAction.CancelEditing:
          this.cancelEdit();
          break;

        case _KeyBindingsManager.MessageComposerAction.EditPrevMessage:
          {
            var _this$editorRef$curre2, _this$editorRef$curre3;

            if ((_this$editorRef$curre2 = this.editorRef.current) !== null && _this$editorRef$curre2 !== void 0 && _this$editorRef$curre2.isModified() || !((_this$editorRef$curre3 = this.editorRef.current) !== null && _this$editorRef$curre3 !== void 0 && _this$editorRef$curre3.isCaretAtStart())) {
              return;
            }

            const previousEvent = (0, _EventUtils.findEditableEvent)({
              events: this.events,
              isForward: false,
              fromEventId: this.props.editState.getEvent().getId()
            });

            if (previousEvent) {
              _dispatcher.default.dispatch({
                action: _actions.Action.EditEvent,
                event: previousEvent,
                timelineRenderingType: this.context.timelineRenderingType
              });

              event.preventDefault();
            }

            break;
          }

        case _KeyBindingsManager.MessageComposerAction.EditNextMessage:
          {
            var _this$editorRef$curre4, _this$editorRef$curre5;

            if ((_this$editorRef$curre4 = this.editorRef.current) !== null && _this$editorRef$curre4 !== void 0 && _this$editorRef$curre4.isModified() || !((_this$editorRef$curre5 = this.editorRef.current) !== null && _this$editorRef$curre5 !== void 0 && _this$editorRef$curre5.isCaretAtEnd())) {
              return;
            }

            const nextEvent = (0, _EventUtils.findEditableEvent)({
              events: this.events,
              isForward: true,
              fromEventId: this.props.editState.getEvent().getId()
            });

            if (nextEvent) {
              _dispatcher.default.dispatch({
                action: _actions.Action.EditEvent,
                event: nextEvent,
                timelineRenderingType: this.context.timelineRenderingType
              });
            } else {
              this.clearStoredEditorState();

              _dispatcher.default.dispatch({
                action: _actions.Action.EditEvent,
                event: null,
                timelineRenderingType: this.context.timelineRenderingType
              });

              _dispatcher.default.dispatch({
                action: _actions.Action.FocusSendMessageComposer,
                context: this.context.timelineRenderingType
              });
            }

            event.preventDefault();
            break;
          }
      }
    });
    (0, _defineProperty2.default)(this, "cancelEdit", () => {
      this.clearStoredEditorState();

      _dispatcher.default.dispatch({
        action: _actions.Action.EditEvent,
        event: null,
        timelineRenderingType: this.context.timelineRenderingType
      });

      _dispatcher.default.dispatch({
        action: _actions.Action.FocusSendMessageComposer,
        context: this.context.timelineRenderingType
      });
    });
    (0, _defineProperty2.default)(this, "saveStoredEditorState", () => {
      const item = _SendHistoryManager.default.createItem(this.model);

      this.clearPreviousEdit();
      localStorage.setItem(this.editorRoomKey, this.props.editState.getEvent().getId());
      localStorage.setItem(this.editorStateKey, JSON.stringify(item));
    });
    (0, _defineProperty2.default)(this, "sendEdit", async () => {
      const startTime = _CountlyAnalytics.default.getTimestamp();

      const editedEvent = this.props.editState.getEvent(); // Replace emoticon at the end of the message

      if (_SettingsStore.default.getValue('MessageComposerInput.autoReplaceEmoji')) {
        var _this$editorRef$curre6, _this$editorRef$curre7;

        const caret = (_this$editorRef$curre6 = this.editorRef.current) === null || _this$editorRef$curre6 === void 0 ? void 0 : _this$editorRef$curre6.getCaret();
        const position = this.model.positionForOffset(caret.offset, caret.atNodeEnd);
        (_this$editorRef$curre7 = this.editorRef.current) === null || _this$editorRef$curre7 === void 0 ? void 0 : _this$editorRef$curre7.replaceEmoticon(position, _BasicMessageComposer.REGEX_EMOTICON);
      }

      const editContent = createEditContent(this.model, editedEvent);
      const newContent = editContent["m.new_content"];
      let shouldSend = true;

      if ((newContent === null || newContent === void 0 ? void 0 : newContent.body) === '') {
        this.cancelPreviousPendingEdit();
        (0, _ConfirmRedactDialog.createRedactEventDialog)({
          mxEvent: editedEvent
        });
        return;
      } // If content is modified then send an updated event into the room


      if (this.isContentModified(newContent)) {
        const roomId = editedEvent.getRoomId();

        if (!(0, _serialize.containsEmote)(this.model) && this.isSlashCommand()) {
          const [cmd, args, commandText] = this.getSlashCommand();

          if (cmd) {
            if (cmd.category === _SlashCommands.CommandCategories.messages) {
              editContent["m.new_content"] = await this.runSlashCommand(cmd, args, roomId);
            } else {
              this.runSlashCommand(cmd, args, roomId);
              shouldSend = false;
            }
          } else {
            // ask the user if their unknown command should be sent as a message
            const {
              finished
            } = _Modal.default.createTrackedDialog("Unknown command", "", _QuestionDialog.default, {
              title: (0, _languageHandler._t)("Unknown Command"),
              description: /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Unrecognised command: %(commandText)s", {
                commandText
              })), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("You can use <code>/help</code> to list available commands. " + "Did you mean to send this as a message?", {}, {
                code: t => /*#__PURE__*/_react.default.createElement("code", null, t)
              })), /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Hint: Begin your message with <code>//</code> to start it with a slash.", {}, {
                code: t => /*#__PURE__*/_react.default.createElement("code", null, t)
              }))),
              button: (0, _languageHandler._t)('Send as message')
            });

            const [sendAnyway] = await finished; // if !sendAnyway bail to let the user edit the composer and try again

            if (!sendAnyway) return;
          }
        }

        if (shouldSend) {
          this.cancelPreviousPendingEdit();
          const event = this.props.editState.getEvent();
          const threadId = event.threadRootId || null;
          const prom = this.props.mxClient.sendMessage(roomId, threadId, editContent);
          this.clearStoredEditorState();

          _dispatcher.default.dispatch({
            action: "message_sent"
          });

          _CountlyAnalytics.default.instance.trackSendMessage(startTime, prom, roomId, true, false, editContent);
        }
      } // close the event editing and focus composer


      _dispatcher.default.dispatch({
        action: _actions.Action.EditEvent,
        event: null,
        timelineRenderingType: this.context.timelineRenderingType
      });

      _dispatcher.default.dispatch({
        action: _actions.Action.FocusSendMessageComposer,
        context: this.context.timelineRenderingType
      });
    });
    (0, _defineProperty2.default)(this, "onChange", () => {
      var _this$editorRef$curre8;

      if (!this.state.saveDisabled || !((_this$editorRef$curre8 = this.editorRef.current) !== null && _this$editorRef$curre8 !== void 0 && _this$editorRef$curre8.isModified())) {
        return;
      }

      this.setState({
        saveDisabled: false
      });
    });
    (0, _defineProperty2.default)(this, "onAction", payload => {
      if (!this.editorRef.current) return;

      if (payload.action === _actions.Action.ComposerInsert) {
        if (payload.timelineRenderingType !== this.context.timelineRenderingType) return;
        if (payload.composerType !== _ComposerInsertPayload.ComposerType.Edit) return;

        if (payload.userId) {
          var _this$editorRef$curre9;

          (_this$editorRef$curre9 = this.editorRef.current) === null || _this$editorRef$curre9 === void 0 ? void 0 : _this$editorRef$curre9.insertMention(payload.userId);
        } else if (payload.event) {
          var _this$editorRef$curre10;

          (_this$editorRef$curre10 = this.editorRef.current) === null || _this$editorRef$curre10 === void 0 ? void 0 : _this$editorRef$curre10.insertQuotedMessage(payload.event);
        } else if (payload.text) {
          var _this$editorRef$curre11;

          (_this$editorRef$curre11 = this.editorRef.current) === null || _this$editorRef$curre11 === void 0 ? void 0 : _this$editorRef$curre11.insertPlaintext(payload.text);
        }
      } else if (payload.action === _actions.Action.FocusEditMessageComposer) {
        this.editorRef.current.focus();
      }
    });
    this.context = context; // otherwise React will only set it prior to render due to type def above

    const isRestored = this.createEditorModel();
    const ev = this.props.editState.getEvent();

    const _editContent = createEditContent(this.model, ev);

    this.state = {
      saveDisabled: !isRestored || !this.isContentModified(_editContent["m.new_content"])
    };
    window.addEventListener("beforeunload", this.saveStoredEditorState);
    this.dispatcherRef = _dispatcher.default.register(this.onAction);
  }

  getRoom() {
    return this.props.mxClient.getRoom(this.props.editState.getEvent().getRoomId());
  }

  get editorRoomKey() {
    return `mx_edit_room_${this.getRoom().roomId}_${this.context.timelineRenderingType}`;
  }

  get editorStateKey() {
    return `mx_edit_state_${this.props.editState.getEvent().getId()}`;
  }

  get events() {
    const liveTimelineEvents = this.context.liveTimeline.getEvents();
    const pendingEvents = this.getRoom().getPendingEvents();
    const isInThread = Boolean(this.props.editState.getEvent().getThread());
    return liveTimelineEvents.concat(isInThread ? [] : pendingEvents);
  }

  get shouldSaveStoredEditorState() {
    return localStorage.getItem(this.editorRoomKey) !== null;
  }

  restoreStoredEditorState(partCreator) {
    const json = localStorage.getItem(this.editorStateKey);

    if (json) {
      try {
        const {
          parts: serializedParts
        } = JSON.parse(json);
        const parts = serializedParts.map(p => partCreator.deserializePart(p));
        return parts;
      } catch (e) {
        _logger.logger.error("Error parsing editing state: ", e);
      }
    }
  }

  clearStoredEditorState() {
    localStorage.removeItem(this.editorRoomKey);
    localStorage.removeItem(this.editorStateKey);
  }

  clearPreviousEdit() {
    if (localStorage.getItem(this.editorRoomKey)) {
      localStorage.removeItem(`mx_edit_state_${localStorage.getItem(this.editorRoomKey)}`);
    }
  }

  isSlashCommand() {
    const parts = this.model.parts;
    const firstPart = parts[0];

    if (firstPart) {
      if (firstPart.type === _parts.Type.Command && firstPart.text.startsWith("/") && !firstPart.text.startsWith("//")) {
        return true;
      }

      if (firstPart.text.startsWith("/") && !firstPart.text.startsWith("//") && (firstPart.type === _parts.Type.Plain || firstPart.type === _parts.Type.PillCandidate)) {
        return true;
      }
    }

    return false;
  }

  isContentModified(newContent) {
    // if nothing has changed then bail
    const oldContent = this.props.editState.getEvent().getContent();

    if (oldContent["msgtype"] === newContent["msgtype"] && oldContent["body"] === newContent["body"] && oldContent["format"] === newContent["format"] && oldContent["formatted_body"] === newContent["formatted_body"]) {
      return false;
    }

    return true;
  }

  getSlashCommand() {
    const commandText = this.model.parts.reduce((text, part) => {
      // use mxid to textify user pills in a command
      if (part.type === _parts.Type.UserPill) {
        return text + part.resourceId;
      }

      return text + part.text;
    }, "");
    const {
      cmd,
      args
    } = (0, _SlashCommands.getCommand)(commandText);
    return [cmd, args, commandText];
  }

  async runSlashCommand(cmd, args, roomId) {
    var _this$props$editState, _this$props$editState2, _this$props$editState3;

    const threadId = ((_this$props$editState = this.props.editState) === null || _this$props$editState === void 0 ? void 0 : (_this$props$editState2 = _this$props$editState.getEvent()) === null || _this$props$editState2 === void 0 ? void 0 : (_this$props$editState3 = _this$props$editState2.getThread()) === null || _this$props$editState3 === void 0 ? void 0 : _this$props$editState3.id) || null;
    const result = cmd.run(roomId, threadId, args);
    let messageContent;
    let error = result.error;

    if (result.promise) {
      try {
        if (cmd.category === _SlashCommands.CommandCategories.messages) {
          messageContent = await result.promise;
        } else {
          await result.promise;
        }
      } catch (err) {
        error = err;
      }
    }

    if (error) {
      _logger.logger.error("Command failure: %s", error); // assume the error is a server error when the command is async


      const isServerError = !!result.promise;
      const title = isServerError ? (0, _languageHandler._td)("Server error") : (0, _languageHandler._td)("Command error");
      let errText;

      if (typeof error === 'string') {
        errText = error;
      } else if (error.message) {
        errText = error.message;
      } else {
        errText = (0, _languageHandler._t)("Server unavailable, overloaded, or something else went wrong.");
      }

      _Modal.default.createTrackedDialog(title, '', _ErrorDialog.default, {
        title: (0, _languageHandler._t)(title),
        description: errText
      });
    } else {
      _logger.logger.log("Command success.");

      if (messageContent) return messageContent;
    }
  }

  cancelPreviousPendingEdit() {
    const originalEvent = this.props.editState.getEvent();
    const previousEdit = originalEvent.replacingEvent();

    if (previousEdit && (previousEdit.status === _event.EventStatus.QUEUED || previousEdit.status === _event.EventStatus.NOT_SENT)) {
      this.props.mxClient.cancelPendingEvent(previousEdit);
    }
  }

  componentWillUnmount() {
    // store caret and serialized parts in the
    // editorstate so it can be restored when the remote echo event tile gets rendered
    // in case we're currently editing a pending event
    const sel = document.getSelection();
    let caret;

    if (sel.focusNode) {
      var _this$editorRef$curre12;

      caret = (0, _dom.getCaretOffsetAndText)((_this$editorRef$curre12 = this.editorRef.current) === null || _this$editorRef$curre12 === void 0 ? void 0 : _this$editorRef$curre12.editorRef.current, sel).caret;
    }

    const parts = this.model.serializeParts(); // if caret is undefined because for some reason there isn't a valid selection,
    // then when mounting the editor again with the same editor state,
    // it will set the cursor at the end.

    this.props.editState.setEditorState(caret, parts);
    window.removeEventListener("beforeunload", this.saveStoredEditorState);

    if (this.shouldSaveStoredEditorState) {
      this.saveStoredEditorState();
    }

    _dispatcher.default.unregister(this.dispatcherRef);
  }

  createEditorModel() {
    const {
      editState
    } = this.props;
    const room = this.getRoom();
    const partCreator = new _parts.CommandPartCreator(room, this.props.mxClient);
    let parts;
    let isRestored = false;

    if (editState.hasEditorState()) {
      // if restoring state from a previous editor,
      // restore serialized parts from the state
      parts = editState.getSerializedParts().map(p => partCreator.deserializePart(p));
    } else {
      // otherwise, either restore serialized parts from localStorage or parse the body of the event
      const restoredParts = this.restoreStoredEditorState(partCreator);
      parts = restoredParts || (0, _deserialize.parseEvent)(editState.getEvent(), partCreator);
      isRestored = !!restoredParts;
    }

    this.model = new _model.default(parts, partCreator);
    this.saveStoredEditorState();
    return isRestored;
  }

  render() {
    var _this$props$editState4, _this$props$editState5, _this$props$editState6;

    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)("mx_EditMessageComposer", this.props.className),
      onKeyDown: this.onKeyDown
    }, /*#__PURE__*/_react.default.createElement(_BasicMessageComposer.default, {
      ref: this.editorRef,
      model: this.model,
      room: this.getRoom(),
      threadId: (_this$props$editState4 = this.props.editState) === null || _this$props$editState4 === void 0 ? void 0 : (_this$props$editState5 = _this$props$editState4.getEvent()) === null || _this$props$editState5 === void 0 ? void 0 : (_this$props$editState6 = _this$props$editState5.getThread()) === null || _this$props$editState6 === void 0 ? void 0 : _this$props$editState6.id,
      initialCaret: this.props.editState.getCaret(),
      label: (0, _languageHandler._t)("Edit message"),
      onChange: this.onChange
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_EditMessageComposer_buttons"
    }, /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      kind: "secondary",
      onClick: this.cancelEdit
    }, (0, _languageHandler._t)("Cancel")), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      kind: "primary",
      onClick: this.sendEdit,
      disabled: this.state.saveDisabled
    }, (0, _languageHandler._t)("Save"))));
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _RoomContext.default), _temp)) || _class);
const EditMessageComposerWithMatrixClient = (0, _MatrixClientContext.withMatrixClientHOC)(EditMessageComposer);
var _default = EditMessageComposerWithMatrixClient;
exports.default = _default;
