// matrix-react-sdk/lib/components/views/rooms/LinkPreviewGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _useStateToggle = require("../../../hooks/useStateToggle");

var _LinkPreviewWidget = _interopRequireDefault(require("./LinkPreviewWidget"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _languageHandler = require("../../../languageHandler");

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _useAsyncMemo = require("../../../hooks/useAsyncMemo");

var _logger = require("matrix-js-sdk/src/logger");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const INITIAL_NUM_PREVIEWS = 2;

const LinkPreviewGroup = ({
  links,
  mxEvent,
  onCancelClick,
  onHeightChanged
}) => {
  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const [expanded, toggleExpanded] = (0, _useStateToggle.useStateToggle)();
  const ts = mxEvent.getTs();
  const previews = (0, _useAsyncMemo.useAsyncMemo)(async () => {
    return fetchPreviews(cli, links, ts);
  }, [links, ts], []);
  (0, _react.useEffect)(() => {
    onHeightChanged();
  }, [onHeightChanged, expanded, previews]);
  const showPreviews = expanded ? previews : previews.slice(0, INITIAL_NUM_PREVIEWS);
  let toggleButton;

  if (previews.length > INITIAL_NUM_PREVIEWS) {
    toggleButton = /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: toggleExpanded
    }, expanded ? (0, _languageHandler._t)("Collapse") : (0, _languageHandler._t)("Show %(count)s other previews", {
      count: previews.length - showPreviews.length
    }));
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_LinkPreviewGroup"
  }, showPreviews.map(([link, preview], i) => /*#__PURE__*/_react.default.createElement(_LinkPreviewWidget.default, {
    key: link,
    link: link,
    preview: preview,
    mxEvent: mxEvent
  }, i === 0 ? /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
    className: "mx_LinkPreviewGroup_hide",
    onClick: onCancelClick,
    "aria-label": (0, _languageHandler._t)("Close preview")
  }, /*#__PURE__*/_react.default.createElement("img", {
    className: "mx_filterFlipColor",
    alt: "",
    role: "presentation",
    src: require("../../../../res/img/cancel.svg"),
    width: "18",
    height: "18"
  })) : undefined)), toggleButton);
};

const fetchPreviews = (cli, links, ts) => {
  return Promise.all(links.map(async link => {
    try {
      const preview = await cli.getUrlPreview(link, ts);

      if (preview && Object.keys(preview).length > 0) {
        return [link, preview];
      }
    } catch (error) {
      _logger.logger.error("Failed to get URL preview: " + error);
    }
  })).then(a => a.filter(Boolean));
};

var _default = LinkPreviewGroup;
exports.default = _default;
