// matrix-react-sdk/lib/components/views/rooms/ReplyPreview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _ReplyTile = _interopRequireDefault(require("./ReplyTile"));

var _RoomContext = _interopRequireDefault(require("../../../contexts/RoomContext"));

var _dec, _class, _class2, _temp;

function cancelQuoting(context) {
  _dispatcher.default.dispatch({
    action: 'reply_to_event',
    event: null,
    context
  });
}

let ReplyPreview = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.ReplyPreview"), _dec(_class = (_temp = _class2 = class ReplyPreview extends _react.default.Component {
  render() {
    if (!this.props.replyToEvent) return null;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview_section"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview_header mx_ReplyPreview_title"
    }, (0, _languageHandler._t)('Replying')), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview_header mx_ReplyPreview_cancel"
    }, /*#__PURE__*/_react.default.createElement("img", {
      className: "mx_filterFlipColor",
      src: require("../../../../res/img/cancel.svg"),
      width: "18",
      height: "18",
      onClick: () => cancelQuoting(this.context.timelineRenderingType)
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview_clear"
    }), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReplyPreview_tile"
    }, /*#__PURE__*/_react.default.createElement(_ReplyTile.default, {
      mxEvent: this.props.replyToEvent,
      permalinkCreator: this.props.permalinkCreator
    }))));
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _RoomContext.default), _temp)) || _class);
exports.default = ReplyPreview;
