// matrix-react-sdk/lib/components/views/rooms/RoomDetailList.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _actions = require("../../../dispatcher/actions");

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _RoomDetailRow = _interopRequireDefault(require("./RoomDetailRow"));

var _dec, _class;

let RoomDetailList = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.RoomDetailList"), _dec(_class = class RoomDetailList extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onDetailsClick", (ev, room) => {
      _dispatcher.default.dispatch({
        action: _actions.Action.ViewRoom,
        room_id: room.roomId,
        room_alias: room.getCanonicalAlias() || (room.getAltAliases() || [])[0]
      });
    });
  }

  getRows() {
    if (!this.props.rooms) return [];
    return this.props.rooms.map((room, index) => {
      return /*#__PURE__*/_react.default.createElement(_RoomDetailRow.default, {
        key: index,
        room: room,
        onClick: this.onDetailsClick
      });
    });
  }

  render() {
    const rows = this.getRows();
    let rooms;

    if (rows.length === 0) {
      rooms = /*#__PURE__*/_react.default.createElement("i", null, (0, _languageHandler._t)('No rooms to show'));
    } else {
      rooms = /*#__PURE__*/_react.default.createElement("table", {
        className: "mx_RoomDirectory_table"
      }, /*#__PURE__*/_react.default.createElement("tbody", null, this.getRows()));
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: (0, _classnames.default)("mx_RoomDetailList", this.props.className)
    }, rooms);
  }

}) || _class);
exports.default = RoomDetailList;
