// matrix-react-sdk/lib/components/views/rooms/SearchResultTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _RoomContext = _interopRequireDefault(require("../../../contexts/RoomContext"));

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _UIFeature = require("../../../settings/UIFeature");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _DateSeparator = _interopRequireDefault(require("../messages/DateSeparator"));

var _EventTile = _interopRequireWildcard(require("./EventTile"));

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let SearchResultTile = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.SearchResultTile"), _dec(_class = (_temp = _class2 = class SearchResultTile extends _react.default.Component {
  render() {
    const result = this.props.searchResult;
    const mxEv = result.context.getEvent();
    const eventId = mxEv.getId();
    const ts1 = mxEv.getTs();
    const ret = [/*#__PURE__*/_react.default.createElement(_DateSeparator.default, {
      key: ts1 + "-search",
      ts: ts1
    })];

    const layout = _SettingsStore.default.getValue("layout");

    const isTwelveHour = _SettingsStore.default.getValue("showTwelveHourTimestamps");

    const alwaysShowTimestamps = _SettingsStore.default.getValue("alwaysShowTimestamps");

    const enableFlair = _SettingsStore.default.getValue(_UIFeature.UIFeature.Flair);

    const timeline = result.context.getTimeline();

    for (let j = 0; j < timeline.length; j++) {
      var _this$context;

      const ev = timeline[j];
      let highlights;
      const contextual = j != result.context.getOurEventIndex();

      if (!contextual) {
        highlights = this.props.searchHighlights;
      }

      if ((0, _EventTile.haveTileForEvent)(ev, (_this$context = this.context) === null || _this$context === void 0 ? void 0 : _this$context.showHiddenEventsInTimeline)) {
        ret.push( /*#__PURE__*/_react.default.createElement(_EventTile.default, {
          key: `${eventId}+${j}`,
          mxEvent: ev,
          layout: layout,
          contextual: contextual,
          highlights: highlights,
          permalinkCreator: this.props.permalinkCreator,
          highlightLink: this.props.resultLink,
          onHeightChanged: this.props.onHeightChanged,
          isTwelveHour: isTwelveHour,
          alwaysShowTimestamps: alwaysShowTimestamps,
          enableFlair: enableFlair
        }));
      }
    }

    return /*#__PURE__*/_react.default.createElement("li", {
      "data-scroll-tokens": eventId
    }, ret);
  }

}, (0, _defineProperty2.default)(_class2, "contextType", _RoomContext.default), _temp)) || _class);
exports.default = SearchResultTile;
