// matrix-react-sdk/lib/components/views/rooms/WhoIsTypingTile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var WhoIsTyping = _interopRequireWildcard(require("../../../WhoIsTyping"));

var _Timer = _interopRequireDefault(require("../../../utils/Timer"));

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _MemberAvatar = _interopRequireDefault(require("../avatars/MemberAvatar"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _strings = require("../../../utils/strings");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let WhoIsTypingTile = (_dec = (0, _replaceableComponent.replaceableComponent)("views.rooms.WhoIsTypingTile"), _dec(_class = (_temp = _class2 = class WhoIsTypingTile extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      usersTyping: WhoIsTyping.usersTypingApartFromMe(this.props.room),
      delayedStopTypingTimers: {}
    });
    (0, _defineProperty2.default)(this, "isVisible", () => {
      return WhoIsTypingTile.isVisible(this.state);
    });
    (0, _defineProperty2.default)(this, "onRoomTimeline", (event, room) => {
      var _this$props$room;

      if ((room === null || room === void 0 ? void 0 : room.roomId) === ((_this$props$room = this.props.room) === null || _this$props$room === void 0 ? void 0 : _this$props$room.roomId)) {
        const userId = event.getSender(); // remove user from usersTyping

        const usersTyping = this.state.usersTyping.filter(m => m.userId !== userId);

        if (usersTyping.length !== this.state.usersTyping.length) {
          this.setState({
            usersTyping
          });
        } // abort timer if any


        this.abortUserTimer(userId);
      }
    });
    (0, _defineProperty2.default)(this, "onRoomMemberTyping", () => {
      const usersTyping = WhoIsTyping.usersTypingApartFromMeAndIgnored(this.props.room);
      this.setState({
        delayedStopTypingTimers: this.updateDelayedStopTypingTimers(usersTyping),
        usersTyping
      });
    });
  }

  componentDidMount() {
    _MatrixClientPeg.MatrixClientPeg.get().on("RoomMember.typing", this.onRoomMemberTyping);

    _MatrixClientPeg.MatrixClientPeg.get().on("Room.timeline", this.onRoomTimeline);
  }

  componentDidUpdate(_, prevState) {
    const wasVisible = WhoIsTypingTile.isVisible(prevState);
    const isVisible = WhoIsTypingTile.isVisible(this.state);

    if (this.props.onShown && !wasVisible && isVisible) {
      this.props.onShown();
    } else if (this.props.onHidden && wasVisible && !isVisible) {
      this.props.onHidden();
    }
  }

  componentWillUnmount() {
    // we may have entirely lost our client as we're logging out before clicking login on the guest bar...
    const client = _MatrixClientPeg.MatrixClientPeg.get();

    if (client) {
      client.removeListener("RoomMember.typing", this.onRoomMemberTyping);
      client.removeListener("Room.timeline", this.onRoomTimeline);
    }

    Object.values(this.state.delayedStopTypingTimers).forEach(t => t.abort());
  }

  static isVisible(state) {
    return state.usersTyping.length !== 0 || Object.keys(state.delayedStopTypingTimers).length !== 0;
  }

  updateDelayedStopTypingTimers(usersTyping) {
    const usersThatStoppedTyping = this.state.usersTyping.filter(a => {
      return !usersTyping.some(b => a.userId === b.userId);
    });
    const usersThatStartedTyping = usersTyping.filter(a => {
      return !this.state.usersTyping.some(b => a.userId === b.userId);
    }); // abort all the timers for the users that started typing again

    usersThatStartedTyping.forEach(m => {
      const timer = this.state.delayedStopTypingTimers[m.userId];

      if (timer) {
        timer.abort();
      }
    }); // prepare new delayedStopTypingTimers object to update state with

    let delayedStopTypingTimers = Object.assign({}, this.state.delayedStopTypingTimers); // remove members that started typing again

    delayedStopTypingTimers = usersThatStartedTyping.reduce((delayedStopTypingTimers, m) => {
      delete delayedStopTypingTimers[m.userId];
      return delayedStopTypingTimers;
    }, delayedStopTypingTimers); // start timer for members that stopped typing

    delayedStopTypingTimers = usersThatStoppedTyping.reduce((delayedStopTypingTimers, m) => {
      if (!delayedStopTypingTimers[m.userId]) {
        const timer = new _Timer.default(5000);
        delayedStopTypingTimers[m.userId] = timer;
        timer.start();
        timer.finished().then(() => this.removeUserTimer(m.userId), // on elapsed
        () => {
          /* aborted */
        });
      }

      return delayedStopTypingTimers;
    }, delayedStopTypingTimers);
    return delayedStopTypingTimers;
  }

  abortUserTimer(userId) {
    const timer = this.state.delayedStopTypingTimers[userId];

    if (timer) {
      timer.abort();
      this.removeUserTimer(userId);
    }
  }

  removeUserTimer(userId) {
    const timer = this.state.delayedStopTypingTimers[userId];

    if (timer) {
      const delayedStopTypingTimers = Object.assign({}, this.state.delayedStopTypingTimers);
      delete delayedStopTypingTimers[userId];
      this.setState({
        delayedStopTypingTimers
      });
    }
  }

  renderTypingIndicatorAvatars(users, limit) {
    let othersCount = 0;

    if (users.length > limit) {
      othersCount = users.length - limit + 1;
      users = users.slice(0, limit - 1);
    }

    const avatars = users.map(u => {
      return /*#__PURE__*/_react.default.createElement(_MemberAvatar.default, {
        key: u.userId,
        member: u,
        width: 24,
        height: 24,
        resizeMethod: "crop",
        viewUserOnClick: true,
        "aria-live": "off"
      });
    });

    if (othersCount > 0) {
      avatars.push( /*#__PURE__*/_react.default.createElement("span", {
        className: "mx_WhoIsTypingTile_remainingAvatarPlaceholder",
        key: "others"
      }, "+", othersCount));
    }

    return avatars;
  }

  render() {
    let usersTyping = this.state.usersTyping;
    const stoppedUsersOnTimer = Object.keys(this.state.delayedStopTypingTimers).map(userId => this.props.room.getMember(userId)); // append the users that have been reported not typing anymore
    // but have a timeout timer running so they can disappear
    // when a message comes in

    usersTyping = usersTyping.concat(stoppedUsersOnTimer); // sort them so the typing members don't change order when
    // moved to delayedStopTypingTimers

    usersTyping.sort((a, b) => (0, _strings.compare)(a.name, b.name));
    const typingString = WhoIsTyping.whoIsTypingString(usersTyping, this.props.whoIsTypingLimit);

    if (!typingString) {
      return null;
    }

    return /*#__PURE__*/_react.default.createElement("li", {
      className: "mx_WhoIsTypingTile",
      "aria-atomic": "true"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_WhoIsTypingTile_avatars"
    }, this.renderTypingIndicatorAvatars(usersTyping, this.props.whoIsTypingLimit)), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_WhoIsTypingTile_label"
    }, typingString));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  whoIsTypingLimit: 3
}), _temp)) || _class);
exports.default = WhoIsTypingTile;
