// matrix-react-sdk/lib/stores/CommunityPrototypeStore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CommunityPrototypeStore = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _AsyncStoreWithClient = require("./AsyncStoreWithClient");

var _dispatcher = _interopRequireDefault(require("../dispatcher/dispatcher"));

var _actions = require("../dispatcher/actions");

var _membership = require("../utils/membership");

var _SettingsStore = _interopRequireDefault(require("../settings/SettingsStore"));

var utils = _interopRequireWildcard(require("matrix-js-sdk/src/utils"));

var _AsyncStore = require("./AsyncStore");

var _FlairStore = _interopRequireDefault(require("./FlairStore"));

var _GroupFilterOrderStore = _interopRequireDefault(require("./GroupFilterOrderStore"));

var _GroupStore = _interopRequireDefault(require("./GroupStore"));

var _logger = require("matrix-js-sdk/src/logger");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class CommunityPrototypeStore extends _AsyncStoreWithClient.AsyncStoreWithClient {
  constructor() {
    super(_dispatcher.default, {});
  }

  static get instance() {
    return CommunityPrototypeStore.internalInstance;
  }

  static getUpdateEventName(roomId) {
    return `${_AsyncStore.UPDATE_EVENT}:${roomId}`;
  }

  getSelectedCommunityId() {
    if (_SettingsStore.default.getValue("feature_communities_v2_prototypes")) {
      return _GroupFilterOrderStore.default.getSelectedTags()[0];
    }

    return null; // no selection as far as this function is concerned
  }

  getSelectedCommunityName() {
    return CommunityPrototypeStore.instance.getCommunityName(this.getSelectedCommunityId());
  }

  getSelectedCommunityGeneralChat() {
    const communityId = this.getSelectedCommunityId();

    if (communityId) {
      return this.getGeneralChat(communityId);
    }
  }

  getCommunityName(communityId) {
    const profile = _FlairStore.default.getGroupProfileCachedFast(this.matrixClient, communityId);

    return (profile === null || profile === void 0 ? void 0 : profile.name) || communityId;
  }

  getCommunityProfile(communityId) {
    return _FlairStore.default.getGroupProfileCachedFast(this.matrixClient, communityId);
  }

  getGeneralChat(communityId) {
    const rooms = _GroupStore.default.getGroupRooms(communityId).map(r => this.matrixClient.getRoom(r.roomId)).filter(r => !!r);

    let chat = rooms.find(r => {
      const idState = r.currentState.getStateEvents("im.vector.general_chat", "");
      if (!idState || idState.getContent()['groupId'] !== communityId) return false;
      return true;
    });
    if (!chat) chat = rooms[0];
    return chat; // can be null
  }

  isAdminOf(communityId) {
    const members = _GroupStore.default.getGroupMembers(communityId);

    const myMember = members.find(m => m.userId === this.matrixClient.getUserId());
    return myMember === null || myMember === void 0 ? void 0 : myMember.isPrivileged;
  }

  canInviteTo(communityId) {
    const generalChat = this.getGeneralChat(communityId);
    if (!generalChat) return this.isAdminOf(communityId);
    const myMember = generalChat.getMember(this.matrixClient.getUserId());
    if (!myMember) return this.isAdminOf(communityId);
    const pl = generalChat.currentState.getStateEvents("m.room.power_levels", "");
    if (!pl) return this.isAdminOf(communityId);
    const plContent = pl.getContent();
    const invitePl = (0, utils.isNullOrUndefined)(plContent.invite) ? 50 : Number(plContent.invite);
    return invitePl <= myMember.powerLevel;
  }

  async onAction(payload) {
    if (!this.matrixClient || !_SettingsStore.default.getValue("feature_communities_v2_prototypes")) {
      return;
    }

    if (payload.action === "MatrixActions.Room.myMembership") {
      const room = payload.room;
      const membership = (0, _membership.getEffectiveMembership)(payload.membership);
      const oldMembership = (0, _membership.getEffectiveMembership)(payload.oldMembership);
      if (membership === oldMembership) return;

      if (membership === _membership.EffectiveMembership.Invite) {
        try {
          const path = utils.encodeUri("/rooms/$roomId/group_info", {
            $roomId: room.roomId
          });
          const profile = await this.matrixClient.http.authedRequest(undefined, "GET", path, undefined, undefined, {
            prefix: "/_matrix/client/unstable/im.vector.custom"
          }); // we use global account data because per-room account data on invites is unreliable

          await this.matrixClient.setAccountData("im.vector.group_info." + room.roomId, profile);
        } catch (e) {
          _logger.logger.warn("Non-fatal error getting group information for invite:", e);
        }
      }
    } else if (payload.action === "MatrixActions.accountData") {
      if (payload.event_type.startsWith("im.vector.group_info.")) {
        const roomId = payload.event_type.substring("im.vector.group_info.".length);
        this.emit(CommunityPrototypeStore.getUpdateEventName(roomId), roomId);
      }
    } else if (payload.action === "select_tag") {
      // Automatically select the general chat when switching communities
      const chat = this.getGeneralChat(payload.tag);

      if (chat) {
        _dispatcher.default.dispatch({
          action: _actions.Action.ViewRoom,
          room_id: chat.roomId
        });
      }
    }
  }

  getInviteProfile(roomId) {
    if (!this.matrixClient) return {
      displayName: null,
      avatarMxc: null
    };
    const room = this.matrixClient.getRoom(roomId);

    if (_SettingsStore.default.getValue("feature_communities_v2_prototypes")) {
      const data = this.matrixClient.getAccountData("im.vector.group_info." + roomId);

      if (data && data.getContent()) {
        return {
          displayName: data.getContent().name,
          avatarMxc: data.getContent().avatar_url
        };
      }
    }

    return {
      displayName: room.name,
      avatarMxc: room.getMxcAvatarUrl()
    };
  }

  async onReady() {
    for (const room of this.matrixClient.getRooms()) {
      const myMember = room.currentState.getMembers().find(m => m.userId === this.matrixClient.getUserId());
      if (!myMember) continue;

      if ((0, _membership.getEffectiveMembership)(myMember.membership) === _membership.EffectiveMembership.Invite) {
        // Fake an update for anything that might have started listening before the invite
        // data was available (eg: RoomPreviewBar after a refresh)
        this.emit(CommunityPrototypeStore.getUpdateEventName(room.roomId), room.roomId);
      }
    }
  }

}

exports.CommunityPrototypeStore = CommunityPrototypeStore;
(0, _defineProperty2.default)(CommunityPrototypeStore, "internalInstance", new CommunityPrototypeStore());
