// matrix-react-sdk/lib/stores/ModalWidgetStore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalWidgetStore = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _AsyncStoreWithClient = require("./AsyncStoreWithClient");

var _dispatcher = _interopRequireDefault(require("../dispatcher/dispatcher"));

var _Modal = _interopRequireDefault(require("../Modal"));

var _ModalWidgetDialog = _interopRequireDefault(require("../components/views/dialogs/ModalWidgetDialog"));

var _WidgetMessagingStore = require("./widgets/WidgetMessagingStore");

var _logger = require("matrix-js-sdk/src/logger");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

class ModalWidgetStore extends _AsyncStoreWithClient.AsyncStoreWithClient {
  constructor() {
    super(_dispatcher.default, {});
    (0, _defineProperty2.default)(this, "modalInstance", null);
    (0, _defineProperty2.default)(this, "openSourceWidgetId", null);
    (0, _defineProperty2.default)(this, "canOpenModalWidget", () => {
      return !this.modalInstance;
    });
    (0, _defineProperty2.default)(this, "openModalWidget", (requestData, sourceWidget, widgetRoomId) => {
      if (this.modalInstance) return;
      this.openSourceWidgetId = sourceWidget.id;
      this.modalInstance = _Modal.default.createTrackedDialog('Modal Widget', '', _ModalWidgetDialog.default, {
        widgetDefinition: _objectSpread({}, requestData),
        widgetRoomId,
        sourceWidgetId: sourceWidget.id,
        onFinished: (success, data) => {
          if (!success) {
            this.closeModalWidget(sourceWidget, {
              "m.exited": true
            });
          } else {
            this.closeModalWidget(sourceWidget, data);
          }

          this.openSourceWidgetId = null;
          this.modalInstance = null;
        }
      }, null,
      /* priority = */
      false,
      /* static = */
      true);
    });
    (0, _defineProperty2.default)(this, "closeModalWidget", (sourceWidget, data) => {
      if (!this.modalInstance) return;

      if (this.openSourceWidgetId === sourceWidget.id) {
        this.openSourceWidgetId = null;
        this.modalInstance.close();
        this.modalInstance = null;

        const sourceMessaging = _WidgetMessagingStore.WidgetMessagingStore.instance.getMessaging(sourceWidget);

        if (!sourceMessaging) {
          _logger.logger.error("No source widget messaging for modal widget");

          return;
        }

        sourceMessaging.notifyModalWidgetClose(data);
      }
    });
  }

  static get instance() {
    return ModalWidgetStore.internalInstance;
  }

  async onAction(payload) {// nothing
  }

}

exports.ModalWidgetStore = ModalWidgetStore;
(0, _defineProperty2.default)(ModalWidgetStore, "internalInstance", new ModalWidgetStore());
window.mxModalWidgetStore = ModalWidgetStore.instance;
