// matrix-react-sdk/lib/stores/RightPanelStorePhases.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RightPanelPhases = exports.RIGHT_PANEL_SPACE_PHASES = exports.RIGHT_PANEL_PHASES_NO_ARGS = void 0;

/*
Copyright 2019 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
// These are in their own file because of circular imports being a problem.
let RightPanelPhases; // These are the phases that are safe to persist (the ones that don't require additional
// arguments).

exports.RightPanelPhases = RightPanelPhases;

(function (RightPanelPhases) {
  RightPanelPhases["RoomMemberList"] = "RoomMemberList";
  RightPanelPhases["FilePanel"] = "FilePanel";
  RightPanelPhases["NotificationPanel"] = "NotificationPanel";
  RightPanelPhases["RoomMemberInfo"] = "RoomMemberInfo";
  RightPanelPhases["EncryptionPanel"] = "EncryptionPanel";
  RightPanelPhases["RoomSummary"] = "RoomSummary";
  RightPanelPhases["Widget"] = "Widget";
  RightPanelPhases["PinnedMessages"] = "PinnedMessages";
  RightPanelPhases["Timeline"] = "Timeline";
  RightPanelPhases["Room3pidMemberInfo"] = "Room3pidMemberInfo";
  RightPanelPhases["GroupMemberList"] = "GroupMemberList";
  RightPanelPhases["GroupRoomList"] = "GroupRoomList";
  RightPanelPhases["GroupRoomInfo"] = "GroupRoomInfo";
  RightPanelPhases["GroupMemberInfo"] = "GroupMemberInfo";
  RightPanelPhases["SpaceMemberList"] = "SpaceMemberList";
  RightPanelPhases["SpaceMemberInfo"] = "SpaceMemberInfo";
  RightPanelPhases["Space3pidMemberInfo"] = "Space3pidMemberInfo";
  RightPanelPhases["ThreadView"] = "ThreadView";
  RightPanelPhases["ThreadPanel"] = "ThreadPanel";
})(RightPanelPhases || (exports.RightPanelPhases = RightPanelPhases = {}));

const RIGHT_PANEL_PHASES_NO_ARGS = [RightPanelPhases.RoomSummary, RightPanelPhases.NotificationPanel, RightPanelPhases.PinnedMessages, RightPanelPhases.FilePanel, RightPanelPhases.RoomMemberList, RightPanelPhases.GroupMemberList, RightPanelPhases.GroupRoomList, RightPanelPhases.Timeline]; // Subset of phases visible in the Space View

exports.RIGHT_PANEL_PHASES_NO_ARGS = RIGHT_PANEL_PHASES_NO_ARGS;
const RIGHT_PANEL_SPACE_PHASES = [RightPanelPhases.SpaceMemberList, RightPanelPhases.Space3pidMemberInfo, RightPanelPhases.SpaceMemberInfo];
exports.RIGHT_PANEL_SPACE_PHASES = RIGHT_PANEL_SPACE_PHASES;
