// matrix-react-sdk/lib/stores/VoiceRecordingStore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VoiceRecordingStore = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _AsyncStoreWithClient = require("./AsyncStoreWithClient");

var _dispatcher = _interopRequireDefault(require("../dispatcher/dispatcher"));

var _VoiceRecording = require("../audio/VoiceRecording");

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
class VoiceRecordingStore extends _AsyncStoreWithClient.AsyncStoreWithClient {
  constructor() {
    super(_dispatcher.default, {});
  }
  /**
   * Gets the active recording instance, if any.
   */


  get activeRecording() {
    return this.state.recording;
  }

  static get instance() {
    if (!VoiceRecordingStore.internalInstance) {
      VoiceRecordingStore.internalInstance = new VoiceRecordingStore();
    }

    return VoiceRecordingStore.internalInstance;
  }

  async onAction(payload) {
    // Nothing to do, but we're required to override the function
    return;
  }
  /**
   * Starts a new recording if one isn't already in progress. Note that this simply
   * creates a recording instance - whether or not recording is actively in progress
   * can be seen via the VoiceRecording class.
   * @returns {VoiceRecording} The recording.
   */


  startRecording() {
    if (!this.matrixClient) throw new Error("Cannot start a recording without a MatrixClient");
    if (this.state.recording) throw new Error("A recording is already in progress");
    const recording = new _VoiceRecording.VoiceRecording(this.matrixClient); // noinspection JSIgnoredPromiseFromCall - we can safely run this async

    this.updateState({
      recording
    });
    return recording;
  }
  /**
   * Disposes of the current recording, no matter the state of it.
   * @returns {Promise<void>} Resolves when complete.
   */


  disposeRecording() {
    if (this.state.recording) {
      this.state.recording.destroy(); // stops internally
    }

    return this.updateState({
      recording: null
    });
  }

}

exports.VoiceRecordingStore = VoiceRecordingStore;
(0, _defineProperty2.default)(VoiceRecordingStore, "internalInstance", void 0);
window.mxVoiceRecordingStore = VoiceRecordingStore.instance;
