// matrix-react-sdk/lib/components/views/terms/InlineTermsAgreement.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _objects = require("../../../utils/objects");

var _StyledCheckbox = _interopRequireDefault(require("../elements/StyledCheckbox"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _dec, _class;

let InlineTermsAgreement = (_dec = (0, _replaceableComponent.replaceableComponent)("views.terms.InlineTermsAgreement"), _dec(_class = class InlineTermsAgreement extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "togglePolicy", index => {
      const policies = (0, _objects.objectClone)(this.state.policies);
      policies[index].checked = !policies[index].checked;
      this.setState({
        policies
      });
    });
    (0, _defineProperty2.default)(this, "onContinue", () => {
      const hasUnchecked = !!this.state.policies.some(p => !p.checked);
      if (hasUnchecked) return;
      this.setState({
        busy: true
      });
      this.props.onFinished(this.state.policies.map(p => p.url));
    });
    this.state = {
      policies: [],
      busy: false
    };
  }

  componentDidMount() {
    // Build all the terms the user needs to accept
    const policies = []; // { checked, url, name }

    for (const servicePolicies of this.props.policiesAndServicePairs) {
      const availablePolicies = Object.values(servicePolicies.policies);

      for (const policy of availablePolicies) {
        const language = (0, _languageHandler.pickBestLanguage)(Object.keys(policy).filter(p => p !== 'version'));
        const renderablePolicy = {
          checked: false,
          url: policy[language].url,
          name: policy[language].name
        };
        policies.push(renderablePolicy);
      }
    }

    this.setState({
      policies
    });
  }

  renderCheckboxes() {
    const rendered = [];

    for (let i = 0; i < this.state.policies.length; i++) {
      const policy = this.state.policies[i];
      const introText = (0, _languageHandler._t)("Accept <policyLink /> to continue:", {}, {
        policyLink: () => {
          return /*#__PURE__*/_react.default.createElement("a", {
            href: policy.url,
            rel: "noreferrer noopener",
            target: "_blank"
          }, policy.name, /*#__PURE__*/_react.default.createElement("span", {
            className: "mx_InlineTermsAgreement_link"
          }));
        }
      });
      rendered.push( /*#__PURE__*/_react.default.createElement("div", {
        key: i,
        className: "mx_InlineTermsAgreement_cbContainer"
      }, /*#__PURE__*/_react.default.createElement("div", null, introText), /*#__PURE__*/_react.default.createElement("div", {
        className: "mx_InlineTermsAgreement_checkbox"
      }, /*#__PURE__*/_react.default.createElement(_StyledCheckbox.default, {
        onChange: () => this.togglePolicy(i),
        checked: policy.checked
      }, (0, _languageHandler._t)("Accept")))));
    }

    return rendered;
  }

  render() {
    const hasUnchecked = !!this.state.policies.some(p => !p.checked);
    return /*#__PURE__*/_react.default.createElement("div", null, this.props.introElement, this.renderCheckboxes(), /*#__PURE__*/_react.default.createElement(_AccessibleButton.default, {
      onClick: this.onContinue,
      disabled: hasUnchecked || this.state.busy,
      kind: "primary_sm"
    }, (0, _languageHandler._t)("Continue")));
  }

}) || _class);
exports.default = InlineTermsAgreement;
