// matrix-react-sdk/lib/toasts/ServerLimitToast.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showToast = exports.hideToast = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../languageHandler");

var _GenericToast = _interopRequireDefault(require("../components/views/toasts/GenericToast"));

var _ToastStore = _interopRequireDefault(require("../stores/ToastStore"));

var _ErrorUtils = require("../utils/ErrorUtils");

/*
Copyright 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const TOAST_KEY = "serverlimit";

const showToast = (limitType, onHideToast, adminContact, syncError) => {
  const errorText = (0, _ErrorUtils.messageForResourceLimitError)(limitType, adminContact, {
    'monthly_active_user': (0, _languageHandler._td)("Your homeserver has exceeded its user limit."),
    'hs_blocked': (0, _languageHandler._td)("This homeserver has been blocked by it's administrator."),
    '': (0, _languageHandler._td)("Your homeserver has exceeded one of its resource limits.")
  });
  const contactText = (0, _ErrorUtils.messageForResourceLimitError)(limitType, adminContact, {
    '': (0, _languageHandler._td)("Contact your <a>server admin</a>.")
  });

  _ToastStore.default.sharedInstance().addOrReplaceToast({
    key: TOAST_KEY,
    title: (0, _languageHandler._t)("Warning"),
    props: {
      description: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, errorText, " ", contactText),
      acceptLabel: (0, _languageHandler._t)("Ok"),
      onAccept: () => {
        hideToast();
        if (onHideToast) onHideToast();
      }
    },
    component: _GenericToast.default,
    priority: 70
  });
};

exports.showToast = showToast;

const hideToast = () => {
  _ToastStore.default.sharedInstance().dismissToast(TOAST_KEY);
};

exports.hideToast = hideToast;
