// matrix-react-sdk/lib/polls/consts.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.POLL_START_EVENT_TYPE = exports.POLL_RESPONSE_EVENT_TYPE = exports.POLL_KIND_UNDISCLOSED = exports.POLL_KIND_DISCLOSED = void 0;
exports.makePollContent = makePollContent;

var _NamespacedValue = require("matrix-js-sdk/src/NamespacedValue");

/*
Copyright 2021 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const POLL_START_EVENT_TYPE = new _NamespacedValue.UnstableValue("m.poll.start", "org.matrix.msc3381.poll.start");
exports.POLL_START_EVENT_TYPE = POLL_START_EVENT_TYPE;
const POLL_RESPONSE_EVENT_TYPE = new _NamespacedValue.UnstableValue("m.poll.response", "org.matrix.msc3381.poll.response");
exports.POLL_RESPONSE_EVENT_TYPE = POLL_RESPONSE_EVENT_TYPE;
const POLL_KIND_DISCLOSED = new _NamespacedValue.UnstableValue("m.poll.disclosed", "org.matrix.msc3381.poll.disclosed");
exports.POLL_KIND_DISCLOSED = POLL_KIND_DISCLOSED;
const POLL_KIND_UNDISCLOSED = new _NamespacedValue.UnstableValue("m.poll.undisclosed", "org.matrix.msc3381.poll.undisclosed"); // TODO: [TravisR] Use extensible events library when ready

exports.POLL_KIND_UNDISCLOSED = POLL_KIND_UNDISCLOSED;
const TEXT_NODE_TYPE = "org.matrix.msc1767.text";

function makePollContent(question, answers, kind) {
  question = question.trim();
  answers = answers.map(a => a.trim()).filter(a => !!a);
  return {
    [TEXT_NODE_TYPE]: `${question}\n${answers.map((a, i) => `${i + 1}. ${a}`).join('\n')}`,
    [POLL_START_EVENT_TYPE.name]: {
      kind: kind,
      question: {
        [TEXT_NODE_TYPE]: question
      },
      answers: answers.map((a, i) => ({
        id: `${i}-${a}`,
        [TEXT_NODE_TYPE]: a
      }))
    }
  };
}
