// matrix-react-sdk/lib/components/views/dialogs/ErrorDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _dec, _class, _class2, _temp;

let ErrorDialog = (_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ErrorDialog"), _dec(_class = (_temp = _class2 = class ErrorDialog extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "onClick", () => {
      this.props.onFinished(true);
    });
  }

  render() {
    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_ErrorDialog",
      onFinished: this.props.onFinished,
      title: this.props.title || (0, _languageHandler._t)('Error'),
      headerImage: this.props.headerImage,
      contentId: "mx_Dialog_content"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_content",
      id: "mx_Dialog_content"
    }, this.props.description || (0, _languageHandler._t)('An error has occurred.')), /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_Dialog_buttons"
    }, /*#__PURE__*/_react.default.createElement("button", {
      className: "mx_Dialog_primary",
      onClick: this.onClick,
      autoFocus: this.props.focus
    }, this.props.button || (0, _languageHandler._t)('OK'))));
  }

}, (0, _defineProperty2.default)(_class2, "defaultProps", {
  focus: true,
  title: null,
  description: null,
  button: null
}), _temp)) || _class);
exports.default = ErrorDialog;
