// matrix-react-sdk/lib/components/views/dialogs/ReportEventDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _createRoom = require("../../../createRoom");

var _MatrixClientPeg = require("../../../MatrixClientPeg");

var _SdkConfig = _interopRequireDefault(require("../../../SdkConfig"));

var _Markdown = _interopRequireDefault(require("../../../Markdown"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

var _StyledRadioButton = _interopRequireDefault(require("../elements/StyledRadioButton"));

var _BaseDialog = _interopRequireDefault(require("./BaseDialog"));

var _DialogButtons = _interopRequireDefault(require("../elements/DialogButtons"));

var _Field = _interopRequireDefault(require("../elements/Field"));

var _Spinner = _interopRequireDefault(require("../elements/Spinner"));

var _dec, _class;

const MODERATED_BY_STATE_EVENT_TYPE = ["org.matrix.msc3215.room.moderation.moderated_by"
/**
 * Unprefixed state event. Not ready for prime time.
 *
 * "m.room.moderation.moderated_by"
 */
];
const ABUSE_EVENT_TYPE = "org.matrix.msc3215.abuse.report"; // Standard abuse natures.

var Nature;

(function (Nature) {
  Nature["Disagreement"] = "org.matrix.msc3215.abuse.nature.disagreement";
  Nature["Toxic"] = "org.matrix.msc3215.abuse.nature.toxic";
  Nature["Illegal"] = "org.matrix.msc3215.abuse.nature.illegal";
  Nature["Spam"] = "org.matrix.msc3215.abuse.nature.spam";
  Nature["Other"] = "org.matrix.msc3215.abuse.nature.other";
})(Nature || (Nature = {}));

var NonStandardValue;

(function (NonStandardValue) {
  NonStandardValue["Admin"] = "non-standard.abuse.nature.admin";
})(NonStandardValue || (NonStandardValue = {}));

let ReportEventDialog = (
/*
 * A dialog for reporting an event.
 *
 * The actual content of the dialog will depend on two things:
 *
 * 1. Is `feature_report_to_moderators` enabled?
 * 2. Does the room support moderation as per MSC3215, i.e. is there
 *    a well-formed state event `m.room.moderation.moderated_by`
 *    /`org.matrix.msc3215.room.moderation.moderated_by`?
 */
_dec = (0, _replaceableComponent.replaceableComponent)("views.dialogs.ReportEventDialog"), _dec(_class = class ReportEventDialog extends _react.default.Component {
  // If the room supports moderation, the moderation information.
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "moderation", void 0);
    (0, _defineProperty2.default)(this, "onReasonChange", ({
      target: {
        value: reason
      }
    }) => {
      this.setState({
        reason
      });
    });
    (0, _defineProperty2.default)(this, "onNatureChosen", e => {
      this.setState({
        nature: e.currentTarget.value
      });
    });
    (0, _defineProperty2.default)(this, "onCancel", () => {
      this.props.onFinished(false);
    });
    (0, _defineProperty2.default)(this, "onSubmit", async () => {
      let reason = this.state.reason || "";
      reason = reason.trim();

      if (this.moderation) {
        // This room supports moderation.
        // We need a nature.
        // If the nature is `NATURE.OTHER` or `NON_STANDARD_NATURE.ADMIN`, we also need a `reason`.
        if (!this.state.nature || (this.state.nature == Nature.Other || this.state.nature == NonStandardValue.Admin) && !reason) {
          this.setState({
            err: (0, _languageHandler._t)("Please fill why you're reporting.")
          });
          return;
        }
      } else {
        // This room does not support moderation.
        // We need a `reason`.
        if (!reason) {
          this.setState({
            err: (0, _languageHandler._t)("Please fill why you're reporting.")
          });
          return;
        }
      }

      this.setState({
        busy: true,
        err: null
      });

      try {
        const client = _MatrixClientPeg.MatrixClientPeg.get();

        const ev = this.props.mxEvent;

        if (this.moderation && this.state.nature != NonStandardValue.Admin) {
          const nature = this.state.nature; // Report to moderators through to the dedicated bot,
          // as configured in the room's state events.

          const dmRoomId = await (0, _createRoom.ensureDMExists)(client, this.moderation.moderationBotUserId);
          await client.sendEvent(dmRoomId, ABUSE_EVENT_TYPE, {
            event_id: ev.getId(),
            room_id: ev.getRoomId(),
            moderated_by_id: this.moderation.moderationRoomId,
            nature,
            reporter: client.getUserId(),
            comment: this.state.reason.trim()
          });
        } else {
          // Report to homeserver admin through the dedicated Matrix API.
          await client.reportEvent(ev.getRoomId(), ev.getId(), -100, this.state.reason.trim());
        }

        this.props.onFinished(true);
      } catch (e) {
        this.setState({
          busy: false,
          err: e.message
        });
      }
    });
    let moderatedByRoomId = null;
    let moderatedByUserId = null;

    if (_SettingsStore.default.getValue("feature_report_to_moderators")) {
      // The client supports reporting to moderators.
      // Does the room support it, too?
      // Extract state events to determine whether we should display
      const client = _MatrixClientPeg.MatrixClientPeg.get();

      const room = client.getRoom(props.mxEvent.getRoomId());

      for (const stateEventType of MODERATED_BY_STATE_EVENT_TYPE) {
        const stateEvent = room.currentState.getStateEvents(stateEventType, stateEventType);

        if (!stateEvent) {
          continue;
        }

        if (Array.isArray(stateEvent)) {
          // Internal error.
          throw new TypeError(`getStateEvents(${stateEventType}, ${stateEventType}) ` + "should return at most one state event");
        }

        const event = stateEvent.event;

        if (!("content" in event) || typeof event["content"] != "object") {
          // The room is improperly configured.
          // Display this debug message for the sake of moderators.
          console.debug("Moderation error", "state event", stateEventType, "should have an object field `content`, got", event);
          continue;
        }

        const content = event["content"];

        if (!("room_id" in content) || typeof content["room_id"] != "string") {
          // The room is improperly configured.
          // Display this debug message for the sake of moderators.
          console.debug("Moderation error", "state event", stateEventType, "should have a string field `content.room_id`, got", event);
          continue;
        }

        if (!("user_id" in content) || typeof content["user_id"] != "string") {
          // The room is improperly configured.
          // Display this debug message for the sake of moderators.
          console.debug("Moderation error", "state event", stateEventType, "should have a string field `content.user_id`, got", event);
          continue;
        }

        moderatedByRoomId = content["room_id"];
        moderatedByUserId = content["user_id"];
      }

      if (moderatedByRoomId && moderatedByUserId) {
        // The room supports moderation.
        this.moderation = {
          moderationRoomId: moderatedByRoomId,
          moderationBotUserId: moderatedByUserId
        };
      }
    }

    this.state = {
      // A free-form text describing the abuse.
      reason: "",
      busy: false,
      err: null,
      // If specified, the nature of the abuse, as specified by MSC3215.
      nature: null
    };
  } // The user has written down a freeform description of the abuse.


  render() {
    let error = null;

    if (this.state.err) {
      error = /*#__PURE__*/_react.default.createElement("div", {
        className: "error"
      }, this.state.err);
    }

    let progress = null;

    if (this.state.busy) {
      progress = /*#__PURE__*/_react.default.createElement("div", {
        className: "progress"
      }, /*#__PURE__*/_react.default.createElement(_Spinner.default, null));
    }

    const adminMessageMD = _SdkConfig.default.get().reportEvent && _SdkConfig.default.get().reportEvent.adminMessageMD;

    let adminMessage;

    if (adminMessageMD) {
      const html = new _Markdown.default(adminMessageMD).toHTML({
        externalLinks: true
      });
      adminMessage = /*#__PURE__*/_react.default.createElement("p", {
        dangerouslySetInnerHTML: {
          __html: html
        }
      });
    }

    if (this.moderation) {
      // Display report-to-moderator dialog.
      // We let the user pick a nature.
      const client = _MatrixClientPeg.MatrixClientPeg.get();

      const homeServerName = _SdkConfig.default.get()["validated_server_config"].hsName;

      let subtitle;

      switch (this.state.nature) {
        case Nature.Disagreement:
          subtitle = (0, _languageHandler._t)("What this user is writing is wrong.\n" + "This will be reported to the room moderators.");
          break;

        case Nature.Toxic:
          subtitle = (0, _languageHandler._t)("This user is displaying toxic behaviour, " + "for instance by insulting other users or sharing " + " adult-only content in a family-friendly room " + " or otherwise violating the rules of this room.\n" + "This will be reported to the room moderators.");
          break;

        case Nature.Illegal:
          subtitle = (0, _languageHandler._t)("This user is displaying illegal behaviour, " + "for instance by doxing people or threatening violence.\n" + "This will be reported to the room moderators who may escalate this to legal authorities.");
          break;

        case Nature.Spam:
          subtitle = (0, _languageHandler._t)("This user is spamming the room with ads, links to ads or to propaganda.\n" + "This will be reported to the room moderators.");
          break;

        case NonStandardValue.Admin:
          if (client.isRoomEncrypted(this.props.mxEvent.getRoomId())) {
            subtitle = (0, _languageHandler._t)("This room is dedicated to illegal or toxic content " + "or the moderators fail to moderate illegal or toxic content.\n" + "This will be reported to the administrators of %(homeserver)s. " + "The administrators will NOT be able to read the encrypted content of this room.", {
              homeserver: homeServerName
            });
          } else {
            subtitle = (0, _languageHandler._t)("This room is dedicated to illegal or toxic content " + "or the moderators fail to moderate illegal or toxic content.\n" + " This will be reported to the administrators of %(homeserver)s.", {
              homeserver: homeServerName
            });
          }

          break;

        case Nature.Other:
          subtitle = (0, _languageHandler._t)("Any other reason. Please describe the problem.\n" + "This will be reported to the room moderators.");
          break;

        default:
          subtitle = (0, _languageHandler._t)("Please pick a nature and describe what makes this message abusive.");
          break;
      }

      return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
        className: "mx_ReportEventDialog",
        onFinished: this.props.onFinished,
        title: (0, _languageHandler._t)('Report Content'),
        contentId: "mx_ReportEventDialog"
      }, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: Nature.Disagreement,
        checked: this.state.nature == Nature.Disagreement,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Disagree')), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: Nature.Toxic,
        checked: this.state.nature == Nature.Toxic,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Toxic Behaviour')), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: Nature.Illegal,
        checked: this.state.nature == Nature.Illegal,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Illegal Content')), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: Nature.Spam,
        checked: this.state.nature == Nature.Spam,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Spam or propaganda')), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: NonStandardValue.Admin,
        checked: this.state.nature == NonStandardValue.Admin,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Report the entire room')), /*#__PURE__*/_react.default.createElement(_StyledRadioButton.default, {
        name: "nature",
        value: Nature.Other,
        checked: this.state.nature == Nature.Other,
        onChange: this.onNatureChosen
      }, (0, _languageHandler._t)('Other')), /*#__PURE__*/_react.default.createElement("p", null, subtitle), /*#__PURE__*/_react.default.createElement(_Field.default, {
        className: "mx_ReportEventDialog_reason",
        element: "textarea",
        label: (0, _languageHandler._t)("Reason"),
        rows: 5,
        onChange: this.onReasonChange,
        value: this.state.reason,
        disabled: this.state.busy
      }), progress, error), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
        primaryButton: (0, _languageHandler._t)("Send report"),
        onPrimaryButtonClick: this.onSubmit,
        focus: true,
        onCancel: this.onCancel,
        disabled: this.state.busy
      }));
    } // Report to homeserver admin.
    // Currently, the API does not support natures.


    return /*#__PURE__*/_react.default.createElement(_BaseDialog.default, {
      className: "mx_ReportEventDialog",
      onFinished: this.props.onFinished,
      title: (0, _languageHandler._t)('Report Content to Your Homeserver Administrator'),
      contentId: "mx_ReportEventDialog"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "mx_ReportEventDialog",
      id: "mx_ReportEventDialog"
    }, /*#__PURE__*/_react.default.createElement("p", null, (0, _languageHandler._t)("Reporting this message will send its unique 'event ID' to the administrator of " + "your homeserver. If messages in this room are encrypted, your homeserver " + "administrator will not be able to read the message text or view any files " + "or images.")), adminMessage, /*#__PURE__*/_react.default.createElement(_Field.default, {
      className: "mx_ReportEventDialog_reason",
      element: "textarea",
      label: (0, _languageHandler._t)("Reason"),
      rows: 5,
      onChange: this.onReasonChange,
      value: this.state.reason,
      disabled: this.state.busy
    }), progress, error), /*#__PURE__*/_react.default.createElement(_DialogButtons.default, {
      primaryButton: (0, _languageHandler._t)("Send report"),
      onPrimaryButtonClick: this.onSubmit,
      focus: true,
      onCancel: this.onCancel,
      disabled: this.state.busy
    }));
  }

}) || _class);
exports.default = ReportEventDialog;
