// matrix-react-sdk/lib/components/views/dialogs/SlashCommandHelpDialog.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _languageHandler = require("../../../languageHandler");

var _SlashCommands = require("../../../SlashCommands");

var _InfoDialog = _interopRequireDefault(require("./InfoDialog"));

/*
Copyright 2019 Michael Telatynski <7t3chguy@gmail.com>

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
const SlashCommandHelpDialog = ({
  onFinished
}) => {
  const categories = {};

  _SlashCommands.Commands.forEach(cmd => {
    if (!cmd.isEnabled()) return;

    if (!categories[cmd.category]) {
      categories[cmd.category] = [];
    }

    categories[cmd.category].push(cmd);
  });

  const body = Object.values(_SlashCommands.CommandCategories).filter(c => categories[c]).map(category => {
    const rows = [/*#__PURE__*/_react.default.createElement("tr", {
      key: "_category_" + category,
      className: "mx_SlashCommandHelpDialog_headerRow"
    }, /*#__PURE__*/_react.default.createElement("td", {
      colSpan: 3
    }, /*#__PURE__*/_react.default.createElement("h2", null, (0, _languageHandler._t)(category))))];
    categories[category].forEach(cmd => {
      rows.push( /*#__PURE__*/_react.default.createElement("tr", {
        key: cmd.command
      }, /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement("strong", null, cmd.getCommand())), /*#__PURE__*/_react.default.createElement("td", null, cmd.args), /*#__PURE__*/_react.default.createElement("td", null, cmd.description)));
    });
    return rows;
  });
  return /*#__PURE__*/_react.default.createElement(_InfoDialog.default, {
    className: "mx_SlashCommandHelpDialog",
    title: (0, _languageHandler._t)("Command Help"),
    description: /*#__PURE__*/_react.default.createElement("table", null, /*#__PURE__*/_react.default.createElement("tbody", null, body)),
    hasCloseButton: true,
    onFinished: onFinished
  });
};

var _default = SlashCommandHelpDialog;
exports.default = _default;
