// matrix-react-sdk/lib/components/views/context_menus/CallContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _languageHandler = require("../../../languageHandler");

var _ContextMenu = _interopRequireWildcard(require("../../structures/ContextMenu"));

var _CallHandler = _interopRequireDefault(require("../../../CallHandler"));

var _InviteDialog = _interopRequireWildcard(require("../dialogs/InviteDialog"));

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let CallContextMenu = (_dec = (0, _replaceableComponent.replaceableComponent)("views.context_menus.CallContextMenu"), _dec(_class = (_temp = _class2 = class CallContextMenu extends _react.default.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onHoldClick", () => {
      this.props.call.setRemoteOnHold(true);
      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "onUnholdClick", () => {
      _CallHandler.default.sharedInstance().setActiveCallRoomId(this.props.call.roomId);

      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "onTransferClick", () => {
      _Modal.default.createTrackedDialog('Transfer Call', '', _InviteDialog.default, {
        kind: _InviteDialog.KIND_CALL_TRANSFER,
        call: this.props.call
      },
      /*className=*/
      "mx_InviteDialog_transferWrapper",
      /*isPriority=*/
      false,
      /*isStatic=*/
      true);

      this.props.onFinished();
    });
  }

  render() {
    const holdUnholdCaption = this.props.call.isRemoteOnHold() ? (0, _languageHandler._t)("Resume") : (0, _languageHandler._t)("Hold");
    const handler = this.props.call.isRemoteOnHold() ? this.onUnholdClick : this.onHoldClick;
    let transferItem;

    if (this.props.call.opponentCanBeTransferred()) {
      transferItem = /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
        className: "mx_CallContextMenu_item",
        onClick: this.onTransferClick
      }, (0, _languageHandler._t)("Transfer"));
    }

    return /*#__PURE__*/_react.default.createElement(_ContextMenu.default, this.props, /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
      className: "mx_CallContextMenu_item",
      onClick: handler
    }, holdUnholdCaption), transferItem);
  }

}, (0, _defineProperty2.default)(_class2, "propTypes", {
  // js-sdk User object. Not required because it might not exist.
  user: _propTypes.default.object
}), _temp)) || _class);
exports.default = CallContextMenu;
