// matrix-react-sdk/lib/components/views/context_menus/DialpadContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _AccessibleButton = _interopRequireDefault(require("../elements/AccessibleButton"));

var _ContextMenu = _interopRequireDefault(require("../../structures/ContextMenu"));

var _Field = _interopRequireDefault(require("../elements/Field"));

var _DialPad = _interopRequireDefault(require("../voip/DialPad"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _dec, _class;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let DialpadContextMenu = (_dec = (0, _replaceableComponent.replaceableComponent)("views.context_menus.DialpadContextMenu"), _dec(_class = class DialpadContextMenu extends React.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "numberEntryFieldRef", /*#__PURE__*/(0, React.createRef)());
    (0, _defineProperty2.default)(this, "onDigitPress", (digit, ev) => {
      this.props.call.sendDtmfDigit(digit);
      this.setState({
        value: this.state.value + digit
      }); // Keep the number field focused so that keyboard entry is still available
      // However, don't focus if this wasn't the result of directly clicking on the button,
      // i.e someone using keyboard navigation.

      if (ev.type === "click") {
        var _this$numberEntryFiel;

        (_this$numberEntryFiel = this.numberEntryFieldRef.current) === null || _this$numberEntryFiel === void 0 ? void 0 : _this$numberEntryFiel.focus();
      }
    });
    (0, _defineProperty2.default)(this, "onCancelClick", () => {
      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "onKeyDown", ev => {
      // Prevent Backspace and Delete keys from functioning in the entry field
      if (ev.code === "Backspace" || ev.code === "Delete") {
        ev.preventDefault();
      }
    });
    (0, _defineProperty2.default)(this, "onChange", ev => {
      this.setState({
        value: ev.target.value
      });
    });
    this.state = {
      value: ''
    };
  }

  render() {
    return /*#__PURE__*/React.createElement(_ContextMenu.default, this.props, /*#__PURE__*/React.createElement("div", {
      className: "mx_DialPadContextMenuWrapper"
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(_AccessibleButton.default, {
      className: "mx_DialPadContextMenu_cancel",
      onClick: this.onCancelClick
    })), /*#__PURE__*/React.createElement("div", {
      className: "mx_DialPadContextMenu_header"
    }, /*#__PURE__*/React.createElement(_Field.default, {
      ref: this.numberEntryFieldRef,
      className: "mx_DialPadContextMenu_dialled",
      value: this.state.value,
      autoFocus: true,
      onKeyDown: this.onKeyDown,
      onChange: this.onChange
    })), /*#__PURE__*/React.createElement("div", {
      className: "mx_DialPadContextMenu_dialPad"
    }, /*#__PURE__*/React.createElement(_DialPad.default, {
      onDigitPress: this.onDigitPress,
      hasDial: false
    }))));
  }

}) || _class);
exports.default = DialpadContextMenu;
