// matrix-react-sdk/lib/components/views/context_menus/GroupInviteTileContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var sdk = _interopRequireWildcard(require("../../../index"));

var _languageHandler = require("../../../languageHandler");

var _Modal = _interopRequireDefault(require("../../../Modal"));

var _group = require("matrix-js-sdk/src/models/group");

var _GroupStore = _interopRequireDefault(require("../../../stores/GroupStore"));

var _ContextMenu = require("../../structures/ContextMenu");

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _logger = require("matrix-js-sdk/src/logger");

var _dec, _class, _class2, _temp;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

let GroupInviteTileContextMenu = (_dec = (0, _replaceableComponent.replaceableComponent)("views.context_menus.GroupInviteTileContextMenu"), _dec(_class = (_temp = _class2 = class GroupInviteTileContextMenu extends _react.default.Component {
  constructor(props) {
    super(props);
    this._onClickReject = this._onClickReject.bind(this);
  }

  componentDidMount() {
    this._unmounted = false;
  }

  componentWillUnmount() {
    this._unmounted = true;
  }

  _onClickReject() {
    const QuestionDialog = sdk.getComponent('dialogs.QuestionDialog');

    _Modal.default.createTrackedDialog('Reject community invite', '', QuestionDialog, {
      title: (0, _languageHandler._t)('Reject invitation'),
      description: (0, _languageHandler._t)('Are you sure you want to reject the invitation?'),
      onFinished: async shouldLeave => {
        if (!shouldLeave) return; // FIXME: controller shouldn't be loading a view :(

        const Loader = sdk.getComponent("elements.Spinner");

        const modal = _Modal.default.createDialog(Loader, null, 'mx_Dialog_spinner');

        try {
          await _GroupStore.default.leaveGroup(this.props.group.groupId);
        } catch (e) {
          _logger.logger.error("Error rejecting community invite: ", e);

          const ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");

          _Modal.default.createTrackedDialog('Error rejecting invite', '', ErrorDialog, {
            title: (0, _languageHandler._t)("Error"),
            description: (0, _languageHandler._t)("Unable to reject invite")
          });
        } finally {
          modal.close();
        }
      }
    }); // Close the context menu


    if (this.props.onFinished) {
      this.props.onFinished();
    }
  }

  render() {
    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
      className: "mx_RoomTileContextMenu_leave",
      onClick: this._onClickReject
    }, /*#__PURE__*/_react.default.createElement("img", {
      className: "mx_RoomTileContextMenu_tag_icon",
      src: require("../../../../res/img/icon_context_delete.svg"),
      width: "15",
      height: "15",
      alt: ""
    }), (0, _languageHandler._t)('Reject')));
  }

}, (0, _defineProperty2.default)(_class2, "propTypes", {
  group: _propTypes.default.instanceOf(_group.Group).isRequired,

  /* callback called when the menu is dismissed */
  onFinished: _propTypes.default.func
}), _temp)) || _class);
exports.default = GroupInviteTileContextMenu;
