// matrix-react-sdk/lib/components/views/context_menus/SpaceContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _event = require("matrix-js-sdk/src/@types/event");

var _IconizedContextMenu = _interopRequireWildcard(require("./IconizedContextMenu"));

var _languageHandler = require("../../../languageHandler");

var _space = require("../../../utils/space");

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _RoomViewStore = _interopRequireDefault(require("../../../stores/RoomViewStore"));

var _actions = require("../../../dispatcher/actions");

var _RightPanelStorePhases = require("../../../stores/RightPanelStorePhases");

var _BetaCard = require("../beta/BetaCard");

var _SettingsStore = _interopRequireDefault(require("../../../settings/SettingsStore"));

const _excluded = ["space", "onFinished"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const SpaceContextMenu = _ref => {
  let {
    space,
    onFinished
  } = _ref,
      props = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  const cli = (0, _react.useContext)(_MatrixClientContext.default);
  const userId = cli.getUserId();
  let inviteOption;

  if (space.getJoinRule() === "public" || space.canInvite(userId)) {
    const onInviteClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.showSpaceInvite)(space);
      onFinished();
    };

    inviteOption = /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      className: "mx_SpacePanel_contextMenu_inviteButton",
      iconClassName: "mx_SpacePanel_iconInvite",
      label: (0, _languageHandler._t)("Invite people"),
      onClick: onInviteClick
    });
  }

  let settingsOption;
  let leaveSection;

  if ((0, _space.shouldShowSpaceSettings)(space)) {
    const onSettingsClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.showSpaceSettings)(space);
      onFinished();
    };

    settingsOption = /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconSettings",
      label: (0, _languageHandler._t)("Settings"),
      onClick: onSettingsClick
    });
  } else {
    const onLeaveClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.leaveSpace)(space);
      onFinished();
    };

    leaveSection = /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOptionList, {
      red: true,
      first: true
    }, /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconLeave",
      label: (0, _languageHandler._t)("Leave space"),
      onClick: onLeaveClick
    }));
  }

  let devtoolsSection;

  if (_SettingsStore.default.getValue("developerMode")) {
    const onViewTimelineClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();

      _dispatcher.default.dispatch({
        action: _actions.Action.ViewRoom,
        room_id: space.roomId,
        forceTimeline: true
      });

      onFinished();
    };

    devtoolsSection = /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOptionList, {
      first: true
    }, /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconSettings",
      label: (0, _languageHandler._t)("See room timeline (devtools)"),
      onClick: onViewTimelineClick
    }));
  }

  const canAddRooms = space.currentState.maySendStateEvent(_event.EventType.SpaceChild, userId);
  let newRoomSection;

  if (space.currentState.maySendStateEvent(_event.EventType.SpaceChild, userId)) {
    const onNewRoomClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.showCreateNewRoom)(space);
      onFinished();
    };

    const onAddExistingRoomClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.showAddExistingRooms)(space);
      onFinished();
    };

    const onNewSubspaceClick = ev => {
      ev.preventDefault();
      ev.stopPropagation();
      (0, _space.showCreateNewSubspace)(space);
      onFinished();
    };

    newRoomSection = /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOptionList, {
      first: true
    }, /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconPlus",
      label: (0, _languageHandler._t)("Create new room"),
      onClick: onNewRoomClick
    }), /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconHash",
      label: (0, _languageHandler._t)("Add existing room"),
      onClick: onAddExistingRoomClick
    }), /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
      iconClassName: "mx_SpacePanel_iconPlus",
      label: (0, _languageHandler._t)("Add space"),
      onClick: onNewSubspaceClick
    }, /*#__PURE__*/_react.default.createElement(_BetaCard.BetaPill, null)));
  }

  const onMembersClick = ev => {
    ev.preventDefault();
    ev.stopPropagation();

    if (!_RoomViewStore.default.getRoomId()) {
      _dispatcher.default.dispatch({
        action: "view_room",
        room_id: space.roomId
      }, true);
    }

    _dispatcher.default.dispatch({
      action: _actions.Action.SetRightPanelPhase,
      phase: _RightPanelStorePhases.RightPanelPhases.SpaceMemberList,
      refireParams: {
        space
      }
    });

    onFinished();
  };

  const onExploreRoomsClick = ev => {
    ev.preventDefault();
    ev.stopPropagation();

    _dispatcher.default.dispatch({
      action: "view_room",
      room_id: space.roomId
    });

    onFinished();
  };

  return /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.default, (0, _extends2.default)({}, props, {
    onFinished: onFinished,
    className: "mx_SpacePanel_contextMenu",
    compact: true
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "mx_SpacePanel_contextMenu_header"
  }, space.name), /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOptionList, {
    first: true
  }, inviteOption, /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
    iconClassName: "mx_SpacePanel_iconMembers",
    label: (0, _languageHandler._t)("Members"),
    onClick: onMembersClick
  }), settingsOption, /*#__PURE__*/_react.default.createElement(_IconizedContextMenu.IconizedContextMenuOption, {
    iconClassName: "mx_SpacePanel_iconExplore",
    label: canAddRooms ? (0, _languageHandler._t)("Manage & explore rooms") : (0, _languageHandler._t)("Explore rooms"),
    onClick: onExploreRoomsClick
  })), newRoomSection, leaveSection, devtoolsSection);
};

var _default = SpaceContextMenu;
exports.default = _default;
