// matrix-react-sdk/lib/components/views/context_menus/TagTileContextMenu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _languageHandler = require("../../../languageHandler");

var _dispatcher = _interopRequireDefault(require("../../../dispatcher/dispatcher"));

var _TagOrderActions = _interopRequireDefault(require("../../../actions/TagOrderActions"));

var _ContextMenu = require("../../structures/ContextMenu");

var _MatrixClientContext = _interopRequireDefault(require("../../../contexts/MatrixClientContext"));

var _replaceableComponent = require("../../../utils/replaceableComponent");

var _GroupFilterOrderStore = _interopRequireDefault(require("../../../stores/GroupFilterOrderStore"));

var _space = require("../../../utils/space");

var _GroupStore = _interopRequireDefault(require("../../../stores/GroupStore"));

var _dec, _class, _class2, _temp;

let TagTileContextMenu = (_dec = (0, _replaceableComponent.replaceableComponent)("views.context_menus.TagTileContextMenu"), _dec(_class = (_temp = _class2 = class TagTileContextMenu extends _react.default.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "_onViewCommunityClick", () => {
      _dispatcher.default.dispatch({
        action: 'view_group',
        group_id: this.props.tag
      });

      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "_onRemoveClick", () => {
      _dispatcher.default.dispatch(_TagOrderActions.default.removeTag(this.context, this.props.tag));

      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "_onCreateSpaceClick", () => {
      (0, _space.createSpaceFromCommunity)(this.context, this.props.tag);
      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "_onMoveUp", () => {
      _dispatcher.default.dispatch(_TagOrderActions.default.moveTag(this.context, this.props.tag, this.props.index - 1));

      this.props.onFinished();
    });
    (0, _defineProperty2.default)(this, "_onMoveDown", () => {
      _dispatcher.default.dispatch(_TagOrderActions.default.moveTag(this.context, this.props.tag, this.props.index + 1));

      this.props.onFinished();
    });
  }

  render() {
    let moveUp;
    let moveDown;

    if (this.props.index > 0) {
      moveUp = /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
        className: "mx_TagTileContextMenu_item mx_TagTileContextMenu_moveUp",
        onClick: this._onMoveUp
      }, (0, _languageHandler._t)("Move up"));
    }

    if (this.props.index < (_GroupFilterOrderStore.default.getOrderedTags() || []).length - 1) {
      moveDown = /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
        className: "mx_TagTileContextMenu_item mx_TagTileContextMenu_moveDown",
        onClick: this._onMoveDown
      }, (0, _languageHandler._t)("Move down"));
    }

    let createSpaceOption;

    if (_GroupStore.default.isUserPrivileged(this.props.tag)) {
      createSpaceOption = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("hr", {
        className: "mx_TagTileContextMenu_separator",
        role: "separator"
      }), /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
        className: "mx_TagTileContextMenu_item mx_TagTileContextMenu_createSpace",
        onClick: this._onCreateSpaceClick
      }, (0, _languageHandler._t)("Create Space")));
    }

    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
      className: "mx_TagTileContextMenu_item mx_TagTileContextMenu_viewCommunity",
      onClick: this._onViewCommunityClick
    }, (0, _languageHandler._t)('View Community')), moveUp || moveDown ? /*#__PURE__*/_react.default.createElement("hr", {
      className: "mx_TagTileContextMenu_separator",
      role: "separator"
    }) : null, moveUp, moveDown, /*#__PURE__*/_react.default.createElement("hr", {
      className: "mx_TagTileContextMenu_separator",
      role: "separator"
    }), /*#__PURE__*/_react.default.createElement(_ContextMenu.MenuItem, {
      className: "mx_TagTileContextMenu_item mx_TagTileContextMenu_hideCommunity",
      onClick: this._onRemoveClick
    }, (0, _languageHandler._t)("Unpin")), createSpaceOption);
  }

}, (0, _defineProperty2.default)(_class2, "propTypes", {
  tag: _propTypes.default.string.isRequired,
  index: _propTypes.default.number.isRequired,

  /* callback called when the menu is dismissed */
  onFinished: _propTypes.default.func.isRequired
}), (0, _defineProperty2.default)(_class2, "contextType", _MatrixClientContext.default), _temp)) || _class);
exports.default = TagTileContextMenu;
